%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional
  \DeclareUnicodeCharacter{"00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{"2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{"2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{"2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{"251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{"2572}{\textbackslash}
 \else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{2572}{\textbackslash}
 \fi
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\captionsenglish{\renewcommand{\literalblockcontinuedname}{continued from previous page}}
\addto\captionsenglish{\renewcommand{\literalblockcontinuesname}{continues on next page}}

\addto\extrasenglish{\def\pageautorefname{page}}




% Jupyter Notebook code cell colors
\definecolor{nbsphinxin}{HTML}{307FC1}
\definecolor{nbsphinxout}{HTML}{BF5B3D}
\definecolor{nbsphinx-code-bg}{HTML}{F5F5F5}
\definecolor{nbsphinx-code-border}{HTML}{E0E0E0}
\definecolor{nbsphinx-stderr}{HTML}{FFDDDD}
% ANSI colors for output streams and traceback highlighting
\definecolor{ansi-black}{HTML}{3E424D}
\definecolor{ansi-black-intense}{HTML}{282C36}
\definecolor{ansi-red}{HTML}{E75C58}
\definecolor{ansi-red-intense}{HTML}{B22B31}
\definecolor{ansi-green}{HTML}{00A250}
\definecolor{ansi-green-intense}{HTML}{007427}
\definecolor{ansi-yellow}{HTML}{DDB62B}
\definecolor{ansi-yellow-intense}{HTML}{B27D12}
\definecolor{ansi-blue}{HTML}{208FFB}
\definecolor{ansi-blue-intense}{HTML}{0065CA}
\definecolor{ansi-magenta}{HTML}{D160C4}
\definecolor{ansi-magenta-intense}{HTML}{A03196}
\definecolor{ansi-cyan}{HTML}{60C6C8}
\definecolor{ansi-cyan-intense}{HTML}{258F8F}
\definecolor{ansi-white}{HTML}{C5C1B4}
\definecolor{ansi-white-intense}{HTML}{A1A6B2}
\definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
\definecolor{ansi-default-inverse-bg}{HTML}{000000}

% Define an environment for non-plain-text code cell outputs (e.g. images)
\makeatletter
\newenvironment{nbsphinxfancyoutput}{%
    % Avoid fatal error with framed.sty if graphics too long to fit on one page
    \let\sphinxincludegraphics\nbsphinxincludegraphics
    \nbsphinx@image@maxheight\textheight
    \advance\nbsphinx@image@maxheight -2\fboxsep   % default \fboxsep 3pt
    \advance\nbsphinx@image@maxheight -2\fboxrule  % default \fboxrule 0.4pt
    \advance\nbsphinx@image@maxheight -\baselineskip
\def\nbsphinxfcolorbox{\spx@fcolorbox{nbsphinx-code-border}{white}}%
\def\FrameCommand{\nbsphinxfcolorbox\nbsphinxfancyaddprompt\@empty}%
\def\FirstFrameCommand{\nbsphinxfcolorbox\nbsphinxfancyaddprompt\sphinxVerbatim@Continues}%
\def\MidFrameCommand{\nbsphinxfcolorbox\sphinxVerbatim@Continued\sphinxVerbatim@Continues}%
\def\LastFrameCommand{\nbsphinxfcolorbox\sphinxVerbatim@Continued\@empty}%
\MakeFramed{\advance\hsize-\width\@totalleftmargin\z@\linewidth\hsize\@setminipage}%
}{\par\unskip\@minipagefalse\endMakeFramed}
\makeatother
\newbox\nbsphinxpromptbox
\def\nbsphinxfancyaddprompt{\ifvoid\nbsphinxpromptbox\else
    \kern\fboxrule\kern\fboxsep
    \copy\nbsphinxpromptbox
    \kern-\ht\nbsphinxpromptbox\kern-\dp\nbsphinxpromptbox
    \kern-\fboxsep\kern-\fboxrule\nointerlineskip
    \fi}
\newlength\nbsphinxcodecellspacing
\setlength{\nbsphinxcodecellspacing}{0pt}

% Define support macros for attaching opening and closing lines to notebooks
\newsavebox\nbsphinxbox
\makeatletter
\newcommand{\nbsphinxstartnotebook}[1]{%
    \par
    % measure needed space
    \setbox\nbsphinxbox\vtop{{#1\par}}
    % reserve some space at bottom of page, else start new page
    \needspace{\dimexpr2.5\baselineskip+\ht\nbsphinxbox+\dp\nbsphinxbox}
    % mimick vertical spacing from \section command
      \addpenalty\@secpenalty
      \@tempskipa 3.5ex \@plus 1ex \@minus .2ex\relax
      \addvspace\@tempskipa
      {\Large\@tempskipa\baselineskip
             \advance\@tempskipa-\prevdepth
             \advance\@tempskipa-\ht\nbsphinxbox
             \ifdim\@tempskipa>\z@
               \vskip \@tempskipa
             \fi}
    \unvbox\nbsphinxbox
    % if notebook starts with a \section, prevent it from adding extra space
    \@nobreaktrue\everypar{\@nobreakfalse\everypar{}}%
    % compensate the parskip which will get inserted by next paragraph
    \nobreak\vskip-\parskip
    % do not break here
    \nobreak
}% end of \nbsphinxstartnotebook

\newcommand{\nbsphinxstopnotebook}[1]{%
    \par
    % measure needed space
    \setbox\nbsphinxbox\vbox{{#1\par}}
    \nobreak % it updates page totals
    \dimen@\pagegoal
    \advance\dimen@-\pagetotal \advance\dimen@-\pagedepth
    \advance\dimen@-\ht\nbsphinxbox \advance\dimen@-\dp\nbsphinxbox
    \ifdim\dimen@<\z@
      % little space left
      \unvbox\nbsphinxbox
      \kern-.8\baselineskip
      \nobreak\vskip\z@\@plus1fil
      \penalty100
      \vskip\z@\@plus-1fil
      \kern.8\baselineskip
    \else
      \unvbox\nbsphinxbox
    \fi
}% end of \nbsphinxstopnotebook

% Ensure height of an included graphics fits in nbsphinxfancyoutput frame
\newdimen\nbsphinx@image@maxheight % set in nbsphinxfancyoutput environment
\newcommand*{\nbsphinxincludegraphics}[2][]{%
    \gdef\spx@includegraphics@options{#1}%
    \setbox\spx@image@box\hbox{\includegraphics[#1,draft]{#2}}%
    \in@false
    \ifdim \wd\spx@image@box>\linewidth
      \g@addto@macro\spx@includegraphics@options{,width=\linewidth}%
      \in@true
    \fi
    % no rotation, no need to worry about depth
    \ifdim \ht\spx@image@box>\nbsphinx@image@maxheight
      \g@addto@macro\spx@includegraphics@options{,height=\nbsphinx@image@maxheight}%
      \in@true
    \fi
    \ifin@
      \g@addto@macro\spx@includegraphics@options{,keepaspectratio}%
    \fi
    \setbox\spx@image@box\box\voidb@x % clear memory
    \expandafter\includegraphics\expandafter[\spx@includegraphics@options]{#2}%
}% end of "\MakeFrame"-safe variant of \sphinxincludegraphics
\makeatother



\title{celloracle}
\date{Sep 28, 2019}
\release{0.1.0}
\author{}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}


CellOracle is a python library for the analysis of Gene Regulatory Network with single cell data.

Source code are available at \sphinxhref{https://github.com/KenjiKamimoto-wustl122/CellOracle}{celloracle  git hub repository}

\begin{sphinxadmonition}{note}{Note:}
\begin{DUlineblock}{0em}
\item[] \sphinxstylestrong{For colleagues in Morris Lab}
\item[] \sphinxstylestrong{Thank you for your kind help!!}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Documentation is also available as a word and pdf file.
\end{DUlineblock}

\sphinxcode{\sphinxupquote{pdf documentation}}

\sphinxcode{\sphinxupquote{word documentation}}

\begin{DUlineblock}{0em}
\item[] \sphinxstylestrong{Could you please add comments on the file and send it back to Kenji?}
\item[] I really appreciate your support!
\item[] 
\item[] Thanks,
\item[] Kenji
\end{DUlineblock}
\begin{quote}

(This message will be deleted when we launch celloracle.)
\end{quote}
\end{sphinxadmonition}

\begin{sphinxadmonition}{warning}{Warning:}
CellOracle is still under development. It is alpha version and functions in this package may change in the future release.
\end{sphinxadmonition}


\chapter{Contents}
\label{\detokenize{index:contents}}

\section{Installation}
\label{\detokenize{installation/index:installation}}\label{\detokenize{installation/index:install}}\label{\detokenize{installation/index::doc}}
\sphinxcode{\sphinxupquote{celloracle}} uses several python libraries and R library. Please follow this guide below to install the dependent software of celloracle.


\subsection{Python Requirements}
\label{\detokenize{installation/index:python-requirements}}\label{\detokenize{installation/index:require}}\begin{itemize}
\item {} 
\sphinxcode{\sphinxupquote{celloracle}} was developed with python 3.6. We do not support python 2.7x or python \textless{}=3.5.

\item {} 
\sphinxcode{\sphinxupquote{celloracle}} was developed in Linux and macOS. We do not guarantee that \sphinxcode{\sphinxupquote{celloracle}} works in Windows OS.

\item {} 
We highly recommend using \sphinxhref{https://www.continuum.io/downloads}{anaconda} to setup python environment.

\item {} 
Please install all dependent libraries before installing \sphinxcode{\sphinxupquote{celloracle}} according to the instructions below.

\item {} 
\sphinxcode{\sphinxupquote{celloracle}}  is still beta version and it is not available through PyPI or anaconda distribution yet. Please install \sphinxcode{\sphinxupquote{celloracle}} from GitHub repository according to the instruction below.

\end{itemize}


\subsubsection{0. (Optional) Make a new environment}
\label{\detokenize{installation/index:optional-make-a-new-environment}}
This step is optional. Please make a new python environment for celloracle and install dependent libraries in it if you get some software conflicts.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{make} \PYG{o}{\PYGZhy{}}\PYG{n}{n} \PYG{n}{celloracle\PYGZus{}env} \PYG{n}{python}\PYG{o}{=}\PYG{l+m+mf}{3.6}
\PYG{n}{conda} \PYG{n}{activate} \PYG{n}{celloracle\PYGZus{}env}
\end{sphinxVerbatim}


\subsubsection{1. Add conda channels}
\label{\detokenize{installation/index:add-conda-channels}}
Installation of some libraries below requires non-default anaconda channels. Please add the channels below. Instead, you can explicitly enter the channel when you install a library.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{config} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{add} \PYG{n}{channels} \PYG{n}{defaults}
\PYG{n}{conda} \PYG{n}{config} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{add} \PYG{n}{channels} \PYG{n}{bioconda}
\PYG{n}{conda} \PYG{n}{config} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{add} \PYG{n}{channels} \PYG{n}{conda}\PYG{o}{\PYGZhy{}}\PYG{n}{forge}
\end{sphinxVerbatim}


\subsubsection{2. Install velocyto}
\label{\detokenize{installation/index:install-velocyto}}
Please install velocyto with the following commands or \sphinxhref{http://velocyto.org/velocyto.py/install/index.html}{the author’s instruction} .
On Mac OS, you may have a compile error during velocyto installation. I recommend you installing \sphinxhref{https://developer.apple.com/xcode/}{Xcode} in that case.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{numpy} \PYG{n}{scipy} \PYG{n}{cython} \PYG{n}{numba} \PYG{n}{matplotlib} \PYG{n}{scikit}\PYG{o}{\PYGZhy{}}\PYG{n}{learn} \PYG{n}{h5py} \PYG{n}{click}
\end{sphinxVerbatim}

Then

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{velocyto}
\end{sphinxVerbatim}


\subsubsection{3. Install scanpy}
\label{\detokenize{installation/index:install-scanpy}}
Please install scanpy with the following commands or \sphinxhref{https://scanpy.readthedocs.io/en/stable/installation.html}{the author’s instruction} .

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{seaborn} \PYG{n}{scikit}\PYG{o}{\PYGZhy{}}\PYG{n}{learn} \PYG{n}{statsmodels} \PYG{n}{numba} \PYG{n}{pytables} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{igraph} \PYG{n}{louvain}
\end{sphinxVerbatim}

Then

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{scanpy}
\end{sphinxVerbatim}


\subsubsection{4. Install gimmemotifs}
\label{\detokenize{installation/index:install-gimmemotifs}}
Please install gimmemotifs with the following commands or \sphinxhref{https://gimmemotifs.readthedocs.io/en/master/installation.html}{the author’s instruction} .

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{gimmemotifs} \PYG{n}{genomepy}\PYG{o}{=}\PYG{l+m+mf}{0.5}\PYG{o}{.}\PYG{l+m+mi}{5}
\end{sphinxVerbatim}


\subsubsection{5. Install other python libraries}
\label{\detokenize{installation/index:install-other-python-libraries}}
Please install other python libraries below with the following commands.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{goatools} \PYG{n}{pyarrow} \PYG{n}{tqdm} \PYG{n}{joblib} \PYG{n}{jupyter}
\end{sphinxVerbatim}


\subsubsection{6. install celloracle from github}
\label{\detokenize{installation/index:install-celloracle-from-github}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{git}\PYG{o}{+}\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{github}\PYG{o}{.}\PYG{n}{com}\PYG{o}{/}\PYG{n}{KenjiKamimoto}\PYG{o}{\PYGZhy{}}\PYG{n}{wustl122}\PYG{o}{/}\PYG{n}{CellOracle}
\end{sphinxVerbatim}


\subsection{R requirements}
\label{\detokenize{installation/index:r-requirements}}
\sphinxcode{\sphinxupquote{celloracle}} use R library for the network analysis and scATAC-seq analysis.
Please install \sphinxhref{https://www.r-project.org}{R} (\textgreater{}=3.5) and R libraries below according to the author’s instruction.


\subsubsection{Seurat}
\label{\detokenize{installation/index:id3}}
Please install \sphinxcode{\sphinxupquote{Seurat}} with the following r-script or \sphinxhref{https://satijalab.org/seurat/install.html}{the author’s instruction} .
\sphinxcode{\sphinxupquote{celloracle}} is compatible with both Seurat V2 and V3.
If you use only \sphinxcode{\sphinxupquote{scanpy}} for the scRNA-seq preprocessing and do not use \sphinxcode{\sphinxupquote{Seurat}} , you can skip installation of \sphinxcode{\sphinxupquote{Seurat}}.

in R console,

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{Seurat\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{Cicero}
\label{\detokenize{installation/index:id5}}
Please install \sphinxcode{\sphinxupquote{Cicero}} with the following r-script or \sphinxhref{https://cole-trapnell-lab.github.io/cicero-release/docs/\#installing-cicero}{the author’s instruction} .
If you have no plan for scATAC-seq analysis and just want to use  \sphinxcode{\sphinxupquote{celloracle}} with a default TF information which was supplied with celloracle, you can skip installation of \sphinxcode{\sphinxupquote{Cicero}}.

in R console,

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{if }\PYG{p}{(}\PYG{o}{!}\PYG{n+nf}{requireNamespace}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{BiocManager\PYGZdq{}}\PYG{p}{,} \PYG{n}{quietly} \PYG{o}{=} \PYG{k+kc}{TRUE}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{BiocManager\PYGZdq{}}\PYG{p}{)}
\PYG{n}{BiocManager}\PYG{o}{::}\PYG{n+nf}{install}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{cicero\PYGZdq{}}\PYG{p}{,} \PYG{n}{version} \PYG{o}{=} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{3.8\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{igraph}
\label{\detokenize{installation/index:id7}}
Please install \sphinxcode{\sphinxupquote{igraph}} with the following r-script or \sphinxhref{https://igraph.org/r/}{the author’s instruction} .

in R console,

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{igraph\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{linkcomm}
\label{\detokenize{installation/index:id9}}
Please install \sphinxcode{\sphinxupquote{linkcomm}} with the following r-script or \sphinxhref{https://cran.r-project.org/web/packages/linkcomm/index.html}{the author’s instruction} .

in R console,

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{linkcomm\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{rnetcarto}
\label{\detokenize{installation/index:id11}}
\sphinxcode{\sphinxupquote{rnetcarto}} installation has to be done with several steps. Please install rnetcarto with \sphinxhref{https://github.com/cran/rnetcarto/blob/master/src/rgraph/README.md}{the author’s instruction} .
You need to install \sphinxhref{https://www.gnu.org/software/gsl/}{the GNU Scientific Libraries} before installing rnetcarto. Detailed instruction can be found \sphinxhref{https://github.com/cran/rnetcarto/blob/master/src/rgraph/README.md}{here} .


\subsubsection{Check installation}
\label{\detokenize{installation/index:check-installation}}
These R libraries above are necessary for the network analysis in celloracle. You can check installation using celloracle’s function.

in python console,

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{celloracle} \PYG{k+kn}{as} \PYG{n+nn}{co}
\PYG{n}{co}\PYG{o}{.}\PYG{n}{network\PYGZus{}analysis}\PYG{o}{.}\PYG{n}{test\PYGZus{}R\PYGZus{}libraries\PYGZus{}installation}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

Please make sure that all R libraries are installed. The following message will be shown when all R libraries are appropriately installed.

\begin{DUlineblock}{0em}
\item[] checking R library installation: gProfileR -\textgreater{} OK
\item[] checking R library installation: igraph -\textgreater{} OK
\item[] checking R library installation: linkcomm -\textgreater{} OK
\item[] checking R library installation: rnetcarto -\textgreater{} OK
\end{DUlineblock}


\section{Tutorial}
\label{\detokenize{tutorials/index:tutorial}}\label{\detokenize{tutorials/index:id1}}\label{\detokenize{tutorials/index::doc}}
The analysis proceeds through multiple steps.
Please run the notebooks following the number of the notebook.
If you do not have ATAC-seq data and want to use the default TF binding information, you can skip the first and second step and start from the third step.

Please refer to the \sphinxcode{\sphinxupquote{celloracle}} paper for scientific premise and the detail of the algorithm of celloracle.

The jupyter notebook files in this tutorial are available \sphinxhref{https://github.com/KenjiKamimoto-wustl122/CellOracle/tree/master/docs/notebooks}{here} .


\subsection{ATAC-seq data preprocessing}
\label{\detokenize{tutorials/atac:atac-seq-data-preprocessing}}\label{\detokenize{tutorials/atac:atac}}\label{\detokenize{tutorials/atac::doc}}
In this step, we process scATAC-seq data (or bulk ATAC-seq data) to get open accessible promoter/enhancer DNA sequence.
We can get active proximal promoter/enhancer genome sequence by picking up ATAC-seq peaks that exist around the transcription starting site (TSS).
Distal cis-regulatory elements can be picked up using  \sphinxhref{https://cole-trapnell-lab.github.io/cicero-release/docs/\#installing-cicero}{Cicero} .
Cicero analyzes scATAC-seq data to calculate a co-accessible score between peaks.
We can identify cis-regulatory elements using Cicero co-access score and TSS information.

If you have bulk ATAC-seq data instead of scATAC-data, we’ll get only proximal promoter/enhancer genome sequence.


\subsubsection{A. Extract TF binding information from scATAC-seq data}
\label{\detokenize{tutorials/atac:a-extract-tf-binding-information-from-scatac-seq-data}}
If you have a scATAC-seq data, you can get information of distal cis-regulatory elements.
This step uses Cicero and does not use celloracle. Please refer to \sphinxhref{https://cole-trapnell-lab.github.io/cicero-release/}{the documentation of Cicero} for the detailed usage.

R notebook


\paragraph{0. Import library}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero:0.-Import-library}}\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n+nf}{library}\PYG{p}{(}\PYG{n}{cicero}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{1. Prepare data}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero:1.-Prepare-data}}
In this tutorial we use acATAC-seq data in 10x genomics database. You do not need to download these data if you analyze your scATAC data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} create folder to store data}
\PYG{n+nf}{dir.create}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{data\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} download demo dataset from 10x genomics}
\PYG{n+nf}{system}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{wget \PYGZhy{}O data/matrix.tar.gz http://cf.10xgenomics.com/samples/cell\PYGZhy{}atac/1.1.0/atac\PYGZus{}v1\PYGZus{}E18\PYGZus{}brain\PYGZus{}fresh\PYGZus{}5k/atac\PYGZus{}v1\PYGZus{}E18\PYGZus{}brain\PYGZus{}fresh\PYGZus{}5k\PYGZus{}filtered\PYGZus{}peak\PYGZus{}bc\PYGZus{}matrix.tar.gz\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} unzip data}
\PYG{n+nf}{system}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{tar \PYGZhy{}xvf data/matrix.tar.gz \PYGZhy{}C data\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} You can substitute the data path below with the data path of your scATAC data.}
\PYG{n}{data\PYGZus{}folder} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{data/filtered\PYGZus{}peak\PYGZus{}bc\PYGZus{}matrix\PYGZdq{}}

\PYG{c+c1}{\PYGZsh{} Create a folder to save results}
\PYG{n}{output\PYGZus{}folder} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{cicero\PYGZus{}output\PYGZdq{}}
\PYG{n+nf}{dir.create}\PYG{p}{(}\PYG{n}{output\PYGZus{}folder}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{2. Load data and make Cell Data Set (CDS) object}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero:2.-Load-data-and-make-Cell-Data-Set-(CDS)-object}}

\subparagraph{2.1. Process data to make CDS object}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero:2.1.-Process-data-to-make-CDS-object}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} read in matrix data using the Matrix package}
\PYG{n}{indata} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{readMM}\PYG{p}{(}\PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{data\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/matrix.mtx\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} binarize the matrix}
\PYG{n}{indata}\PYG{o}{@}\PYG{n}{x}\PYG{n}{[indata}\PYG{o}{@}\PYG{n}{x} \PYG{o}{\PYGZgt{}} \PYG{l+m}{0}\PYG{n}{]} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+m}{1}

\PYG{c+c1}{\PYGZsh{} format cell info}
\PYG{n}{cellinfo} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{read.table}\PYG{p}{(}\PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{data\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/barcodes.tsv\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{cellinfo}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{cellinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{V1}
\PYG{n+nf}{names}\PYG{p}{(}\PYG{n}{cellinfo}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{cells\PYGZdq{}}

\PYG{c+c1}{\PYGZsh{} format peak info}
\PYG{n}{peakinfo} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{read.table}\PYG{p}{(}\PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{data\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/peaks.bed\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{names}\PYG{p}{(}\PYG{n}{peakinfo}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{c}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{chr\PYGZdq{}}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{bp1\PYGZdq{}}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{bp2\PYGZdq{}}\PYG{p}{)}
\PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{site\PYGZus{}name} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{paste}\PYG{p}{(}\PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{chr}\PYG{p}{,} \PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{bp1}\PYG{p}{,} \PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{bp2}\PYG{p}{,} \PYG{n}{sep}\PYG{o}{=}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{\PYGZus{}\PYGZdq{}}\PYG{p}{)}
\PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{peakinfo}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{site\PYGZus{}name}

\PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{indata}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{peakinfo}\PYG{p}{)}
\PYG{n+nf}{colnames}\PYG{p}{(}\PYG{n}{indata}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{cellinfo}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} make CDS}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}}  \PYG{n+nf}{suppressWarnings}\PYG{p}{(}\PYG{n+nf}{newCellDataSet}\PYG{p}{(}\PYG{n}{indata}\PYG{p}{,}
                            \PYG{n}{phenoData} \PYG{o}{=} \PYG{n}{methods}\PYG{o}{::}\PYG{n+nf}{new}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{AnnotatedDataFrame\PYGZdq{}}\PYG{p}{,} \PYG{n}{data} \PYG{o}{=} \PYG{n}{cellinfo}\PYG{p}{)}\PYG{p}{,}
                            \PYG{n}{featureData} \PYG{o}{=} \PYG{n}{methods}\PYG{o}{::}\PYG{n+nf}{new}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{AnnotatedDataFrame\PYGZdq{}}\PYG{p}{,} \PYG{n}{data} \PYG{o}{=} \PYG{n}{peakinfo}\PYG{p}{)}\PYG{p}{,}
                            \PYG{n}{expressionFamily}\PYG{o}{=}\PYG{n}{VGAM}\PYG{o}{::}\PYG{n+nf}{binomialff}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,}
                            \PYG{n}{lowerDetectionLimit}\PYG{o}{=}\PYG{l+m}{0}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{input\PYGZus{}cds}\PYG{o}{@}\PYG{n}{expressionFamily}\PYG{o}{@}\PYG{n}{vfamily} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{binomialff\PYGZdq{}}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{monocle}\PYG{o}{::}\PYG{n+nf}{detectGenes}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{}Ensure there are no peaks included with zero reads}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{input\PYGZus{}cds}\PYG{n}{[Matrix}\PYG{o}{::}\PYG{n+nf}{rowSums}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)} \PYG{o}{\PYGZgt{}=} \PYG{l+m}{100}\PYG{p}{,}\PYG{n}{]}
\end{sphinxVerbatim}
}


\paragraph{3. Qauality check and Filtering}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero:3.-Qauality-check-and-Filtering}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Visualize peak\PYGZus{}count\PYGZus{}per\PYGZus{}cell}
\PYG{n+nf}{hist}\PYG{p}{(}\PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{colSums}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=420\sphinxpxdimen,height=420\sphinxpxdimen]{{notebooks_01_ATAC-seq_data_processing_option1_scATAC-seq_data_analysis_with_cicero_01_atacdata_to_cicero_8_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} filter cells by peak\PYGZus{}count}
\PYG{n}{max\PYGZus{}count} \PYG{o}{\PYGZlt{}\PYGZhy{}}  \PYG{l+m}{15000} \PYG{c+c1}{\PYGZsh{} Please change the threshold value according to the distribution of the peak\PYGZus{}count of your data}
\PYG{n}{min\PYGZus{}count} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+m}{2000} \PYG{c+c1}{\PYGZsh{} Please change the threshold value according to the distribution of the peak\PYGZus{}count of your data}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{input\PYGZus{}cds}\PYG{n}{[}\PYG{p}{,}\PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{colSums}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)} \PYG{o}{\PYGZgt{}=} \PYG{n}{min\PYGZus{}count}\PYG{n}{]}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{input\PYGZus{}cds}\PYG{n}{[}\PYG{p}{,}\PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{colSums}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)} \PYG{o}{\PYGZlt{}=} \PYG{n}{max\PYGZus{}count}\PYG{n}{]}
\end{sphinxVerbatim}
}


\paragraph{4. Process cicero-CDS object}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero:4.-Process-cicero-CDS-object}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Run cicero to get cis\PYGZhy{}regulatory networks}
\PYG{n+nf}{set.seed}\PYG{p}{(}\PYG{l+m}{2017}\PYG{p}{)}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{detectGenes}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{estimateSizeFactors}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}

\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{reduceDimension}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{,} \PYG{n}{max\PYGZus{}components} \PYG{o}{=} \PYG{l+m}{2}\PYG{p}{,} \PYG{n}{verbose}\PYG{o}{=}\PYG{n+nb+bp}{T}\PYG{p}{,}\PYG{n}{scaling} \PYG{o}{=} \PYG{k+kc}{FALSE}\PYG{p}{,}\PYG{n}{relative\PYGZus{}expr}\PYG{o}{=}\PYG{k+kc}{FALSE}\PYG{p}{,}
                      \PYG{n}{reduction\PYGZus{}method} \PYG{o}{=} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{tSNE\PYGZsq{}}\PYG{p}{,} \PYG{n}{norm\PYGZus{}method} \PYG{o}{=} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{none\PYGZdq{}}\PYG{p}{)}

\PYG{n}{tsne\PYGZus{}coords} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{t}\PYG{p}{(}\PYG{n+nf}{reducedDimA}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{tsne\PYGZus{}coords}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n+nf}{pData}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{cicero\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{make\PYGZus{}cicero\PYGZus{}cds}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{,} \PYG{n}{reduced\PYGZus{}coordinates} \PYG{o}{=} \PYG{n}{tsne\PYGZus{}coords}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Save cicero\PYGZhy{}CDS object if you want.}
\PYG{c+c1}{\PYGZsh{}saveRDS(cicero\PYGZus{}cds, paste0(output\PYGZus{}folder, \PYGZdq{}/cicero\PYGZus{}cds.Rds\PYGZdq{}))}

\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Remove noise by PCA {\ldots}

Reduce dimension by tSNE {\ldots}

Overlap QC metrics:
Cells per bin: 50
Maximum shared cells bin-bin: 44
Mean shared cells bin-bin: 0.76256263875674
Median shared cells bin-bin: 0

\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\paragraph{5. Run cicero to get cis-regulatory connection scores}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero:5.-Run-cicero-to-get-cis-regulatory-connection-scores}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} import genome length, which is needed for the function, run\PYGZus{}cicero}
\PYG{n}{mm10\PYGZus{}chromosome\PYGZus{}length} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{read.table}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{./mm10\PYGZus{}chromosome\PYGZus{}length.txt\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} runc the main function}
\PYG{n}{conns} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{run\PYGZus{}cicero}\PYG{p}{(}\PYG{n}{cicero\PYGZus{}cds}\PYG{p}{,} \PYG{n}{mm10\PYGZus{}chromosome\PYGZus{}length}\PYG{p}{)} \PYG{c+c1}{\PYGZsh{} Takes a few minutes to run}

\PYG{c+c1}{\PYGZsh{} check results}
\PYG{n+nf}{head}\PYG{p}{(}\PYG{n}{conns}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
[1] "Starting Cicero"
[1] "Calculating distance\_parameter value"
[1] "Running models"
[1] "Assembling connections"
[1] "Done"
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}
A data.frame: 6 × 3
\begin{tabular}{r|lll}
  & Peak1 & Peak2 & coaccess\\
  & <fct> & <fct> & <dbl>\\
\hline
    2 & chr1\_3094484\_3095479 & chr1\_3113499\_3113979 & -0.316289004\\
    3 & chr1\_3094484\_3095479 & chr1\_3119478\_3121690 & -0.419240532\\
    4 & chr1\_3094484\_3095479 & chr1\_3399730\_3400368 & -0.050867246\\
    5 & chr1\_3113499\_3113979 & chr1\_3094484\_3095479 & -0.316289004\\
    7 & chr1\_3113499\_3113979 & chr1\_3119478\_3121690 &  0.370342744\\
    8 & chr1\_3113499\_3113979 & chr1\_3399730\_3400368 & -0.009276026\\
\end{tabular}
\end{nbsphinxfancyoutput}


\paragraph{6. Save results for next step}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero:6.-Save-results-for-next-step}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{all\PYGZus{}peaks} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{write.csv}\PYG{p}{(}\PYG{n}{x} \PYG{o}{=} \PYG{n}{all\PYGZus{}peaks}\PYG{p}{,} \PYG{n}{file} \PYG{o}{=} \PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{output\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/all\PYGZus{}peaks.csv\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{write.csv}\PYG{p}{(}\PYG{n}{x} \PYG{o}{=} \PYG{n}{conns}\PYG{p}{,} \PYG{n}{file} \PYG{o}{=} \PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{output\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/cicero\PYGZus{}connections.csv\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}
}

Next, the results of Cicero analysis will be processed to make TSS annotations.

Python notebook

In this notebook, we process the results of cicero analysis to get active promoter/enhancer DNA peaks. First, we pick up peaks around the transcription starting site (TSS). Second, we merge cicero data with the peaks around TSS. Then we remove peaks that have a weak connection to TSS peak so that the final product includes TSS peaks and peaks that have a strong connection with the TSS peaks. We use this information as an active promoter/enhancer elements.


\paragraph{0. Import libraries}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:0.-Import-libraries}}\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}
\PYG{c+c1}{\PYGZsh{}import time}

\PYG{k+kn}{import} \PYG{n+nn}{os}\PYG{o}{,} \PYG{n+nn}{sys}\PYG{o}{,} \PYG{n+nn}{shutil}\PYG{o}{,} \PYG{n+nn}{importlib}\PYG{o}{,} \PYG{n+nn}{glob}
\PYG{k+kn}{from} \PYG{n+nn}{tqdm} \PYG{k}{import} \PYG{n}{tqdm\PYGZus{}notebook} \PYG{k}{as} \PYG{n}{tqdm}

\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{300}


\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{from} \PYG{n+nn}{celloracle} \PYG{k}{import} \PYG{n}{motif\PYGZus{}analysis} \PYG{k}{as} \PYG{n}{ma}
\end{sphinxVerbatim}
}


\paragraph{1. Load data made with cicero}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:1.-Load-data-made-with-cicero}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} load all peaks}
\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cicero\PYGZus{}output/all\PYGZus{}peaks.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{index\PYGZus{}col}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{peaks}\PYG{o}{.}\PYG{n}{x}\PYG{o}{.}\PYG{n}{values}
\PYG{n}{peaks}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}array([\PYGZsq{}chr1\PYGZus{}3094484\PYGZus{}3095479\PYGZsq{}, \PYGZsq{}chr1\PYGZus{}3113499\PYGZus{}3113979\PYGZsq{},
       \PYGZsq{}chr1\PYGZus{}3119478\PYGZus{}3121690\PYGZsq{}, ..., \PYGZsq{}chrY\PYGZus{}90804622\PYGZus{}90805450\PYGZsq{},
       \PYGZsq{}chrY\PYGZus{}90808626\PYGZus{}90809117\PYGZsq{}, \PYGZsq{}chrY\PYGZus{}90810560\PYGZus{}90811167\PYGZsq{}], dtype=object)
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} load cicero results}
\PYG{n}{cicero\PYGZus{}connections} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cicero\PYGZus{}output/cicero\PYGZus{}connections.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{index\PYGZus{}col}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{n}{cicero\PYGZus{}connections}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                  Peak1                 Peak2  coaccess
2  chr1\PYGZus{}3094484\PYGZus{}3095479  chr1\PYGZus{}3113499\PYGZus{}3113979 \PYGZhy{}0.316289
3  chr1\PYGZus{}3094484\PYGZus{}3095479  chr1\PYGZus{}3119478\PYGZus{}3121690 \PYGZhy{}0.419241
4  chr1\PYGZus{}3094484\PYGZus{}3095479  chr1\PYGZus{}3399730\PYGZus{}3400368 \PYGZhy{}0.050867
5  chr1\PYGZus{}3113499\PYGZus{}3113979  chr1\PYGZus{}3094484\PYGZus{}3095479 \PYGZhy{}0.316289
7  chr1\PYGZus{}3113499\PYGZus{}3113979  chr1\PYGZus{}3119478\PYGZus{}3121690  0.370343
\end{sphinxVerbatim}
}


\paragraph{2. Make TSS annotation}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:2.-Make-TSS-annotation}}

\subparagraph{!! Please make sure that you are setting correct reference genoms.}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:!!-Please-make-sure-that-you-are-setting-correct-reference-genoms.}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{tss\PYGZus{}annotated} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{get\PYGZus{}tss\PYGZus{}info}\PYG{p}{(}\PYG{n}{peak\PYGZus{}str\PYGZus{}list}\PYG{o}{=}\PYG{n}{peaks}\PYG{p}{,} \PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mm10}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} check results}
\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{tail}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
que bed peaks: 72402
tss peaks in que: 16987
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}         chr      start        end gene\PYGZus{}short\PYGZus{}name strand
16982   chr1   55130650   55132118            Mob4      +
16983   chr6   94499875   94500767        Slc25a26      +
16984  chr19   45659222   45660823           Fbxw4      \PYGZhy{}
16985  chr12  100898848  100899597           Gpr68      \PYGZhy{}
16986   chr4  129491262  129492047         Fam229a      \PYGZhy{}
\end{sphinxVerbatim}
}


\paragraph{3. Integrate TSS info and cicero connections}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:3.-Integrate-TSS-info-and-cicero-connections}}
The output file after the integration process has three columns; “peak\_id”, “gene\_short\_name”, and “coaccess”. “peak\_id” is either of TSS peak or peaks that have a connection with a TSS peak. “gene\_short\_name” is a gene name of the TSS site. “coaccess” is a co-access score between a peak and TSS peak. If the score is 1, it means that the peak is TSS itself.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{integrated} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{integrate\PYGZus{}tss\PYGZus{}peak\PYGZus{}with\PYGZus{}cicero}\PYG{p}{(}\PYG{n}{tss\PYGZus{}peak}\PYG{o}{=}\PYG{n}{tss\PYGZus{}annotated}\PYG{p}{,}
                                               \PYG{n}{cicero\PYGZus{}connections}\PYG{o}{=}\PYG{n}{cicero\PYGZus{}connections}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{integrated}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{integrated}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
(263279, 3)
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\PYGZus{}id gene\PYGZus{}short\PYGZus{}name  coaccess
0  chr10\PYGZus{}100015291\PYGZus{}100017830            Kitl  1.000000
1  chr10\PYGZus{}100018677\PYGZus{}100020384            Kitl  0.086299
2  chr10\PYGZus{}100050858\PYGZus{}100051762            Kitl  0.034558
3  chr10\PYGZus{}100052829\PYGZus{}100053395            Kitl  0.167188
4  chr10\PYGZus{}100128086\PYGZus{}100128882           Tmtc3  0.022341
\end{sphinxVerbatim}
}


\paragraph{4. Filter peaks}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:4.-Filter-peaks}}
Remove peaks that have weak coaccess score.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{peak} \PYG{o}{=} \PYG{n}{integrated}\PYG{p}{[}\PYG{n}{integrated}\PYG{o}{.}\PYG{n}{coaccess} \PYG{o}{\PYGZgt{}}\PYG{o}{=} \PYG{l+m+mf}{0.8}\PYG{p}{]}
\PYG{n}{peak} \PYG{o}{=} \PYG{n}{peak}\PYG{p}{[}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}id}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{gene\PYGZus{}short\PYGZus{}name}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{]}\PYG{o}{.}\PYG{n}{reset\PYGZus{}index}\PYG{p}{(}\PYG{n}{drop}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{peak}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{peak}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
(15680, 2)
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\PYGZus{}id gene\PYGZus{}short\PYGZus{}name
0  chr10\PYGZus{}100015291\PYGZus{}100017830            Kitl
1  chr10\PYGZus{}100486534\PYGZus{}100488209           Tmtc3
2  chr10\PYGZus{}100588641\PYGZus{}100589556   4930430F08Rik
3  chr10\PYGZus{}100741247\PYGZus{}100742505         Gm35722
4  chr10\PYGZus{}101681379\PYGZus{}101682124          Mgat4c
\end{sphinxVerbatim}
}


\paragraph{5. Save data}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:5.-Save-data}}
Save the promoter/enhancer peak.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{peak}\PYG{o}{.}\PYG{n}{to\PYGZus{}parquet}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}file.parquet}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

-\textgreater{} go to next notebook


\subsubsection{B. Extract TF binding information from bulk ATAC-seq data or Chip-seq data}
\label{\detokenize{tutorials/atac:b-extract-tf-binding-information-from-bulk-atac-seq-data-or-chip-seq-data}}
Bulk DNA-seq data can be used to get open accessible promoter/enhancer sequence.

Python notebook


\paragraph{0. Import libraries}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data:0.-Import-libraries}}\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}
\PYG{c+c1}{\PYGZsh{}import time}

\PYG{k+kn}{import} \PYG{n+nn}{os}\PYG{o}{,} \PYG{n+nn}{sys}\PYG{o}{,} \PYG{n+nn}{shutil}\PYG{o}{,} \PYG{n+nn}{importlib}\PYG{o}{,} \PYG{n+nn}{glob}
\PYG{k+kn}{from} \PYG{n+nn}{tqdm} \PYG{k}{import} \PYG{n}{tqdm\PYGZus{}notebook} \PYG{k}{as} \PYG{n}{tqdm}

\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{300}

\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} import celloracle function}
\PYG{k+kn}{from} \PYG{n+nn}{celloracle} \PYG{k}{import} \PYG{n}{motif\PYGZus{}analysis} \PYG{k}{as} \PYG{n}{ma}
\end{sphinxVerbatim}
}


\paragraph{1. Load bed file}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data:1.-Load-bed-file}}
Import bed file of ATAC-seq data. This script can also be used DNase-seq or Chip-seq data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{file\PYGZus{}path\PYGZus{}of\PYGZus{}bed\PYGZus{}file} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{data/all\PYGZus{}peaks.bed}\PYG{l+s+s2}{\PYGZdq{}}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} load bed\PYGZus{}file}
\PYG{n}{bed} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{read\PYGZus{}bed}\PYG{p}{(}\PYG{n}{file\PYGZus{}path\PYGZus{}of\PYGZus{}bed\PYGZus{}file}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{bed}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{bed}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
(436206, 4)
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}  chrom    start      end               seqname
0  chr1  3002478  3002968  chr1\PYGZus{}3002478\PYGZus{}3002968
1  chr1  3084739  3085712  chr1\PYGZus{}3084739\PYGZus{}3085712
2  chr1  3103576  3104022  chr1\PYGZus{}3103576\PYGZus{}3104022
3  chr1  3106871  3107210  chr1\PYGZus{}3106871\PYGZus{}3107210
4  chr1  3108932  3109158  chr1\PYGZus{}3108932\PYGZus{}3109158
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} convert bed file into peak name list}
\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{process\PYGZus{}bed\PYGZus{}file}\PYG{o}{.}\PYG{n}{df\PYGZus{}to\PYGZus{}list\PYGZus{}peakstr}\PYG{p}{(}\PYG{n}{bed}\PYG{p}{)}
\PYG{n}{peaks}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}array([\PYGZsq{}chr1\PYGZus{}3002478\PYGZus{}3002968\PYGZsq{}, \PYGZsq{}chr1\PYGZus{}3084739\PYGZus{}3085712\PYGZsq{},
       \PYGZsq{}chr1\PYGZus{}3103576\PYGZus{}3104022\PYGZsq{}, ..., \PYGZsq{}chrY\PYGZus{}631222\PYGZus{}631480\PYGZsq{},
       \PYGZsq{}chrY\PYGZus{}795887\PYGZus{}796426\PYGZsq{}, \PYGZsq{}chrY\PYGZus{}2397419\PYGZus{}2397628\PYGZsq{}], dtype=object)
\end{sphinxVerbatim}
}


\paragraph{2. Make TSS annotation}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data:2.-Make-TSS-annotation}}
!! Please make sure that you are setting correct ref genoms.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{tss\PYGZus{}annotated} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{get\PYGZus{}tss\PYGZus{}info}\PYG{p}{(}\PYG{n}{peak\PYGZus{}str\PYGZus{}list}\PYG{o}{=}\PYG{n}{peaks}\PYG{p}{,} \PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mm9}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} check results}
\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{tail}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
que bed peaks: 436206
tss peaks in que: 24822
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}         chr     start       end gene\PYGZus{}short\PYGZus{}name strand
24817   chr2  60560211  60561602           Itgb6      \PYGZhy{}
24818  chr15   3975177   3978654        BC037032      \PYGZhy{}
24819  chr14  67690701  67692101         Ppp2r2a      \PYGZhy{}
24820  chr17  48455247  48455773   B430306N03Rik      +
24821  chr10  59861192  59861608         Gm17455      +
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} change format}
\PYG{n}{peak\PYGZus{}id\PYGZus{}tss} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{process\PYGZus{}bed\PYGZus{}file}\PYG{o}{.}\PYG{n}{df\PYGZus{}to\PYGZus{}list\PYGZus{}peakstr}\PYG{p}{(}\PYG{n}{tss\PYGZus{}annotated}\PYG{p}{)}
\PYG{n}{tss\PYGZus{}annotated} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}id}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{peak\PYGZus{}id\PYGZus{}tss}\PYG{p}{,}
                              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{gene\PYGZus{}short\PYGZus{}name}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{gene\PYGZus{}short\PYGZus{}name}\PYG{o}{.}\PYG{n}{values}\PYG{p}{\PYGZcb{}}\PYG{p}{)}
\PYG{n}{tss\PYGZus{}annotated} \PYG{o}{=} \PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{reset\PYGZus{}index}\PYG{p}{(}\PYG{n}{drop}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
(24822, 2)
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                   peak\PYGZus{}id gene\PYGZus{}short\PYGZus{}name
0   chr7\PYGZus{}50691730\PYGZus{}50692032            Nkg7
1   chr7\PYGZus{}50692077\PYGZus{}50692785            Nkg7
2  chr13\PYGZus{}93564413\PYGZus{}93564836           Thbs4
3  chr13\PYGZus{}14613429\PYGZus{}14615645           Hecw1
4   chr3\PYGZus{}99688753\PYGZus{}99689665          Spag17
\end{sphinxVerbatim}
}


\paragraph{3. Save data}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data:3.-Save-data}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{to\PYGZus{}parquet}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}file.parquet}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

-\textgreater{} go to next notebook


\subsection{Transcription factor binding motif scan}
\label{\detokenize{tutorials/motifscan:transcription-factor-binding-motif-scan}}\label{\detokenize{tutorials/motifscan:motifscan}}\label{\detokenize{tutorials/motifscan::doc}}
We identified open-accessible Promoter/enhancer DNAs using ATAC-seq data.
Next, we scan the regulatory genomic sequence searching for the TF-binding motifs to get the list of TFs for each target gene.
In the later GRN inference process, this TF list per target gene will be used as a potential regulatory connection.

Python notebook


\subsubsection{0. Import libraries}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:0.-Import-libraries}}\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}
\PYG{c+c1}{\PYGZsh{}import time}

\PYG{k+kn}{import} \PYG{n+nn}{os}\PYG{o}{,} \PYG{n+nn}{sys}\PYG{o}{,} \PYG{n+nn}{shutil}\PYG{o}{,} \PYG{n+nn}{importlib}\PYG{o}{,} \PYG{n+nn}{glob}
\PYG{k+kn}{from} \PYG{n+nn}{tqdm} \PYG{k}{import} \PYG{n}{tqdm\PYGZus{}notebook} \PYG{k}{as} \PYG{n}{tqdm}

\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{15}\PYG{p}{,}\PYG{l+m+mi}{7}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{600}

\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{from} \PYG{n+nn}{celloracle} \PYG{k}{import} \PYG{n}{motif\PYGZus{}analysis} \PYG{k}{as} \PYG{n}{ma}
\PYG{k+kn}{from} \PYG{n+nn}{celloracle}\PYG{n+nn}{.}\PYG{n+nn}{utility} \PYG{k}{import} \PYG{n}{save\PYGZus{}as\PYGZus{}pickled\PYGZus{}object}
\end{sphinxVerbatim}
}


\subsubsection{1. Load data}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:1.-Load-data}}
load annotated peak data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}parquet}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{../01\PYGZus{}ATAC\PYGZhy{}seq\PYGZus{}data\PYGZus{}processing/option1\PYGZus{}scATAC\PYGZhy{}seq\PYGZus{}data\PYGZus{}analysis\PYGZus{}with\PYGZus{}cicero/peak\PYGZus{}file.parquet}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{peaks}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\PYGZus{}id gene\PYGZus{}short\PYGZus{}name
0  chr10\PYGZus{}100015291\PYGZus{}100017830            Kitl
1  chr10\PYGZus{}100486534\PYGZus{}100488209           Tmtc3
2  chr10\PYGZus{}100588641\PYGZus{}100589556   4930430F08Rik
3  chr10\PYGZus{}100741247\PYGZus{}100742505         Gm35722
4  chr10\PYGZus{}101681379\PYGZus{}101682124          Mgat4c
\end{sphinxVerbatim}
}


\subsubsection{2. Check data}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:2.-Check-data}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check data}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{number of peak: }\PYG{l+s+s2}{\PYGZob{}}\PYG{l+s+s2}{len(peaks.peak\PYGZus{}id.unique())\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{mean\PYGZus{}} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{peaks}\PYG{o}{.}\PYG{n}{groupby}\PYG{p}{(}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{gene\PYGZus{}short\PYGZus{}name}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}id}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{)}\PYG{o}{.}\PYG{n}{count}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{o}{/}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{peaks}\PYG{o}{.}\PYG{n}{gene\PYGZus{}short\PYGZus{}name}\PYG{o}{.}\PYG{n}{unique}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mean peaks per gene: }\PYG{l+s+si}{\PYGZob{}mean\PYGZus{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{getLength}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{a}\PYG{p}{,} \PYG{n}{b}\PYG{p}{,} \PYG{n}{c} \PYG{o}{=} \PYG{n}{x}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}id}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{o}{.}\PYG{n}{split}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZus{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{k}{return} \PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{c}\PYG{p}{)} \PYG{o}{\PYGZhy{}} \PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{b}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{}}
\PYG{n}{df} \PYG{o}{=} \PYG{n}{peaks}\PYG{o}{.}\PYG{n}{apply}\PYG{p}{(}\PYG{k}{lambda} \PYG{n}{x}\PYG{p}{:} \PYG{n}{getLength}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mean peak length: }\PYG{l+s+s2}{\PYGZob{}}\PYG{l+s+s2}{df.values.mean()\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
number of peak: 13919
mean peaks per gene: 1.0591731964333964
mean peak length: 1756.1744260204082
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\paragraph{2.1. Remove short peaks}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:2.1.-Remove-short-peaks}}
Short DNA fragment, such as less than 5 bases, cannot be used for motif scan. Remove short DNA fragment.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{peaks}\PYG{p}{[}\PYG{n}{df}\PYG{o}{\PYGZgt{}}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{]}
\end{sphinxVerbatim}
}


\subsubsection{3. Instantiate TFinfo object and search for TF binding motifs}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs}}
The motif analysis module has a custom class; TFinfo. TFinfo object converts a peak data into a DNA sequence and scans the DNA sequence searching for TF binding motifs. Then the results of motif scan will be filtered and converted into several files, such as python dictionary or dataframe. This TF information is necessary for GRN inference.


\subsubsection{3.1 check reference genome installation}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:3.1-check-reference-genome-installation}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} PLEASE make sure that you are setting correct ref genome.}
\PYG{n}{ref\PYGZus{}genome} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mm10}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{n}{ma}\PYG{o}{.}\PYG{n}{is\PYGZus{}genome\PYGZus{}installed}\PYG{p}{(}\PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{n}{ref\PYGZus{}genome}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
genome mm10 is not installed in this environment.
Please install genome using genomepy.
e.g.
    >>> import genomepy
    >>> genomepy.install\_genome("mm9", "UCSC")
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}False
\end{sphinxVerbatim}
}


\paragraph{3.2. Install reference genome (if refgenome is not installed)}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:3.2.-Install-reference-genome-(if-refgenome-is-not-installed)}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{genomepy}
\PYG{n}{genomepy}\PYG{o}{.}\PYG{n}{install\PYGZus{}genome}\PYG{p}{(}\PYG{n}{ref\PYGZus{}genome}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{UCSC}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
downloading from http://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz{\ldots}
done{\ldots}
name: mm10
local name: mm10
fasta: /home/k/.local/share/genomes/mm10/mm10.fa
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check again}
\PYG{n}{ma}\PYG{o}{.}\PYG{n}{is\PYGZus{}genome\PYGZus{}installed}\PYG{p}{(}\PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{n}{ref\PYGZus{}genome}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}True
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[14]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Instantiate TFinfo object}
\PYG{n}{tfi} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{TFinfo}\PYG{p}{(}\PYG{n}{peak\PYGZus{}data\PYGZus{}frame}\PYG{o}{=}\PYG{n}{peaks}\PYG{p}{,} \PYG{c+c1}{\PYGZsh{} peak info calculated from ATAC\PYGZhy{}seq data}
                \PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{n}{ref\PYGZus{}genome}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{4. Scan motifs and save object}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:4.-Scan-motifs-and-save-object}}
This step requires computational resource and may take long time

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[15]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{o}{\PYGZpc{}\PYGZpc{}time}
\PYG{c+c1}{\PYGZsh{} Scan motifs}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{scan}\PYG{p}{(}\PYG{n}{fpr}\PYG{o}{=}\PYG{l+m+mf}{0.02}\PYG{p}{,} \PYG{n}{verbose}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} sace tfinfo object}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}hdf5}\PYG{p}{(}\PYG{n}{file\PYGZus{}path}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{test.celloracle.tfinfo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
initiating scanner {\ldots}
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
2019-09-22 23:00:18,604 - INFO - Using background: genome mm10 with length 200
2019-09-22 23:00:18,986 - INFO - Determining FPR-based threshold
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
getting DNA sequences {\ldots}
scanning motifs {\ldots}
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=1, bar\PYGZus{}style=\PYGZsq{}info\PYGZsq{}, max=1), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

CPU times: user 52min 23s, sys: 36.8 s, total: 53min
Wall time: 52min 58s
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check motif scan results}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{scanned\PYGZus{}df}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     seqname                      motif\PYGZus{}id factors\PYGZus{}direct  \PYGZbs{}
0  chr10\PYGZus{}100015291\PYGZus{}100017830       GM.5.0.Homeodomain.0001          TGIF1
1  chr10\PYGZus{}100015291\PYGZus{}100017830             GM.5.0.Mixed.0001
2  chr10\PYGZus{}100015291\PYGZus{}100017830             GM.5.0.Mixed.0001
3  chr10\PYGZus{}100015291\PYGZus{}100017830             GM.5.0.Mixed.0001
4  chr10\PYGZus{}100015291\PYGZus{}100017830  GM.5.0.Nuclear\PYGZus{}receptor.0002          NR2C2

         factors\PYGZus{}indirect      score   pos  strand
0  ENSG00000234254, TGIF1  10.311002  1003       1
1               SRF, EGR1   7.925873   481       1
2               SRF, EGR1   7.321375   911      \PYGZhy{}1
3               SRF, EGR1   7.276585   811      \PYGZhy{}1
4            NR2C2, Nr2c2   9.067331   449      \PYGZhy{}1
\end{sphinxVerbatim}
}

We have the score for each sequence and motif\_id pair. In the next step we will filter the motifs with low score.


\subsubsection{5. Filtering motifs}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:5.-Filtering-motifs}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[17]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} reset filtering}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{reset\PYGZus{}filtering}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} do filtering}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{filter\PYGZus{}motifs\PYGZus{}by\PYGZus{}score}\PYG{p}{(}\PYG{n}{threshold}\PYG{o}{=}\PYG{l+m+mf}{10.5}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} do post filtering process. Convert results into several file format.}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{make\PYGZus{}TFinfo\PYGZus{}dataframe\PYGZus{}and\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{verbose}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
peaks were filtered: 12934005 -> 2285279
1. converting scanned results into one-hot encoded dataframe.
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=13919), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

2. converting results into dictionaries.
converting scan results into dictionaries{\ldots}
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=14804), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1090), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\subsubsection{6. Get Final results}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:6.-Get-Final-results}}

\paragraph{6.1. Get resutls as a dictionary}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:6.1.-Get-resutls-as-a-dictionary}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[18]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{td} \PYG{o}{=} \PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{dictionary\PYGZus{}type}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{targetgene2TFs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\end{sphinxVerbatim}
}


\paragraph{6.2. Get results as a dataframe}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:6.2.-Get-results-as-a-dataframe}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[20]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{df} \PYG{o}{=} \PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}dataframe}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{df}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[20]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\PYGZus{}id gene\PYGZus{}short\PYGZus{}name  9430076c15rik  Ac002126.6  \PYGZbs{}
0  chr10\PYGZus{}100015291\PYGZus{}100017830            Kitl              0           0
1  chr10\PYGZus{}100486534\PYGZus{}100488209           Tmtc3              0           0
2  chr10\PYGZus{}100588641\PYGZus{}100589556   4930430F08Rik              0           0
3  chr10\PYGZus{}100741247\PYGZus{}100742505         Gm35722              0           0
4  chr10\PYGZus{}101681379\PYGZus{}101682124          Mgat4c              0           0

   Ac012531.1  Ac226150.2  Afp  Ahr  Ahrr  Aire  ...  Znf784  Znf8  Znf816  \PYGZbs{}
0           0           0    0    1     1     0  ...       0     0       0
1           0           0    0    0     0     0  ...       1     0       0
2           1           0    0    1     1     0  ...       0     0       0
3           0           0    0    0     0     0  ...       0     0       0
4           0           0    0    0     0     0  ...       0     0       0

   Znf85  Zscan10  Zscan16  Zscan22  Zscan26  Zscan31  Zscan4
0      0        0        0        0        0        1       0
1      0        0        0        1        0        0       0
2      0        0        0        0        0        0       0
3      0        0        0        0        0        0       0
4      0        0        0        0        0        0       1

[5 rows x 1092 columns]
\end{sphinxVerbatim}
}


\subsubsection{7. Save TFinfo as dictionary and dataframe}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:7.-Save-TFinfo-as-dictionary-and-dataframe}}
We’ll use this information in the later step to make GRN. Save the results.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[21]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{folder} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{TFinfo\PYGZus{}outputs}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{folder}\PYG{p}{,} \PYG{n}{exist\PYGZus{}ok}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} save TFinfo as a dictionary}
\PYG{n}{td} \PYG{o}{=} \PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{dictionary\PYGZus{}type}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{targetgene2TFs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{save\PYGZus{}as\PYGZus{}pickled\PYGZus{}object}\PYG{p}{(}\PYG{n}{td}\PYG{p}{,} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{folder}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{TFinfo\PYGZus{}targetgene2TFs.pickled}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} save TFinfo as a dataframe}
\PYG{n}{df} \PYG{o}{=} \PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}dataframe}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{df}\PYG{o}{.}\PYG{n}{to\PYGZus{}parquet}\PYG{p}{(}\PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{folder}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{TFinfo\PYGZus{}dataframe.parquet}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsection{Single-cell RNA-seq data preprocessing}
\label{\detokenize{tutorials/scrnaprocess:single-cell-rna-seq-data-preprocessing}}\label{\detokenize{tutorials/scrnaprocess:scrnaprocess}}\label{\detokenize{tutorials/scrnaprocess::doc}}\begin{description}
\item[{Network analysis and simulation in celloracle will be performed using scRNA-seq data. The scRNA-seq data should include the compontnes below.}] \leavevmode\begin{itemize}
\item {} 
Gene expression matrix; mRNA counts before scaling and transformation.

\item {} 
Clustering results.

\item {} 
Dimensional reduction results.

\end{itemize}

\item[{In addition to these minimum requirements, we highly recommend doing these analyses below in the preprocessing step.}] \leavevmode\begin{itemize}
\item {} 
Data quality check and Cell/Gene filtering.

\item {} 
Normalization

\item {} 
Identification of highly variable genes

\end{itemize}

\end{description}

We recommend processing scRNA-seq data using either scanpy or Seurat.
If you are not familiar with the general workflow of scRNA-seq data processing, please go to \sphinxhref{https://scanpy.readthedocs.io/en/stable/}{the documentation of scanpy} and \sphinxhref{https://satijalab.org/seurat/vignettes.html}{the documentation of Seurat} .

If you already have preprocessed scRNA-seq data, which includes necessary information above, you can skip this part.


\subsubsection{A. scRNA-seq data preprocessing with scanpy}
\label{\detokenize{tutorials/scrnaprocess:a-scrna-seq-data-preprocessing-with-scanpy}}
\sphinxcode{\sphinxupquote{scanpy}} is a python library for the analysis of scRNA-seq data.

In this tutorial, we introduce an example of scRNA-seq preprocessing for celloracle with \sphinxcode{\sphinxupquote{scanpy}}.
We wrote the notebook based on \sphinxhref{https://scanpy-tutorials.readthedocs.io/en/latest/paga-paul15.html}{one of scanpy’s tutorials} with some modification.

Python notebook


\paragraph{0. Import libraries}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:0.-Import-libraries}}\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{scanpy} \PYG{k}{as} \PYG{n+nn}{sc}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline
\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{300}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}


\paragraph{1. Load data}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:1.-Load-data}}
In this notebook, we will show an example of how to process scRNA-seq data using a scRNA-seq data of hematopoiesis (Paul, F., Arkin, Y., Giladi, A., Jaitin, D. A., Kenigsberg, E., Keren-Shaul, H., et al. (2015). Transcriptional Heterogeneity and Lineage Commitment in Myeloid Progenitors. Cell, 163(7), 1663\textendash{}1677. \sphinxurl{http://doi.org/10.1016/j.cell.2015.11.013}). You can easily download this scRNA-seq data with a scanpy’s function.

Please change the code below if you want to use your data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} download dataset. You can change the code blow if you use another data.}
\PYG{n}{adata} \PYG{o}{=} \PYG{n}{sc}\PYG{o}{.}\PYG{n}{datasets}\PYG{o}{.}\PYG{n}{paul15}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
WARNING: In Scanpy 0.*, this returned logarithmized data. Now it returns non-logarithmized data.
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
{\ldots} storing 'paul15\_clusters' as categorical
Trying to set attribute `.uns` of view, making a copy.
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\paragraph{2. Filtering}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:2.-Filtering}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} only consider genes with more than 1 count}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{filter\PYGZus{}genes}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{min\PYGZus{}counts}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}

\end{sphinxVerbatim}
}


\paragraph{3. Normalization}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:3.-Normalization}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} normalize gene expression matrix with total UMI count per cell}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{normalize\PYGZus{}per\PYGZus{}cell}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{key\PYGZus{}n\PYGZus{}counts}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{n\PYGZus{}counts\PYGZus{}all}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{4. Identification of highly variable genes}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:4.-Identification-of-highly-variable-genes}}
Removing non-variable genes reduces calculation time in the GRN reconstruction and simulation. We recommend using top 2000\textasciitilde{}3000 variable genes.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} select top 2000 highly\PYGZhy{}variable genes}
\PYG{n}{filter\PYGZus{}result} \PYG{o}{=} \PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{filter\PYGZus{}genes\PYGZus{}dispersion}\PYG{p}{(}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{X}\PYG{p}{,}
                                              \PYG{n}{flavor}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}ranger}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                                              \PYG{n}{n\PYGZus{}top\PYGZus{}genes}\PYG{o}{=}\PYG{l+m+mi}{2000}\PYG{p}{,}
                                              \PYG{n}{log}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} subset the genes}
\PYG{n}{adata} \PYG{o}{=} \PYG{n}{adata}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{filter\PYGZus{}result}\PYG{o}{.}\PYG{n}{gene\PYGZus{}subset}\PYG{p}{]}

\PYG{c+c1}{\PYGZsh{} renormalize after filtering}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{normalize\PYGZus{}per\PYGZus{}cell}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Trying to set attribute `.obs` of view, making a copy.
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\paragraph{5. Log transformation}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:5.-Log-transformation}}
We will do log transformation scaling because these are necessary for PCA, clustering, differential gene calculation. However, we also need non-transformed gene expression data in the celloracle analysis. Thus we keep raw count in anndata with the following command before log transformation.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} keep raw cont data before log transformation}
\PYG{n}{adata}\PYG{o}{.}\PYG{n}{raw} \PYG{o}{=} \PYG{n}{adata}

\PYG{c+c1}{\PYGZsh{} Log transformation and scaling}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{log1p}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{scale}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} (optional) Regressing out some values help reduce batch effects.}

\PYG{c+c1}{\PYGZsh{}sc.pp.regress\PYGZus{}out(adata, [\PYGZsq{}n\PYGZus{}counts\PYGZsq{}, \PYGZsq{}percent\PYGZus{}mito\PYGZsq{}])}
\PYG{c+c1}{\PYGZsh{}sc.pp.regress\PYGZus{}out(adata, [\PYGZsq{}n\PYGZus{}counts\PYGZsq{}])}
\end{sphinxVerbatim}
}


\paragraph{6. Dimensional reduction}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:6.-Dimensional-reduction}}
Dimensional reduction is one of the most important parts of the scRNA-seq analysis. celloracle needs dimensional reduction embeddings to simulate cell transition.

Please choose a proper algorithm of dimensional reduction so that the embedding appropriately represents the data structure. We recommend using one of the dimensional reduction algorithms (or trajectory inference algorithms); UMAP, tSNE, diffusion map, Force-directed graph drawing, PAGA.

In this example, we use a combination of four algorithms; diffusion map, force-directed graph drawing, and PAGA.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} PCA}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{pca}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{svd\PYGZus{}solver}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{arpack}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} diffusion map}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{neighbors}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{n\PYGZus{}neighbors}\PYG{o}{=}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{n}{n\PYGZus{}pcs}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}

\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{diffmap}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} calculate neihbors again based on diffusionmap}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{neighbors}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{n\PYGZus{}neighbors}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{n}{use\PYGZus{}rep}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{X\PYGZus{}diffmap}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{7. Clustering}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:7.-Clustering}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{louvain}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{resolution}\PYG{o}{=}\PYG{l+m+mf}{0.8}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{(Optional) Re-calculate Dimensional reduction graph}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:(Optional)-Re-calculate-Dimensional-reduction-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[12]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} PAGA graph construction}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{paga}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{groups}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[13]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check current cluster name}
\PYG{n}{cluster\PYGZus{}list} \PYG{o}{=} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{louvain}\PYG{o}{.}\PYG{n}{unique}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{cluster\PYGZus{}list}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[13]:\,\hspace{\fboxrule}\hspace{\fboxsep}}[5, 2, 12, 13, 0, ..., 6, 20, 14, 15, 21]
Length: 23
Categories (23, object): [5, 2, 12, 13, ..., 20, 14, 15, 21]
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[14]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[15]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{paga}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=383\sphinxpxdimen,height=277\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_23_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{init\PYGZus{}pos}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{paga}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{123}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[17]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=364\sphinxpxdimen,height=288\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_25_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{8. Check data}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:8.-Check-data}}

\subparagraph{8.1. Visualize marker gene expression}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:8.1.-Visualize-marker-gene-expression}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[18]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[19]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{markers} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Erythroids}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Klf1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gypa}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Hba\PYGZhy{}a2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
           \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Megakaryocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Itga2b}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Pbx1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Sdpr}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Vwf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
            \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Granulocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Elane}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Cebpe}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ctsg}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mpo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gfi1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
            \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Monocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Irf8}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Csf1r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ctsg}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mpo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
            \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mast\PYGZus{}cells}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Cma1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gzmb}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Kit}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
            \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Basophils}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mcpt8}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Prss34}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}
            \PYG{p}{\PYGZcb{}}

\PYG{k}{for} \PYG{n}{cell\PYGZus{}type}\PYG{p}{,} \PYG{n}{genes} \PYG{o+ow}{in} \PYG{n}{markers}\PYG{o}{.}\PYG{n}{items}\PYG{p}{(}\PYG{p}{)}\PYG{p}{:}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{marker gene of }\PYG{l+s+si}{\PYGZob{}cell\PYGZus{}type\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{n}{genes}\PYG{p}{,} \PYG{n}{use\PYGZus{}raw}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{n}{ncols}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}


\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Erythroids
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=656\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_1}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Megakaryocytes
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=644\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_3}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Granulocytes
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=655\sphinxpxdimen,height=856\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_5}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Monocytes
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=656\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_7}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Mast\_cells
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=644\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_9}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Basophils
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=644\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_11}.png}

\end{nbsphinxfancyoutput}


\paragraph{8. Make annotation for cluster}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:8.-Make-annotation-for-cluster}}
Based on the marker gene expression and previous report, we will manually make annotation for each cluster.


\subparagraph{8.1. Make annotation (1)}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:8.1.-Make-annotation-(1)}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[20]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{paul15\PYGZus{}clusters}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_32_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[21]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check current cluster name}
\PYG{n}{cluster\PYGZus{}list} \PYG{o}{=} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{louvain}\PYG{o}{.}\PYG{n}{unique}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{cluster\PYGZus{}list}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[21]:\,\hspace{\fboxrule}\hspace{\fboxsep}}[5, 2, 12, 13, 0, ..., 6, 20, 14, 15, 21]
Length: 23
Categories (23, object): [5, 2, 12, 13, ..., 20, 14, 15, 21]
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[22]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} make anottation dictionary}
\PYG{n}{annotation} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{5}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Erythroids}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+m+mi}{15}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{16}\PYG{p}{,} \PYG{l+m+mi}{9}\PYG{p}{,} \PYG{l+m+mi}{8}\PYG{p}{,} \PYG{l+m+mi}{14}\PYG{p}{,} \PYG{l+m+mi}{19}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{12}\PYG{p}{,} \PYG{l+m+mi}{18}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Megakaryocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{17}\PYG{p}{,} \PYG{l+m+mi}{22}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{GMP}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{late\PYGZus{}GMP}\PYG{l+s+s2}{\PYGZdq{}} \PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Granulocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{7}\PYG{p}{,} \PYG{l+m+mi}{13}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Monocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DC}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{21}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lymphoid}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{20}\PYG{p}{]}\PYG{p}{\PYGZcb{}}

\PYG{c+c1}{\PYGZsh{} change dictionary format}
\PYG{n}{annotation\PYGZus{}rev} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{p}{\PYGZcb{}}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n}{cluster\PYGZus{}list}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{k} \PYG{o+ow}{in} \PYG{n}{annotation}\PYG{p}{:}
        \PYG{k}{if} \PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)} \PYG{o+ow}{in} \PYG{n}{annotation}\PYG{p}{[}\PYG{n}{k}\PYG{p}{]}\PYG{p}{:}
            \PYG{n}{annotation\PYGZus{}rev}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{k}

\PYG{c+c1}{\PYGZsh{} check dictionary}
\PYG{n}{annotation\PYGZus{}rev}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[22]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYGZob{}\PYGZsq{}5\PYGZsq{}: \PYGZsq{}MEP\PYGZsq{},
 \PYGZsq{}2\PYGZsq{}: \PYGZsq{}Monocytes\PYGZsq{},
 \PYGZsq{}12\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}13\PYGZsq{}: \PYGZsq{}Granulocytes\PYGZsq{},
 \PYGZsq{}0\PYGZsq{}: \PYGZsq{}late\PYGZus{}GMP\PYGZsq{},
 \PYGZsq{}10\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}3\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}18\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}11\PYGZsq{}: \PYGZsq{}GMP\PYGZsq{},
 \PYGZsq{}7\PYGZsq{}: \PYGZsq{}Granulocytes\PYGZsq{},
 \PYGZsq{}8\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}22\PYGZsq{}: \PYGZsq{}Megakaryocytes\PYGZsq{},
 \PYGZsq{}16\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}1\PYGZsq{}: \PYGZsq{}GMP\PYGZsq{},
 \PYGZsq{}17\PYGZsq{}: \PYGZsq{}Megakaryocytes\PYGZsq{},
 \PYGZsq{}4\PYGZsq{}: \PYGZsq{}Granulocytes\PYGZsq{},
 \PYGZsq{}19\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}9\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}6\PYGZsq{}: \PYGZsq{}Monocytes\PYGZsq{},
 \PYGZsq{}20\PYGZsq{}: \PYGZsq{}Lymphoid\PYGZsq{},
 \PYGZsq{}14\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}15\PYGZsq{}: \PYGZsq{}Erythroids\PYGZsq{},
 \PYGZsq{}21\PYGZsq{}: \PYGZsq{}DC\PYGZsq{}\PYGZcb{}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[23]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cell\PYGZus{}type}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{n}{annotation\PYGZus{}rev}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{louvain}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[24]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check results}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}type}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{paul15\PYGZus{}clusters}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
{\ldots} storing 'cell\_type' as categorical
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_36_1}.png}

\end{nbsphinxfancyoutput}


\subparagraph{8.2. Make annotation (2)}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:8.2.-Make-annotation-(2)}}
We’ll make another annotation manually for each louvain clusters

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[25]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}type}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_38_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[26]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{annotation\PYGZus{}2} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{15}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{10}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{16}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{14}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{9}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}4}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{8}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{19}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}6}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}7}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{12}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}8}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{18}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}9}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{17}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mk\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{22}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mk\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{11}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMPl\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{7}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{13}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{4}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{6}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{21}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{DC\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{20}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Lym\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{\PYGZcb{}}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[27]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{n}{annotation\PYGZus{}2}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{louvain}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[28]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check result}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain\PYGZus{}annot}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}type}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
{\ldots} storing 'louvain\_annot' as categorical
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_41_1}.png}

\end{nbsphinxfancyoutput}

We’ve done secveral scRNA-preprocessing steps; filtering, normalization, clustering, dimensional reduction. In the next step, we’ll do GRN inference, network analysis, and in silico simulation based on these information.


\paragraph{9. (Option) Subset cells}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:9.-(Option)-Subset-cells}}
In this tutorial, we are using scRNA-seq data of hematopoiesis. In the latter part, we will focus on the cell fate decision in the myeloid lineage. So we remove non-myeloid cell cluster, DC and Lymphoid now.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[29]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{cell\PYGZus{}type}\PYG{o}{.}\PYG{n}{unique}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[29]:\,\hspace{\fboxrule}\hspace{\fboxsep}}[MEP, Monocytes, Erythroids, Granulocytes, late\PYGZus{}GMP, GMP, Megakaryocytes, Lymphoid, DC]
Categories (9, object): [MEP, Monocytes, Erythroids, Granulocytes, ..., GMP, Megakaryocytes, Lymphoid, DC]
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[30]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{cell\PYGZus{}of\PYGZus{}interest} \PYG{o}{=} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{index}\PYG{p}{[}\PYG{o}{\PYGZti{}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{cell\PYGZus{}type}\PYG{o}{.}\PYG{n}{isin}\PYG{p}{(}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lymphoid}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DC}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{)}\PYG{p}{]}
\PYG{n}{adata} \PYG{o}{=} \PYG{n}{adata}\PYG{p}{[}\PYG{n}{cell\PYGZus{}of\PYGZus{}interest}\PYG{p}{,} \PYG{p}{:}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[31]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check result}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain\PYGZus{}annot}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}type}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_46_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{10. Save data}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:10.-Save-data}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[32]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{}adata.write\PYGZus{}h5ad(\PYGZdq{}data/Paul\PYGZus{}etal\PYGZus{}15.h5ad\PYGZdq{})}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}
\end{sphinxVerbatim}
}


\subsubsection{B. scRNA-seq data preprocessing with Seurat}
\label{\detokenize{tutorials/scrnaprocess:b-scrna-seq-data-preprocessing-with-seurat}}
R notebook … comming soon

\begin{sphinxadmonition}{note}{Note:}
If you use \sphinxcode{\sphinxupquote{Seurat}} for preprocessing, you need to convert the scRNA-seq data (Seurat object) into anndata to analyze the data with \sphinxcode{\sphinxupquote{celloracle}}.
\sphinxcode{\sphinxupquote{celloracle}} have python API and command-line API to convert Seurat object into anndata.
Please go to the documentation of celloracle API for more information.
\end{sphinxadmonition}


\subsection{Network analysis}
\label{\detokenize{tutorials/networkanalysis:network-analysis}}\label{\detokenize{tutorials/networkanalysis:networkanalysis}}\label{\detokenize{tutorials/networkanalysis::doc}}
\sphinxcode{\sphinxupquote{celloracle}} import scRNA-seq dataset and TF binding information to find active regulatory connections for all genes, generating sample-specific GRNs.

The inferred GRN is analyzed with several network algorithms to get various network scores. The network score is useful to identify key regulatory genes.

celloracle reconstructs GRN for each cluster enabling us to compare GRNs each other. It is also possible to analyze how the GRN change over differentiation.
The dynamics of the GRN structure provide us insight into the
context-dependent regulatory mechanism.

Python notebook


\subsubsection{0. Import libraries}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:0.-Import-libraries}}\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} 0. Import}

\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{sys}

\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{scanpy} \PYG{k}{as} \PYG{n+nn}{sc}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{celloracle} \PYG{k}{as} \PYG{n+nn}{co}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} visualization settings}
\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{300}

\end{sphinxVerbatim}
}


\paragraph{0.1. Check installation}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:0.1.-Check-installation}}
Celloracle uses some R libraries in network analysis. Please make sure that all dependent R libraries are installed on your computer. You can test the installation with the following command.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{co}\PYG{o}{.}\PYG{n}{network\PYGZus{}analysis}\PYG{o}{.}\PYG{n}{test\PYGZus{}R\PYGZus{}libraries\PYGZus{}installation}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
checking R library installation: igraph -> OK
checking R library installation: linkcomm -> OK
checking R library installation: rnetcarto -> OK
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\paragraph{0.2. Make a folder to save graph}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:0.2.-Make-a-folder-to-save-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{save\PYGZus{}folder} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figures}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{save\PYGZus{}folder}\PYG{p}{,} \PYG{n}{exist\PYGZus{}ok}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{1. Load data}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:1.-Load-data}}

\paragraph{1.1. Load processed gene expression data (anndata)}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:1.1.-Load-processed-gene-expression-data-(anndata)}}
Please refer to the previous notebook in the tutorial for an example of how to process scRNA-seq data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} load data. !!Replace the data path below when you use another data.}
\PYG{n}{adata} \PYG{o}{=} \PYG{n}{sc}\PYG{o}{.}\PYG{n}{read\PYGZus{}h5ad}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{../03\PYGZus{}scRNA\PYGZhy{}seq\PYGZus{}data\PYGZus{}preprocessing/data/Paul\PYGZus{}etal\PYGZus{}15.h5ad}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{1.2. Load TF data.}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:1.2.-Load-TF-data.}}
For the GRN inference, celloracle needs TF information, which contains lists of the regulatory candidate gene. There are several ways to make such TF information. We can generate TF information from scATAC-seq data or Bulk ATAC-seq data. Please refer to the first step of the tutorial for the details of how to make TF information.

If you do not have your scATAC-seq data, you can use some built-in data in celloracle. celloracle have a TFinfo made with various kind of tissue/cell-types of mouse ATAC-seq atlas dataset (\sphinxurl{http://atlas.gs.washington.edu/mouse-atac/}).

You can load and use the data with the following command.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Load TF info which was made from mouse cell atlas dataset.}
\PYG{n}{TFinfo\PYGZus{}df} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{data}\PYG{o}{.}\PYG{n}{load\PYGZus{}TFinfo\PYGZus{}df\PYGZus{}mm9\PYGZus{}mouse\PYGZus{}atac\PYGZus{}atlas}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} check data}
\PYG{n}{TFinfo\PYGZus{}df}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\PYGZus{}id gene\PYGZus{}short\PYGZus{}name  9430076c15rik  Ac002126.6  \PYGZbs{}
0  chr10\PYGZus{}100050979\PYGZus{}100052296   4930430F08Rik            0.0         0.0
1  chr10\PYGZus{}101006922\PYGZus{}101007748         SNORA17            0.0         0.0
2  chr10\PYGZus{}101144061\PYGZus{}101145000          Mgat4c            0.0         0.0
3    chr10\PYGZus{}10148873\PYGZus{}10149183   9130014G24Rik            0.0         0.0
4    chr10\PYGZus{}10149425\PYGZus{}10149815   9130014G24Rik            0.0         0.0

   Ac012531.1  Ac226150.2  Afp  Ahr  Ahrr  Aire  ...  Znf784  Znf8  Znf816  \PYGZbs{}
0         1.0         0.0  0.0  0.0   0.0   0.0  ...     0.0   0.0     0.0
1         0.0         0.0  0.0  0.0   0.0   0.0  ...     0.0   0.0     0.0
2         0.0         0.0  0.0  0.0   0.0   0.0  ...     0.0   0.0     0.0
3         0.0         0.0  0.0  0.0   0.0   0.0  ...     0.0   0.0     0.0
4         0.0         0.0  0.0  0.0   0.0   0.0  ...     0.0   0.0     0.0

   Znf85  Zscan10  Zscan16  Zscan22  Zscan26  Zscan31  Zscan4
0    0.0      0.0      0.0      0.0      0.0      0.0     0.0
1    0.0      0.0      0.0      0.0      0.0      1.0     0.0
2    0.0      0.0      0.0      0.0      0.0      0.0     1.0
3    0.0      0.0      0.0      0.0      0.0      0.0     0.0
4    0.0      0.0      0.0      0.0      0.0      0.0     0.0

[5 rows x 1095 columns]
\end{sphinxVerbatim}
}


\subsubsection{2. Initiate Oracle object}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:2.-Initiate-Oracle-object}}
celloracle have a custom class, Oracle. We can use Oracle for the data preprocessing and GRN inference. Oracle object stores all information and do the calculation with its internal functions. We instantiate an Oracle object, then input a gene expression data (anndata) and a TFinfo into the Oracle object.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} make Oracle object}
\PYG{n}{oracle} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{Oracle}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{2.1. load gene expression data into oracle object.}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:2.1.-load-gene-expression-data-into-oracle-object.}}
When you load scRNA-seq data, please enter the name of clustering data and dimensional reduction data. Clustering data suppose to be stored in the attribute of “obs” in the anndata. Dimensional reduction data suppose to be stored in the attribute of “obsm” in the anndata. You can check these data by the following command.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[18]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} show data name in anndata}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{metadata columns :}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{list}\PYG{p}{(}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{columns}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{dimensional reduction: }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{list}\PYG{p}{(}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obsm}\PYG{o}{.}\PYG{n}{keys}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
metadata columns : ['paul15\_clusters', 'n\_counts\_all', 'n\_counts', 'louvain', 'cell\_type', 'louvain\_annot']
dimensional reduction:  ['X\_diffmap', 'X\_draw\_graph\_fa', 'X\_pca']
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[14]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} The anndata shoud include (1) gene expression count, (2) clustering information, (3) trajectory (dimensional reduction embeddings) data.}
\PYG{c+c1}{\PYGZsh{} Please refer to another notebook for the detail of anndata preprocessing.}

\PYG{c+c1}{\PYGZsh{} In this notebook, we use raw mRNA count as an input of Oracle object.}
\PYG{n}{adata}\PYG{o}{.}\PYG{n}{X} \PYG{o}{=} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{raw}\PYG{o}{.}\PYG{n}{X}\PYG{o}{.}\PYG{n}{copy}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Instantiate Oracle object. All calculation will be done in this object.}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{import\PYGZus{}anndata\PYGZus{}as\PYGZus{}raw\PYGZus{}count}\PYG{p}{(}\PYG{n}{adata}\PYG{o}{=}\PYG{n}{adata}\PYG{p}{,}
                                   \PYG{n}{cluster\PYGZus{}column\PYGZus{}name}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                                   \PYG{n}{embedding\PYGZus{}name}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{X\PYGZus{}draw\PYGZus{}graph\PYGZus{}fa}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{2.2. load TFinfo into oracle object}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:2.2.-load-TFinfo-into-oracle-object}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[15]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} If you can load TF info dataframe with the following code.}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{import\PYGZus{}TF\PYGZus{}data}\PYG{p}{(}\PYG{n}{TF\PYGZus{}info\PYGZus{}matrix}\PYG{o}{=}\PYG{n}{TFinfo\PYGZus{}df}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Instead of the way avove, you can use TF \PYGZdq{}dictionary\PYGZdq{} with the following code.}
\PYG{c+c1}{\PYGZsh{} oracle.import\PYGZus{}TF\PYGZus{}data(TFdict=TFinfo\PYGZus{}dictionary)}
\end{sphinxVerbatim}
}


\paragraph{2.3. (Optional) Add TF info manually}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:2.3.-(Optional)-Add-TF-info-manually}}
While we mainly use TF info data made from scATAC-seq data, we can also add arbitrary information about the TF-target gene pair by the manual way.

For example, if there is a study or database about TF-target pair, you can use such information in the following way.


\subparagraph{2.3.1. Make TF info dictionary manually}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:2.3.1.-Make-TF-info-dictionary-manually}}
Here, we introduce a way to add TF binding information.

We have a TF binding data in about some TFs. The information is the supplemental table 4 in (\sphinxurl{http://doi.org/10.1016/j.cell.2015.11.013}).

In order to import TF data into Oracle object, we need to convert them into a python dictionary, in which dictionary keys are target gene, and values are regulatory candidate TFs.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} We have TF and its target gene information. This is from a supplemental Fig of Paul et. al, (2015).}
\PYG{n}{Paul\PYGZus{}15\PYGZus{}data} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{TF\PYGZus{}data\PYGZus{}in\PYGZus{}Paul15.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{Paul\PYGZus{}15\PYGZus{}data}

\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}      TF                                       Target\PYGZus{}genes
0  Cebpa  Abcb1b, Acot1, C3, Cnpy3, Dhrs7, Dtx4, Edem2, ...
1   Irf8  Abcd1, Aif1, BC017643, Cbl, Ccdc109b, Ccl6, d6...
2   Irf8  1100001G20Rik, 4732418C07Rik, 9230105E10Rik, A...
3   Klf1  2010011I20Rik, 5730469M10Rik, Acsl6, Add2, Ank...
4  Sfpi1  0910001L09Rik, 2310014H01Rik, 4632428N05Rik, A...
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[17]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Make dictionary: Key is TF, Value is list of Target gene}
\PYG{n}{TF\PYGZus{}to\PYGZus{}TG\PYGZus{}dictionary} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{p}{\PYGZcb{}}

\PYG{k}{for} \PYG{n}{TF}\PYG{p}{,} \PYG{n}{TGs} \PYG{o+ow}{in} \PYG{n+nb}{zip}\PYG{p}{(}\PYG{n}{Paul\PYGZus{}15\PYGZus{}data}\PYG{o}{.}\PYG{n}{TF}\PYG{p}{,} \PYG{n}{Paul\PYGZus{}15\PYGZus{}data}\PYG{o}{.}\PYG{n}{Target\PYGZus{}genes}\PYG{p}{)}\PYG{p}{:}
    \PYG{c+c1}{\PYGZsh{} convert target gene to list}
    \PYG{n}{TG\PYGZus{}list} \PYG{o}{=} \PYG{n}{TGs}\PYG{o}{.}\PYG{n}{replace}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{o}{.}\PYG{n}{split}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{,}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{c+c1}{\PYGZsh{} store target gene list in a dictionary}
    \PYG{n}{TF\PYGZus{}to\PYGZus{}TG\PYGZus{}dictionary}\PYG{p}{[}\PYG{n}{TF}\PYG{p}{]} \PYG{o}{=} \PYG{n}{TG\PYGZus{}list}

\PYG{c+c1}{\PYGZsh{} We have to make a dictionary, in which a Key is Target gene and value is TF.}
\PYG{c+c1}{\PYGZsh{} We inverse the dictionary above using a utility function in celloracle.}
\PYG{n}{TG\PYGZus{}to\PYGZus{}TF\PYGZus{}dictionary} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{utility}\PYG{o}{.}\PYG{n}{inverse\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{TF\PYGZus{}to\PYGZus{}TG\PYGZus{}dictionary}\PYG{p}{)}

\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=178), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\subparagraph{2.3.2. Add TF informatio dictionary into the oracle object}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:2.3.2.-Add-TF-informatio-dictionary-into-the-oracle-object}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[18]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Add TF information}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{addTFinfo\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{TG\PYGZus{}to\PYGZus{}TF\PYGZus{}dictionary}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{3. Knn imputation}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:3.-Knn-imputation}}
Celloracle uses almost the same strategy as velocyto for visualization of cell transition. This process requires knn imputation.

For the Knn imputation, we need PCA and PC selection first.


\paragraph{3.1. PCA}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:3.1.-PCA}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[67]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} perform PCA}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{perform\PYGZus{}PCA}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} select important PCs}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{cumsum}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{pca}\PYG{o}{.}\PYG{n}{explained\PYGZus{}variance\PYGZus{}ratio\PYGZus{}}\PYG{p}{)}\PYG{p}{[}\PYG{p}{:}\PYG{l+m+mi}{100}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{n\PYGZus{}comps} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{where}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{diff}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{diff}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{cumsum}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{pca}\PYG{o}{.}\PYG{n}{explained\PYGZus{}variance\PYGZus{}ratio\PYGZus{}}\PYG{p}{)}\PYG{p}{)}\PYG{o}{\PYGZgt{}}\PYG{l+m+mf}{0.002}\PYG{p}{)}\PYG{p}{)}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axvline}\PYG{p}{(}\PYG{n}{n\PYGZus{}comps}\PYG{p}{,} \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{k}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{n\PYGZus{}comps}\PYG{p}{)}
\PYG{n}{n\PYGZus{}comps} \PYG{o}{=} \PYG{n+nb}{min}\PYG{p}{(}\PYG{n}{n\PYGZus{}comps}\PYG{p}{,} \PYG{l+m+mi}{50}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
45
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=380\sphinxpxdimen,height=278\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_27_1}.png}

\end{nbsphinxfancyoutput}


\paragraph{3.2. Knn imputation}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:3.2.-Knn-imputation}}
Estimate number of nearest neighbor for k-nn imputation.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[20]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{n\PYGZus{}cell} \PYG{o}{=} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cell number is :}\PYG{l+s+si}{\PYGZob{}n\PYGZus{}cell\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
cell number is :2671
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[21]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{k} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{l+m+mf}{0.025}\PYG{o}{*}\PYG{n}{n\PYGZus{}cell}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Auto\PYGZhy{}selected k is :}\PYG{l+s+si}{\PYGZob{}k\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Auto-selected k is :66
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[22]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{knn\PYGZus{}imputation}\PYG{p}{(}\PYG{n}{n\PYGZus{}pca\PYGZus{}dims}\PYG{o}{=}\PYG{n}{n\PYGZus{}comps}\PYG{p}{,} \PYG{n}{k}\PYG{o}{=}\PYG{n}{k}\PYG{p}{,} \PYG{n}{balanced}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{b\PYGZus{}sight}\PYG{o}{=}\PYG{n}{k}\PYG{o}{*}\PYG{l+m+mi}{8}\PYG{p}{,}
                      \PYG{n}{b\PYGZus{}maxl}\PYG{o}{=}\PYG{n}{k}\PYG{o}{*}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{n}{n\PYGZus{}jobs}\PYG{o}{=}\PYG{l+m+mi}{4}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{4. Save and Load.}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:4.-Save-and-Load.}}
Celloracle has some custom-class: Links, Oracle, TFinfo. You can save such an object using “to\_hdf5”.

Pleasae use “load\_hdf5” function to load the file.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[23]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} save oracle object.}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{to\PYGZus{}hdf5}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Paul\PYGZus{}15\PYGZus{}data.celloracle.oracle}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[24]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} load file.}
\PYG{c+c1}{\PYGZsh{} oracle = co.load\PYGZus{}hdf5(\PYGZdq{}Paul\PYGZus{}15\PYGZus{}data.celloracle.oracle\PYGZdq{})}
\end{sphinxVerbatim}
}


\subsubsection{4. GRN calculation}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:4.-GRN-calculation}}
Next step is constructing a cluster-specific GRN for all clusters.

You can calculate GRNs with “get\_links” function, and the function returns GRNs as a Links object. Links object stores inferred GRNs and its metadata. You can do network analysis with Links object.

In the example below, we construct GRNs based on “louvain\_annot” clustering unit.

GRN can be calculated at any arbitrary unit as long as the clustering information is stored in anndata.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[43]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check data}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=516\sphinxpxdimen,height=288\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_37_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{4.1. Get GRNs}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:4.1.-Get-GRNs}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[28]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} calculate GRN for each population in \PYGZdq{}louvain\PYGZus{}annot\PYGZdq{} clustering unit.}
\PYG{c+c1}{\PYGZsh{} This step may take long time.}
\PYG{n}{links} \PYG{o}{=} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{get\PYGZus{}links}\PYG{p}{(}\PYG{n}{cluster\PYGZus{}name\PYGZus{}for\PYGZus{}GRN\PYGZus{}unit}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,}
                         \PYG{n}{verbose\PYGZus{}level}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{test\PYGZus{}mode}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}

\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=20), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
inferring GRN for Ery\_0{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_1{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_2{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_3{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_4{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_5{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_6{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_7{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_8{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Ery\_9{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for GMP\_0{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for GMP\_1{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for GMPl\_0{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Gran\_0{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Gran\_1{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Gran\_2{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for MEP\_0{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Mk\_0{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Mo\_0{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

inferring GRN for Mo\_1{\ldots}
method: bagging\_ridge
alpha: 10
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1850), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]


\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\paragraph{4.2. (Optional) Export GRNs}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:4.2.-(Optional)-Export-GRNs}}
Although celloracle has many functions for network analysis, you can analyze GRNs by yourself. The raw GRN data is stored in the attribute of “links\_dict”.

For example, you can get GRNs for “Ery\_0” cluster with the following commands.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{links\PYGZus{}dict}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ery\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}       source         target  coef\PYGZus{}mean  coef\PYGZus{}abs             p      \PYGZhy{}logp
0         Id2  0610007L01Rik   0.000552  0.000552  5.724488e\PYGZhy{}01   0.242263
1        Klf2  0610007L01Rik   0.000000  0.000000           NaN  \PYGZhy{}0.000000
2      Stat5a  0610007L01Rik  \PYGZhy{}0.005179  0.005179  5.294497e\PYGZhy{}04   3.276175
3        Elf1  0610007L01Rik   0.002107  0.002107  1.498884e\PYGZhy{}01   0.824232
4       Gata1  0610007L01Rik  \PYGZhy{}0.000700  0.000700  5.731063e\PYGZhy{}01   0.241765
...       ...            ...        ...       ...           ...        ...
74460   Cxxc1            Zyx  \PYGZhy{}0.004999  0.004999  1.578852e\PYGZhy{}02   1.801659
74461   Mef2c            Zyx   0.017708  0.017708  3.011616e\PYGZhy{}07   6.521200
74462    Nfe2            Zyx   0.034433  0.034433  2.548244e\PYGZhy{}12  11.593759
74463   Nr3c1            Zyx  \PYGZhy{}0.022663  0.022663  2.265408e\PYGZhy{}08   7.644854
74464    Ets1            Zyx   0.012826  0.012826  4.813285e\PYGZhy{}09   8.317558

[74465 rows x 6 columns]
\end{sphinxVerbatim}
}

You can export the file as follows.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} set cluster name}
\PYG{n}{cluster} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ery\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{c+c1}{\PYGZsh{} save as csv}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{links\PYGZus{}dict}\PYG{p}{[}\PYG{n}{cluster}\PYG{p}{]}\PYG{o}{.}\PYG{n}{to\PYGZus{}csv}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{raw\PYGZus{}GRN\PYGZus{}for\PYGZus{}}\PYG{l+s+si}{\PYGZob{}cluster\PYGZcb{}}\PYG{l+s+s2}{.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{4.3. (Optional) Change order}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:4.3.-(Optional)-Change-order}}
Links object have a color information in a attribute, “palette”. This information is used for the visualization

The sample will be visualized in the order. Here we change the order.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Show the contents of pallete}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{palette}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}        palette
Ery\PYGZus{}0   \PYGZsh{}7D87B9
Ery\PYGZus{}1   \PYGZsh{}BEC1D4
Ery\PYGZus{}2   \PYGZsh{}D6BCC0
Ery\PYGZus{}3   \PYGZsh{}BB7784
Ery\PYGZus{}4   \PYGZsh{}8E063B
Ery\PYGZus{}5   \PYGZsh{}4A6FE3
Ery\PYGZus{}6   \PYGZsh{}8595E1
Ery\PYGZus{}7   \PYGZsh{}B5BBE3
Ery\PYGZus{}8   \PYGZsh{}E6AFB9
Ery\PYGZus{}9   \PYGZsh{}E07B91
GMP\PYGZus{}0   \PYGZsh{}D33F6A
GMP\PYGZus{}1   \PYGZsh{}11C638
GMPl\PYGZus{}0  \PYGZsh{}8DD593
Gran\PYGZus{}0  \PYGZsh{}C6DEC7
Gran\PYGZus{}1  \PYGZsh{}EAD3C6
Gran\PYGZus{}2  \PYGZsh{}F0B98D
MEP\PYGZus{}0   \PYGZsh{}0FCFC0
Mk\PYGZus{}0    \PYGZsh{}9CDED6
Mo\PYGZus{}0    \PYGZsh{}D5EAE7
Mo\PYGZus{}1    \PYGZsh{}F3E1EB
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[17]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} change the order of pallete}
\PYG{n}{order} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mk\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}4}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}6}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}7}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}8}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}9}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMPl\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{palette} \PYG{o}{=} \PYG{n}{links}\PYG{o}{.}\PYG{n}{palette}\PYG{o}{.}\PYG{n}{loc}\PYG{p}{[}\PYG{n}{order}\PYG{p}{]}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{palette}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[17]:\,\hspace{\fboxrule}\hspace{\fboxsep}}        palette
MEP\PYGZus{}0   \PYGZsh{}0FCFC0
Mk\PYGZus{}0    \PYGZsh{}9CDED6
Ery\PYGZus{}0   \PYGZsh{}7D87B9
Ery\PYGZus{}1   \PYGZsh{}BEC1D4
Ery\PYGZus{}2   \PYGZsh{}D6BCC0
Ery\PYGZus{}3   \PYGZsh{}BB7784
Ery\PYGZus{}4   \PYGZsh{}8E063B
Ery\PYGZus{}5   \PYGZsh{}4A6FE3
Ery\PYGZus{}6   \PYGZsh{}8595E1
Ery\PYGZus{}7   \PYGZsh{}B5BBE3
Ery\PYGZus{}8   \PYGZsh{}E6AFB9
Ery\PYGZus{}9   \PYGZsh{}E07B91
GMP\PYGZus{}0   \PYGZsh{}D33F6A
GMP\PYGZus{}1   \PYGZsh{}11C638
GMPl\PYGZus{}0  \PYGZsh{}8DD593
Mo\PYGZus{}0    \PYGZsh{}D5EAE7
Mo\PYGZus{}1    \PYGZsh{}F3E1EB
Gran\PYGZus{}0  \PYGZsh{}C6DEC7
Gran\PYGZus{}1  \PYGZsh{}EAD3C6
Gran\PYGZus{}2  \PYGZsh{}F0B98D
\end{sphinxVerbatim}
}


\subsubsection{5. Network preprocessing}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:5.-Network-preprocessing}}

\paragraph{5.1. Filter network edges}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:5.1.-Filter-network-edges}}
Celloracle leverages bagging ridge or Bayesian ridge regression for network inference. These methods provide a network edge strength as a distribution rather than point value. We can use the distribution to know the certainness of the connection.

We filter the network edges as follows.
\begin{enumerate}
\item {} 
Remove uncertain network edge based on the p-value.

\item {} 
Remove weak network edge. In this tutorial, we pick up top 2000 edges in terms of network strength.

\end{enumerate}

The raw network data is stored as an attribute, “links\_dict,” while filtered network data is stored in “filtered\_links.” So the filtering function keeps raw network information rather than overwriting data. You can come back to the filtering process to filter the data with different parameters if you want.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[32]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{filter\PYGZus{}links}\PYG{p}{(}\PYG{n}{p}\PYG{o}{=}\PYG{l+m+mf}{0.001}\PYG{p}{,} \PYG{n}{weight}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{coef\PYGZus{}abs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{thread\PYGZus{}number}\PYG{o}{=}\PYG{l+m+mi}{2000}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{5.2. Degree distribution}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:5.2.-Degree-distribution}}
For the first step, we examine network degree distribution. Network degree, which is the number of edges for each node, is one of the important metrics to investigate the network structure (\sphinxurl{https://en.wikipedia.org/wiki/Degree\_distribution}).

Please keep in mind that the degree distribution may change depending on the filtering threshold.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[50]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{9}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[51]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}degree\PYGZus{}distributions}\PYG{p}{(}\PYG{n}{plot\PYGZus{}model}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{/degree\PYGZus{}distribution/}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_1}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_3}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_2
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_5}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_3
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_7}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_4
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_9}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_5
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=562\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_11}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_6
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_13}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_7
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_15}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_8
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_17}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_9
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_19}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMP\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_21}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMP\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_23}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMPl\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_25}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_27}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_29}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_2
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_31}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
MEP\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_33}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mk\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_35}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mo\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_37}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mo\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_39}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[52]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}


\paragraph{5.3. Calculate netowrk score}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:5.3.-Calculate-netowrk-score}}
Next, we calculate several network score using some R libraries. Please make sure that R libraries are installed in your PC before running the command below.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[39]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} calculate network scores. It takes several minuts.}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{get\PYGZus{}score}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
processing{\ldots} batch 1/5
Ery\_0: finished.
Ery\_1: finished.
Ery\_2: finished.
Ery\_3: finished.
processing{\ldots} batch 2/5
Ery\_4: finished.
Ery\_5: finished.
Ery\_6: finished.
Ery\_7: finished.
processing{\ldots} batch 3/5
Ery\_8: finished.
Ery\_9: finished.
GMP\_0: finished.
GMP\_1: finished.
processing{\ldots} batch 4/5
GMPl\_0: finished.
Gran\_0: finished.
Gran\_1: finished.
Gran\_2: finished.
processing{\ldots} batch 5/5
MEP\_0: finished.
Mk\_0: finished.
Mo\_0: finished.
Mo\_1: finished.
the scores are saved in ./louvain\_annot/
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

The score is stored as a attribute “merged\_score”, and score will also be saved in a folder in your computer.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[57]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{merged\PYGZus{}score}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[57]:\,\hspace{\fboxrule}\hspace{\fboxsep}}       degree\PYGZus{}all  degree\PYGZus{}in  degree\PYGZus{}out  clustering\PYGZus{}coefficient  \PYGZbs{}
Stat3          91          0          91                0.019780
Mycn           28          0          28                0.002646
Zbtb1          27          0          27                0.005698
E2f4          186          3         183                0.009474
Ybx1           71          9          62                0.027364

       clustering\PYGZus{}coefficient\PYGZus{}weighted  degree\PYGZus{}centrality\PYGZus{}all  \PYGZbs{}
Stat3                         0.020122               0.167279
Mycn                          0.001828               0.051471
Zbtb1                         0.008576               0.049632
E2f4                          0.011623               0.341912
Ybx1                          0.027485               0.130515

       degree\PYGZus{}centrality\PYGZus{}in  degree\PYGZus{}centrality\PYGZus{}out  betweenness\PYGZus{}centrality  \PYGZbs{}
Stat3              0.000000               0.167279                       0
Mycn               0.000000               0.051471                       0
Zbtb1              0.000000               0.049632                       0
E2f4               0.005515               0.336397                    2788
Ybx1               0.016544               0.113971                    1153

       closeness\PYGZus{}centrality  ...  assortative\PYGZus{}coefficient  \PYGZbs{}
Stat3              0.000013  ...                \PYGZhy{}0.172207
Mycn               0.000004  ...                \PYGZhy{}0.172207
Zbtb1              0.000011  ...                \PYGZhy{}0.172207
E2f4               0.000010  ...                \PYGZhy{}0.172207
Ybx1               0.000004  ...                \PYGZhy{}0.172207

       average\PYGZus{}path\PYGZus{}length  community\PYGZus{}edge\PYGZus{}betweenness  community\PYGZus{}random\PYGZus{}walk  \PYGZbs{}
Stat3             2.475244                           1                      1
Mycn              2.475244                           2                      1
Zbtb1             2.475244                           3                      1
E2f4              2.475244                           4                      1
Ybx1              2.475244                           5                     11

       community\PYGZus{}eigenvector  module  connectivity  participation  \PYGZbs{}
Stat3                      1       0      4.052023       0.632291
Mycn                       5       2      2.347146       0.625000
Zbtb1                      5       1      2.275431       0.666667
E2f4                       3       0      8.496354       0.637760
Ybx1                       4       3      5.433857       0.564967

                role  cluster
Stat3  Connector Hub    Ery\PYGZus{}0
Mycn       Connector    Ery\PYGZus{}0
Zbtb1      Connector    Ery\PYGZus{}0
E2f4   Connector Hub    Ery\PYGZus{}0
Ybx1   Connector Hub    Ery\PYGZus{}0

[5 rows x 22 columns]
\end{sphinxVerbatim}
}


\paragraph{5.4. Save}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:5.4.-Save}}
Save processed GRN. We use this file in the next notebook; “in silico simulation with GRNs”.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[42]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Save Links object.}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{to\PYGZus{}hdf5}\PYG{p}{(}\PYG{n}{file\PYGZus{}path}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{links.celloracle.links}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} You can load files with the following command.}
\PYG{n}{links} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{load\PYGZus{}hdf5}\PYG{p}{(}\PYG{n}{file\PYGZus{}path}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{links.celloracle.links}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\end{sphinxVerbatim}
}

If you are not interested in the network score analysis below and just want to do simulation, you can skip them and go to the next notebook.


\subsubsection{6. Network analysis; Network score for each gene}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:6.-Network-analysis;-Network-score-for-each-gene}}
Links class has many functions to visualize network score. See the documentation for the details of the functions.


\paragraph{6.1. Network score in each cluster}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:6.1.-Network-score-in-each-cluster}}
We have calculated several network scores for centrality. We can use the centrality score to identify key regulatory genes because centrality is one of the important indicators of network structure (\sphinxurl{https://en.wikipedia.org/wiki/Centrality}).

Let’s visualize genes with high network centrality.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[44]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check cluster name}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{cluster}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[44]:\,\hspace{\fboxrule}\hspace{\fboxsep}}[\PYGZsq{}Ery\PYGZus{}0\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}1\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}2\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}3\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}4\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}5\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}6\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}7\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}8\PYGZsq{},
 \PYGZsq{}Ery\PYGZus{}9\PYGZsq{},
 \PYGZsq{}GMP\PYGZus{}0\PYGZsq{},
 \PYGZsq{}GMP\PYGZus{}1\PYGZsq{},
 \PYGZsq{}GMPl\PYGZus{}0\PYGZsq{},
 \PYGZsq{}Gran\PYGZus{}0\PYGZsq{},
 \PYGZsq{}Gran\PYGZus{}1\PYGZsq{},
 \PYGZsq{}Gran\PYGZus{}2\PYGZsq{},
 \PYGZsq{}MEP\PYGZus{}0\PYGZsq{},
 \PYGZsq{}Mk\PYGZus{}0\PYGZsq{},
 \PYGZsq{}Mo\PYGZus{}0\PYGZsq{},
 \PYGZsq{}Mo\PYGZus{}1\PYGZsq{}]
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[53]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Visualize top n\PYGZhy{}th genes that have high scores.}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}scores\PYGZus{}as\PYGZus{}rank}\PYG{p}{(}\PYG{n}{cluster}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{n\PYGZus{}gene}\PYG{o}{=}\PYG{l+m+mi}{30}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{/ranked\PYGZus{}score}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=304\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_0}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=341\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_1}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=304\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_2}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=305\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_3}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=296\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_4}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=342\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_5}.png}

\end{nbsphinxfancyoutput}


\paragraph{6.2. Network score comparison between two clusters}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:6.2.-Network-score-comparison-between-two-clusters}}
By comparing network scores between two clusters, we can analyze the difference of GRN structure.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[54]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ticklabel\PYGZus{}format}\PYG{p}{(}\PYG{n}{style}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sci}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{axis}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{y}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{scilimits}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}comparison\PYGZus{}2D}\PYG{p}{(}\PYG{n}{value}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{eigenvector\PYGZus{}centrality}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{cluster1}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{cluster2}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{GMPl\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{percentile}\PYG{o}{=}\PYG{l+m+mi}{98}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{/score\PYGZus{}comparison}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=407\sphinxpxdimen,height=304\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_68_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[55]:\,\hspace{\fboxrule}\hspace{\fboxsep}}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ticklabel\PYGZus{}format}\PYG{p}{(}\PYG{n}{style}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sci}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{axis}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{y}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{scilimits}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}comparison\PYGZus{}2D}\PYG{p}{(}\PYG{n}{value}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{betweenness\PYGZus{}centrality}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{cluster1}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{cluster2}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{GMPl\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{percentile}\PYG{o}{=}\PYG{l+m+mi}{98}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{/score\PYGZus{}comparison}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=416\sphinxpxdimen,height=304\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_69_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[56]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ticklabel\PYGZus{}format}\PYG{p}{(}\PYG{n}{style}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sci}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{axis}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{y}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{scilimits}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}comparison\PYGZus{}2D}\PYG{p}{(}\PYG{n}{value}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{degree\PYGZus{}centrality\PYGZus{}all}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{cluster1}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{cluster2}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{GMPl\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{percentile}\PYG{o}{=}\PYG{l+m+mi}{98}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{/score\PYGZus{}comparison}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=392\sphinxpxdimen,height=304\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_70_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{6.3. Network score dynamics}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:6.3.-Network-score-dynamics}}
In the following session, we focus on how network score for gene changes during the differentiation.

We make some graph to show an example of how to use functions using Gata2 gene.

Gata2 is known to play an essential role in the early progenitor population; MEP, GMP. The following results can recapitulate the Gata2 feature.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[57]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Visualize Gata2 network score dynamics}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}per\PYGZus{}cluster}\PYG{p}{(}\PYG{n}{goi}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{/network\PYGZus{}score\PYGZus{}per\PYGZus{}gene/}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gata2
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=408\sphinxpxdimen,height=301\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_72_1}.png}

\end{nbsphinxfancyoutput}


\paragraph{6.4. Gene cartography analysis}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:6.4.-Gene-cartography-analysis}}
Gene cartography is a method for gene network analysis. The method classifies gene into several groups using the network module structure and connections. It provides us an insight about the role and regulatory mechanism per each gene. Please read the paper for the details of gene cartography (\sphinxurl{https://www.nature.com/articles/nature03288})

The gene cartography will be calculated for the GRN in each cluster. Thus we can know how it change by comparing the the score between clusters.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[58]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} plot cartography as a scatter plot}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}cartography\PYGZus{}scatter\PYGZus{}per\PYGZus{}cluster}\PYG{p}{(}\PYG{n}{scatter}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,}
                                           \PYG{n}{kde}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,}
                                           \PYG{n}{gois}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Sfpi1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
                                           \PYG{n}{auto\PYGZus{}gene\PYGZus{}annot}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,}
                                           \PYG{n}{args\PYGZus{}dot}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{n\PYGZus{}levels}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{l+m+mi}{105}\PYG{p}{\PYGZcb{}}\PYG{p}{,}
                                           \PYG{n}{args\PYGZus{}line}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{c}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{gray}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{\PYGZcb{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{/cartography}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_1}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_3}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_2
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_5}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_3
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_7}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_4
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_9}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_5
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_11}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_6
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_13}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_7
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_15}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_8
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_17}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_9
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_19}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMP\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_21}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMP\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_23}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMPl\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_25}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_27}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_29}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_2
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_31}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
MEP\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_33}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mk\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_35}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mo\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_37}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mo\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_39}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[66]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} plot the summary of cartography analysis}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}cartography\PYGZus{}term}\PYG{p}{(}\PYG{n}{goi}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{/cartography}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gata2
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=408\sphinxpxdimen,height=346\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_75_1}.png}

\end{nbsphinxfancyoutput}


\subsubsection{7. Network analysis; Network score distribution}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:7.-Network-analysis;-Network-score-distribution}}
Next, we visualize the distribution of network score to get insight into the global trend of GRNs.


\paragraph{7.1. Distribution of network degree}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:7.1.-Distribution-of-network-degree}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[60]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.15}\PYG{p}{,} \PYG{n}{bottom}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylim}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mf}{0.040}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}discributions}\PYG{p}{(}\PYG{n}{values}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{degree\PYGZus{}centrality\PYGZus{}all}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{method}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{boxplot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}


\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
degree\_centrality\_all
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=390\sphinxpxdimen,height=263\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_78_1}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[61]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.15}\PYG{p}{,} \PYG{n}{bottom}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylim}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{0.40}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}discributions}\PYG{p}{(}\PYG{n}{values}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{eigenvector\PYGZus{}centrality}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{method}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{boxplot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}



\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
eigenvector\_centrality
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=383\sphinxpxdimen,height=263\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_79_1}.png}

\end{nbsphinxfancyoutput}


\paragraph{7.2. Distribution of netowrk entolopy}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data:7.2.-Distribution-of-netowrk-entolopy}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[62]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.15}\PYG{p}{,} \PYG{n}{bottom}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}network\PYGZus{}entropy\PYGZus{}distributions}\PYG{p}{(}\PYG{n}{save}\PYG{o}{=}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}save\PYGZus{}folder\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}


\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/scipy/stats/\_distn\_infrastructure.py:2614: RuntimeWarning: invalid value encountered in true\_divide
  pk = 1.0*pk / np.sum(pk, axis=0)
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/celloracle/network\_analysis/links\_object.py:345: RuntimeWarning: divide by zero encountered in log
  ent\_norm.append(en/np.log(k[i]))
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/celloracle/network\_analysis/links\_object.py:345: RuntimeWarning: invalid value encountered in double\_scalars
  ent\_norm.append(en/np.log(k[i]))
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=393\sphinxpxdimen,height=263\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_81_1}.png}

\end{nbsphinxfancyoutput}

Using the network score, we could pick up cluster-specific key TFs. “Gata2”, “Gata1”, “Klf1”, “E2f1”, for example, are known to play an essential role in MEP, and these TFs showed high network score in our GRN.

In this step, however, we cannot know the specific function or relationship between cell fate.

In the next analysis, we investigate their function and relationship between cell fate by the simulation with GRNs.


\subsection{Simulation with GRNs}
\label{\detokenize{tutorials/simulation:simulation-with-grns}}\label{\detokenize{tutorials/simulation:simulation}}\label{\detokenize{tutorials/simulation::doc}}
\sphinxcode{\sphinxupquote{celloracle}} leverage GRNs to simulate signal propagation inside a cell.
We can estimate the effect of gene perturbation by the simulation with GRNs.

\sphinxcode{\sphinxupquote{celloracle}} leverage GRNs to simulate signal propagation inside a cell. We can estimate the effect of gene perturbation by the simulation with GRNs.

Besides, we will combine the simulation for the signal propagation and the simulation of cell state transition, which is performed by a python library for RNA-velocity analysis, \sphinxcode{\sphinxupquote{velocyto}} .
A series of simulations visualizes a complex system in which TF controls enormous target genes to determines cell fate.

In short, celloracle provides insight into the regulatory mechanism of cell state from the viewpoint of TF and GRN.

Python notebook


\subsubsection{0. Import libraries}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:0.-Import-libraries}}\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data::doc}}

\paragraph{0.1. Import public libraries}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:0.1.-Import-public-libraries}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{sys}

\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{colors} \PYG{k}{as} \PYG{n+nn}{colors}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{scanpy} \PYG{k}{as} \PYG{n+nn}{sc}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{celloracle} \PYG{k}{as} \PYG{n+nn}{co}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[15]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{}plt.rcParams[\PYGZdq{}font.family\PYGZdq{}] = \PYGZdq{}arial\PYGZdq{}}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{9}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{]}
\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{600}

\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline
\end{sphinxVerbatim}
}


\paragraph{0.1. Make a folder to save graph}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:0.1.-Make-a-folder-to-save-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} make folder to save plots}
\PYG{n}{save\PYGZus{}folder} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figures}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{save\PYGZus{}folder}\PYG{p}{,} \PYG{n}{exist\PYGZus{}ok}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{1. Load data}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:1.-Load-data}}

\paragraph{1.1. Load processed oracle object}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:1.1.-Load-processed-oracle-object}}
Load the oracle object. See the previous notebook for the detil of how to prepare oracle object.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{oracle} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{load\PYGZus{}hdf5}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{../04\PYGZus{}Network\PYGZus{}analysis/Paul\PYGZus{}15\PYGZus{}data.celloracle.oracle}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{1.2. Load inferred GRNs}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:1.2.-Load-inferred-GRNs}}
In the previous notebook, we already calculated GRNs. We use this GRNs for simulation.

We import GRNs which was saved in Links object.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{load\PYGZus{}hdf5}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{../04\PYGZus{}Network\PYGZus{}analysis/links.celloracle.links}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{3. Make predictive models for simulation}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:3.-Make-predictive-models-for-simulation}}
We construct ridge regression models again based on the filtered GRN structure. This process takes less time than the GRN inference in the previous notebook because we use only significant TFs to predict target gene instead of all regulatory candidate TFs.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[12]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{filter\PYGZus{}links}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{get\PYGZus{}cluster\PYGZus{}specific\PYGZus{}TFdict\PYGZus{}from\PYGZus{}Links}\PYG{p}{(}\PYG{n}{links\PYGZus{}object}\PYG{o}{=}\PYG{n}{links}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{fit\PYGZus{}GRN\PYGZus{}for\PYGZus{}simulation}\PYG{p}{(}\PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{n}{use\PYGZus{}cluster\PYGZus{}specific\PYGZus{}TFdict}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
calculating GRN using cluster specicif TF dict{\ldots}
calculating GRN in Ery\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1074 genes
calculating GRN in Ery\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1092 genes
calculating GRN in Ery\_2
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1064 genes
calculating GRN in Ery\_3
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1105 genes
calculating GRN in Ery\_4
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1102 genes
calculating GRN in Ery\_5
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1116 genes
calculating GRN in Ery\_6
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1097 genes
calculating GRN in Ery\_7
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1062 genes
calculating GRN in Ery\_8
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1117 genes
calculating GRN in Ery\_9
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1121 genes
calculating GRN in GMP\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1107 genes
calculating GRN in GMP\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1104 genes
calculating GRN in GMPl\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1089 genes
calculating GRN in Gran\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1067 genes
calculating GRN in Gran\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1076 genes
calculating GRN in Gran\_2
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1105 genes
calculating GRN in MEP\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1152 genes
calculating GRN in Mk\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1114 genes
calculating GRN in Mo\_0
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1085 genes
calculating GRN in Mo\_1
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value=\PYGZsq{}\PYGZsq{})))
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1074 genes
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\subsubsection{4. in silico Perturbation-simulation}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.-in-silico-Perturbation-simulation}}
Next, we simulate an effect of perturbation of a TF to investigate the function and regulatory mechanism of the TF.

See the celloracle paper for the details and scientific remnants on the algorithm.

In this notebook, we’ll show an example of the simulation; we’ll simulate knock-out of Gata1 gene in the hematopoiesis.

Celloracle’s simulation supposes to recapitulate previous findings of this gene.

The previous studies have shown that Gata1 is one of the decisive TFs that regulate cell fate in the myeloid progenitors and Gata1 have a positive effect on the erythroid cell differentiation.


\paragraph{4.1. Check gene expression pattern.}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.1.-Check-gene-expression-pattern.}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[26]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check gene expression}
\PYG{n}{goi} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata1}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{n}{goi}\PYG{p}{,} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{cluster\PYGZus{}column\PYGZus{}name}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{layer}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{imputed\PYGZus{}count}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{use\PYGZus{}raw}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=935\sphinxpxdimen,height=265\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_15_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[33]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} plot gene expression in histogram}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{get}\PYG{o}{.}\PYG{n}{obs\PYGZus{}df}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{keys}\PYG{o}{=}\PYG{p}{[}\PYG{n}{goi}\PYG{p}{]}\PYG{p}{,} \PYG{n}{layer}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{imputed\PYGZus{}count}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{o}{.}\PYG{n}{hist}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=383\sphinxpxdimen,height=263\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_16_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{4.1. calculate future gene expression after perturbation.}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.1.-calculate-future-gene-expression-after-perturbation.}}
Although any arbitrary gene expression value can be used for input of perturbation, we recommend avoiding extreme value which is far from natural gene expression range. If you set Gata1 gene expression to 100, for example, it may lead to biologically infeasible results.

Here we simulate Gata1 KO; we predict what happens to the cells if Gata1 gene expression changed into 0.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[34]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Enter perturbation conditions to simulate signal propagation after the perturbation.}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{simulate\PYGZus{}shift}\PYG{p}{(}\PYG{n}{perturb\PYGZus{}condition}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{n}{goi}\PYG{p}{:} \PYG{l+m+mf}{0.0}\PYG{p}{\PYGZcb{}}\PYG{p}{,}
                      \PYG{n}{n\PYGZus{}propagation}\PYG{o}{=}\PYG{l+m+mi}{3}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{4.2. calculate transition probability between cells}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.2.-calculate-transition-probability-between-cells}}
In the step avobe, we simulated simulated future gene expression after perturbation. This prediction is based on itelative calculations of signal propagations inside the GRN. Here we performed just small number of the calculation, thus the we simulated short future.

Next step, we will calculate the propability of cell state transition based on the simulated data. Using this the transition probability between cells, we can predict how a cell changes after perturbation.

This transition probability will be used in two ways.
\begin{enumerate}
\item {} 
Visualization of directed trjectory graph.

\item {} 
Markof simulation.

\end{enumerate}

In Step 4.2 and 4.3, we use some functions imported from velocytoloom class in velocyto.py. Please see the documentation of VelocytoLoom for the detail of parameters in the functions. \sphinxurl{http://velocyto.org/velocyto.py/fullapi/api\_analysis.html}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[35]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} get transition probability}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{estimate\PYGZus{}transition\PYGZus{}prob}\PYG{p}{(}\PYG{n}{n\PYGZus{}neighbors}\PYG{o}{=}\PYG{l+m+mi}{200}\PYG{p}{,} \PYG{n}{knn\PYGZus{}random}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{sampled\PYGZus{}fraction}\PYG{o}{=}\PYG{l+m+mf}{0.5}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} calculate embedding}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{calculate\PYGZus{}embedding\PYGZus{}shift}\PYG{p}{(}\PYG{n}{sigma\PYGZus{}corr} \PYG{o}{=} \PYG{l+m+mf}{0.05}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} calculate global trend of cell transition}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{calculate\PYGZus{}grid\PYGZus{}arrows}\PYG{p}{(}\PYG{n}{smooth}\PYG{o}{=}\PYG{l+m+mf}{0.8}\PYG{p}{,} \PYG{n}{steps}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{40}\PYG{p}{,} \PYG{l+m+mi}{40}\PYG{p}{)}\PYG{p}{,} \PYG{n}{n\PYGZus{}neighbors}\PYG{o}{=}\PYG{l+m+mi}{300}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3326: FutureWarning: arrays to stack must be passed as a "sequence" type such as list or tuple. Support for non-sequence iterables such as generators is deprecated as of NumPy 1.16 and will raise an error in the future.
  exec(code\_obj, self.user\_global\_ns, self.user\_ns)
WARNING:root:Nans encountered in corrcoef and corrected to 1s. If not identical cells were present it is probably a small isolated cluster converging after imputation.
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\paragraph{4.3. Visualization}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.3.-Visualization}}

\subparagraph{4.3.1. Detailed directed trajectory graph}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.3.1.-Detailed-directed-trajectory-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[36]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{k+kc}{None}\PYG{p}{,}\PYG{p}{(}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{quiver\PYGZus{}scale} \PYG{o}{=} \PYG{l+m+mi}{40}


\PYG{n}{ix\PYGZus{}choice} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{size}\PYG{o}{=}\PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{/}\PYG{l+m+mf}{1.}\PYG{p}{)}\PYG{p}{,} \PYG{n}{replace}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}

\PYG{n}{embedding} \PYG{o}{=} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obsm}\PYG{p}{[}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{embedding\PYGZus{}name}\PYG{p}{]}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,}
            \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{0.8}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{,} \PYG{n}{s}\PYG{o}{=}\PYG{l+m+mi}{38}\PYG{p}{,} \PYG{n}{edgecolor}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{,} \PYG{n}{rasterized}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}

\PYG{n}{quiver\PYGZus{}kwargs}\PYG{o}{=}\PYG{n+nb}{dict}\PYG{p}{(}\PYG{n}{headaxislength}\PYG{o}{=}\PYG{l+m+mi}{7}\PYG{p}{,} \PYG{n}{headlength}\PYG{o}{=}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{n}{headwidth}\PYG{o}{=}\PYG{l+m+mi}{8}\PYG{p}{,}
                   \PYG{n}{linewidths}\PYG{o}{=}\PYG{l+m+mf}{0.35}\PYG{p}{,} \PYG{n}{width}\PYG{o}{=}\PYG{l+m+mf}{0.0045}\PYG{p}{,}\PYG{n}{edgecolors}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{k}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                   \PYG{n}{color}\PYG{o}{=}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{colorandum}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{quiver}\PYG{p}{(}\PYG{n}{embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,}
           \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{delta\PYGZus{}embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{delta\PYGZus{}embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,}
           \PYG{n}{scale}\PYG{o}{=}\PYG{n}{quiver\PYGZus{}scale}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{quiver\PYGZus{}kwargs}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{off}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}plt.savefig(f\PYGZdq{}\PYGZob{}save\PYGZus{}folder\PYGZcb{}/full\PYGZus{}arrows\PYGZob{}goi\PYGZcb{}.png\PYGZdq{},  transparent=True)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimColor={named}{white}}

\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[36]:\,\hspace{\fboxrule}\hspace{\fboxsep}}(\PYGZhy{}10815.27020913708, 10950.84121716522, \PYGZhy{}10711.36365432337, 10949.477199695968)
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=398\sphinxpxdimen,height=357\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_22_1}.png}

\end{nbsphinxfancyoutput}


\subparagraph{4.3.2. Grid graph}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.3.2.-Grid-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[37]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} plot whole graph}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{k+kc}{None}\PYG{p}{,}\PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}grid\PYGZus{}arrows}\PYG{p}{(}\PYG{n}{quiver\PYGZus{}scale}\PYG{o}{=}\PYG{l+m+mf}{2.0}\PYG{p}{,}
                        \PYG{n}{scatter\PYGZus{}kwargs\PYGZus{}dict}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{alpha}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+m+mf}{0.35}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{lw}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+m+mf}{0.35}\PYG{p}{,}
                                              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{edgecolor}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{0.4}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{s}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+m+mi}{38}\PYG{p}{,}
                                              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{rasterized}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{k+kc}{True}\PYG{p}{\PYGZcb{}}\PYG{p}{,}
                        \PYG{n}{min\PYGZus{}mass}\PYG{o}{=}\PYG{l+m+mf}{0.015}\PYG{p}{,} \PYG{n}{angles}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{xy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{scale\PYGZus{}units}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{xy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                        \PYG{n}{headaxislength}\PYG{o}{=}\PYG{l+m+mf}{2.75}\PYG{p}{,}
                        \PYG{n}{headlength}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{n}{headwidth}\PYG{o}{=}\PYG{l+m+mf}{4.8}\PYG{p}{,} \PYG{n}{minlength}\PYG{o}{=}\PYG{l+m+mf}{1.5}\PYG{p}{,}
                        \PYG{n}{plot\PYGZus{}random}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{n}{scale\PYGZus{}type}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{relative}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}plt.savefig(f\PYGZdq{}\PYGZob{}save\PYGZus{}folder\PYGZcb{}/vectorfield\PYGZus{}\PYGZob{}goi\PYGZcb{}.png\PYGZdq{}, transparent=True)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=619\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_24_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{4.4. Markov simulation to analyze the effects of perturbation on cell fate transition}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.4.-Markov-simulation-to-analyze-the-effects-of-perturbation-on-cell-fate-transition}}
We can also simulate cell state transition with Markof simulation with the transition probability.


\subparagraph{4.4.1. Do simulation}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.4.1.-Do-simulation}}
We simulate with the parameters, “n\_steps=200” and “n\_duplication=5” in the following example.

It means that
\begin{enumerate}
\item {} 
we will do 200 times of iterative simulations to predict how cell changes over time

\item {} 
and we will do 5 round of simulations for each cells.

\end{enumerate}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[83]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{o}{\PYGZpc{}\PYGZpc{}time}
\PYG{c+c1}{\PYGZsh{} n\PYGZus{}steps is the number of steps in markov simulation.}
\PYG{c+c1}{\PYGZsh{} n\PYGZus{}duplication is the number of technical duplicate for the simulation}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{run\PYGZus{}markov\PYGZus{}chain\PYGZus{}simulation}\PYG{p}{(}\PYG{n}{n\PYGZus{}steps}\PYG{o}{=}\PYG{l+m+mi}{200}\PYG{p}{,} \PYG{n}{n\PYGZus{}duplication}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
CPU times: user 1.33 s, sys: 0 ns, total: 1.33 s
Wall time: 1.33 s
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax


\subparagraph{4.4.2. Check the results of simulation focusing on a specific cells}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.4.2.-Check-the-results-of-simulation-focusing-on-a-specific-cells}}
Check the results of simulation. Pick up some cells and visualize their transition trajectory.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[88]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{}np.random.seed(12)}
\PYG{c+c1}{\PYGZsh{} randomly pick up 3 cells}
\PYG{n}{cells} \PYG{o}{=} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{index}\PYG{o}{.}\PYG{n}{values}\PYG{p}{[}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{ixs\PYGZus{}mcmc}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{)}\PYG{p}{]}

\PYG{c+c1}{\PYGZsh{} visualize the simulated results of cell transition after perturbation}
\PYG{k}{for} \PYG{n}{k} \PYG{o+ow}{in} \PYG{n}{cells}\PYG{p}{:}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cell }\PYG{l+s+si}{\PYGZob{}k\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{9}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{]}\PYG{p}{)}
    \PYG{k}{for} \PYG{n}{j}\PYG{p}{,} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{20}\PYG{p}{,} \PYG{l+m+mi}{50}\PYG{p}{]}\PYG{p}{)}\PYG{p}{:} \PYG{c+c1}{\PYGZsh{} time points}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{p}{(}\PYG{n}{j}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
        \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}mc\PYGZus{}result\PYGZus{}as\PYGZus{}trajectory}\PYG{p}{(}\PYG{n}{k}\PYG{p}{,} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{i}\PYG{p}{)}\PYG{p}{)}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{n}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{simulation step: 0\PYGZti{}}\PYG{l+s+si}{\PYGZob{}i\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{off}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
cell 1961
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=573\sphinxpxdimen,height=209\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_1}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
cell 43
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=573\sphinxpxdimen,height=209\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_3}.png}

\end{nbsphinxfancyoutput}



%
{
\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\sphinxsetup{VerbatimColor={named}{white}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
cell 1567
\end{sphinxVerbatim}
}
% The following \relax is needed to avoid problems with adjacent ANSI
% cells and some other stuff (e.g. bullet lists) following ANSI cells.
% See https://github.com/sphinx-doc/sphinx/issues/3594
\relax

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=573\sphinxpxdimen,height=209\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_5}.png}

\end{nbsphinxfancyoutput}


\subparagraph{4.4.3. Summarize the results of simulation by plotting sankey diagram}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data:4.4.3.-Summarize-the-results-of-simulation-by-plotting-sankey-diagram}}
Sankey diagram is useful when you want to visualize cell transition between some groups.

The function below can make a Sankey diagram with any arbitrary cluster unit.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[89]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} plot sankey diagram}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{cl} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}mc\PYGZus{}resutls\PYGZus{}as\PYGZus{}sankey}\PYG{p}{(}\PYG{n}{cluster\PYGZus{}use}\PYG{o}{=}\PYG{n}{cl}\PYG{p}{,} \PYG{n}{start}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{end}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=357\sphinxpxdimen,height=357\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_35_0}.png}

\end{nbsphinxfancyoutput}

The sankey diagram above looks messy because cluster order is random.

Change cluster order and make plot again

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[90]:\,\hspace{\fboxrule}\hspace{\fboxsep}}
\PYG{n}{cl} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{order} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mk\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}4}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}6}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}7}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}8}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}9}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMPl\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMPl\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{,} \PYG{n}{right}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}mc\PYGZus{}resutls\PYGZus{}as\PYGZus{}sankey}\PYG{p}{(}\PYG{n}{cluster\PYGZus{}use}\PYG{o}{=}\PYG{n}{cl}\PYG{p}{,}  \PYG{n}{start}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{end}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,} \PYG{n}{order}\PYG{o}{=}\PYG{n}{order}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}plt.savefig(f\PYGZdq{}\PYGZob{}save\PYGZus{}folder\PYGZcb{}/mcmc\PYGZus{}\PYGZob{}cl\PYGZcb{}.png\PYGZdq{})}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=226\sphinxpxdimen,height=357\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_37_0}.png}

\end{nbsphinxfancyoutput}

Make another saneky diagram with different cluster unit.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[92]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{order} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Megakaryocytes}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEP}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Erythroids}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{late\PYGZus{}GMP}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Monocytes}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Granulocytes}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{cl} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cell\PYGZus{}type}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.35}\PYG{p}{,} \PYG{n}{right}\PYG{o}{=}\PYG{l+m+mf}{0.65}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}mc\PYGZus{}resutls\PYGZus{}as\PYGZus{}sankey}\PYG{p}{(}\PYG{n}{cluster\PYGZus{}use}\PYG{o}{=}\PYG{n}{cl}\PYG{p}{,} \PYG{n}{start}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{end}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,} \PYG{n}{order}\PYG{o}{=}\PYG{n}{order}\PYG{p}{,} \PYG{n}{font\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mi}{14}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}plt.savefig(f\PYGZdq{}\PYGZob{}save\PYGZus{}folder\PYGZcb{}/mcmc\PYGZus{}\PYGZob{}cl\PYGZcb{}\PYGZob{}goi\PYGZcb{}.png\PYGZdq{}, transparent=True)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=346\sphinxpxdimen,height=360\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_39_0}.png}

\end{nbsphinxfancyoutput}

Based on the results, we may conclude several things as follows.

Gata1 KO induced cell state transition from Erythroids to MEP, and from MEP to GMP.
\begin{enumerate}
\item {} 
These results suggest that Gata1 may play a role in the progression of Erythroid differentiation and cell state determination between MEP and GMP lineage.

\item {} 
Gata1 KO also induced cell state transition from granulocytes to late GMP, suggesting Gata1’s involvement in Granulocytes differentiation.

\end{enumerate}

These results agree with previous reports about Gata1. We could recapitulate Gata1’s cell-type-specific function regarding the cell fate decisions in hematopoiesis.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}
\end{sphinxVerbatim}
}


\section{API}
\label{\detokenize{modules/index:api}}\label{\detokenize{modules/index::doc}}

\subsection{Command Line API}
\label{\detokenize{modules/index:command-line-api}}
CellOracle have one command line API. This command can be used to convert scRNA-seq data.
If you have a scRNA-seq data which was processed with seurat and saved as Rds file, you can use the following command to make anndata from seurat object.
The anndata produced by this command can be used for an input of celloracle. Please go to step3 and step4 of the tutorial to see an example.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{seuratToAnndata} \PYG{n}{YOUR\PYGZus{}SEURAT\PYGZus{}OBJECT}\PYG{o}{.}\PYG{n}{Rds} \PYG{n}{OUTPUT\PYGZus{}PATH}
\end{sphinxVerbatim}


\subsection{Python API}
\label{\detokenize{modules/index:python-api}}

\subsubsection{Custom class in celloracle}
\label{\detokenize{modules/celloracle:custom-class-in-celloracle}}\label{\detokenize{modules/celloracle::doc}}
We define some custom classes in celloracle.

\phantomsection\label{\detokenize{modules/celloracle:module-celloracle}}\index{celloracle (module)}\index{Oracle (class in celloracle)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle}}\pysigline{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.}}\sphinxbfcode{\sphinxupquote{Oracle}}}
Bases: \sphinxcode{\sphinxupquote{celloracle.trajectory.modified\_VelocytoLoom\_class.modified\_VelocytoLoom}}

Oracle is the main class in CellOracle.
Oracle object imports scRNA-seq data (anndata) and TF information to infer cluster-specific GRNs.
It can predict future gene expression patterns and cell state transition after perturbations of TFs.
Please see the paper of CellOracle for details.

The code of the Oracle class was made of three components below.

(1) Anndata:  Gene expression matrix and metadata in single-cell RNA-seq are stored in anndata object.
Processed values, such as normalized counts and simulated values, are stored as layers of anndata.
Metadata (i.e., Cluster info) are saved in anndata.obs. Refer to scanpy/anndata documentation for detail.

(2) Net: Net is a custom class in celloracle.
Net object process several data to infer GRN.
See the documentation of Net class for detail.

(3) VelycytoLoom: Calculation of transition probability and visualization of directed trajectory graph will be performed in the same way as velocytoloom.
VelocytoLoom is class for the Velocyto, which is a python library for RNA-velocity analysis.
In the celloracle, we use almost the same functions as velocytoloom for the visualization, but celloracle use simulated gene expression values instead of RNA-velocity data.

Some CellOracle’s methods were inspired by velocyto analysis and some codes were made by modifying VelocytoLoom class.
\index{adata (celloracle.Oracle attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.adata}}\pysigline{\sphinxbfcode{\sphinxupquote{adata}}}
\sphinxstyleemphasis{anndata} \textendash{} Imported anndata object

\end{fulllineitems}

\index{cluster\_column\_name (celloracle.Oracle attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.cluster_column_name}}\pysigline{\sphinxbfcode{\sphinxupquote{cluster\_column\_name}}}
\sphinxstyleemphasis{str} \textendash{} The column name in adata.obs about cluster info

\end{fulllineitems}

\index{embedding\_name (celloracle.Oracle attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.embedding_name}}\pysigline{\sphinxbfcode{\sphinxupquote{embedding\_name}}}
\sphinxstyleemphasis{str} \textendash{} The key name in adata.obsm about dimensional reduction cordinates

\end{fulllineitems}

\index{addTFinfo\_dictionary() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.addTFinfo_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addTFinfo\_dictionary}}}{\emph{TFdict}}{}
Add new TF info to pre-existing TFdict.
Values in the old TF dictionary will remain.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Python dictionary of TF info.

\end{description}\end{quote}

\end{fulllineitems}

\index{copy() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.copy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{copy}}}{}{}
Deepcopy itself.

\end{fulllineitems}

\index{fit\_GRN\_for\_simulation() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.fit_GRN_for_simulation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_GRN\_for\_simulation}}}{\emph{GRN\_unit='cluster'}, \emph{alpha=1}, \emph{use\_cluster\_specific\_TFdict=False}}{}
Do GRN inference.
Please see the paper of CellOracle for details.

GRN can be constructed at an arbitrary cell group.
If you want to infer cluster-specific GRN, please set {[}GRN\_unit=”cluster”{]}.
GRN will be inferred for each cluster. You can select Cluster information when you import data (not when you run this method.).

If you set {[}GRN\_unit=”whole”{]}, GRN will be made using all cells.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{GRN\_unit}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} select “cluster” or “whole”

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{ or }}\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the strength of regularization.
If you set a lower value, the sensitivity increase, and you can detect a weak network connection, but it might get more noize.
With a higher value of alpha may reduce the chance of overfitting.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_cluster\_specific\_TFdict\_from\_Links() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.get_cluster_specific_TFdict_from_Links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_cluster\_specific\_TFdict\_from\_Links}}}{\emph{links\_object}}{}
Extract TF and its target gene information from Links object.
This function can be used to reconstruct GRNs based on pre-existing GRNs saved in Links object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{links\_object}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} Please see the explanation of Links class.

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_links() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.get_links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_links}}}{\emph{cluster\_name\_for\_GRN\_unit=None}, \emph{alpha=10}, \emph{bagging\_number=20}, \emph{verbose\_level=1}, \emph{test\_mode=False}}{}
Make GRN for each cluster and returns results as a Links object.
Several preprocessing should be done before using this function.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_name\_for\_GRN\_unit}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name for GRN calculation. The cluster information should be stored in Oracle.adata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{ or }}\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the strength of regularization.
If you set a lower value, the sensitivity increase, and you can detect a weak network connection, but it might get more noize.
With a higher value of alpha may reduce the chance of overfitting.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number for bagging calculation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose\_level}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} if {[}verbose\_level\textgreater{}1{]}, most detailed progress information will be shown.
if {[}verbose\_level \textgreater{} 0{]}, one progress bar will be shown.
if {[}verbose\_level == 0{]}, no progress bar will be shown.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{test\_mode}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} If test\_mode is True, GRN calculation will be done for only one cluster rather than all clusters.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{import\_TF\_data() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.import_TF_data}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{import\_TF\_data}}}{\emph{TF\_info\_matrix=None}, \emph{TF\_info\_matrix\_path=None}, \emph{TFdict=None}}{}
Load data about potential-regulatory TFs.
You can import either TF\_info\_matrix or TFdict.
See the tutorial of celloracle or motif\_analysis module for an example to make such files.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{TF\_info\_matrix}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.DataFrame}}) \textendash{} TF\_info\_matrix.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{TF\_info\_matrix\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} File path for TF\_info\_matrix (pandas.DataFrame).

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Python dictionary of TF info.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{import\_anndata\_as\_normalized\_count() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.import_anndata_as_normalized_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{import\_anndata\_as\_normalized\_count}}}{\emph{adata}, \emph{cluster\_column\_name=None}, \emph{embedding\_name=None}}{}
Load scRNA-seq data. scRNA-seq data should be prepared as an anndata.
Preprocessing (cell and gene filtering, calculate DR and cluster, etc.) should be done before loading data.
The method will import NORMALIZED and LOG TRANSFORMED but NOT SCALED and NOT CENTERED DATA.
See the tutorial for the details for an example of how to process scRNA-seq data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{adata}} (\sphinxstyleliteralemphasis{\sphinxupquote{anndata}}) \textendash{} anndata object that store scRNA-seq data.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_column\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the column name about cluster info in anndata.obs.
Clustering data suppose to be in anndata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{embedding\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the key name about a dimensional reduction in anndata.obsm.
Dimensional reduction (or 2D trajectory graph) should be in anndata.obsm.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{transform}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The method for log-transformation. Chose one from “natural\_log” or “log2”.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{import\_anndata\_as\_raw\_count() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.import_anndata_as_raw_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{import\_anndata\_as\_raw\_count}}}{\emph{adata}, \emph{cluster\_column\_name=None}, \emph{embedding\_name=None}, \emph{transform='natural\_log'}}{}
Load scRNA-seq data. scRNA-seq data should be prepared as an anndata.
Preprocessing (cell and gene filtering, calculate DR and cluster, etc.) should be done before loading data.
The method imports RAW GENE COUNTS because unscaled and uncentered gene expression data are required for the GRN inference and simulation.
See tutorial notebook for the details about how to process scRNA-seq data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{adata}} (\sphinxstyleliteralemphasis{\sphinxupquote{anndata}}) \textendash{} anndata object that stores scRNA-seq data.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_column\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the column name about cluster info in anndata.obs.
Clustering data suppose to be in anndata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{embedding\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the key name about a dimensional reduction in anndata.obsm.
Dimensional reduction (or 2D trajectory graph) should be in anndata.obsm.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{transform}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The method for log-transformation. Chose one from “natural\_log” or “log2”.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_mc\_result\_as\_kde() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.plot_mc_result_as_kde}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_mc\_result\_as\_kde}}}{\emph{n\_time}, \emph{args=\{\}}}{}
Pick up one timepoint in the cell state-transition simulation and plot as a kde plot.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_time}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the number in Markov simulation

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} An argument for seaborn.kdeplot.
See seaborn documentation for detail (\sphinxurl{https://seaborn.pydata.org/generated/seaborn.kdeplot.html\#seaborn.kdeplot}).

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_mc\_result\_as\_trajectory() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.plot_mc_result_as_trajectory}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_mc\_result\_as\_trajectory}}}{\emph{cell\_name}, \emph{time\_range}, \emph{args=\{\}}}{}
Pick up several timepoints in the cell state-transition simulation and plot as a line plot.
This function can be used to visualize how cell-state changes after perturbation focusing on a specific cell.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cell\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} cell name. chose from adata.obs.index

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{time\_range}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of int}}) \textendash{} the number in markov simulation

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} dictionary for the arguments for matplotlib.pyplit.plot.
See matplotlib documentation for detail (\sphinxurl{https://matplotlib.org/api/\_as\_gen/matplotlib.pyplot.plot.html\#matplotlib.pyplot.plot}).

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_mc\_resutls\_as\_sankey() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.plot_mc_resutls_as_sankey}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_mc\_resutls\_as\_sankey}}}{\emph{cluster\_use}, \emph{start=0}, \emph{end=-1}, \emph{order=None}, \emph{font\_size=10}}{}
Plot the simulated cell state-transition as a Sankey-diagram after groping by the cluster.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_use}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} cluster information name in anndata.obs.
You can use any arbitrary cluster information in anndata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{start}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The starting point of Sankey-diagram. Please select a  step in the Markov simulation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{end}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The end point of Sankey-diagram. Please select a  step in the Markov simulation.
if you set {[}end=-1{]}, the final step of Markov simulation will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{order}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The order of cluster name in sankey-diagram.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{font\_size}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Font size for cluster name label in Sankey diagram.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{prepare\_markov\_simulation() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.prepare_markov_simulation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{prepare\_markov\_simulation}}}{\emph{verbose=False}}{}
Pick up cells for Markov simulation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} If True, it plots selected cells.

\end{description}\end{quote}

\end{fulllineitems}

\index{run\_markov\_chain\_simulation() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.run_markov_chain_simulation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{run\_markov\_chain\_simulation}}}{\emph{n\_steps=500}, \emph{n\_duplication=5}, \emph{seed=123}}{}
Do Markov simlation to predict cell transition after perturbation.
The transition probability between cells has been calculated
based on simulated gene expression values in the signal propagation process.
The cell state transition will be simulated based on the probability.
You can simulate the process for multiple times to get a robust outcome.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_steps}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} steps for Markov simulation. This value is equivalent to the time after perturbation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_duplication}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the number for multiple calculations.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{simulate\_shift() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.simulate_shift}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{simulate\_shift}}}{\emph{perturb\_condition=None}, \emph{GRN\_unit='whole'}, \emph{n\_propagation=3}}{}
Simulate signal propagation with GRNs. Please see the paper of CellOracle for details.
This function simulates a gene expression pattern in the near future.
Simulated values will be stored in anndata.layers: {[}“simulated\_count”{]}

Three data below are used for the simulation.
(1) GRN inference results (coef\_matrix).
(2) perturb\_condition: You can set arbitrary perturbation condition.
(3) gene expression matrix: simulation starts from imputed gene expression data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{perturb\_condition}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} condition for perturbation.
if you want to simulate knockout for GeneX, please set {[}perturb\_condition=\{“GeneX”: 0.0\}{]}
Although you can set any non-negative values for the gene condition, avoid setting biologically unfeasible values for the perturb condition.
It is strongly recommended to check actual gene expression values in your data before selecting perturb condition.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_propagation}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Calculation will be performed iteratively to simulate signal propagation in GRN.
you can set the number of steps for this calculation.
With a higher number, the results may recapitulate signal propagation for many genes.
However, a higher number of propagation may cause more error/noise.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{summarize\_mc\_results\_by\_cluster() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.summarize_mc_results_by_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{summarize\_mc\_results\_by\_cluster}}}{\emph{cluster\_use}}{}
This function summarizes the simulated cell state-transition by groping the results into each cluster.
It returns sumarized results as a pandas.DataFrame.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_use}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} cluster information name in anndata.obs.
You can use any arbitrary cluster information in anndata.obs.

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{file\_path}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.oracle’

\end{description}\end{quote}

\end{fulllineitems}

\index{updateTFinfo\_dictionary() (celloracle.Oracle method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.updateTFinfo_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{updateTFinfo\_dictionary}}}{\emph{TFdict}}{}
Update a TF dictionary.
If a key in the new TF dictionary already existed in the old TF dictionary, old values will be replaced with a new one.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Python dictionary of TF info.

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{Links (class in celloracle)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.}}\sphinxbfcode{\sphinxupquote{Links}}}{\emph{name}, \emph{links\_dict=\{\}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

This is a class for the processing and visualization of GRNs.
Links object stores cluster-specific GRNs and metadata.
Please use “get\_links” function in Oracle object to generate Links object.
\index{links\_dict (celloracle.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.links_dict}}\pysigline{\sphinxbfcode{\sphinxupquote{links\_dict}}}
\sphinxstyleemphasis{dictionary} \textendash{} Dictionary that store unprocessed network data.

\end{fulllineitems}

\index{filtered\_links (celloracle.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.filtered_links}}\pysigline{\sphinxbfcode{\sphinxupquote{filtered\_links}}}
\sphinxstyleemphasis{dictionary} \textendash{} Dictionary that store filtered network data.

\end{fulllineitems}

\index{merged\_score (celloracle.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.merged_score}}\pysigline{\sphinxbfcode{\sphinxupquote{merged\_score}}}
\sphinxstyleemphasis{pandas.dataframe} \textendash{} Network scores.

\end{fulllineitems}

\index{cluster (celloracle.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.cluster}}\pysigline{\sphinxbfcode{\sphinxupquote{cluster}}}
\sphinxstyleemphasis{list of str} \textendash{} List of cluster name.

\end{fulllineitems}

\index{name (celloracle.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.name}}\pysigline{\sphinxbfcode{\sphinxupquote{name}}}
\sphinxstyleemphasis{str} \textendash{} Name of clustering unit.

\end{fulllineitems}

\index{palette (celloracle.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.palette}}\pysigline{\sphinxbfcode{\sphinxupquote{palette}}}
\sphinxstyleemphasis{pandas.dataframe} \textendash{} DataFrame that store color information.

\end{fulllineitems}

\index{add\_palette() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.add_palette}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_palette}}}{\emph{cluster\_name\_list}, \emph{color\_list}}{}
Add color information for each cluster.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_name\_list}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} cluster names.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{color\_list}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of stf}}) \textendash{} list of color for each cluster.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{filter\_links() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.filter_links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{filter\_links}}}{\emph{p=0.001}, \emph{weight='coef\_abs'}, \emph{thread\_number=10000}, \emph{genelist\_source=None}, \emph{genelist\_target=None}}{}
Filter network edges.
In most case inferred GRN has non-significant random edges.
We have to remove such random edges before analyzing network structure.
You can do the filtering either of three ways below.
\begin{enumerate}
\item {} 
Do filtering based on the p-value of the network edge.
Please enter p-value for thresholding.

\item {} 
Do filtering based on network edge number.
If you set the number, network edges will be filtered based on the order of a network score. The top n-th network edges with network weight will remain, and the other edges will be removed.
The network data have several types of network weight, so you have to select which network weight do you want to use.

\item {} 
Do filtering based on an arbitrary gene list. You can set a gene list for source nodes or target nodes.

\end{enumerate}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{p}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} threshold for p-value of the network edge.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{weight}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Please select network weight name for the filtering

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{genelist\_source}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list to remain in regulatory gene nodes. Default is None.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{genelist\_target}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list to remain in target gene nodes. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_network\_entropy() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.get_network_entropy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_network\_entropy}}}{\emph{value='coef\_abs'}}{}
Calculate network entropy scores.
Please select a type of network weight for the entropy calculation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{value}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Default is “coef\_abs”.

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_score() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.get_score}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_score}}}{\emph{test\_mode=False}}{}
Get several network sores using R libraries.
Make sure all dependant R libraries are installed in your environment before running this function.
You can check the installation for the R libraries by running test\_installation() in network\_analysis module.

\end{fulllineitems}

\index{plot\_cartography\_scatter\_per\_cluster() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_cartography_scatter_per_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cartography\_scatter\_per\_cluster}}}{\emph{gois=None}, \emph{clusters=None}, \emph{scatter=False}, \emph{kde=True}, \emph{auto\_gene\_annot=False}, \emph{percentile=98}, \emph{args\_dot=\{\}}, \emph{args\_line=\{\}}, \emph{args\_annot=\{\}}, \emph{save=None}}{}
Make a plot of gene network cartography.
Please read the original paper of gene network cartography for the principle of gene network cartography.
\sphinxurl{https://www.nature.com/articles/nature03288}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gois}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of srt}}) \textendash{} List of Gene name to highlight.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{clusters}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} List of cluster name to analyze. If None, all clusters in Links object will be analyzed.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scatter}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to make a scatter plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{auto\_gene\_annot}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to pick up genes to make an annotation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{percentile}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Genes with a network score above the percentile will be shown with annotation. Default is 98.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_dot}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for scatter plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_line}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for lines in cartography plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_annot}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for annotation in plots.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_cartography\_term() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_cartography_term}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cartography\_term}}}{\emph{goi}, \emph{save=None}}{}
Plot the summary of gene network cartography like a heatmap.
Please read the original paper of gene network cartography for the principle of gene network cartography.
\sphinxurl{https://www.nature.com/articles/nature03288}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gois}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of srt}}) \textendash{} List of Gene name to highlight.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_degree\_distributions() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_degree_distributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_degree\_distributions}}}{\emph{plot\_model=False}, \emph{save=None}}{}
Plot the distribution of network degree (the number of edge per gene).
The network degree will be visualized in both linear scale and log scale.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plot\_model}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to plot linear approximation line.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_network\_entropy\_distributions() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_network_entropy_distributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_network\_entropy\_distributions}}}{\emph{update\_network\_entropy=False}, \emph{save=None}}{}
Plot the distribution of network entropy.
See the CellOracle paper for the detail.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} (\sphinxstyleliteralemphasis{\sphinxupquote{Links object}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{values}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The list of score to visualize. If it is None, all network score (listed above) will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{update\_network\_entropy}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to recalculate network entropy.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_comparison\_2D() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_score_comparison_2D}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_comparison\_2D}}}{\emph{value}, \emph{cluster1}, \emph{cluster2}, \emph{percentile=99}, \emph{annot\_shifts=None}, \emph{save=None}}{}
Make a scatter plot that shows the relationship of a specific network score in two groups.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{value}} (\sphinxstyleliteralemphasis{\sphinxupquote{srt}}) \textendash{} The network score type.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster1}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name. Network scores in the cluster1 will be visualized in the x-axis.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster2}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name. Network scores in the cluster2 will be visualized in the y-axis.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{percentile}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Genes with a network score above the percentile will be shown with annotation. Default is 99.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{annot\_shifts}} (\sphinxstyleliteralemphasis{\sphinxupquote{(}}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{)}}) \textendash{} Annotation visualization setting.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_discributions() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_score_discributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_discributions}}}{\emph{values=None}, \emph{method='boxplot'}, \emph{save=None}}{}
Plot the distribution of network scores.
An individual data point is a network edge (gene) of GRN in each cluster.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{values}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The list of score to visualize. If it is None, all network score will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Plotting method. Select either of “boxplot” or “barplot”.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_per\_cluster() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_score_per_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_per\_cluster}}}{\emph{goi}, \emph{save=None}}{}
Plot network score for a gene.
This function visualizes the network score of a specific gene between clusters to get an insight into the dynamics of the gene.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{goi}} (\sphinxstyleliteralemphasis{\sphinxupquote{srt}}) \textendash{} Gene name.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_scores\_as\_rank() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_scores_as_rank}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_scores\_as\_rank}}}{\emph{cluster}, \emph{n\_gene=50}, \emph{save=None}}{}
Pick up top n-th genes wich high-network scores and make plots.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster nome to analyze.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_gene}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Number of genes to plot. Default is 50.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{file\_path}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.links’

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{Net (class in celloracle)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.}}\sphinxbfcode{\sphinxupquote{Net}}}{\emph{gene\_expression\_matrix}, \emph{gem\_standerdized=None}, \emph{TFinfo\_matrix=None}, \emph{cellstate=None}, \emph{TFinfo\_dic=None}, \emph{annotation=None}, \emph{verbose=True}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

Net is a custom class for inferring sample-specific GRN from scRNA-seq data.
This class is used inside Oracle class for GRN inference.
This class requires two information below.
\begin{enumerate}
\item {} 
single-cell RNA-seq data
The Net class needs processed scRNA-seq data.
Gene and cell filtering, quality check, normalization, log-transformation (but not scaling and centering) have to be done before starting GRN calculation with this class.
You can also use any arbitrary metadata (i.e., mRNA count, cell-cycle phase) for GRN input.

\item {} 
potential regulatory connection
This method uses the list of potential regulatory TFs as input.
This information can be calculated from Atac-seq data using the motif-analysis module.
If sample-specific ATAC-seq data is not available,
you can use general TF-binding info made from public ATAC-seq dataset of various tissue/cell type.

\end{enumerate}
\index{linkList (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.linkList}}\pysigline{\sphinxbfcode{\sphinxupquote{linkList}}}
\sphinxstyleemphasis{pandas.DataFrame} \textendash{} the result of GRN inference.

\end{fulllineitems}

\index{all\_genes (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.all_genes}}\pysigline{\sphinxbfcode{\sphinxupquote{all\_genes}}}
\sphinxstyleemphasis{numpy.array} \textendash{} an array of all genes that exist in input gene expression matrix

\end{fulllineitems}

\index{embedding\_name (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.embedding_name}}\pysigline{\sphinxbfcode{\sphinxupquote{embedding\_name}}}
\sphinxstyleemphasis{str} \textendash{} the key name name in adata.obsm about dimensional reduction coordinates

\end{fulllineitems}

\index{annotation (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.annotation}}\pysigline{\sphinxbfcode{\sphinxupquote{annotation}}}
\sphinxstyleemphasis{dictionary} \textendash{} annotation. you can add an arbitrary annotation.

\end{fulllineitems}

\index{coefs\_dict (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.coefs_dict}}\pysigline{\sphinxbfcode{\sphinxupquote{coefs\_dict}}}
\sphinxstyleemphasis{dictionary} \textendash{} coefs of linear regression.

\end{fulllineitems}

\index{stats\_dict (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.stats_dict}}\pysigline{\sphinxbfcode{\sphinxupquote{stats\_dict}}}
\sphinxstyleemphasis{dictionary} \textendash{} statistic values about coefs.

\end{fulllineitems}

\index{fitted\_genes (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.fitted_genes}}\pysigline{\sphinxbfcode{\sphinxupquote{fitted\_genes}}}
\sphinxstyleemphasis{list of str} \textendash{} list of genes with which regression model was successfully calculated.

\end{fulllineitems}

\index{failed\_genes (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.failed_genes}}\pysigline{\sphinxbfcode{\sphinxupquote{failed\_genes}}}
\sphinxstyleemphasis{list of str} \textendash{} list of genes that were not able to get coefs

\end{fulllineitems}

\index{cellstate (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.cellstate}}\pysigline{\sphinxbfcode{\sphinxupquote{cellstate}}}
\sphinxstyleemphasis{pandas.DataFrame} \textendash{} metadata for GRN input

\end{fulllineitems}

\index{TFinfo (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.TFinfo}}\pysigline{\sphinxbfcode{\sphinxupquote{TFinfo}}}
\sphinxstyleemphasis{pandas.DataFrame} \textendash{} information about potential regulatory TFs.

\end{fulllineitems}

\index{gem (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.gem}}\pysigline{\sphinxbfcode{\sphinxupquote{gem}}}
\sphinxstyleemphasis{pandas.DataFrame} \textendash{} merged matrix made with gene\_expression\_matrix and cellstate matrix.

\end{fulllineitems}

\index{gem\_standerdized (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.gem_standerdized}}\pysigline{\sphinxbfcode{\sphinxupquote{gem\_standerdized}}}
\sphinxstyleemphasis{pandas.DataFrame} \textendash{} almost the same as gem, but the gene\_expression\_matrix was standardized.

\end{fulllineitems}

\index{library\_last\_update\_date (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.library_last_update_date}}\pysigline{\sphinxbfcode{\sphinxupquote{library\_last\_update\_date}}}
\sphinxstyleemphasis{str} \textendash{} last update date of this code (this info is for code development. it will be deleted in the future)

\end{fulllineitems}

\index{object\_initiation\_date (celloracle.Net attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.object_initiation_date}}\pysigline{\sphinxbfcode{\sphinxupquote{object\_initiation\_date}}}
\sphinxstyleemphasis{str} \textendash{} the date when this object is made.

\end{fulllineitems}

\index{addAnnotation() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.addAnnotation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addAnnotation}}}{\emph{annotation\_dictionary}}{}
Add a new annotation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{annotation\_dictionary}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} e.g. \{“sample\_name”: “NIH 3T3 cell”\}

\end{description}\end{quote}

\end{fulllineitems}

\index{addTFinfo\_dictionary() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.addTFinfo_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addTFinfo\_dictionary}}}{\emph{TFdict}}{}
Add a new TF info to pre-exiting TFdict.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} python dictionary of TF info.

\end{description}\end{quote}

\end{fulllineitems}

\index{addTFinfo\_matrix() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.addTFinfo_matrix}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addTFinfo\_matrix}}}{\emph{TFinfo\_matrix}}{}
Load TF info dataframe.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFinfo}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.DataFrame}}) \textendash{} information about potential regulatory TFs.

\end{description}\end{quote}

\end{fulllineitems}

\index{copy() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.copy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{copy}}}{}{}
Deepcopy itself

\end{fulllineitems}

\index{fit\_All\_genes() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.fit_All_genes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_All\_genes}}}{\emph{bagging\_number=200}, \emph{scaling=True}, \emph{model\_method='bagging\_ridge'}, \emph{command\_line\_mode=False}, \emph{log=None}, \emph{alpha=1}, \emph{verbose=True}}{}
Make ML models for all genes.
The calculation will be performed in parallel using scikit-learn bagging function.
You can select a modeling method (bagging\_ridge or bayesian\_ridge).  The calculation usually takes a long time.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of estimators for bagging.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scaling}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to scale regulatory gene expression values.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{model\_method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} ML model name. “bagging\_ridge” or “bayesian\_ridge”

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{command\_line\_mode}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Please select False if the calculation is performed on jupyter notebook.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log}} (\sphinxstyleliteralemphasis{\sphinxupquote{logging object}}) \textendash{} log object to output log

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Strength of regularization.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{fit\_All\_genes\_parallel() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.fit_All_genes_parallel}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_All\_genes\_parallel}}}{\emph{bagging\_number=200}, \emph{scaling=True}, \emph{log=None}, \emph{verbose=10}}{}
!! This function has some bug now and currently unavailable.

Make ML models for all genes.
The calculation will be performed in parallel using joblib parallel module.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of estimators for bagging.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scaling}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to scale regulatory gene expression values.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log}} (\sphinxstyleliteralemphasis{\sphinxupquote{logging object}}) \textendash{} log object to output log

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} verbose for joblib parallel

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{fit\_genes() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.fit_genes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_genes}}}{\emph{target\_genes}, \emph{bagging\_number=200}, \emph{scaling=True}, \emph{model\_method='bagging\_ridge'}, \emph{save\_coefs=False}, \emph{command\_line\_mode=False}, \emph{log=None}, \emph{alpha=1}, \emph{verbose=True}}{}
Make ML models for genes of interest.
The calculation will be performed in parallel using scikit-learn bagging function.
You can select a modeling method (bagging\_ridge or bayesian\_ridge).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{target\_genes}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of estimators for bagging.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scaling}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to scale regulatory gene expression values.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{model\_method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} ML model name. “bagging\_ridge” or “bayesian\_ridge”

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save\_coefs}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to store details of coef values in bagging model.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{command\_line\_mode}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Please select False if the calculation is performed on jupyter notebook.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log}} (\sphinxstyleliteralemphasis{\sphinxupquote{logging object}}) \textendash{} log object to output log

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Strength of regularization.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plotCoefs() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.plotCoefs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plotCoefs}}}{\emph{target\_gene}, \emph{sort=True}, \emph{threshold\_p=None}}{}
Plot the distribution of Coef values (network edge weights).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{target\_gene}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} gene name

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sort}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to sort genes by its strength

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of estimators for bagging.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{threshold\_p}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} the threshold for p-values. TFs will be filtered based on the p-value.
if None, no filtering is applied.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{file\_path}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.net’

\end{description}\end{quote}

\end{fulllineitems}

\index{updateLinkList() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.updateLinkList}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{updateLinkList}}}{\emph{verbose=True}}{}
Update linkList.
LinkList is a data frame that store information about inferred GRN.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar

\end{description}\end{quote}

\end{fulllineitems}

\index{updateTFinfo\_dictionary() (celloracle.Net method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.updateTFinfo_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{updateTFinfo\_dictionary}}}{\emph{TFdict}}{}
Update TF info matrix
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} A python dictionary in which a key is Target gene, value are potential regulatory genes for the target gene.

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{load\_hdf5() (in module celloracle)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.load_hdf5}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.}}\sphinxbfcode{\sphinxupquote{load\_hdf5}}}{\emph{file\_path}, \emph{object\_class\_name=None}}{}
Load an object of celloracle’s custom class that was saved as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file\_path.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{object\_class\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Types of object.
If it is None, object class will be identified from the extension of file\_name.
Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}



\subsubsection{Modules for ATAC-seq analysis}
\label{\detokenize{modules/celloracle:modules-for-atac-seq-analysis}}\begin{quote}


\paragraph{celloracle.motif\_analysis module}
\label{\detokenize{modules/celloracle.motif_analysis:module-celloracle.motif_analysis}}\label{\detokenize{modules/celloracle.motif_analysis:celloracle-motif-analysis-module}}\label{\detokenize{modules/celloracle.motif_analysis::doc}}\index{celloracle.motif\_analysis (module)}
The {\hyperref[\detokenize{modules/celloracle.motif_analysis:module-celloracle.motif_analysis}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{motif\_analysis}}}}} module implements transcription factor motif scan.

Genomic activity information (peak of ATAC-seq or Chip-seq) is extracted first.
Then the peak DNA sequence will be subjected to TF motif scan.
Finally we will get list of TFs that potentially binds to a specific gene.
\index{is\_genome\_installed() (in module celloracle.motif\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.is_genome_installed}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{is\_genome\_installed}}}{\emph{ref\_genome}}{}
Celloracle motif\_analysis module uses gimmemotifs and genomepy internally.
Reference genome files should be installed in the PC to use gimmemotifs and genomepy.
This function checks the installation status of the reference genome.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ref\_genome}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} names of reference genome. i.e., “mm10”, “hg19”

\end{description}\end{quote}

\end{fulllineitems}

\index{peak2fasta() (in module celloracle.motif\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.peak2fasta}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{peak2fasta}}}{\emph{peak\_ids}, \emph{ref\_genome}}{}
Convert peak\_id into fasta object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peak\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{ or }}\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} Peak\_id.  e.g. “chr5\_0930303\_9499409”
or it can be a list of peak\_id.  e.g. {[}“chr5\_0930303\_9499409”, “chr11\_123445555\_123445577”{]}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ref\_genome}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Reference genome name.   e.g. “mm9”, “mm10”, “hg19” etc

\end{itemize}

\item[{Returns}] \leavevmode
DNA sequence in fasta format

\item[{Return type}] \leavevmode
gimmemotifs fasta object

\end{description}\end{quote}

\end{fulllineitems}

\index{read\_bed() (in module celloracle.motif\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.read_bed}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{read\_bed}}}{\emph{bed\_path}}{}
Load bed file and return as dataframe.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{bed\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} File path.

\item[{Returns}] \leavevmode
bed file in dataframe.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_TFinfo\_from\_parquets() (in module celloracle.motif\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.load_TFinfo_from_parquets}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{load\_TFinfo\_from\_parquets}}}{\emph{folder\_path}}{}
Load TFinfo object which was saved with the function; “save\_as\_parquet”.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{folder\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} folder path

\item[{Returns}] \leavevmode
Loaded TFinfo object.

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo}]{\sphinxcrossref{TFinfo}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{make\_TFinfo\_from\_scanned\_file() (in module celloracle.motif\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.make_TFinfo_from_scanned_file}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{make\_TFinfo\_from\_scanned\_file}}}{\emph{path\_to\_raw\_bed}, \emph{path\_to\_scanned\_result\_bed}, \emph{ref\_genome}}{}
This function is currently an available.

\end{fulllineitems}

\index{TFinfo (class in celloracle.motif\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{TFinfo}}}{\emph{peak\_data\_frame}, \emph{ref\_genome}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

This is a custom class for motif analysis in celloracle.
TFinfo object performs motif scan using the TF motif database in gimmemotifs and several functions of genomepy.
Analysis results can be exported as a python dictionary or dataframe.
These files; python dictionary of dataframe of TF binding information, are needed in GRN inference.
\index{peak\_df (celloracle.motif\_analysis.TFinfo attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.peak_df}}\pysigline{\sphinxbfcode{\sphinxupquote{peak\_df}}}
\sphinxstyleemphasis{pandas.dataframe} \textendash{} dataframe about DNA peak and target gene data.

\end{fulllineitems}

\index{all\_target\_gene (celloracle.motif\_analysis.TFinfo attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.all_target_gene}}\pysigline{\sphinxbfcode{\sphinxupquote{all\_target\_gene}}}
\sphinxstyleemphasis{array of str} \textendash{} target genes.

\end{fulllineitems}

\index{ref\_genome (celloracle.motif\_analysis.TFinfo attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.ref_genome}}\pysigline{\sphinxbfcode{\sphinxupquote{ref\_genome}}}
\sphinxstyleemphasis{str} \textendash{} reference genome name that was used in DNA peak generation.

\end{fulllineitems}

\index{scanned\_df (celloracle.motif\_analysis.TFinfo attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.scanned_df}}\pysigline{\sphinxbfcode{\sphinxupquote{scanned\_df}}}
\sphinxstyleemphasis{dictionary} \textendash{} Results of motif scan. Key is a peak name. Value is a dataframe of motif scan.

\end{fulllineitems}

\index{dic\_targetgene2TFs (celloracle.motif\_analysis.TFinfo attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.dic_targetgene2TFs}}\pysigline{\sphinxbfcode{\sphinxupquote{dic\_targetgene2TFs}}}
\sphinxstyleemphasis{dictionary} \textendash{} Final product of Motif scan. Key is a target gene. Value is a list of regulatory candidate genes.

\end{fulllineitems}

\index{dic\_peak2Targetgene (celloracle.motif\_analysis.TFinfo attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.dic_peak2Targetgene}}\pysigline{\sphinxbfcode{\sphinxupquote{dic\_peak2Targetgene}}}
\sphinxstyleemphasis{dictionary} \textendash{} Dictionary. Key is a peak name. Value is a list of the target gene.

\end{fulllineitems}

\index{dic\_TF2targetgenes (celloracle.motif\_analysis.TFinfo attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.dic_TF2targetgenes}}\pysigline{\sphinxbfcode{\sphinxupquote{dic\_TF2targetgenes}}}
\sphinxstyleemphasis{dictionary} \textendash{} Final product of Motif scan. Key is a TF. Value is a list of potential target genes of the TF.

\end{fulllineitems}

\index{copy() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.copy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{copy}}}{}{}
Deepcoty itself.

\end{fulllineitems}

\index{filter\_motifs\_by\_score() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.filter_motifs_by_score}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{filter\_motifs\_by\_score}}}{\emph{threshold}, \emph{method='cumlative\_score'}}{}
Remove motifs with low binding scores.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} thresholding method. Select either of {[}“indivisual\_score”, “cumlative\_score”{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{filter\_peaks() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.filter_peaks}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{filter\_peaks}}}{\emph{peaks\_to\_be\_remained}}{}
Filter peaks.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{peaks\_to\_be\_remained}} (\sphinxstyleliteralemphasis{\sphinxupquote{array of str}}) \textendash{} list of peaks. Peaks that are NOT in the list will be removed.

\end{description}\end{quote}

\end{fulllineitems}

\index{make\_TFinfo\_dataframe\_and\_dictionary() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.make_TFinfo_dataframe_and_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{make\_TFinfo\_dataframe\_and\_dictionary}}}{\emph{verbose=True}}{}
This is the final step of motif\_analysis.
Convert scanned results into a data frame and dictionaries.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar.

\end{description}\end{quote}

\end{fulllineitems}

\index{reset\_dictionary\_and\_df() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.reset_dictionary_and_df}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{reset\_dictionary\_and\_df}}}{}{}
Reset TF dictionary and TF dataframe.
The following attributes will be erased; TF\_onehot, dic\_targetgene2TFs, dic\_peak2Targetgene, dic\_TF2targetgenes.

\end{fulllineitems}

\index{reset\_filtering() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.reset_filtering}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{reset\_filtering}}}{}{}
Reset filtering information.
You can use this function to stat over the filtering step with new conditions.
The following attributes will be erased; TF\_onehot, dic\_targetgene2TFs, dic\_peak2Targetgene, dic\_TF2targetgenes.

\end{fulllineitems}

\index{save\_as\_parquet() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.save_as_parquet}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{save\_as\_parquet}}}{\emph{folder\_path=None}}{}
Save itself. Some attributes are saved as parquet file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{folder\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} folder path

\end{description}\end{quote}

\end{fulllineitems}

\index{scan() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.scan}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{scan}}}{\emph{background\_length=200}, \emph{fpr=0.02}, \emph{n\_cpus=-1}, \emph{verbose=True}}{}
Scan DNA sequences searching for TF binding motifs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{background\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} background length. This is used for the calculation of the binding score.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{fpr}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} False positive rate for motif identification.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_cpus}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} number of CPUs for parallel calculation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_dataframe() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.to_dataframe}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_dataframe}}}{\emph{verbose=True}}{}
Return results as a dataframe.
Rows are peak\_id, and columns are TFs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar.

\item[{Returns}] \leavevmode
TFinfo matrix.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_dictionary() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.to_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_dictionary}}}{\emph{dictionary\_type='targetgene2TFs'}, \emph{verbose=True}}{}
Return TF information as a python dictionary.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{dictionary\_type}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Type of dictionary. Select from {[}“targetgene2TFs”, “TF2targetgenes”{]}.
If you chose “targetgene2TFs”, it returns a dictionary in which a key is a target gene, and a value is a list of regulatory candidate genes (TFs) of the target.
If you chose “TF2targetgenes”, it returns a dictionary in which a key is a TF and a value is a list of potential target genes of the TF.

\item[{Returns}] \leavevmode
dictionary.

\item[{Return type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.motif\_analysis.TFinfo method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{file\_path}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.tfinfo’

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{get\_tss\_info() (in module celloracle.motif\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.get_tss_info}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{get\_tss\_info}}}{\emph{peak\_str\_list}, \emph{ref\_genome}, \emph{verbose=True}}{}
Get annotation about Transcription Starting Site (TSS).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peak\_str\_list}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} list of peak\_id. e.g., {[}“chr5\_0930303\_9499409”, “chr11\_123445555\_123445577”{]}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ref\_genome}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} reference genome name.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} verbosity.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{integrate\_tss\_peak\_with\_cicero() (in module celloracle.motif\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.integrate_tss_peak_with_cicero}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{integrate\_tss\_peak\_with\_cicero}}}{\emph{tss\_peak}, \emph{cicero\_connections}}{}
Process output of cicero data and returns DNA peak information for motif analysis in celloracle.
Please see the tutorial of celloracle documentation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tss\_peak}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.dataframe}}) \textendash{} dataframe about TSS information. Please use the function, “get\_tss\_info” to get this dataframe.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cicero\_connections}} (\sphinxstyleliteralemphasis{\sphinxupquote{dataframe}}) \textendash{} dataframe that stores the results of cicero analysis.

\end{itemize}

\item[{Returns}] \leavevmode
DNA peak about promoter/enhancer and its annotation about target gene.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}


\subsubsection{Modules for Network analysis}
\label{\detokenize{modules/celloracle:modules-for-network-analysis}}\begin{quote}


\paragraph{celloracle.network\_analysis module}
\label{\detokenize{modules/celloracle.network_analysis:module-celloracle.network_analysis}}\label{\detokenize{modules/celloracle.network_analysis:celloracle-network-analysis-module}}\label{\detokenize{modules/celloracle.network_analysis::doc}}\index{celloracle.network\_analysis (module)}
The {\hyperref[\detokenize{modules/celloracle.network_analysis:module-celloracle.network_analysis}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{network\_analysis}}}}} module implements Network analysis.
\index{get\_links() (in module celloracle.network\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.get_links}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{get\_links}}}{\emph{oracle\_object}, \emph{cluster\_name\_for\_GRN\_unit=None}, \emph{alpha=10}, \emph{bagging\_number=20}, \emph{verbose\_level=1}, \emph{test\_mode=False}}{}
Make GRN for each cluster and returns results as a Links object.
Several preprocessing should be done before using this function.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{oracle\_object}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Oracle}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Oracle}}}}}) \textendash{} See Oracle module for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_name\_for\_GRN\_unit}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name for GRN calculation. The cluster information should be stored in Oracle.adata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{ or }}\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the strength of regularization.
If you set a lower value, the sensitivity increase, and you can detect a weak network connection, but it might get more noize.
With a higher value of alpha may reduce the chance of overfitting.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number for bagging calculation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose\_level}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} if {[}verbose\_level\textgreater{}1{]}, most detailed progress information will be shown.
if {[}verbose\_level \textgreater{} 0{]}, one progress bar will be shown.
if {[}verbose\_level == 0{]}, no progress bar will be shown.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{test\_mode}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} If test\_mode is True, GRN calculation will be done for only one cluster rather than all clusters.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{test\_R\_libraries\_installation() (in module celloracle.network\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.test_R_libraries_installation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{test\_R\_libraries\_installation}}}{}{}
CellOracle.network\_analysis use some R libraries for network analysis.
This is a test function to check instalation of necessary R libraries.

\end{fulllineitems}

\index{load\_links() (in module celloracle.network\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.load_links}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{load\_links}}}{\emph{file\_path}}{}
Load links object saved as a hdf5 file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path.

\item[{Returns}] \leavevmode
loaded links object.

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{Links}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{Links (class in celloracle.network\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{Links}}}{\emph{name}, \emph{links\_dict=\{\}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

This is a class for the processing and visualization of GRNs.
Links object stores cluster-specific GRNs and metadata.
Please use “get\_links” function in Oracle object to generate Links object.
\index{links\_dict (celloracle.network\_analysis.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.links_dict}}\pysigline{\sphinxbfcode{\sphinxupquote{links\_dict}}}
\sphinxstyleemphasis{dictionary} \textendash{} Dictionary that store unprocessed network data.

\end{fulllineitems}

\index{filtered\_links (celloracle.network\_analysis.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.filtered_links}}\pysigline{\sphinxbfcode{\sphinxupquote{filtered\_links}}}
\sphinxstyleemphasis{dictionary} \textendash{} Dictionary that store filtered network data.

\end{fulllineitems}

\index{merged\_score (celloracle.network\_analysis.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.merged_score}}\pysigline{\sphinxbfcode{\sphinxupquote{merged\_score}}}
\sphinxstyleemphasis{pandas.dataframe} \textendash{} Network scores.

\end{fulllineitems}

\index{cluster (celloracle.network\_analysis.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.cluster}}\pysigline{\sphinxbfcode{\sphinxupquote{cluster}}}
\sphinxstyleemphasis{list of str} \textendash{} List of cluster name.

\end{fulllineitems}

\index{name (celloracle.network\_analysis.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.name}}\pysigline{\sphinxbfcode{\sphinxupquote{name}}}
\sphinxstyleemphasis{str} \textendash{} Name of clustering unit.

\end{fulllineitems}

\index{palette (celloracle.network\_analysis.Links attribute)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.palette}}\pysigline{\sphinxbfcode{\sphinxupquote{palette}}}
\sphinxstyleemphasis{pandas.dataframe} \textendash{} DataFrame that store color information.

\end{fulllineitems}

\index{add\_palette() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.add_palette}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_palette}}}{\emph{cluster\_name\_list}, \emph{color\_list}}{}
Add color information for each cluster.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_name\_list}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} cluster names.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{color\_list}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of stf}}) \textendash{} list of color for each cluster.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{filter\_links() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.filter_links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{filter\_links}}}{\emph{p=0.001}, \emph{weight='coef\_abs'}, \emph{thread\_number=10000}, \emph{genelist\_source=None}, \emph{genelist\_target=None}}{}
Filter network edges.
In most case inferred GRN has non-significant random edges.
We have to remove such random edges before analyzing network structure.
You can do the filtering either of three ways below.
\begin{enumerate}
\item {} 
Do filtering based on the p-value of the network edge.
Please enter p-value for thresholding.

\item {} 
Do filtering based on network edge number.
If you set the number, network edges will be filtered based on the order of a network score. The top n-th network edges with network weight will remain, and the other edges will be removed.
The network data have several types of network weight, so you have to select which network weight do you want to use.

\item {} 
Do filtering based on an arbitrary gene list. You can set a gene list for source nodes or target nodes.

\end{enumerate}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{p}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} threshold for p-value of the network edge.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{weight}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Please select network weight name for the filtering

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{genelist\_source}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list to remain in regulatory gene nodes. Default is None.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{genelist\_target}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list to remain in target gene nodes. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_network\_entropy() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.get_network_entropy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_network\_entropy}}}{\emph{value='coef\_abs'}}{}
Calculate network entropy scores.
Please select a type of network weight for the entropy calculation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{value}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Default is “coef\_abs”.

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_score() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.get_score}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_score}}}{\emph{test\_mode=False}}{}
Get several network sores using R libraries.
Make sure all dependant R libraries are installed in your environment before running this function.
You can check the installation for the R libraries by running test\_installation() in network\_analysis module.

\end{fulllineitems}

\index{plot\_cartography\_scatter\_per\_cluster() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_cartography_scatter_per_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cartography\_scatter\_per\_cluster}}}{\emph{gois=None}, \emph{clusters=None}, \emph{scatter=False}, \emph{kde=True}, \emph{auto\_gene\_annot=False}, \emph{percentile=98}, \emph{args\_dot=\{\}}, \emph{args\_line=\{\}}, \emph{args\_annot=\{\}}, \emph{save=None}}{}
Make a plot of gene network cartography.
Please read the original paper of gene network cartography for the principle of gene network cartography.
\sphinxurl{https://www.nature.com/articles/nature03288}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gois}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of srt}}) \textendash{} List of Gene name to highlight.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{clusters}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} List of cluster name to analyze. If None, all clusters in Links object will be analyzed.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scatter}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to make a scatter plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{auto\_gene\_annot}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to pick up genes to make an annotation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{percentile}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Genes with a network score above the percentile will be shown with annotation. Default is 98.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_dot}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for scatter plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_line}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for lines in cartography plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_annot}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for annotation in plots.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_cartography\_term() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_cartography_term}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cartography\_term}}}{\emph{goi}, \emph{save=None}}{}
Plot the summary of gene network cartography like a heatmap.
Please read the original paper of gene network cartography for the principle of gene network cartography.
\sphinxurl{https://www.nature.com/articles/nature03288}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gois}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of srt}}) \textendash{} List of Gene name to highlight.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_degree\_distributions() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_degree_distributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_degree\_distributions}}}{\emph{plot\_model=False}, \emph{save=None}}{}
Plot the distribution of network degree (the number of edge per gene).
The network degree will be visualized in both linear scale and log scale.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plot\_model}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to plot linear approximation line.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_network\_entropy\_distributions() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_network_entropy_distributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_network\_entropy\_distributions}}}{\emph{update\_network\_entropy=False}, \emph{save=None}}{}
Plot the distribution of network entropy.
See the CellOracle paper for the detail.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} (\sphinxstyleliteralemphasis{\sphinxupquote{Links object}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{values}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The list of score to visualize. If it is None, all network score (listed above) will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{update\_network\_entropy}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to recalculate network entropy.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_comparison\_2D() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_score_comparison_2D}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_comparison\_2D}}}{\emph{value}, \emph{cluster1}, \emph{cluster2}, \emph{percentile=99}, \emph{annot\_shifts=None}, \emph{save=None}}{}
Make a scatter plot that shows the relationship of a specific network score in two groups.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{value}} (\sphinxstyleliteralemphasis{\sphinxupquote{srt}}) \textendash{} The network score type.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster1}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name. Network scores in the cluster1 will be visualized in the x-axis.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster2}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name. Network scores in the cluster2 will be visualized in the y-axis.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{percentile}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Genes with a network score above the percentile will be shown with annotation. Default is 99.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{annot\_shifts}} (\sphinxstyleliteralemphasis{\sphinxupquote{(}}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{)}}) \textendash{} Annotation visualization setting.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_discributions() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_score_discributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_discributions}}}{\emph{values=None}, \emph{method='boxplot'}, \emph{save=None}}{}
Plot the distribution of network scores.
An individual data point is a network edge (gene) of GRN in each cluster.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{values}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The list of score to visualize. If it is None, all network score will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Plotting method. Select either of “boxplot” or “barplot”.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_per\_cluster() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_score_per_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_per\_cluster}}}{\emph{goi}, \emph{save=None}}{}
Plot network score for a gene.
This function visualizes the network score of a specific gene between clusters to get an insight into the dynamics of the gene.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{goi}} (\sphinxstyleliteralemphasis{\sphinxupquote{srt}}) \textendash{} Gene name.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_scores\_as\_rank() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_scores_as_rank}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_scores\_as\_rank}}}{\emph{cluster}, \emph{n\_gene=50}, \emph{save=None}}{}
Pick up top n-th genes wich high-network scores and make plots.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analisis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster nome to analyze.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_gene}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Number of genes to plot. Default is 50.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.network\_analysis.Links method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{file\_path}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.links’

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{transfer\_scores\_from\_links\_to\_adata() (in module celloracle.network\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.transfer_scores_from_links_to_adata}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{transfer\_scores\_from\_links\_to\_adata}}}{\emph{adata}, \emph{links}, \emph{method='median'}}{}
Transfer the summary of network scores (median or mean) per group from Links object into adata.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{adata}} (\sphinxstyleliteralemphasis{\sphinxupquote{anndata}}) \textendash{} anndata

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} likns object

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The method to summarize data.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{linkList\_to\_networkgraph() (in module celloracle.network\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.linkList_to_networkgraph}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{linkList\_to\_networkgraph}}}{\emph{filteredlinkList}}{}
Convert linkList into Graph object in NetworkX.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{filteredlinkList}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.DataFrame}}) \textendash{} GRN saved as linkList.

\item[{Returns}] \leavevmode
Network X graph objenct.

\item[{Return type}] \leavevmode
Graph object

\end{description}\end{quote}

\end{fulllineitems}

\index{draw\_network() (in module celloracle.network\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.draw_network}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{draw\_network}}}{\emph{linkList}, \emph{return\_graph=False}}{}
Plot network graph.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{linkList}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.DataFrame}}) \textendash{} GRN saved as linkList.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{return\_graph}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to return graph object.

\end{itemize}

\item[{Returns}] \leavevmode
Network X graph objenct.

\item[{Return type}] \leavevmode
Graph object

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}


\subsubsection{Other modules}
\label{\detokenize{modules/celloracle:other-modules}}\begin{quote}


\paragraph{celloracle.go\_analysis module}
\label{\detokenize{modules/celloracle.go_analysis:celloracle-go-analysis-module}}\label{\detokenize{modules/celloracle.go_analysis::doc}}\begin{quote}
\phantomsection\label{\detokenize{modules/celloracle.go_analysis:module-celloracle.go_analysis}}\index{celloracle.go\_analysis (module)}
The {\hyperref[\detokenize{modules/celloracle.go_analysis:module-celloracle.go_analysis}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{go\_analysis}}}}} module implements Gene Ontology analysis.
This module use goatools internally.
\index{geneSymbol2ID() (in module celloracle.go\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.go_analysis:celloracle.go_analysis.geneSymbol2ID}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.go\_analysis.}}\sphinxbfcode{\sphinxupquote{geneSymbol2ID}}}{\emph{symbols}, \emph{species='mouse'}}{}
Convert gene symbol into Entrez gene id.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{symbols}} (\sphinxstyleliteralemphasis{\sphinxupquote{array of str}}) \textendash{} gene symbol

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{species}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Select species. Either “mouse” or “human”

\end{itemize}

\item[{Returns}] \leavevmode
Entrez gene id

\item[{Return type}] \leavevmode
list of str

\end{description}\end{quote}

\end{fulllineitems}

\index{geneID2Symbol() (in module celloracle.go\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.go_analysis:celloracle.go_analysis.geneID2Symbol}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.go\_analysis.}}\sphinxbfcode{\sphinxupquote{geneID2Symbol}}}{\emph{IDs}, \emph{species='mouse'}}{}
Convert Entrez gene id into gene symbol.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IDs}} (\sphinxstyleliteralemphasis{\sphinxupquote{array of str}}) \textendash{} Entrez gene id.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{species}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Select species. Either “mouse” or “human”.

\end{itemize}

\item[{Returns}] \leavevmode
Gene symbol

\item[{Return type}] \leavevmode
list of str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_GO() (in module celloracle.go\_analysis)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.go_analysis:celloracle.go_analysis.get_GO}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.go\_analysis.}}\sphinxbfcode{\sphinxupquote{get\_GO}}}{\emph{gene\_query}, \emph{species='mouse'}}{}
Get Gene Ontologies (GOs).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_query}} (\sphinxstyleliteralemphasis{\sphinxupquote{array of str}}) \textendash{} gene list.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{species}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Select species. Either “mouse” or “human”

\end{itemize}

\item[{Returns}] \leavevmode
GO analysis results as dataframe.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}


\paragraph{celloracle.utility module}
\label{\detokenize{modules/celloracle.utility:celloracle-utility-module}}\label{\detokenize{modules/celloracle.utility::doc}}\begin{quote}
\phantomsection\label{\detokenize{modules/celloracle.utility:module-celloracle.utility}}\index{celloracle.utility (module)}
The {\hyperref[\detokenize{modules/celloracle.utility:module-celloracle.utility}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{utility}}}}} module has several functions that support celloracle.
\index{makelog (class in celloracle.utility)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.makelog}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{makelog}}}{\emph{file\_name=None}, \emph{directory=None}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

This is a class for making log.
\index{info() (celloracle.utility.makelog method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.makelog.info}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{info}}}{\emph{comment}}{}
Add comment into the log file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{comment}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} comment.

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{save\_as\_pickled\_object() (in module celloracle.utility)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.save_as_pickled_object}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{save\_as\_pickled\_object}}}{\emph{obj}, \emph{filepath}}{}
Save any object using pickle.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{obj}} (\sphinxstyleliteralemphasis{\sphinxupquote{any python object}}) \textendash{} python object.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{filepath}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_pickled\_object() (in module celloracle.utility)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.load_pickled_object}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{load\_pickled\_object}}}{\emph{filepath}}{}
Load pickled object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{filepath}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path.

\item[{Returns}] \leavevmode
loaded object.

\item[{Return type}] \leavevmode
python object

\end{description}\end{quote}

\end{fulllineitems}

\index{intersect() (in module celloracle.utility)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.intersect}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{intersect}}}{\emph{list1}, \emph{list2}}{}
Intersect two list and get components that exists in both list.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{list1}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} input list.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{list2}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} input list.

\end{itemize}

\item[{Returns}] \leavevmode
intersected list.

\item[{Return type}] \leavevmode
list

\end{description}\end{quote}

\end{fulllineitems}

\index{exec\_process() (in module celloracle.utility)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.exec_process}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{exec\_process}}}{\emph{commands}, \emph{message=True}, \emph{wait\_finished=True}, \emph{return\_process=True}}{}
Excute a command. This is a wrapper of “subprocess.Popen”
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{commands}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} command.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{message}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to return a message or not.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{wait\_finished}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to wait for the process finished. If False, the function finish immediately.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{return\_process}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to return “process”.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{standard() (in module celloracle.utility)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.standard}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{standard}}}{\emph{df}}{}
Standerdize value.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{df}} (\sphinxstyleliteralemphasis{\sphinxupquote{padas.dataframe}}) \textendash{} dataframe.

\item[{Returns}] \leavevmode
data after standerdization.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_hdf5() (in module celloracle.utility)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.load_hdf5}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{load\_hdf5}}}{\emph{file\_path}, \emph{object\_class\_name=None}}{}
Load an object of celloracle’s custom class that was saved as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file\_path.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{object\_class\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Types of object.
If it is None, object class will be identified from the extension of file\_name.
Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{inverse\_dictionary() (in module celloracle.utility)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.inverse_dictionary}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{inverse\_dictionary}}}{\emph{dictionary}, \emph{verbose=True}, \emph{return\_value\_as\_numpy=False}}{}
Make inversed dictionary.
See examples below for detail.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{dictionary}} (\sphinxstyleliteralemphasis{\sphinxupquote{dict}}) \textendash{} python dictionary

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show progress bar.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{return\_value\_as\_numpy}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to convert values into numpy array.

\end{itemize}

\item[{Returns}] \leavevmode
Python dictionary.

\item[{Return type}] \leavevmode
dict

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{dic} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{a}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{b}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{]}\PYG{p}{\PYGZcb{}}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{inverse\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{dic}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}1: [\PYGZsq{}a\PYGZsq{}], 2: [\PYGZsq{}a\PYGZsq{}, \PYGZsq{}b\PYGZsq{}], 3: [\PYGZsq{}a\PYGZsq{}, \PYGZsq{}b\PYGZsq{}], 4: [\PYGZsq{}b\PYGZsq{}]\PYGZcb{}}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{dic} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{a}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{b}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{]}\PYG{p}{\PYGZcb{}}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{inverse\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{dic}\PYG{p}{,} \PYG{n}{return\PYGZus{}value\PYGZus{}as\PYGZus{}numpy}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}1: array([\PYGZsq{}a\PYGZsq{}], dtype=\PYGZsq{}\PYGZlt{}U1\PYGZsq{}),}
\PYG{g+go}{ 2: array([\PYGZsq{}a\PYGZsq{}, \PYGZsq{}b\PYGZsq{}], dtype=\PYGZsq{}\PYGZlt{}U1\PYGZsq{}),}
\PYG{g+go}{ 3: array([\PYGZsq{}a\PYGZsq{}, \PYGZsq{}b\PYGZsq{}], dtype=\PYGZsq{}\PYGZlt{}U1\PYGZsq{}),}
\PYG{g+go}{ 4: array([\PYGZsq{}b\PYGZsq{}], dtype=\PYGZsq{}\PYGZlt{}U1\PYGZsq{})\PYGZcb{}}
\end{sphinxVerbatim}

\end{fulllineitems}

\end{quote}


\paragraph{celloracle.data module}
\label{\detokenize{modules/celloracle.data:celloracle-data-module}}\label{\detokenize{modules/celloracle.data::doc}}\begin{quote}
\phantomsection\label{\detokenize{modules/celloracle.data:module-celloracle.data}}\index{celloracle.data (module)}
The {\hyperref[\detokenize{modules/celloracle.data:module-celloracle.data}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{data}}}}} module implements data download and loading.
\index{load\_TFinfo\_df\_mm9\_mouse\_atac\_atlas() (in module celloracle.data)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.data:celloracle.data.load_TFinfo_df_mm9_mouse_atac_atlas}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.data.}}\sphinxbfcode{\sphinxupquote{load\_TFinfo\_df\_mm9\_mouse\_atac\_atlas}}}{}{}
Load Transcription factor binding information made from mouse scATAC-seq atlas dataset.
mm9 genome was used for the reference genome.

Args:
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
TF binding info.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}


\paragraph{celloracle.data\_conversion module}
\label{\detokenize{modules/celloracle.data_conversion:celloracle-data-conversion-module}}\label{\detokenize{modules/celloracle.data_conversion::doc}}\begin{quote}
\phantomsection\label{\detokenize{modules/celloracle.data_conversion:module-celloracle.data_conversion}}\index{celloracle.data\_conversion (module)}
The {\hyperref[\detokenize{modules/celloracle.data_conversion:module-celloracle.data_conversion}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{data\_conversion}}}}} module implements data conversion between different platform.
\index{seurat\_object\_to\_anndata() (in module celloracle.data\_conversion)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.data_conversion:celloracle.data_conversion.seurat_object_to_anndata}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.data\_conversion.}}\sphinxbfcode{\sphinxupquote{seurat\_object\_to\_anndata}}}{\emph{file\_path\_seurat\_object}, \emph{delete\_tmp\_file=True}}{}
Convert seurat object into anndata.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{file\_path\_seurat\_object}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} File path of seurat object. Seurat object should be saved as Rds format.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{delete\_tmp\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to delete temporary file.

\end{itemize}

\item[{Returns}] \leavevmode
anndata object.

\item[{Return type}] \leavevmode
anndata

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}
\end{quote}


\section{Changelog}
\label{\detokenize{changelog:changelog}}\label{\detokenize{changelog:id1}}\label{\detokenize{changelog::doc}}

\section{License}
\label{\detokenize{license/index:license}}\label{\detokenize{license/index:id1}}\label{\detokenize{license/index::doc}}
The software is provided under xxxxxxx License.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{License} \PYG{n}{xxxxxxxx}

\PYG{n}{Copyright} \PYG{p}{(}\PYG{n}{c}\PYG{p}{)} \PYG{l+m+mi}{2019} \PYG{n}{Kenji} \PYG{n}{Kamimoto}\PYG{p}{,} \PYG{n}{Christy} \PYG{n}{Hoffmann}\PYG{p}{,} \PYG{n}{Samantha} \PYG{n}{Morris}


\PYG{n}{This} \PYG{o+ow}{is} \PYG{n}{a} \PYG{n}{placeholder}\PYG{o}{.} \PYG{n}{License} \PYG{n}{description} \PYG{n}{will} \PYG{n}{be} \PYG{n}{shown} \PYG{n}{here}\PYG{o}{.}
\end{sphinxVerbatim}


\section{Authors and citations}
\label{\detokenize{citation/index:authors-and-citations}}\label{\detokenize{citation/index:citing}}\label{\detokenize{citation/index::doc}}

\subsection{Cite celloracle}
\label{\detokenize{citation/index:cite-celloracle}}
If you use velocyto please cite our bioarxiv preprint \sphinxcode{\sphinxupquote{TITLE}}.


\subsection{celloracle software development}
\label{\detokenize{citation/index:celloracle-software-development}}
celloracle is developed and maintained by the effort of celloracle team in Samantha Morris Lab.
Please post troubles or questions on  \sphinxhref{https://cole-trapnell-lab.github.io/cicero-release/}{the Github repository}


\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{c}
\item {\sphinxstyleindexentry{celloracle}}\sphinxstyleindexpageref{modules/celloracle:\detokenize{module-celloracle}}
\item {\sphinxstyleindexentry{celloracle.data}}\sphinxstyleindexpageref{modules/celloracle.data:\detokenize{module-celloracle.data}}
\item {\sphinxstyleindexentry{celloracle.data\_conversion}}\sphinxstyleindexpageref{modules/celloracle.data_conversion:\detokenize{module-celloracle.data_conversion}}
\item {\sphinxstyleindexentry{celloracle.go\_analysis}}\sphinxstyleindexpageref{modules/celloracle.go_analysis:\detokenize{module-celloracle.go_analysis}}
\item {\sphinxstyleindexentry{celloracle.motif\_analysis}}\sphinxstyleindexpageref{modules/celloracle.motif_analysis:\detokenize{module-celloracle.motif_analysis}}
\item {\sphinxstyleindexentry{celloracle.network\_analysis}}\sphinxstyleindexpageref{modules/celloracle.network_analysis:\detokenize{module-celloracle.network_analysis}}
\item {\sphinxstyleindexentry{celloracle.utility}}\sphinxstyleindexpageref{modules/celloracle.utility:\detokenize{module-celloracle.utility}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}