%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
  \ifdefined\DeclareUnicodeCharacterAsOptional
    \def\sphinxDUC#1{\DeclareUnicodeCharacter{"#1}}
  \else
    \let\sphinxDUC\DeclareUnicodeCharacter
  \fi
  \sphinxDUC{00A0}{\nobreakspace}
  \sphinxDUC{2500}{\sphinxunichar{2500}}
  \sphinxDUC{2502}{\sphinxunichar{2502}}
  \sphinxDUC{2514}{\sphinxunichar{2514}}
  \sphinxDUC{251C}{\sphinxunichar{251C}}
  \sphinxDUC{2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}



\usepackage{times}
\expandafter\ifx\csname T@LGR\endcsname\relax
\else
% LGR was declared as font encoding
  \substitutefont{LGR}{\rmdefault}{cmr}
  \substitutefont{LGR}{\sfdefault}{cmss}
  \substitutefont{LGR}{\ttdefault}{cmtt}
\fi
\expandafter\ifx\csname T@X2\endcsname\relax
  \expandafter\ifx\csname T@T2A\endcsname\relax
  \else
  % T2A was declared as font encoding
    \substitutefont{T2A}{\rmdefault}{cmr}
    \substitutefont{T2A}{\sfdefault}{cmss}
    \substitutefont{T2A}{\ttdefault}{cmtt}
  \fi
\else
% X2 was declared as font encoding
  \substitutefont{X2}{\rmdefault}{cmr}
  \substitutefont{X2}{\sfdefault}{cmss}
  \substitutefont{X2}{\ttdefault}{cmtt}
\fi


\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\fvset{fontsize=\small}
\usepackage{geometry}


% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}


\usepackage{sphinxmessages}



% Jupyter Notebook code cell colors
\definecolor{nbsphinxin}{HTML}{307FC1}
\definecolor{nbsphinxout}{HTML}{BF5B3D}
\definecolor{nbsphinx-code-bg}{HTML}{F5F5F5}
\definecolor{nbsphinx-code-border}{HTML}{E0E0E0}
\definecolor{nbsphinx-stderr}{HTML}{FFDDDD}
% ANSI colors for output streams and traceback highlighting
\definecolor{ansi-black}{HTML}{3E424D}
\definecolor{ansi-black-intense}{HTML}{282C36}
\definecolor{ansi-red}{HTML}{E75C58}
\definecolor{ansi-red-intense}{HTML}{B22B31}
\definecolor{ansi-green}{HTML}{00A250}
\definecolor{ansi-green-intense}{HTML}{007427}
\definecolor{ansi-yellow}{HTML}{DDB62B}
\definecolor{ansi-yellow-intense}{HTML}{B27D12}
\definecolor{ansi-blue}{HTML}{208FFB}
\definecolor{ansi-blue-intense}{HTML}{0065CA}
\definecolor{ansi-magenta}{HTML}{D160C4}
\definecolor{ansi-magenta-intense}{HTML}{A03196}
\definecolor{ansi-cyan}{HTML}{60C6C8}
\definecolor{ansi-cyan-intense}{HTML}{258F8F}
\definecolor{ansi-white}{HTML}{C5C1B4}
\definecolor{ansi-white-intense}{HTML}{A1A6B2}
\definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
\definecolor{ansi-default-inverse-bg}{HTML}{000000}

% Define an environment for non-plain-text code cell outputs (e.g. images)
\makeatletter
\newenvironment{nbsphinxfancyoutput}{%
    % Avoid fatal error with framed.sty if graphics too long to fit on one page
    \let\sphinxincludegraphics\nbsphinxincludegraphics
    \nbsphinx@image@maxheight\textheight
    \advance\nbsphinx@image@maxheight -2\fboxsep   % default \fboxsep 3pt
    \advance\nbsphinx@image@maxheight -2\fboxrule  % default \fboxrule 0.4pt
    \advance\nbsphinx@image@maxheight -\baselineskip
\def\nbsphinxfcolorbox{\spx@fcolorbox{nbsphinx-code-border}{white}}%
\def\FrameCommand{\nbsphinxfcolorbox\nbsphinxfancyaddprompt\@empty}%
\def\FirstFrameCommand{\nbsphinxfcolorbox\nbsphinxfancyaddprompt\sphinxVerbatim@Continues}%
\def\MidFrameCommand{\nbsphinxfcolorbox\sphinxVerbatim@Continued\sphinxVerbatim@Continues}%
\def\LastFrameCommand{\nbsphinxfcolorbox\sphinxVerbatim@Continued\@empty}%
\MakeFramed{\advance\hsize-\width\@totalleftmargin\z@\linewidth\hsize\@setminipage}%
\lineskip=1ex\lineskiplimit=1ex\raggedright%
}{\par\unskip\@minipagefalse\endMakeFramed}
\makeatother
\newbox\nbsphinxpromptbox
\def\nbsphinxfancyaddprompt{\ifvoid\nbsphinxpromptbox\else
    \kern\fboxrule\kern\fboxsep
    \copy\nbsphinxpromptbox
    \kern-\ht\nbsphinxpromptbox\kern-\dp\nbsphinxpromptbox
    \kern-\fboxsep\kern-\fboxrule\nointerlineskip
    \fi}
\newlength\nbsphinxcodecellspacing
\setlength{\nbsphinxcodecellspacing}{0pt}

% Define support macros for attaching opening and closing lines to notebooks
\newsavebox\nbsphinxbox
\makeatletter
\newcommand{\nbsphinxstartnotebook}[1]{%
    \par
    % measure needed space
    \setbox\nbsphinxbox\vtop{{#1\par}}
    % reserve some space at bottom of page, else start new page
    \needspace{\dimexpr2.5\baselineskip+\ht\nbsphinxbox+\dp\nbsphinxbox}
    % mimick vertical spacing from \section command
      \addpenalty\@secpenalty
      \@tempskipa 3.5ex \@plus 1ex \@minus .2ex\relax
      \addvspace\@tempskipa
      {\Large\@tempskipa\baselineskip
             \advance\@tempskipa-\prevdepth
             \advance\@tempskipa-\ht\nbsphinxbox
             \ifdim\@tempskipa>\z@
               \vskip \@tempskipa
             \fi}
    \unvbox\nbsphinxbox
    % if notebook starts with a \section, prevent it from adding extra space
    \@nobreaktrue\everypar{\@nobreakfalse\everypar{}}%
    % compensate the parskip which will get inserted by next paragraph
    \nobreak\vskip-\parskip
    % do not break here
    \nobreak
}% end of \nbsphinxstartnotebook

\newcommand{\nbsphinxstopnotebook}[1]{%
    \par
    % measure needed space
    \setbox\nbsphinxbox\vbox{{#1\par}}
    \nobreak % it updates page totals
    \dimen@\pagegoal
    \advance\dimen@-\pagetotal \advance\dimen@-\pagedepth
    \advance\dimen@-\ht\nbsphinxbox \advance\dimen@-\dp\nbsphinxbox
    \ifdim\dimen@<\z@
      % little space left
      \unvbox\nbsphinxbox
      \kern-.8\baselineskip
      \nobreak\vskip\z@\@plus1fil
      \penalty100
      \vskip\z@\@plus-1fil
      \kern.8\baselineskip
    \else
      \unvbox\nbsphinxbox
    \fi
}% end of \nbsphinxstopnotebook

% Ensure height of an included graphics fits in nbsphinxfancyoutput frame
\newdimen\nbsphinx@image@maxheight % set in nbsphinxfancyoutput environment
\newcommand*{\nbsphinxincludegraphics}[2][]{%
    \gdef\spx@includegraphics@options{#1}%
    \setbox\spx@image@box\hbox{\includegraphics[#1,draft]{#2}}%
    \in@false
    \ifdim \wd\spx@image@box>\linewidth
      \g@addto@macro\spx@includegraphics@options{,width=\linewidth}%
      \in@true
    \fi
    % no rotation, no need to worry about depth
    \ifdim \ht\spx@image@box>\nbsphinx@image@maxheight
      \g@addto@macro\spx@includegraphics@options{,height=\nbsphinx@image@maxheight}%
      \in@true
    \fi
    \ifin@
      \g@addto@macro\spx@includegraphics@options{,keepaspectratio}%
    \fi
    \setbox\spx@image@box\box\voidb@x % clear memory
    \expandafter\includegraphics\expandafter[\spx@includegraphics@options]{#2}%
}% end of "\MakeFrame"-safe variant of \sphinxincludegraphics
\makeatother

\makeatletter
\renewcommand*\sphinx@verbatim@nolig@list{\do\'\do\`}
\begingroup
\catcode`'=\active
\let\nbsphinx@noligs\@noligs
\g@addto@macro\nbsphinx@noligs{\let'\PYGZsq}
\endgroup
\makeatother
\renewcommand*\sphinxbreaksbeforeactivelist{\do\<\do\"\do\'}
\renewcommand*\sphinxbreaksafteractivelist{\do\.\do\,\do\:\do\;\do\?\do\!\do\/\do\>\do\-}
\makeatletter
\fvset{codes*=\sphinxbreaksattexescapedchars\do\^\^\let\@noligs\nbsphinx@noligs}
\makeatother



\title{celloracle}
\date{Jul 15, 2020}
\release{0.4.1}
\author{Samantha Morris Lab}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex
\begin{document}

\pagestyle{empty}
\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{index::doc}}


CellOracle is a python library for the analysis of Gene Regulatory Network with single cell data.

Source code is available at \sphinxhref{https://github.com/morris-lab/CellOracle}{celloracle  GitHub repository}

For more information, please read our bioarxiv preprint: \sphinxhref{https://www.biorxiv.org/content/10.1101/2020.02.17.947416v3}{CellOracle: Dissecting cell identity via network inference and in silico gene perturbation}

\begin{sphinxadmonition}{note}{Note:}
\begin{DUlineblock}{0em}
\item[] Documentation is also available as a pdf file.
\end{DUlineblock}

\sphinxcode{\sphinxupquote{pdf documentation}}
\end{sphinxadmonition}

\begin{sphinxadmonition}{warning}{Warning:}
CellOracle is still under development. It is beta version and functions in this package may change in the future release.
\end{sphinxadmonition}


\chapter{Contents}
\label{\detokenize{index:contents}}

\section{Installation}
\label{\detokenize{installation/index:installation}}\label{\detokenize{installation/index:install}}\label{\detokenize{installation/index::doc}}
\sphinxcode{\sphinxupquote{celloracle}} uses several python libraries and R library. Please follow this guide below to install the dependent software of celloracle.


\subsection{Docker image}
\label{\detokenize{installation/index:docker-image}}\label{\detokenize{installation/index:require}}\begin{itemize}
\item {} 
Not available now. Comming soon.

\end{itemize}


\subsection{System Requirements}
\label{\detokenize{installation/index:system-requirements}}\begin{itemize}
\item {} 
Operating system: macOS or linux are highly recommended. \sphinxcode{\sphinxupquote{celloracle}} was developed and tested in Linux and macOS.

\item {} 
We found that the celloracle calculation may be EXTREMELY SLOW under an environment of Windows Subsystem for Linux (WSL). We do not recommend using WSL.

\item {} 
While you can install \sphinxcode{\sphinxupquote{celloracle}} in Windows OS, please do so at your own risk and responsibility. We DO NOT provide any support for the use in the Windows OS.

\item {} 
Memory: 8 G byte or more.  Memory usage also depends on your scRNA\sphinxhyphen{}seq data. Especially in silico perturbation requires large amount of memory.

\item {} 
CPU: Core i5 or better processor. GRN inference supports multicore calculation. Higer number of CPU cores enables faster calculation.

\end{itemize}


\subsection{Python Requirements}
\label{\detokenize{installation/index:python-requirements}}\begin{itemize}
\item {} 
\sphinxcode{\sphinxupquote{celloracle}} was developed with python 3.6. We do not support python 2.7x or python \textless{}=3.5.

\item {} 
Please install all dependent libraries before installing \sphinxcode{\sphinxupquote{celloracle}} according to the instructions below.

\item {} 
\sphinxcode{\sphinxupquote{celloracle}}  is still beta version and it is not available through PyPI or anaconda distribution yet. Please install \sphinxcode{\sphinxupquote{celloracle}} from GitHub repository according to the instruction below.

\end{itemize}


\subsubsection{0. (Optional) Make a new environment}
\label{\detokenize{installation/index:optional-make-a-new-environment}}
This step is optional. Please make a new python environment for celloracle and install dependent libraries in it if you get some software conflicts.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{create} \PYG{o}{\PYGZhy{}}\PYG{n}{n} \PYG{n}{celloracle\PYGZus{}env} \PYG{n}{python}\PYG{o}{=}\PYG{l+m+mf}{3.6}
\PYG{n}{conda} \PYG{n}{activate} \PYG{n}{celloracle\PYGZus{}env}
\end{sphinxVerbatim}


\subsubsection{1. Add conda channels}
\label{\detokenize{installation/index:add-conda-channels}}
Installation of some libraries requires non\sphinxhyphen{}default anaconda channels. Please add the channels below. Instead, you can explicitly enter the channel when you install a library.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{config} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{add} \PYG{n}{channels} \PYG{n}{defaults}
\PYG{n}{conda} \PYG{n}{config} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{add} \PYG{n}{channels} \PYG{n}{bioconda}
\PYG{n}{conda} \PYG{n}{config} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{add} \PYG{n}{channels} \PYG{n}{conda}\PYG{o}{\PYGZhy{}}\PYG{n}{forge}
\end{sphinxVerbatim}


\subsubsection{2. Install velocyto}
\label{\detokenize{installation/index:install-velocyto}}
Please install velocyto with the following commands or \sphinxhref{http://velocyto.org/velocyto.py/install/index.html}{the author’s instruction} .

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{numpy} \PYG{n}{scipy} \PYG{n}{cython} \PYG{n}{numba} \PYG{n}{matplotlib} \PYG{n}{scikit}\PYG{o}{\PYGZhy{}}\PYG{n}{learn} \PYG{n}{h5py} \PYG{n}{click} \PYG{n}{pysam} \PYG{n}{llvm} \PYG{n}{louvain}
\end{sphinxVerbatim}

Then

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{velocyto}
\end{sphinxVerbatim}

It was reported that some compile errors might occur during the installation of velocyto on MacOS.
Various errors were reported and you need to find the best solution depending on your error.
You may find the solution with these links below.
\begin{itemize}
\item {} 
\sphinxhref{https://developer.apple.com/xcode/}{Solution 1: Install Xcode}. Please try this first.

\item {} 
\sphinxhref{https://stackoverflow.com/a/53057706/10641716}{Solution 2: Install macOS\_SDK\_headers}. This solution is needed in addition to Solution\sphinxhyphen{}1 if your OS is MacOS Mojave.

\item {} 
\sphinxhref{https://github.com/morris-lab/CellOracle/issues/3}{Solution 3}. This is the solution reported by a CellOracle user. Thank you very much!

\item {} 
\sphinxhref{https://github.com/velocyto-team/velocyto.py/issues?q=}{Other solutions on Velocyto github issue page}

\end{itemize}


\subsubsection{3. Install scanpy}
\label{\detokenize{installation/index:install-scanpy}}
Please install scanpy with the following commands or \sphinxhref{https://scanpy.readthedocs.io/en/stable/installation.html}{the author’s instruction} .

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{scanpy}
\end{sphinxVerbatim}


\subsubsection{4. Install gimmemotifs}
\label{\detokenize{installation/index:install-gimmemotifs}}
Please install gimmemotifs with the following commands or \sphinxhref{https://gimmemotifs.readthedocs.io/en/master/installation.html}{the author’s instruction} .

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{genomepy}\PYG{o}{=}\PYG{l+m+mf}{0.5}\PYG{o}{.}\PYG{l+m+mi}{5} \PYG{n}{gimmemotifs}\PYG{o}{=}\PYG{l+m+mf}{0.13}\PYG{o}{.}\PYG{l+m+mi}{1}
\end{sphinxVerbatim}


\subsubsection{5. Install other python libraries}
\label{\detokenize{installation/index:install-other-python-libraries}}
Please install other python libraries below with the following commands.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{goatools} \PYG{n}{pyarrow} \PYG{n}{tqdm} \PYG{n}{joblib} \PYG{n}{jupyter}
\end{sphinxVerbatim}


\subsubsection{6. install celloracle from github}
\label{\detokenize{installation/index:install-celloracle-from-github}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{git}\PYG{o}{+}\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{github}\PYG{o}{.}\PYG{n}{com}\PYG{o}{/}\PYG{n}{morris}\PYG{o}{\PYGZhy{}}\PYG{n}{lab}\PYG{o}{/}\PYG{n}{CellOracle}\PYG{o}{.}\PYG{n}{git}
\end{sphinxVerbatim}


\subsection{R requirements}
\label{\detokenize{installation/index:r-requirements}}
\sphinxcode{\sphinxupquote{celloracle}} use R libraries for the network analysis and scATAC\sphinxhyphen{}seq analysis.
Please install \sphinxhref{https://www.r-project.org}{R} (\textgreater{}=3.5) and R libraries below according to the author’s instruction.


\subsubsection{Seurat}
\label{\detokenize{installation/index:id3}}
Please install \sphinxcode{\sphinxupquote{Seurat}} with the following r\sphinxhyphen{}script or \sphinxhref{https://satijalab.org/seurat/install.html}{the author’s instruction} .
\sphinxcode{\sphinxupquote{celloracle}} is compatible with both Seurat V2 and V3.
If you use only \sphinxcode{\sphinxupquote{scanpy}} for the scRNA\sphinxhyphen{}seq preprocessing and do not use \sphinxcode{\sphinxupquote{Seurat}} , you can skip installation of \sphinxcode{\sphinxupquote{Seurat}}.

In R console,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{Seurat\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{Cicero}
\label{\detokenize{installation/index:id5}}
Please install \sphinxcode{\sphinxupquote{Cicero}} and \sphinxcode{\sphinxupquote{Monocle3}} with the following r\sphinxhyphen{}script or \sphinxhref{https://cole-trapnell-lab.github.io/cicero-release/docs\_m3/\#installing-cicero}{the author’s instruction} .
If you do not have scATAC\sphinxhyphen{}seq data and plan to use celloracle’s base GRN, you do not need to install \sphinxcode{\sphinxupquote{Cicero}}.

In R console,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{if }\PYG{p}{(}\PYG{o}{!}\PYG{n+nf}{requireNamespace}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{BiocManager\PYGZdq{}}\PYG{p}{,} \PYG{n}{quietly} \PYG{o}{=} \PYG{k+kc}{TRUE}\PYG{p}{)}\PYG{p}{)}
 \PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{BiocManager\PYGZdq{}}\PYG{p}{)}
\PYG{n}{BiocManager}\PYG{o}{::}\PYG{n+nf}{install}\PYG{p}{(}\PYG{n+nf}{c}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{Gviz\PYGZdq{}}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{GenomicRanges\PYGZdq{}}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{rtracklayer\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}

\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{devtools\PYGZdq{}}\PYG{p}{)}
\PYG{n}{devtools}\PYG{o}{::}\PYG{n+nf}{install\PYGZus{}github}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{cole\PYGZhy{}trapnell\PYGZhy{}lab/cicero\PYGZhy{}release\PYGZdq{}}\PYG{p}{,} \PYG{n}{ref} \PYG{o}{=} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{monocle3\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{igraph}
\label{\detokenize{installation/index:id7}}
Please install \sphinxcode{\sphinxupquote{igraph}} with the following r\sphinxhyphen{}script or \sphinxhref{https://igraph.org/r/}{the author’s instruction} .

In R console,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{igraph\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{linkcomm}
\label{\detokenize{installation/index:id9}}
Please install \sphinxcode{\sphinxupquote{linkcomm}} with the following r\sphinxhyphen{}script or \sphinxhref{https://cran.r-project.org/web/packages/linkcomm/index.html}{the author’s instruction} .

In R console,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{linkcomm\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{rnetcarto}
\label{\detokenize{installation/index:id11}}
Please install \sphinxcode{\sphinxupquote{rnetcarto}} with the following r\sphinxhyphen{}script or \sphinxhref{https://github.com/cran/rnetcarto/blob/master/src/rgraph/README.md}{the author’s instruction} .

In R console,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nf}{install.packages}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{rnetcarto\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{Check installation}
\label{\detokenize{installation/index:check-installation}}
These R libraries above are necessary for the network analysis in celloracle. You can check installation using celloracle’s function.

In python console,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{celloracle} \PYG{k}{as} \PYG{n+nn}{co}
\PYG{n}{co}\PYG{o}{.}\PYG{n}{network\PYGZus{}analysis}\PYG{o}{.}\PYG{n}{test\PYGZus{}R\PYGZus{}libraries\PYGZus{}installation}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

Please make sure that all R libraries are installed. The following message will be shown when all R libraries are appropriately installed.

\begin{DUlineblock}{0em}
\item[] R path: /usr/lib/R/bin/R
\item[] checking R library installation: igraph \sphinxhyphen{}\textgreater{} OK
\item[] checking R library installation: linkcomm \sphinxhyphen{}\textgreater{} OK
\item[] checking R library installation: rnetcarto \sphinxhyphen{}\textgreater{} OK
\end{DUlineblock}

The first line above is your R path. If you want to use another R program that was installed at the different place, you can set new R path with the following command.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{co}\PYG{o}{.}\PYG{n}{network\PYGZus{}analysis}\PYG{o}{.}\PYG{n}{set\PYGZus{}R\PYGZus{}path}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ENTER YOUR R PATH HERE}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}

If you changed R path settings, please check installation again to make sure everything works.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{co}\PYG{o}{.}\PYG{n}{network\PYGZus{}analysis}\PYG{o}{.}\PYG{n}{test\PYGZus{}R\PYGZus{}libraries\PYGZus{}installation}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\section{Tutorial}
\label{\detokenize{tutorials/index:tutorial}}\label{\detokenize{tutorials/index:id1}}\label{\detokenize{tutorials/index::doc}}
The analysis proceeds through multiple steps.
Please run the notebooks sequentially.
If you do not have ATAC\sphinxhyphen{}seq data and want to use the default TF binding information, you can skip the first and second step and start from the third step.

Please refer to the \sphinxcode{\sphinxupquote{celloracle}} paper for scientific premise and the detail of the algorithm of celloracle.

The jupyter notebook files in this tutorial are available \sphinxhref{https://github.com/morris-lab/CellOracle/tree/master/docs/notebooks}{here} .


\subsection{ATAC\sphinxhyphen{}seq data preprocessing}
\label{\detokenize{tutorials/atac:atac-seq-data-preprocessing}}\label{\detokenize{tutorials/atac:atac}}\label{\detokenize{tutorials/atac::doc}}
In this step, we process scATAC\sphinxhyphen{}seq data (or bulk ATAC\sphinxhyphen{}seq data) to obtain the accessible promoter/enhancer DNA sequence.
We can get the active proximal promoter/enhancer genome sequences by picking up the ATAC\sphinxhyphen{}seq peaks that exist around the transcription starting site (TSS).
Distal cis\sphinxhyphen{}regulatory elements can be picked up using  \sphinxhref{https://cole-trapnell-lab.github.io/cicero-release/docs/\#installing-cicero}{Cicero} .
Cicero analyzes scATAC\sphinxhyphen{}seq data to calculate a co\sphinxhyphen{}accessible score between peaks.
We can identify cis\sphinxhyphen{}regulatory elements using Cicero’s co\sphinxhyphen{}access score and TSS information.

If you have bulk ATAC\sphinxhyphen{}seq data instead of scATAC\sphinxhyphen{}data, we’ll get only the proximal promoter/enhancer genome sequences.


\subsubsection{A. Extract TF binding information from scATAC\sphinxhyphen{}seq data}
\label{\detokenize{tutorials/atac:a-extract-tf-binding-information-from-scatac-seq-data}}
If you have scATAC\sphinxhyphen{}seq data, you can get information on the distal cis\sphinxhyphen{}regulatory elements.
This step uses Cicero and does not use celloracle.
You need to get co\sphinxhyphen{}accessibility table in this analysis.
Although we provide an example notebook here, you can analyze your data with Cicero in a different way if you are familiar with Cicero.
If you have a question about Cicero, please read \sphinxhref{https://cole-trapnell-lab.github.io/cicero-release/}{the documentation of Cicero} for the detailed usage.


\paragraph{scATAC\sphinxhyphen{}seq analysis with Cicero and Monocle3}
\label{\detokenize{tutorials/atac:scatac-seq-analysis-with-cicero-and-monocle3}}
R notebook

This is an example R script for Cicero analysis. In this R notebook, we’ll use Cicero and Monocle3.

Please make sure that you installed these packages in adavance.

You can download notebook file and additional data files from celloracle github page. \sphinxurl{https://github.com/morris-lab/CellOracle/tree/master/docs/notebooks/01\_ATAC-seq\_data\_processing/option1\_scATAC-seq\_data\_analysis\_with\_cicero}

Another tutorial notebook that uses Monocle2 is also available in the celloracle github page above.


\subparagraph{0. Import library}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:0.-Import-library}}\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n+nf}{library}\PYG{p}{(}\PYG{n}{cicero}\PYG{p}{)}
\PYG{n+nf}{library}\PYG{p}{(}\PYG{n}{monocle3}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subparagraph{1. Prepare data}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:1.-Prepare-data}}
In this tutorial we’ll use acATAC\sphinxhyphen{}seq data from the 10x genomics database. You do not need to download these data if you analyze your own scATAC\sphinxhyphen{}seq data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Create folder to store data}
\PYG{n+nf}{dir.create}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{data\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Download demo dataset from 10x genomics}
\PYG{n+nf}{system}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{wget \PYGZhy{}O data/matrix.tar.gz http://cf.10xgenomics.com/samples/cell\PYGZhy{}atac/1.1.0/atac\PYGZus{}v1\PYGZus{}E18\PYGZus{}brain\PYGZus{}fresh\PYGZus{}5k/atac\PYGZus{}v1\PYGZus{}E18\PYGZus{}brain\PYGZus{}fresh\PYGZus{}5k\PYGZus{}filtered\PYGZus{}peak\PYGZus{}bc\PYGZus{}matrix.tar.gz\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Unzip data}
\PYG{n+nf}{system}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{tar \PYGZhy{}xvf data/matrix.tar.gz \PYGZhy{}C data\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} You can substitute the data path below with the data path of your scATAC data.}
\PYG{n}{data\PYGZus{}folder} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{data/filtered\PYGZus{}peak\PYGZus{}bc\PYGZus{}matrix\PYGZdq{}}

\PYG{c+c1}{\PYGZsh{} Create a folder to save results}
\PYG{n}{output\PYGZus{}folder} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{cicero\PYGZus{}output\PYGZdq{}}
\PYG{n+nf}{dir.create}\PYG{p}{(}\PYG{n}{output\PYGZus{}folder}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subparagraph{2. Load data and make Cell Data Set (CDS) object}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:2.-Load-data-and-make-Cell-Data-Set-(CDS)-object}}

\subparagraph{2.1. Process data to make CDS object}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:2.1.-Process-data-to-make-CDS-object}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} read in matrix data using the Matrix package}
\PYG{n}{indata} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{readMM}\PYG{p}{(}\PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{data\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/matrix.mtx\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} binarize the matrix}
\PYG{n}{indata}\PYG{o}{@}\PYG{n}{x}\PYG{p}{[}\PYG{n}{indata}\PYG{o}{@}\PYG{n}{x} \PYG{o}{\PYGZgt{}} \PYG{l+m}{0}\PYG{p}{]} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+m}{1}

\PYG{c+c1}{\PYGZsh{} format cell info}
\PYG{n}{cellinfo} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{read.table}\PYG{p}{(}\PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{data\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/barcodes.tsv\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{cellinfo}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{cellinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{V1}
\PYG{n+nf}{names}\PYG{p}{(}\PYG{n}{cellinfo}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{cells\PYGZdq{}}

\PYG{c+c1}{\PYGZsh{} format peak info}
\PYG{n}{peakinfo} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{read.table}\PYG{p}{(}\PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{data\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/peaks.bed\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{names}\PYG{p}{(}\PYG{n}{peakinfo}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{c}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{chr\PYGZdq{}}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{bp1\PYGZdq{}}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{bp2\PYGZdq{}}\PYG{p}{)}
\PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{site\PYGZus{}name} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{paste}\PYG{p}{(}\PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{chr}\PYG{p}{,} \PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{bp1}\PYG{p}{,} \PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{bp2}\PYG{p}{,} \PYG{n}{sep}\PYG{o}{=}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{\PYGZus{}\PYGZdq{}}\PYG{p}{)}
\PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{peakinfo}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{peakinfo}\PYG{o}{\PYGZdl{}}\PYG{n}{site\PYGZus{}name}

\PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{indata}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{peakinfo}\PYG{p}{)}
\PYG{n+nf}{colnames}\PYG{p}{(}\PYG{n}{indata}\PYG{p}{)} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n}{cellinfo}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} make CDS}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}}  \PYG{n+nf}{suppressWarnings}\PYG{p}{(}\PYG{n+nf}{new\PYGZus{}cell\PYGZus{}data\PYGZus{}set}\PYG{p}{(}\PYG{n}{indata}\PYG{p}{,}
\PYG{n}{cell\PYGZus{}metadata} \PYG{o}{=} \PYG{n}{cellinfo}\PYG{p}{,}
\PYG{n}{gene\PYGZus{}metadata} \PYG{o}{=} \PYG{n}{peakinfo}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{monocle3}\PYG{o}{::}\PYG{n+nf}{detect\PYGZus{}genes}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{}Ensure there are no peaks included with zero reads}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{input\PYGZus{}cds}\PYG{p}{[}\PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{rowSums}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)} \PYG{o}{!=} \PYG{l+m}{0}\PYG{p}{,}\PYG{p}{]}
\end{sphinxVerbatim}
}


\subparagraph{3. Qauality check and Filtering}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:3.-Qauality-check-and-Filtering}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Visualize peak\PYGZus{}count\PYGZus{}per\PYGZus{}cell}
\PYG{n+nf}{hist}\PYG{p}{(}\PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{colSums}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=420\sphinxpxdimen,height=420\sphinxpxdimen]{{notebooks_01_ATAC-seq_data_processing_option1_scATAC-seq_data_analysis_with_cicero_01_atacdata_analysis_with_cicero_and_monocle3_9_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} filter cells by peak\PYGZus{}count}
\PYG{c+c1}{\PYGZsh{} PLEASE SET APPROPRIATE THRESHOLD VALUES according to your data}
\PYG{n}{max\PYGZus{}count} \PYG{o}{\PYGZlt{}\PYGZhy{}}  \PYG{l+m}{15000}
\PYG{n}{min\PYGZus{}count} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{l+m}{2000}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{input\PYGZus{}cds}\PYG{p}{[}\PYG{p}{,}\PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{colSums}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)} \PYG{o}{\PYGZgt{}=} \PYG{n}{min\PYGZus{}count}\PYG{p}{]}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n}{input\PYGZus{}cds}\PYG{p}{[}\PYG{p}{,}\PYG{n}{Matrix}\PYG{o}{::}\PYG{n+nf}{colSums}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)} \PYG{o}{\PYGZlt{}=} \PYG{n}{max\PYGZus{}count}\PYG{p}{]}

\end{sphinxVerbatim}
}


\subparagraph{4. Process cicero\sphinxhyphen{}CDS object}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:4.-Process-cicero-CDS-object}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Data preprocessing}
\PYG{n+nf}{set.seed}\PYG{p}{(}\PYG{l+m}{2017}\PYG{p}{)}

\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{detect\PYGZus{}genes}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{estimate\PYGZus{}size\PYGZus{}factors}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{preprocess\PYGZus{}cds}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{,} \PYG{n}{method} \PYG{o}{=} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{LSI\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Dimensional reduction with umap}
\PYG{n}{input\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{reduce\PYGZus{}dimension}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{,} \PYG{n}{reduction\PYGZus{}method} \PYG{o}{=} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{UMAP\PYGZsq{}}\PYG{p}{,}
                              \PYG{n}{preprocess\PYGZus{}method} \PYG{o}{=} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{LSI\PYGZdq{}}\PYG{p}{)}
\PYG{n}{umap\PYGZus{}coords} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{reducedDims}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{o}{\PYGZdl{}}\PYG{n}{UMAP}


\PYG{n}{cicero\PYGZus{}cds} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{make\PYGZus{}cicero\PYGZus{}cds}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{,} \PYG{n}{reduced\PYGZus{}coordinates} \PYG{o}{=} \PYG{n}{umap\PYGZus{}coords}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Save cds object if you want}
\PYG{n+nf}{saveRDS}\PYG{p}{(}\PYG{n}{cicero\PYGZus{}cds}\PYG{p}{,} \PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{output\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/cicero\PYGZus{}cds.Rds\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}


\end{sphinxVerbatim}
}


\subparagraph{5. Load reference genome information}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:5.-Load-reference-genome-information}}
To run cicero, we need to get a genome coordinates files, which contains the lengths of each chromosomes. You can read mm10 genome information with the following command. The text file, mm10\_chromosome\_length.txt is included in the celloracle notebook folder.

\sphinxurl{https://github.com/morris-lab/CellOracle/tree/master/docs/notebooks/01\_ATAC-seq\_data\_processing/option1\_scATAC-seq\_data\_analysis\_with\_cicero}

If your scATAC\sphinxhyphen{}seq data use different reference genome, you need to get a genome coordinates files for your reference genome. Please see the Cicero documentation for more information.

\sphinxurl{https://cole-trapnell-lab.github.io/cicero-release/docs\_m3/\#installing-cicero}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} !!Please make sure that the reference genome information below match the reference genome of your scATAC\PYGZhy{}seq data.}

\PYG{c+c1}{\PYGZsh{} If your scATAC\PYGZhy{}seq uses mm10 reference genome, you can read chromosome length file with the following command.}
\PYG{n}{chromosome\PYGZus{}length} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{read.table}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{./mm10\PYGZus{}chromosome\PYGZus{}length.txt\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} For mm9 genome, you can use the following command.}
\PYG{c+c1}{\PYGZsh{}data(\PYGZdq{}mouse.mm9.genome\PYGZdq{})}
\PYG{c+c1}{\PYGZsh{}chromosome\PYGZus{}length \PYGZlt{}\PYGZhy{} mouse.mm9.genome}

\PYG{c+c1}{\PYGZsh{} For hg19 genome, you can use the following command.}
\PYG{c+c1}{\PYGZsh{}data(\PYGZdq{}human.hg19.genome\PYGZdq{})}
\PYG{c+c1}{\PYGZsh{}chromosome\PYGZus{}length \PYGZlt{}\PYGZhy{} mhuman.hg19.genome}


\end{sphinxVerbatim}
}


\subparagraph{6. Run Cicero}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:6.-Run-Cicero}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} runc the main function}
\PYG{n}{conns} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{run\PYGZus{}cicero}\PYG{p}{(}\PYG{n}{cicero\PYGZus{}cds}\PYG{p}{,} \PYG{n}{chromosome\PYGZus{}length}\PYG{p}{)} \PYG{c+c1}{\PYGZsh{} Takes a few minutes to run}

\PYG{c+c1}{\PYGZsh{} save results}
\PYG{n+nf}{saveRDS}\PYG{p}{(}\PYG{n}{conns}\PYG{p}{,} \PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{output\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/cicero\PYGZus{}connections.Rds\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} check results}
\PYG{n+nf}{head}\PYG{p}{(}\PYG{n}{conns}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
[1] "Starting Cicero"
[1] "Calculating distance\_parameter value"
[1] "Running models"
[1] "Assembling connections"
[1] "Done"
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}
A data.frame: 6 × 3
\begin{tabular}{r|lll}
  & Peak1 & Peak2 & coaccess\\
  & <fct> & <fct> & <dbl>\\
\hline
    2 & chr1\_3094484\_3095479 & chr1\_3113499\_3113979 & -0.316289004\\
    3 & chr1\_3094484\_3095479 & chr1\_3119478\_3121690 & -0.419240532\\
    4 & chr1\_3094484\_3095479 & chr1\_3399730\_3400368 & -0.050867246\\
    5 & chr1\_3113499\_3113979 & chr1\_3094484\_3095479 & -0.316289004\\
    7 & chr1\_3113499\_3113979 & chr1\_3119478\_3121690 &  0.370342744\\
    8 & chr1\_3113499\_3113979 & chr1\_3399730\_3400368 & -0.009276026\\
\end{tabular}
\end{nbsphinxfancyoutput}


\subparagraph{6. Save results for the next step}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_analysis_with_cicero_and_monocle3:6.-Save-results-for-the-next-step}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{all\PYGZus{}peaks} \PYG{o}{\PYGZlt{}\PYGZhy{}} \PYG{n+nf}{row.names}\PYG{p}{(}\PYG{n+nf}{exprs}\PYG{p}{(}\PYG{n}{input\PYGZus{}cds}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{write.csv}\PYG{p}{(}\PYG{n}{x} \PYG{o}{=} \PYG{n}{all\PYGZus{}peaks}\PYG{p}{,} \PYG{n}{file} \PYG{o}{=} \PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{output\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/all\PYGZus{}peaks.csv\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nf}{write.csv}\PYG{p}{(}\PYG{n}{x} \PYG{o}{=} \PYG{n}{conns}\PYG{p}{,} \PYG{n}{file} \PYG{o}{=} \PYG{n+nf}{paste0}\PYG{p}{(}\PYG{n}{output\PYGZus{}folder}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{/cicero\PYGZus{}connections.csv\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}

\end{sphinxVerbatim}
}


\paragraph{TSS annotation}
\label{\detokenize{tutorials/atac:tss-annotation}}
Python notebook

In this notebook, we process the results of cicero analysis to get active promoter/enhancer DNA peaks. First, we pick up peaks around the transcription starting site (TSS). Second, we merge cicero data with the peaks around TSS. Then we remove peaks that have a weak connection to TSS peak so that the final product includes TSS peaks and peaks that have a strong connection with the TSS peaks. We use this information as an active promoter/enhancer elements.


\subparagraph{0. Import libraries}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:0.-Import-libraries}}\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}


\PYG{k+kn}{import} \PYG{n+nn}{os}\PYG{o}{,} \PYG{n+nn}{sys}\PYG{o}{,} \PYG{n+nn}{shutil}\PYG{o}{,} \PYG{n+nn}{importlib}\PYG{o}{,} \PYG{n+nn}{glob}
\PYG{k+kn}{from} \PYG{n+nn}{tqdm}\PYG{n+nn}{.}\PYG{n+nn}{notebook} \PYG{k+kn}{import} \PYG{n}{tqdm}

\PYG{k+kn}{from} \PYG{n+nn}{celloracle} \PYG{k+kn}{import} \PYG{n}{motif\PYGZus{}analysis} \PYG{k}{as} \PYG{n}{ma}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{300}
\end{sphinxVerbatim}
}


\subparagraph{1. Load data made with cicero}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:1.-Load-data-made-with-cicero}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Load all peaks}
\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cicero\PYGZus{}output/all\PYGZus{}peaks.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{index\PYGZus{}col}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{peaks}\PYG{o}{.}\PYG{n}{x}\PYG{o}{.}\PYG{n}{values}
\PYG{n}{peaks}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}array(['chr1\_3094484\_3095479', 'chr1\_3113499\_3113979',
       'chr1\_3119478\_3121690', {\ldots}, 'chrY\_90804622\_90805450',
       'chrY\_90808626\_90809117', 'chrY\_90810560\_90811167'], dtype=object)
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Load cicero results}
\PYG{n}{cicero\PYGZus{}connections} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cicero\PYGZus{}output/cicero\PYGZus{}connections.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{index\PYGZus{}col}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{n}{cicero\PYGZus{}connections}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                  Peak1                 Peak2  coaccess
2  chr1\_3094484\_3095479  chr1\_3113499\_3113979 -0.316289
3  chr1\_3094484\_3095479  chr1\_3119478\_3121690 -0.419241
4  chr1\_3094484\_3095479  chr1\_3399730\_3400368 -0.050867
5  chr1\_3113499\_3113979  chr1\_3094484\_3095479 -0.316289
7  chr1\_3113499\_3113979  chr1\_3119478\_3121690  0.370343
\end{sphinxVerbatim}
}


\subparagraph{2. Make TSS annotation}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:2.-Make-TSS-annotation}}

\subparagraph{IMPORTANT: Please make sure that you are setting correct reference genoms.}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:IMPORTANT:-Please-make-sure-that-you-are-setting-correct-reference-genoms.}}
If your scATAC\sphinxhyphen{}seq data was generated with mm10 reference genome, you can set ref\_genome=“mm10”. If you used hg19 human reference genome, please set ref\_genome==“hg19”

Currently we support refgenomes below. \{“Human”: {[}‘hg38’, ‘hg19’{]}, “Mouse”: {[}‘mm10’, ‘mm9’{]}, “S.cerevisiae”: {[}“sacCer2”, “sacCer3”{]}\}

If your reference genome is not in the list, please send a request through github issue page.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}
\PYG{n}{tss\PYGZus{}annotated} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{get\PYGZus{}tss\PYGZus{}info}\PYG{p}{(}\PYG{n}{peak\PYGZus{}str\PYGZus{}list}\PYG{o}{=}\PYG{n}{peaks}\PYG{p}{,} \PYG{n}{ref\PYGZus{}genome}\PYG{o}{=} \PYG{p}{)} \PYG{c+c1}{\PYGZsh{}\PYGZsh{}!! Set reference genome here}


\PYG{c+c1}{\PYGZsh{} Check results}
\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{tail}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
que bed peaks: 72402
tss peaks in que: 16987
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}         chr      start        end gene\_short\_name strand
16982   chr1   55130650   55132118            Mob4      +
16983   chr6   94499875   94500767        Slc25a26      +
16984  chr19   45659222   45660823           Fbxw4      -
16985  chr12  100898848  100899597           Gpr68      -
16986   chr4  129491262  129492047         Fam229a      -
\end{sphinxVerbatim}
}


\subparagraph{3. Integrate TSS info and cicero connections}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:3.-Integrate-TSS-info-and-cicero-connections}}
The output file after the integration process has three columns; “peak\_id”, “gene\_short\_name”, and “coaccess”. “peak\_id” is either the TSS peak or the peaks that have a connection with the TSS peak. “gene\_short\_name” is the gene name that associated with the TSS site. “coaccess” is the co\sphinxhyphen{}access score between a peak and TSS peak. Note, the TSS peak is indicated by a score of 1.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{integrated} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{integrate\PYGZus{}tss\PYGZus{}peak\PYGZus{}with\PYGZus{}cicero}\PYG{p}{(}\PYG{n}{tss\PYGZus{}peak}\PYG{o}{=}\PYG{n}{tss\PYGZus{}annotated}\PYG{p}{,}
                                               \PYG{n}{cicero\PYGZus{}connections}\PYG{o}{=}\PYG{n}{cicero\PYGZus{}connections}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{integrated}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{integrated}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
(263279, 3)
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\_id gene\_short\_name  coaccess
0  chr10\_100015291\_100017830            Kitl  1.000000
1  chr10\_100018677\_100020384            Kitl  0.086299
2  chr10\_100050858\_100051762            Kitl  0.034558
3  chr10\_100052829\_100053395            Kitl  0.167188
4  chr10\_100128086\_100128882           Tmtc3  0.022341
\end{sphinxVerbatim}
}


\subparagraph{4. Filter peaks}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:4.-Filter-peaks}}
Remove peaks that have weak coaccess score.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{peak} \PYG{o}{=} \PYG{n}{integrated}\PYG{p}{[}\PYG{n}{integrated}\PYG{o}{.}\PYG{n}{coaccess} \PYG{o}{\PYGZgt{}}\PYG{o}{=} \PYG{l+m+mf}{0.8}\PYG{p}{]}
\PYG{n}{peak} \PYG{o}{=} \PYG{n}{peak}\PYG{p}{[}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}id}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{gene\PYGZus{}short\PYGZus{}name}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{]}\PYG{o}{.}\PYG{n}{reset\PYGZus{}index}\PYG{p}{(}\PYG{n}{drop}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{peak}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{peak}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
(15680, 2)
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\_id gene\_short\_name
0  chr10\_100015291\_100017830            Kitl
1  chr10\_100486534\_100488209           Tmtc3
2  chr10\_100588641\_100589556   4930430F08Rik
3  chr10\_100741247\_100742505         Gm35722
4  chr10\_101681379\_101682124          Mgat4c
\end{sphinxVerbatim}
}


\subparagraph{5. Save data}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data:5.-Save-data}}
Save the promoter/enhancer peak.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{peak}\PYG{o}{.}\PYG{n}{to\PYGZus{}parquet}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}file.parquet}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\sphinxhyphen{}\textgreater{} go to next notebook


\subsubsection{B. Extract TF binding information from bulk ATAC\sphinxhyphen{}seq data or Chip\sphinxhyphen{}seq data}
\label{\detokenize{tutorials/atac:b-extract-tf-binding-information-from-bulk-atac-seq-data-or-chip-seq-data}}
Bulk DNA\sphinxhyphen{}seq data can be used to get the accessible promoter/enhancer sequences.

Python notebook


\paragraph{0. Import libraries}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data:0.-Import-libraries}}\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\PYG{k+kn}{import} \PYG{n+nn}{os}\PYG{o}{,} \PYG{n+nn}{sys}\PYG{o}{,} \PYG{n+nn}{shutil}\PYG{o}{,} \PYG{n+nn}{importlib}\PYG{o}{,} \PYG{n+nn}{glob}
\PYG{k+kn}{from} \PYG{n+nn}{tqdm} \PYG{k+kn}{import} \PYG{n}{tqdm\PYGZus{}notebook} \PYG{k}{as} \PYG{n}{tqdm}

\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{300}

\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Import celloracle function}
\PYG{k+kn}{from} \PYG{n+nn}{celloracle} \PYG{k+kn}{import} \PYG{n}{motif\PYGZus{}analysis} \PYG{k}{as} \PYG{n}{ma}
\end{sphinxVerbatim}
}


\paragraph{1. Load bed file}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data:1.-Load-bed-file}}
Import ATAC\sphinxhyphen{}seq bed file. This script can also be used with DNase\sphinxhyphen{}seq or Chip\sphinxhyphen{}seq data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{file\PYGZus{}path\PYGZus{}of\PYGZus{}bed\PYGZus{}file} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{data/all\PYGZus{}peaks.bed}\PYG{l+s+s2}{\PYGZdq{}}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Load bed\PYGZus{}file}
\PYG{n}{bed} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{read\PYGZus{}bed}\PYG{p}{(}\PYG{n}{file\PYGZus{}path\PYGZus{}of\PYGZus{}bed\PYGZus{}file}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{bed}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{bed}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
(436206, 4)
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}  chrom    start      end               seqname
0  chr1  3002478  3002968  chr1\_3002478\_3002968
1  chr1  3084739  3085712  chr1\_3084739\_3085712
2  chr1  3103576  3104022  chr1\_3103576\_3104022
3  chr1  3106871  3107210  chr1\_3106871\_3107210
4  chr1  3108932  3109158  chr1\_3108932\_3109158
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Convert bed file into peak name list}
\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{process\PYGZus{}bed\PYGZus{}file}\PYG{o}{.}\PYG{n}{df\PYGZus{}to\PYGZus{}list\PYGZus{}peakstr}\PYG{p}{(}\PYG{n}{bed}\PYG{p}{)}
\PYG{n}{peaks}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}array(['chr1\_3002478\_3002968', 'chr1\_3084739\_3085712',
       'chr1\_3103576\_3104022', {\ldots}, 'chrY\_631222\_631480',
       'chrY\_795887\_796426', 'chrY\_2397419\_2397628'], dtype=object)
\end{sphinxVerbatim}
}


\paragraph{2. Make TSS annotation}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data:2.-Make-TSS-annotation}}
IMPORTANT: Please make sure that you are setting the correct ref genome!

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{tss\PYGZus{}annotated} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{get\PYGZus{}tss\PYGZus{}info}\PYG{p}{(}\PYG{n}{peak\PYGZus{}str\PYGZus{}list}\PYG{o}{=}\PYG{n}{peaks}\PYG{p}{,} \PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mm9}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Check results}
\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{tail}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
que bed peaks: 436206
tss peaks in que: 24822
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}         chr     start       end gene\_short\_name strand
24817   chr2  60560211  60561602           Itgb6      -
24818  chr15   3975177   3978654        BC037032      -
24819  chr14  67690701  67692101         Ppp2r2a      -
24820  chr17  48455247  48455773   B430306N03Rik      +
24821  chr10  59861192  59861608         Gm17455      +
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Change format}
\PYG{n}{peak\PYGZus{}id\PYGZus{}tss} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{process\PYGZus{}bed\PYGZus{}file}\PYG{o}{.}\PYG{n}{df\PYGZus{}to\PYGZus{}list\PYGZus{}peakstr}\PYG{p}{(}\PYG{n}{tss\PYGZus{}annotated}\PYG{p}{)}
\PYG{n}{tss\PYGZus{}annotated} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}id}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{peak\PYGZus{}id\PYGZus{}tss}\PYG{p}{,}
                              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{gene\PYGZus{}short\PYGZus{}name}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{gene\PYGZus{}short\PYGZus{}name}\PYG{o}{.}\PYG{n}{values}\PYG{p}{\PYGZcb{}}\PYG{p}{)}
\PYG{n}{tss\PYGZus{}annotated} \PYG{o}{=} \PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{reset\PYGZus{}index}\PYG{p}{(}\PYG{n}{drop}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
(24822, 2)
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                   peak\_id gene\_short\_name
0   chr7\_50691730\_50692032            Nkg7
1   chr7\_50692077\_50692785            Nkg7
2  chr13\_93564413\_93564836           Thbs4
3  chr13\_14613429\_14615645           Hecw1
4   chr3\_99688753\_99689665          Spag17
\end{sphinxVerbatim}
}


\paragraph{3. Save data}
\label{\detokenize{notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data:3.-Save-data}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{tss\PYGZus{}annotated}\PYG{o}{.}\PYG{n}{to\PYGZus{}parquet}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}file.parquet}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\sphinxhyphen{}\textgreater{} go to next notebook


\subsection{Transcription factor binding motif scan}
\label{\detokenize{tutorials/motifscan:transcription-factor-binding-motif-scan}}\label{\detokenize{tutorials/motifscan:motifscan}}\label{\detokenize{tutorials/motifscan::doc}}
We identified accessible Promoter/enhancer DNA regions using ATAC\sphinxhyphen{}seq data.
Next, we will obtain a list of TFs for each target gene by scanning the regulatory genomic sequences for TF\sphinxhyphen{}binding motifs.
In the later GRN inference process, this list will be used to define potential regulatory connections.

Python notebook


\subsubsection{0. Import libraries}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:0.-Import-libraries}}\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\PYG{k+kn}{import} \PYG{n+nn}{os}\PYG{o}{,} \PYG{n+nn}{sys}\PYG{o}{,} \PYG{n+nn}{shutil}\PYG{o}{,} \PYG{n+nn}{importlib}\PYG{o}{,} \PYG{n+nn}{glob}
\PYG{k+kn}{from} \PYG{n+nn}{tqdm}\PYG{n+nn}{.}\PYG{n+nn}{notebook} \PYG{k+kn}{import} \PYG{n}{tqdm}

\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{15}\PYG{p}{,}\PYG{l+m+mi}{7}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{600}

\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{from} \PYG{n+nn}{celloracle} \PYG{k+kn}{import} \PYG{n}{motif\PYGZus{}analysis} \PYG{k}{as} \PYG{n}{ma}
\PYG{k+kn}{from} \PYG{n+nn}{celloracle}\PYG{n+nn}{.}\PYG{n+nn}{utility} \PYG{k+kn}{import} \PYG{n}{save\PYGZus{}as\PYGZus{}pickled\PYGZus{}object}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/pandas1/lib/python3.6/site-packages/gimmemotifs/plot.py:19: MatplotlibDeprecationWarning: The 'warn' parameter of use() is deprecated since Matplotlib 3.1 and will be removed in 3.3.  If any parameter follows 'warn', they should be pass as keyword, not positionally.
  mpl.use("Agg", warn=False)
\end{sphinxVerbatim}
}


\subsubsection{1. Load data}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:1.-Load-data}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Load annotated peak data.}
\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}parquet}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{../01\PYGZus{}ATAC\PYGZhy{}seq\PYGZus{}data\PYGZus{}processing/option1\PYGZus{}scATAC\PYGZhy{}seq\PYGZus{}data\PYGZus{}analysis\PYGZus{}with\PYGZus{}cicero/peak\PYGZus{}file.parquet}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{peaks}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\_id gene\_short\_name
0  chr10\_100015291\_100017830            Kitl
1  chr10\_100486534\_100488209           Tmtc3
2  chr10\_100588641\_100589556   4930430F08Rik
3  chr10\_100741247\_100742505         Gm35722
4  chr10\_101681379\_101682124          Mgat4c
\end{sphinxVerbatim}
}


\subsubsection{2. Check data}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:2.-Check-data}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Check data}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{number of peak: }\PYG{l+s+si}{\PYGZob{}}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{peaks}\PYG{o}{.}\PYG{n}{peak\PYGZus{}id}\PYG{o}{.}\PYG{n}{unique}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{getLength}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{a}\PYG{p}{,} \PYG{n}{b}\PYG{p}{,} \PYG{n}{c} \PYG{o}{=} \PYG{n}{x}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{peak\PYGZus{}id}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{o}{.}\PYG{n}{split}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZus{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{k}{return} \PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{c}\PYG{p}{)} \PYG{o}{\PYGZhy{}} \PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{b}\PYG{p}{)}

\PYG{n}{df} \PYG{o}{=} \PYG{n}{peaks}\PYG{o}{.}\PYG{n}{apply}\PYG{p}{(}\PYG{k}{lambda} \PYG{n}{x}\PYG{p}{:} \PYG{n}{getLength}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mean peak length: }\PYG{l+s+si}{\PYGZob{}}\PYG{n}{df}\PYG{o}{.}\PYG{n}{values}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{)}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
number of peak: 13919
mean peak length: 1756.1744260204082
\end{sphinxVerbatim}
}


\paragraph{2.1. Remove short peaks}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:2.1.-Remove-short-peaks}}
Short DNA fragment that are less than 5 bases, cannot be used for motif scanning. Therefore, we will remove the short DNA fragments.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{peaks} \PYG{o}{=} \PYG{n}{peaks}\PYG{p}{[}\PYG{n}{df}\PYG{o}{\PYGZgt{}}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{]}
\end{sphinxVerbatim}
}


\subsubsection{3. Instantiate TFinfo object and search for TF binding motifs}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs}}
The motif analysis module has a custom class; TFinfo. The TFinfo object converts a peak data into a DNA sequences and scans the DNA sequences searching for TF binding motifs. Then, the results of motif scan will be filtered and converted into either a python dictionary or a depending on your preference. This TF information is necessary for GRN inference.


\paragraph{3.1 check reference genome installation}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:3.1-check-reference-genome-installation}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} PLEASE make sure that you are setting correct ref genome.}
\PYG{n}{ref\PYGZus{}genome} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mm10}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{n}{ma}\PYG{o}{.}\PYG{n}{is\PYGZus{}genome\PYGZus{}installed}\PYG{p}{(}\PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{n}{ref\PYGZus{}genome}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
genome mm10 is not installed in this environment.
Please install genome using genomepy.
e.g.
    >>> import genomepy
    >>> genomepy.install\_genome("mm9", "UCSC")
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}False
\end{sphinxVerbatim}
}


\paragraph{3.2. Install reference genome (if refgenome is not installed)}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:3.2.-Install-reference-genome-(if-refgenome-is-not-installed)}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check again}
\PYG{n}{ma}\PYG{o}{.}\PYG{n}{is\PYGZus{}genome\PYGZus{}installed}\PYG{p}{(}\PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{n}{ref\PYGZus{}genome}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}True
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[14]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Instantiate TFinfo object}
\PYG{n}{tfi} \PYG{o}{=} \PYG{n}{ma}\PYG{o}{.}\PYG{n}{TFinfo}\PYG{p}{(}\PYG{n}{peak\PYGZus{}data\PYGZus{}frame}\PYG{o}{=}\PYG{n}{peaks}\PYG{p}{,} \PYG{c+c1}{\PYGZsh{} peak info calculated from ATAC\PYGZhy{}seq data}
                \PYG{n}{ref\PYGZus{}genome}\PYG{o}{=}\PYG{n}{ref\PYGZus{}genome}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{4. Scan motifs and save object}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:4.-Scan-motifs-and-save-object}}
This step may take long time

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[15]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{o}{\PYGZpc{}\PYGZpc{}time}
\PYG{c+c1}{\PYGZsh{} Scan motifs}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{scan}\PYG{p}{(}\PYG{n}{fpr}\PYG{o}{=}\PYG{l+m+mf}{0.02}\PYG{p}{,} \PYG{n}{verbose}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Save tfinfo object}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}hdf5}\PYG{p}{(}\PYG{n}{file\PYGZus{}path}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{test.celloracle.tfinfo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
initiating scanner {\ldots}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
2019-09-22 23:00:18,604 - INFO - Using background: genome mm10 with length 200
2019-09-22 23:00:18,986 - INFO - Determining FPR-based threshold
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
getting DNA sequences {\ldots}
scanning motifs {\ldots}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=1, bar\_style='info', max=1), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

CPU times: user 52min 23s, sys: 36.8 s, total: 53min
Wall time: 52min 58s
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Check motif scan results}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{scanned\PYGZus{}df}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     seqname                      motif\_id factors\_direct  \textbackslash{}
0  chr10\_100015291\_100017830       GM.5.0.Homeodomain.0001          TGIF1
1  chr10\_100015291\_100017830             GM.5.0.Mixed.0001
2  chr10\_100015291\_100017830             GM.5.0.Mixed.0001
3  chr10\_100015291\_100017830             GM.5.0.Mixed.0001
4  chr10\_100015291\_100017830  GM.5.0.Nuclear\_receptor.0002          NR2C2

         factors\_indirect      score   pos  strand
0  ENSG00000234254, TGIF1  10.311002  1003       1
1               SRF, EGR1   7.925873   481       1
2               SRF, EGR1   7.321375   911      -1
3               SRF, EGR1   7.276585   811      -1
4            NR2C2, Nr2c2   9.067331   449      -1
\end{sphinxVerbatim}
}

We have the score for each sequence and motif\_id pair. In the next step we will filter the motifs with low score.


\subsubsection{5. Filtering motifs}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:5.-Filtering-motifs}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[17]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Reset filtering}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{reset\PYGZus{}filtering}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Do filtering}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{filter\PYGZus{}motifs\PYGZus{}by\PYGZus{}score}\PYG{p}{(}\PYG{n}{threshold}\PYG{o}{=}\PYG{l+m+mf}{10.5}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Do post filtering process. Convert results into several file format.}
\PYG{n}{tfi}\PYG{o}{.}\PYG{n}{make\PYGZus{}TFinfo\PYGZus{}dataframe\PYGZus{}and\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{verbose}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
peaks were filtered: 12934005 -> 2285279
1. converting scanned results into one-hot encoded dataframe.
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=13919), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

2. converting results into dictionaries.
converting scan results into dictionaries{\ldots}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=14804), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1090), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

\end{sphinxVerbatim}
}


\subsubsection{6. Get Final results}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:6.-Get-Final-results}}

\paragraph{6.1. Get resutls as a dictionary}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:6.1.-Get-resutls-as-a-dictionary}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[18]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{td} \PYG{o}{=} \PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{dictionary\PYGZus{}type}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{targetgene2TFs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\end{sphinxVerbatim}
}


\paragraph{6.2. Get results as a dataframe}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:6.2.-Get-results-as-a-dataframe}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[20]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{df} \PYG{o}{=} \PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}dataframe}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{df}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[20]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\_id gene\_short\_name  9430076c15rik  Ac002126.6  \textbackslash{}
0  chr10\_100015291\_100017830            Kitl              0           0
1  chr10\_100486534\_100488209           Tmtc3              0           0
2  chr10\_100588641\_100589556   4930430F08Rik              0           0
3  chr10\_100741247\_100742505         Gm35722              0           0
4  chr10\_101681379\_101682124          Mgat4c              0           0

   Ac012531.1  Ac226150.2  Afp  Ahr  Ahrr  Aire  {\ldots}  Znf784  Znf8  Znf816  \textbackslash{}
0           0           0    0    1     1     0  {\ldots}       0     0       0
1           0           0    0    0     0     0  {\ldots}       1     0       0
2           1           0    0    1     1     0  {\ldots}       0     0       0
3           0           0    0    0     0     0  {\ldots}       0     0       0
4           0           0    0    0     0     0  {\ldots}       0     0       0

   Znf85  Zscan10  Zscan16  Zscan22  Zscan26  Zscan31  Zscan4
0      0        0        0        0        0        1       0
1      0        0        0        1        0        0       0
2      0        0        0        0        0        0       0
3      0        0        0        0        0        0       0
4      0        0        0        0        0        0       1

[5 rows x 1092 columns]
\end{sphinxVerbatim}
}


\subsubsection{7. Save TFinfo as dictionary or dataframe}
\label{\detokenize{notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901:7.-Save-TFinfo-as-dictionary-or-dataframe}}
We’ll use this information when making the GRNs. Save the results.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[21]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{folder} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{TFinfo\PYGZus{}outputs}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{folder}\PYG{p}{,} \PYG{n}{exist\PYGZus{}ok}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} save TFinfo as a dictionary}
\PYG{n}{td} \PYG{o}{=} \PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{dictionary\PYGZus{}type}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{targetgene2TFs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{save\PYGZus{}as\PYGZus{}pickled\PYGZus{}object}\PYG{p}{(}\PYG{n}{td}\PYG{p}{,} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{folder}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{TFinfo\PYGZus{}targetgene2TFs.pickled}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} save TFinfo as a dataframe}
\PYG{n}{df} \PYG{o}{=} \PYG{n}{tfi}\PYG{o}{.}\PYG{n}{to\PYGZus{}dataframe}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{df}\PYG{o}{.}\PYG{n}{to\PYGZus{}parquet}\PYG{p}{(}\PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{folder}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{TFinfo\PYGZus{}dataframe.parquet}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsection{Single\sphinxhyphen{}cell RNA\sphinxhyphen{}seq data preprocessing}
\label{\detokenize{tutorials/scrnaprocess:single-cell-rna-seq-data-preprocessing}}\label{\detokenize{tutorials/scrnaprocess:scrnaprocess}}\label{\detokenize{tutorials/scrnaprocess::doc}}
Network analysis and simulation in celloracle will be performed using scRNA\sphinxhyphen{}seq data. The scRNA\sphinxhyphen{}seq data should include the components below.
\begin{itemize}
\item {} 
Gene expression matrix; mRNA counts before scaling and transformation.

\item {} 
Clustering results.

\item {} 
Dimensional reduction results.

\end{itemize}

In addition to these minimum requirements, we highly recommend doing these analyses below in the preprocessing step.
\begin{itemize}
\item {} 
Data quality check and cell/gene filtering.

\item {} 
Normalization

\item {} 
Identification of highly variable genes

\end{itemize}

We recommend processing scRNA\sphinxhyphen{}seq data using either Scanpy or Seurat.
If you are not familiar with the general workflow of scRNA\sphinxhyphen{}seq data processing, please go to \sphinxhref{https://scanpy.readthedocs.io/en/stable/}{the documentation for scanpy} and \sphinxhref{https://satijalab.org/seurat/vignettes.html}{the documentation for Seurat} before celloracle analysis.

If you already have preprocessed scRNA\sphinxhyphen{}seq data, which includes the necessary information above, you can skip this part.


\subsubsection{A. scRNA\sphinxhyphen{}seq data preprocessing with scanpy}
\label{\detokenize{tutorials/scrnaprocess:a-scrna-seq-data-preprocessing-with-scanpy}}
\sphinxcode{\sphinxupquote{scanpy}} is a python library for the analysis of scRNA\sphinxhyphen{}seq data.

In this tutorial, we introduce an example of scRNA\sphinxhyphen{}seq preprocessing for celloracle with \sphinxcode{\sphinxupquote{scanpy}}.
We wrote the notebook based on \sphinxhref{https://scanpy-tutorials.readthedocs.io/en/latest/paga-paul15.html}{one of scanpy’s tutorials} with some modifications.

Python notebook


\paragraph{0. Import libraries}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:0.-Import-libraries}}\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{scanpy} \PYG{k}{as} \PYG{n+nn}{sc}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline
\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{300}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}


\paragraph{1. Load data}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:1.-Load-data}}
In this notebook, we will show an example of how to process scRNA\sphinxhyphen{}seq data using a scRNA\sphinxhyphen{}seq data of hematopoiesis (Paul, F., Arkin, Y., Giladi, A., Jaitin, D. A., Kenigsberg, E., Keren\sphinxhyphen{}Shaul, H., et al. (2015). Transcriptional Heterogeneity and Lineage Commitment in Myeloid Progenitors. Cell, 163(7), 1663\textendash{}1677. \sphinxurl{http://doi.org/10.1016/j.cell.2015.11.013}). You can easily download this scRNA\sphinxhyphen{}seq data with a scanpy function.

Please change the code below if you want to use your data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Download dataset. You can change the code blow if you use another data.}
\PYG{n}{adata} \PYG{o}{=} \PYG{n}{sc}\PYG{o}{.}\PYG{n}{datasets}\PYG{o}{.}\PYG{n}{paul15}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
WARNING: In Scanpy 0.*, this returned logarithmized data. Now it returns non-logarithmized data.
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
{\ldots} storing 'paul15\_clusters' as categorical
Trying to set attribute `.uns` of view, making a copy.
\end{sphinxVerbatim}
}


\paragraph{2. Filtering}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:2.-Filtering}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Only consider genes with more than 1 count}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{filter\PYGZus{}genes}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{min\PYGZus{}counts}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}

\end{sphinxVerbatim}
}


\paragraph{3. Normalization}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:3.-Normalization}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Normalize gene expression matrix with total UMI count per cell}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{normalize\PYGZus{}per\PYGZus{}cell}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{key\PYGZus{}n\PYGZus{}counts}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{n\PYGZus{}counts\PYGZus{}all}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{4. Identification of highly variable genes}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:4.-Identification-of-highly-variable-genes}}
Removing non\sphinxhyphen{}variable genes not only reduces the calculation time during the GRN reconstruction and simulation, but also improve the accuracy of GRN inference. We recommend using the top 2000\textasciitilde{}3000 variable genes.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Select top 2000 highly\PYGZhy{}variable genes}
\PYG{n}{filter\PYGZus{}result} \PYG{o}{=} \PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{filter\PYGZus{}genes\PYGZus{}dispersion}\PYG{p}{(}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{X}\PYG{p}{,}
                                              \PYG{n}{flavor}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}ranger}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                                              \PYG{n}{n\PYGZus{}top\PYGZus{}genes}\PYG{o}{=}\PYG{l+m+mi}{2000}\PYG{p}{,}
                                              \PYG{n}{log}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Subset the genes}
\PYG{n}{adata} \PYG{o}{=} \PYG{n}{adata}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{filter\PYGZus{}result}\PYG{o}{.}\PYG{n}{gene\PYGZus{}subset}\PYG{p}{]}

\PYG{c+c1}{\PYGZsh{} Renormalize after filtering}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{normalize\PYGZus{}per\PYGZus{}cell}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Trying to set attribute `.obs` of view, making a copy.
\end{sphinxVerbatim}
}


\paragraph{5. Log transformation}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:5.-Log-transformation}}
We will do log transformation scaling because these are necessary for PCA, clustering, and differential gene calculations. However, we also need non\sphinxhyphen{}transformed gene expression data in the celloracle analysis. Thus we keep raw count in anndata using the following command before the log transformation.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} keep raw cont data before log transformation}
\PYG{n}{adata}\PYG{o}{.}\PYG{n}{raw} \PYG{o}{=} \PYG{n}{adata}

\PYG{c+c1}{\PYGZsh{} Log transformation and scaling}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{log1p}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{scale}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{6. Dimensional reduction}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:6.-Dimensional-reduction}}
Dimensional reduction is one of the most important parts of the scRNA\sphinxhyphen{}seq analysis. Celloracle needs dimensional reduction embeddings to simulate cell transition.

Please choose a proper algorithm for dimensional reduction so that the embedding appropriately represents the data structure. We recommend using one of these dimensional reduction algorithms (or trajectory inference algorithms); UMAP, tSNE, diffusion map, force\sphinxhyphen{}directed graph drawing or PAGA.

In this example, we use a combination of four algorithms; diffusion map, force\sphinxhyphen{}directed graph drawing, and PAGA.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} PCA}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{pca}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{svd\PYGZus{}solver}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{arpack}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[10]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Diffusion map}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{neighbors}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{n\PYGZus{}neighbors}\PYG{o}{=}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{n}{n\PYGZus{}pcs}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}

\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{diffmap}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Calculate neihbors again based on diffusionmap}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pp}\PYG{o}{.}\PYG{n}{neighbors}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{n\PYGZus{}neighbors}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{n}{use\PYGZus{}rep}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{X\PYGZus{}diffmap}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{7. Clustering}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:7.-Clustering}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{louvain}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{resolution}\PYG{o}{=}\PYG{l+m+mf}{0.8}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{(Optional) Re\sphinxhyphen{}calculate Dimensional reduction graph}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:(Optional)-Re-calculate-Dimensional-reduction-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[12]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} PAGA graph construction}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{paga}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{groups}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[13]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Check current cluster name}
\PYG{n}{cluster\PYGZus{}list} \PYG{o}{=} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{louvain}\PYG{o}{.}\PYG{n}{unique}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{cluster\PYGZus{}list}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[13]:\,\hspace{\fboxrule}\hspace{\fboxsep}}[5, 2, 12, 13, 0, {\ldots}, 6, 20, 14, 15, 21]
Length: 23
Categories (23, object): [5, 2, 12, 13, {\ldots}, 20, 14, 15, 21]
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[14]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[15]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{paga}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=383\sphinxpxdimen,height=277\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_22_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[16]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{tl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{init\PYGZus{}pos}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{paga}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{123}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[17]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=364\sphinxpxdimen,height=288\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_24_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{8. Check data}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:8.-Check-data}}

\subparagraph{8.1. Visualize marker gene expression}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:8.1.-Visualize-marker-gene-expression}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[18]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[19]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{markers} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Erythroids}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Klf1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gypa}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Hba\PYGZhy{}a2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
           \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Megakaryocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Itga2b}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Pbx1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Sdpr}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Vwf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
            \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Granulocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Elane}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Cebpe}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ctsg}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mpo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gfi1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
            \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Monocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Irf8}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Csf1r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ctsg}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mpo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
            \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mast\PYGZus{}cells}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Cma1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gzmb}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Kit}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
            \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Basophils}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mcpt8}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Prss34}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}
            \PYG{p}{\PYGZcb{}}

\PYG{k}{for} \PYG{n}{cell\PYGZus{}type}\PYG{p}{,} \PYG{n}{genes} \PYG{o+ow}{in} \PYG{n}{markers}\PYG{o}{.}\PYG{n}{items}\PYG{p}{(}\PYG{p}{)}\PYG{p}{:}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{marker gene of }\PYG{l+s+si}{\PYGZob{}}\PYG{n}{cell\PYGZus{}type}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{n}{genes}\PYG{p}{,} \PYG{n}{use\PYGZus{}raw}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{n}{ncols}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}


\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Erythroids
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=656\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_27_1}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Megakaryocytes
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=644\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_27_3}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Granulocytes
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=655\sphinxpxdimen,height=856\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_27_5}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Monocytes
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=656\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_27_7}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Mast\_cells
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=644\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_27_9}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
marker gene of Basophils
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=644\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_27_11}.png}

\end{nbsphinxfancyoutput}


\paragraph{9. Make annotation for cluster}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:9.-Make-annotation-for-cluster}}
Based on the marker gene expression and previous reports, we will manually annotate each cluster. When using your own data, you will need to annotate the clusters appropriately.


\subparagraph{9.1. Make annotation (1)}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:9.1.-Make-annotation-(1)}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[20]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{paul15\PYGZus{}clusters}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_31_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[21]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Check current cluster name}
\PYG{n}{cluster\PYGZus{}list} \PYG{o}{=} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{louvain}\PYG{o}{.}\PYG{n}{unique}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{cluster\PYGZus{}list}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[21]:\,\hspace{\fboxrule}\hspace{\fboxsep}}[5, 2, 12, 13, 0, {\ldots}, 6, 20, 14, 15, 21]
Length: 23
Categories (23, object): [5, 2, 12, 13, {\ldots}, 20, 14, 15, 21]
\end{sphinxVerbatim}
}


\subparagraph{!! Please change the dictionary below depending on the clustering results. The results may change depending on the execution environment.}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:!!-Please-change-the-dictionary-below-depending-on-the-clustering-results.-The-results-may-change-depending-on-the-execution-environment.}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[22]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Make anottation dictionary}
\PYG{n}{annotation} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{5}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Erythroids}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+m+mi}{15}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{16}\PYG{p}{,} \PYG{l+m+mi}{9}\PYG{p}{,} \PYG{l+m+mi}{8}\PYG{p}{,} \PYG{l+m+mi}{14}\PYG{p}{,} \PYG{l+m+mi}{19}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{12}\PYG{p}{,} \PYG{l+m+mi}{18}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Megakaryocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{17}\PYG{p}{,} \PYG{l+m+mi}{22}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{GMP}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{late\PYGZus{}GMP}\PYG{l+s+s2}{\PYGZdq{}} \PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Granulocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{7}\PYG{p}{,} \PYG{l+m+mi}{13}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Monocytes}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DC}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{21}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lymphoid}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+m+mi}{20}\PYG{p}{]}\PYG{p}{\PYGZcb{}}

\PYG{c+c1}{\PYGZsh{} change dictionary format}
\PYG{n}{annotation\PYGZus{}rev} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{p}{\PYGZcb{}}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n}{cluster\PYGZus{}list}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{k} \PYG{o+ow}{in} \PYG{n}{annotation}\PYG{p}{:}
        \PYG{k}{if} \PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)} \PYG{o+ow}{in} \PYG{n}{annotation}\PYG{p}{[}\PYG{n}{k}\PYG{p}{]}\PYG{p}{:}
            \PYG{n}{annotation\PYGZus{}rev}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{k}

\PYG{c+c1}{\PYGZsh{} check dictionary}
\PYG{n}{annotation\PYGZus{}rev}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[22]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\{'5': 'MEP',
 '2': 'Monocytes',
 '12': 'Erythroids',
 '13': 'Granulocytes',
 '0': 'late\_GMP',
 '10': 'Erythroids',
 '3': 'Erythroids',
 '18': 'Erythroids',
 '11': 'GMP',
 '7': 'Granulocytes',
 '8': 'Erythroids',
 '22': 'Megakaryocytes',
 '16': 'Erythroids',
 '1': 'GMP',
 '17': 'Megakaryocytes',
 '4': 'Granulocytes',
 '19': 'Erythroids',
 '9': 'Erythroids',
 '6': 'Monocytes',
 '20': 'Lymphoid',
 '14': 'Erythroids',
 '15': 'Erythroids',
 '21': 'DC'\}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[23]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cell\PYGZus{}type}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{n}{annotation\PYGZus{}rev}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{louvain}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[24]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check results}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}type}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{paul15\PYGZus{}clusters}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
{\ldots} storing 'cell\_type' as categorical
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_36_1}.png}

\end{nbsphinxfancyoutput}


\subparagraph{9.2. Make annotation (2)}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:9.2.-Make-annotation-(2)}}
We’ll make another annotation manually for each Louvain clusters.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[25]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}type}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_38_0}.png}

\end{nbsphinxfancyoutput}


\subparagraph{!! Please change the dictionary below depending on the clustering results. The results may change depending on the execution environment.}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:id1}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[26]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{annotation\PYGZus{}2} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{15}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{10}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{16}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{14}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{9}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}4}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{8}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{19}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}6}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}7}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{12}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}8}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{18}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}9}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{17}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mk\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{22}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mk\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{11}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMPl\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{7}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{13}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{4}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{6}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{21}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{DC\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{20}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Lym\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{\PYGZcb{}}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[27]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{n}{annotation\PYGZus{}2}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{louvain}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[28]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Check result}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain\PYGZus{}annot}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}type}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
{\ldots} storing 'louvain\_annot' as categorical
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_42_1}.png}

\end{nbsphinxfancyoutput}

We’ve done secveral scRNA\sphinxhyphen{}preprocessing steps; filtering, normalization, clustering, and dimensional reduction. In the next step, we’ll do the GRN inference, network analysis, and in silico simulation based on this information.


\paragraph{10. (Option) Subset cells}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:10.-(Option)-Subset-cells}}
In this tutorial, we are using scRNA\sphinxhyphen{}seq data of hematopoiesis. In the latter part, we will focus on the cell fate decision in the myeloid lineage. So we will remove non\sphinxhyphen{}myeloid cell cluster; DC and Lymphoid cell cluster.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[29]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{cell\PYGZus{}type}\PYG{o}{.}\PYG{n}{unique}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[29]:\,\hspace{\fboxrule}\hspace{\fboxsep}}[MEP, Monocytes, Erythroids, Granulocytes, late\_GMP, GMP, Megakaryocytes, Lymphoid, DC]
Categories (9, object): [MEP, Monocytes, Erythroids, Granulocytes, {\ldots}, GMP, Megakaryocytes, Lymphoid, DC]
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[30]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{cell\PYGZus{}of\PYGZus{}interest} \PYG{o}{=} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{index}\PYG{p}{[}\PYG{o}{\PYGZti{}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{cell\PYGZus{}type}\PYG{o}{.}\PYG{n}{isin}\PYG{p}{(}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lymphoid}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DC}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{)}\PYG{p}{]}
\PYG{n}{adata} \PYG{o}{=} \PYG{n}{adata}\PYG{p}{[}\PYG{n}{cell\PYGZus{}of\PYGZus{}interest}\PYG{p}{,} \PYG{p}{:}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[31]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check result}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{louvain\PYGZus{}annot}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cell\PYGZus{}type}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{legend\PYGZus{}loc}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{on data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=615\sphinxpxdimen,height=292\sphinxpxdimen]{{notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_47_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{11. Save data}
\label{\detokenize{notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data:11.-Save-data}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[32]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{write\PYGZus{}h5ad}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{data/Paul\PYGZus{}etal\PYGZus{}15.h5ad}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{B. scRNA\sphinxhyphen{}seq data preprocessing with Seurat}
\label{\detokenize{tutorials/scrnaprocess:b-scrna-seq-data-preprocessing-with-seurat}}
R notebook … comming in the future update.

\begin{sphinxadmonition}{note}{Note:}
If you use \sphinxcode{\sphinxupquote{Seurat}} for preprocessing, you need to convert the scRNA\sphinxhyphen{}seq data (Seurat object) into anndata to analyze the data with \sphinxcode{\sphinxupquote{celloracle}}.
\sphinxcode{\sphinxupquote{celloracle}} has a python API and command\sphinxhyphen{}line API to convert a Seurat object into an anndata.
Please go to the documentation of celloracle’s API documentation for more information.
\end{sphinxadmonition}


\subsection{Network analysis}
\label{\detokenize{tutorials/networkanalysis:network-analysis}}\label{\detokenize{tutorials/networkanalysis:networkanalysis}}\label{\detokenize{tutorials/networkanalysis::doc}}
\sphinxcode{\sphinxupquote{celloracle}} imports the scRNA\sphinxhyphen{}seq dataset and TF binding information to find active regulatory connections for all genes, generating sample\sphinxhyphen{}specific GRNs.

The inferred GRN is analyzed with several network algorithms to get various network scores. The network score is useful to identify key regulatory genes.

Celloracle reconstructs a GRN for each cluster, enabling us to compare GRNs to each other. It is also possible to analyze how the GRN changes over differentiation.
The dynamics of the GRN structure can provide us insight into the
context\sphinxhyphen{}dependent regulatory mechanisms.

Python notebook


\subsubsection{0. Import libraries}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:0.-Import-libraries}}\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data::doc}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} 0. Import}

\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{sys}

\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{scanpy} \PYG{k}{as} \PYG{n+nn}{sc}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{celloracle} \PYG{k}{as} \PYG{n+nn}{co}
\PYG{n}{co}\PYG{o}{.}\PYG{n}{\PYGZus{}\PYGZus{}version\PYGZus{}\PYGZus{}}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/pandas1/lib/python3.6/site-packages/gimmemotifs/plot.py:19: MatplotlibDeprecationWarning: The 'warn' parameter of use() is deprecated since Matplotlib 3.1 and will be removed in 3.3.  If any parameter follows 'warn', they should be pass as keyword, not positionally.
  mpl.use("Agg", warn=False)
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}'0.4.0'
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} visualization settings}
\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{300}

\end{sphinxVerbatim}
}


\paragraph{0.1. Check installation}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:0.1.-Check-installation}}
Celloracle uses some R libraries in network analysis. Please make sure that all dependent R libraries are installed on your computer. You can test the installation with the following command.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[4]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{co}\PYG{o}{.}\PYG{n}{network\PYGZus{}analysis}\PYG{o}{.}\PYG{n}{test\PYGZus{}R\PYGZus{}libraries\PYGZus{}installation}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
R path: /usr/bin/R
checking R library installation: igraph -> OK
checking R library installation: linkcomm -> OK
checking R library installation: rnetcarto -> OK
\end{sphinxVerbatim}
}


\paragraph{0.2. Make a folder to save graph}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:0.2.-Make-a-folder-to-save-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{save\PYGZus{}folder} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figures}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{save\PYGZus{}folder}\PYG{p}{,} \PYG{n}{exist\PYGZus{}ok}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{1. Load data}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:1.-Load-data}}

\paragraph{1.1. Load processed gene expression data (anndata)}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:1.1.-Load-processed-gene-expression-data-(anndata)}}
Please refer to the previous notebook in the tutorial for an example of how to process scRNA\sphinxhyphen{}seq data.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[6]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Load data. !!Replace the data path below when you use another data.}
\PYG{n}{adata} \PYG{o}{=} \PYG{n}{sc}\PYG{o}{.}\PYG{n}{read\PYGZus{}h5ad}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{../03\PYGZus{}scRNA\PYGZhy{}seq\PYGZus{}data\PYGZus{}preprocessing/data/Paul\PYGZus{}etal\PYGZus{}15.h5ad}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/pandas1/lib/python3.6/site-packages/anndata/compat/\_\_init\_\_.py:161: FutureWarning: Moving element from .uns['neighbors']['distances'] to .obsp['distances'].

This is where adjacency matrices should go now.
  FutureWarning,
/home/k/anaconda3/envs/pandas1/lib/python3.6/site-packages/anndata/compat/\_\_init\_\_.py:161: FutureWarning: Moving element from .uns['neighbors']['connectivities'] to .obsp['connectivities'].

This is where adjacency matrices should go now.
  FutureWarning,
\end{sphinxVerbatim}
}


\paragraph{1.2. Load TF data.}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:1.2.-Load-TF-data.}}
For the GRN inference, celloracle needs TF information, which contains lists of the regulatory candidate genes. There are several ways to make such TF information. We can generate TF information from scATAC\sphinxhyphen{}seq data or bulk ATAC\sphinxhyphen{}seq data. Please refer to the first step of the tutorial for the details of this process.

If you do not have your scATAC\sphinxhyphen{}seq data, you can use some built\sphinxhyphen{}in data in celloracle. The built\sphinxhyphen{}in TFinfo wqs made using various tissue/cell\sphinxhyphen{}types from the mouse ATAC\sphinxhyphen{}seq atlas dataset (\sphinxurl{http://atlas.gs.washington.edu/mouse-atac/}).

You can load and use the data with the following command.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Load TF info which was made from mouse cell atlas dataset.}
\PYG{n}{TFinfo\PYGZus{}df} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{data}\PYG{o}{.}\PYG{n}{load\PYGZus{}TFinfo\PYGZus{}df\PYGZus{}mm9\PYGZus{}mouse\PYGZus{}atac\PYGZus{}atlas}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Check data}
\PYG{n}{TFinfo\PYGZus{}df}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[11]:\,\hspace{\fboxrule}\hspace{\fboxsep}}                     peak\_id gene\_short\_name  9430076c15rik  Ac002126.6  \textbackslash{}
0  chr10\_100050979\_100052296   4930430F08Rik            0.0         0.0
1  chr10\_101006922\_101007748         SNORA17            0.0         0.0
2  chr10\_101144061\_101145000          Mgat4c            0.0         0.0
3    chr10\_10148873\_10149183   9130014G24Rik            0.0         0.0
4    chr10\_10149425\_10149815   9130014G24Rik            0.0         0.0

   Ac012531.1  Ac226150.2  Afp  Ahr  Ahrr  Aire  {\ldots}  Znf784  Znf8  Znf816  \textbackslash{}
0         1.0         0.0  0.0  0.0   0.0   0.0  {\ldots}     0.0   0.0     0.0
1         0.0         0.0  0.0  0.0   0.0   0.0  {\ldots}     0.0   0.0     0.0
2         0.0         0.0  0.0  0.0   0.0   0.0  {\ldots}     0.0   0.0     0.0
3         0.0         0.0  0.0  0.0   0.0   0.0  {\ldots}     0.0   0.0     0.0
4         0.0         0.0  0.0  0.0   0.0   0.0  {\ldots}     0.0   0.0     0.0

   Znf85  Zscan10  Zscan16  Zscan22  Zscan26  Zscan31  Zscan4
0    0.0      0.0      0.0      0.0      0.0      0.0     0.0
1    0.0      0.0      0.0      0.0      0.0      1.0     0.0
2    0.0      0.0      0.0      0.0      0.0      0.0     1.0
3    0.0      0.0      0.0      0.0      0.0      0.0     0.0
4    0.0      0.0      0.0      0.0      0.0      0.0     0.0

[5 rows x 1095 columns]
\end{sphinxVerbatim}
}


\subsubsection{2. Initiate Oracle object}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:2.-Initiate-Oracle-object}}
Celloracle has a custom called Oracle. We can use Oracle for the data preprocessing and GRN inference steps. The Oracle object stores all of necessary information and does the calculations with its internal functions. We instantiate an Oracle object, then input the gene expression data (anndata) and a TFinfo into the Oracle object.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Instantiate Oracle object}
\PYG{n}{oracle} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{Oracle}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{2.1. load gene expression data into oracle object.}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:2.1.-load-gene-expression-data-into-oracle-object.}}
When you load a scRNA\sphinxhyphen{}seq data, please enter the name of clustering data and dimensional reduction data. The clustering data should be to be stored in the attribute of “obs” in the anndata. Dimensional reduction data suppose to be stored in the attribute of “obsm” in the anndata. You can check these data by the following command.

If you are not familiar with anndata, please look at the documentation of annata (\sphinxurl{https://anndata.readthedocs.io/en/stable/}) or Scanpy (\sphinxurl{https://scanpy.readthedocs.io/en/stable/}).

For the celloracle analysis, the anndata shoud include (1) gene expression count, (2) clustering information, (3) trajectory (dimensional reduction embeddings) data. Please refer to another notebook for more information on anndata preprocessing.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} show data name in anndata}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{metadata columns :}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{list}\PYG{p}{(}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{columns}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{dimensional reduction: }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{list}\PYG{p}{(}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obsm}\PYG{o}{.}\PYG{n}{keys}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
metadata columns : ['paul15\_clusters', 'n\_counts\_all', 'n\_counts', 'louvain', 'cell\_type', 'louvain\_annot']
dimensional reduction:  ['X\_diffmap', 'X\_draw\_graph\_fa', 'X\_pca']
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[9]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} In this notebook, we use raw mRNA count as an input of Oracle object.}
\PYG{n}{adata}\PYG{o}{.}\PYG{n}{X} \PYG{o}{=} \PYG{n}{adata}\PYG{o}{.}\PYG{n}{raw}\PYG{o}{.}\PYG{n}{X}\PYG{o}{.}\PYG{n}{copy}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Instantiate Oracle object.}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{import\PYGZus{}anndata\PYGZus{}as\PYGZus{}raw\PYGZus{}count}\PYG{p}{(}\PYG{n}{adata}\PYG{o}{=}\PYG{n}{adata}\PYG{p}{,}
                                   \PYG{n}{cluster\PYGZus{}column\PYGZus{}name}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                                   \PYG{n}{embedding\PYGZus{}name}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{X\PYGZus{}draw\PYGZus{}graph\PYGZus{}fa}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{2.2. Load TFinfo into oracle object}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:2.2.-Load-TFinfo-into-oracle-object}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[13]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} You can load TF info dataframe with the following code.}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{import\PYGZus{}TF\PYGZus{}data}\PYG{p}{(}\PYG{n}{TF\PYGZus{}info\PYGZus{}matrix}\PYG{o}{=}\PYG{n}{TFinfo\PYGZus{}df}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Alternatively, if you saved the informmation as a dictionary, you can use the code below.}
\PYG{c+c1}{\PYGZsh{} oracle.import\PYGZus{}TF\PYGZus{}data(TFdict=TFinfo\PYGZus{}dictionary)}
\end{sphinxVerbatim}
}


\paragraph{2.3. (Optional) Add TF info manually}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:2.3.-(Optional)-Add-TF-info-manually}}
While we mainly use TF info data made from scATAC\sphinxhyphen{}seq data, we can also add additional information about the TF\sphinxhyphen{}target gene pair manually.

For example, if there is a study or database that includes specific TF\sphinxhyphen{}target pairs, you can use such information in the following way.


\subparagraph{2.3.1. Make TF info dictionary manually}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:2.3.1.-Make-TF-info-dictionary-manually}}
Here, we will introduce how to add TF binding information.

We will start with TF binding data from supplemental table 4 in (\sphinxurl{http://doi.org/10.1016/j.cell.2015.11.013}).

In order to import TF data into the Oracle object, we need to convert them into a python dictionary. The dictionary keys will be the target genes, and the values will be the regulatory candidate TFs.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[50]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} We have TF and its target gene information. This is from a supplemental Fig of Paul et. al, (2015).}
\PYG{n}{Paul\PYGZus{}15\PYGZus{}data} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{TF\PYGZus{}data\PYGZus{}in\PYGZus{}Paul15.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{Paul\PYGZus{}15\PYGZus{}data}

\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[50]:\,\hspace{\fboxrule}\hspace{\fboxsep}}      TF                                       Target\_genes
0  Cebpa  Abcb1b, Acot1, C3, Cnpy3, Dhrs7, Dtx4, Edem2, {\ldots}
1   Irf8  Abcd1, Aif1, BC017643, Cbl, Ccdc109b, Ccl6, d6{\ldots}
2   Irf8  1100001G20Rik, 4732418C07Rik, 9230105E10Rik, A{\ldots}
3   Klf1  2010011I20Rik, 5730469M10Rik, Acsl6, Add2, Ank{\ldots}
4  Sfpi1  0910001L09Rik, 2310014H01Rik, 4632428N05Rik, A{\ldots}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[51]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Make dictionary: dictionary Key is TF, dictionary Value is list of target genes}
\PYG{n}{TF\PYGZus{}to\PYGZus{}TG\PYGZus{}dictionary} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{p}{\PYGZcb{}}

\PYG{k}{for} \PYG{n}{TF}\PYG{p}{,} \PYG{n}{TGs} \PYG{o+ow}{in} \PYG{n+nb}{zip}\PYG{p}{(}\PYG{n}{Paul\PYGZus{}15\PYGZus{}data}\PYG{o}{.}\PYG{n}{TF}\PYG{p}{,} \PYG{n}{Paul\PYGZus{}15\PYGZus{}data}\PYG{o}{.}\PYG{n}{Target\PYGZus{}genes}\PYG{p}{)}\PYG{p}{:}
    \PYG{c+c1}{\PYGZsh{} convert target gene to list}
    \PYG{n}{TG\PYGZus{}list} \PYG{o}{=} \PYG{n}{TGs}\PYG{o}{.}\PYG{n}{replace}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{o}{.}\PYG{n}{split}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{,}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{c+c1}{\PYGZsh{} store target gene list in a dictionary}
    \PYG{n}{TF\PYGZus{}to\PYGZus{}TG\PYGZus{}dictionary}\PYG{p}{[}\PYG{n}{TF}\PYG{p}{]} \PYG{o}{=} \PYG{n}{TG\PYGZus{}list}

\PYG{c+c1}{\PYGZsh{} We have to make a dictionary, in which a Key is Target gene and value is TF.}
\PYG{c+c1}{\PYGZsh{} We invert the dictionary above using a utility function in celloracle.}
\PYG{n}{TG\PYGZus{}to\PYGZus{}TF\PYGZus{}dictionary} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{utility}\PYG{o}{.}\PYG{n}{inverse\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{TF\PYGZus{}to\PYGZus{}TG\PYGZus{}dictionary}\PYG{p}{)}

\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(FloatProgress(value=0.0, max=178.0), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

\end{sphinxVerbatim}
}


\subparagraph{2.3.2. Add TF informatio dictionary into the oracle object}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:2.3.2.-Add-TF-informatio-dictionary-into-the-oracle-object}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[53]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Add TF information}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{addTFinfo\PYGZus{}dictionary}\PYG{p}{(}\PYG{n}{TG\PYGZus{}to\PYGZus{}TF\PYGZus{}dictionary}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{3. Knn imputation}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:3.-Knn-imputation}}
Celloracle uses almost the same strategy as velocyto for visualizing cell transitions. This process requires KNN imputation in advance.

For the KNN imputation, we need PCA and PC selection first.


\paragraph{3.1. PCA}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:3.1.-PCA}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[60]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Perform PCA}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{perform\PYGZus{}PCA}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Select important PCs}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{cumsum}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{pca}\PYG{o}{.}\PYG{n}{explained\PYGZus{}variance\PYGZus{}ratio\PYGZus{}}\PYG{p}{)}\PYG{p}{[}\PYG{p}{:}\PYG{l+m+mi}{100}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{n\PYGZus{}comps} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{where}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{diff}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{diff}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{cumsum}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{pca}\PYG{o}{.}\PYG{n}{explained\PYGZus{}variance\PYGZus{}ratio\PYGZus{}}\PYG{p}{)}\PYG{p}{)}\PYG{o}{\PYGZgt{}}\PYG{l+m+mf}{0.002}\PYG{p}{)}\PYG{p}{)}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axvline}\PYG{p}{(}\PYG{n}{n\PYGZus{}comps}\PYG{p}{,} \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{k}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{n\PYGZus{}comps}\PYG{p}{)}
\PYG{n}{n\PYGZus{}comps} \PYG{o}{=} \PYG{n+nb}{min}\PYG{p}{(}\PYG{n}{n\PYGZus{}comps}\PYG{p}{,} \PYG{l+m+mi}{50}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
45
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=380\sphinxpxdimen,height=278\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_27_1}.png}

\end{nbsphinxfancyoutput}


\paragraph{3.2. KNN imputation}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:3.2.-KNN-imputation}}
Estimate the optimal number of nearest neighbors for KNN imputation.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[63]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{n\PYGZus{}cell} \PYG{o}{=} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cell number is :}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{n\PYGZus{}cell}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
cell number is :2671
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[64]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{k} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{l+m+mf}{0.025}\PYG{o}{*}\PYG{n}{n\PYGZus{}cell}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Auto\PYGZhy{}selected k is :}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{k}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Auto-selected k is :66
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[65]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{knn\PYGZus{}imputation}\PYG{p}{(}\PYG{n}{n\PYGZus{}pca\PYGZus{}dims}\PYG{o}{=}\PYG{n}{n\PYGZus{}comps}\PYG{p}{,} \PYG{n}{k}\PYG{o}{=}\PYG{n}{k}\PYG{p}{,} \PYG{n}{balanced}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{b\PYGZus{}sight}\PYG{o}{=}\PYG{n}{k}\PYG{o}{*}\PYG{l+m+mi}{8}\PYG{p}{,}
                      \PYG{n}{b\PYGZus{}maxl}\PYG{o}{=}\PYG{n}{k}\PYG{o}{*}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{n}{n\PYGZus{}jobs}\PYG{o}{=}\PYG{l+m+mi}{4}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{4. Save and Load.}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:4.-Save-and-Load.}}
Celloracle has some custom\sphinxhyphen{}classes: Links, Oracle and TFinfo. You can save such an object using “to\_hdf5”.

Pleasae use “load\_hdf5” function to load the file.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[66]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Save oracle object.}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{to\PYGZus{}hdf5}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Paul\PYGZus{}15\PYGZus{}data.celloracle.oracle}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[19]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Load file.}
\PYG{c+c1}{\PYGZsh{}oracle = co.load\PYGZus{}hdf5(\PYGZdq{}Paul\PYGZus{}15\PYGZus{}data.celloracle.oracle\PYGZdq{})}
\end{sphinxVerbatim}
}


\subsubsection{5. GRN calculation}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:5.-GRN-calculation}}
The next step is constructing a cluster\sphinxhyphen{}specific GRN for all clusters.

You can calculate GRNs with the “get\_links” function, and the function returns GRNs as a Links object. The Links object stores inferred GRNs and the corresponding metadata. You can do network analysis with the Links object.

The GRN will be calculated for each cluster/sub\sphinxhyphen{}group. In the example below, we construct GRN for each unit of the “louvain\_annot” clustering.

The GRNs can be calculated at any arbitrary unit as long as the clustering information is stored in anndata.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[67]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} check data}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=516\sphinxpxdimen,height=288\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_37_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{5.1. Get GRNs}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:5.1.-Get-GRNs}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{o}{\PYGZpc{}\PYGZpc{}time}
\PYG{c+c1}{\PYGZsh{} Calculate GRN for each population in \PYGZdq{}louvain\PYGZus{}annot\PYGZdq{} clustering unit.}
\PYG{c+c1}{\PYGZsh{} This step may take long time.}
\PYG{n}{links} \PYG{o}{=} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{get\PYGZus{}links}\PYG{p}{(}\PYG{n}{cluster\PYGZus{}name\PYGZus{}for\PYGZus{}GRN\PYGZus{}unit}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,}
                         \PYG{n}{verbose\PYGZus{}level}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{n}{test\PYGZus{}mode}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}



\end{sphinxVerbatim}
}


\paragraph{5.2. (Optional) Export GRNs}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:5.2.-(Optional)-Export-GRNs}}
Although celloracle has many functions for network analysis, you can analyze GRNs by hand if you choose. The raw GRN data is stored in the attribute of “links\_dict”.

For example, you can get the GRN for the “Ery\_0” cluster with the following commands.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[72]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{links\PYGZus{}dict}\PYG{o}{.}\PYG{n}{keys}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[72]:\,\hspace{\fboxrule}\hspace{\fboxsep}}dict\_keys(['Ery\_0', 'Ery\_1', 'Ery\_2', 'Ery\_3', 'Ery\_4', 'Ery\_5', 'Ery\_6', 'Ery\_7', 'Ery\_8', 'Ery\_9', 'GMP\_0', 'GMP\_1', 'GMPl\_0', 'Gran\_0', 'Gran\_1', 'Gran\_2', 'MEP\_0', 'Mk\_0', 'Mo\_0', 'Mo\_1'])
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[73]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{links\PYGZus{}dict}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ery\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[73]:\,\hspace{\fboxrule}\hspace{\fboxsep}}        source         target  coef\_mean  coef\_abs             p      -logp
0        Stat3  0610007L01Rik  -0.010275  0.010275  3.476931e-07   6.458804
1        Gata1  0610007L01Rik  -0.000380  0.000380  7.598357e-01   0.119280
2        Zbtb1  0610007L01Rik   0.004452  0.004452  1.018526e-03   2.992028
3         Rara  0610007L01Rik  -0.000669  0.000669  7.065405e-01   0.150863
4          Myc  0610007L01Rik  -0.010705  0.010705  1.696471e-05   4.770454
{\ldots}        {\ldots}            {\ldots}        {\ldots}       {\ldots}           {\ldots}        {\ldots}
74420  Smarcc2            Zyx  -0.003475  0.003475  2.754236e-02   1.559999
74421     Nfe2            Zyx   0.031430  0.031430  1.461503e-11  10.835200
74422    Zbtb4            Zyx   0.001684  0.001684  1.915555e-01   0.717705
74423  Smarcc1            Zyx   0.011356  0.011356  1.843519e-04   3.734352
74424    Nfkb1            Zyx   0.010803  0.010803  1.805959e-06   5.743292

[74425 rows x 6 columns]
\end{sphinxVerbatim}
}

You can export the file as follows.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Set cluster name}
\PYG{n}{cluster} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ery\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{c+c1}{\PYGZsh{} Save as csv}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{links\PYGZus{}dict}\PYG{p}{[}\PYG{n}{cluster}\PYG{p}{]}\PYG{o}{.}\PYG{n}{to\PYGZus{}csv}\PYG{p}{(}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{raw\PYGZus{}GRN\PYGZus{}for\PYGZus{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{cluster}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{5.3. (Optional) Change order}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:5.3.-(Optional)-Change-order}}
The links object has a color information in an attribute, “palette”. This information is used for the visualization

The sample will be visualized in that order. Here we can change the order.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[75]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Show the contents of pallete}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{palette}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[75]:\,\hspace{\fboxrule}\hspace{\fboxsep}}        palette
MEP\_0   \#0FCFC0
Mk\_0    \#9CDED6
Ery\_0   \#7D87B9
Ery\_1   \#BEC1D4
Ery\_2   \#D6BCC0
Ery\_3   \#BB7784
Ery\_4   \#8E063B
Ery\_5   \#4A6FE3
Ery\_6   \#8595E1
Ery\_7   \#B5BBE3
Ery\_8   \#E6AFB9
Ery\_9   \#E07B91
GMP\_0   \#D33F6A
GMP\_1   \#11C638
GMPl\_0  \#8DD593
Mo\_0    \#D5EAE7
Mo\_1    \#F3E1EB
Gran\_0  \#C6DEC7
Gran\_1  \#EAD3C6
Gran\_2  \#F0B98D
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[76]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Change the order of pallete}
\PYG{n}{order} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mk\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}4}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}6}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}7}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}8}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}9}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMPl\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{palette} \PYG{o}{=} \PYG{n}{links}\PYG{o}{.}\PYG{n}{palette}\PYG{o}{.}\PYG{n}{loc}\PYG{p}{[}\PYG{n}{order}\PYG{p}{]}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{palette}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[76]:\,\hspace{\fboxrule}\hspace{\fboxsep}}        palette
MEP\_0   \#0FCFC0
Mk\_0    \#9CDED6
Ery\_0   \#7D87B9
Ery\_1   \#BEC1D4
Ery\_2   \#D6BCC0
Ery\_3   \#BB7784
Ery\_4   \#8E063B
Ery\_5   \#4A6FE3
Ery\_6   \#8595E1
Ery\_7   \#B5BBE3
Ery\_8   \#E6AFB9
Ery\_9   \#E07B91
GMP\_0   \#D33F6A
GMP\_1   \#11C638
GMPl\_0  \#8DD593
Mo\_0    \#D5EAE7
Mo\_1    \#F3E1EB
Gran\_0  \#C6DEC7
Gran\_1  \#EAD3C6
Gran\_2  \#F0B98D
\end{sphinxVerbatim}
}


\subsubsection{6. Network preprocessing}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:6.-Network-preprocessing}}

\paragraph{6.1. Filter network edges}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:6.1.-Filter-network-edges}}
Celloracle utilizes bagging ridge or Bayesian ridge regression to infer gene regulatory networks. These methods provide a network edge strength as a distribution rather than a point value. We can use the distribution to know the certainness of the connection.

We filter the network edges as follows.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{(}{)}%
\item {} 
Remove uncertain network edges based on the p\sphinxhyphen{}value.

\item {} 
Remove weak network edge. In this tutorial, we pick up the top 2000 edges in terms of network strength.

\end{enumerate}

The raw network data is stored as an attribute, “links\_dict,” while filtered network data is stored in “filtered\_links.” Thus the filtering function keeps raw network information rather than overwriting the data. You can come back to the filtering process to filter the data with different parameters if you want.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[78]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{filter\PYGZus{}links}\PYG{p}{(}\PYG{n}{p}\PYG{o}{=}\PYG{l+m+mf}{0.001}\PYG{p}{,} \PYG{n}{weight}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{coef\PYGZus{}abs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{thread\PYGZus{}number}\PYG{o}{=}\PYG{l+m+mi}{2000}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{6.2. Degree distribution}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:6.2.-Degree-distribution}}
In the first step, we examine the network degree distribution. Network degree, which is the number of edges for each node, is one of the important metrics used to investigate the network structure (\sphinxurl{https://en.wikipedia.org/wiki/Degree\_distribution}).

Please keep in mind that the degree distribution may change depending on the filtering threshold.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[79]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{9}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[80]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}degree\PYGZus{}distributions}\PYG{p}{(}\PYG{n}{plot\PYGZus{}model}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{/degree\PYGZus{}distribution/}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_1}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_1
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_3}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_2
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_5}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_3
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_7}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_4
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_9}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_5
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=562\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_11}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_6
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_13}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_7
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_15}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_8
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_17}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_9
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_19}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMP\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_21}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMP\_1
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_23}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMPl\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_25}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_27}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_1
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_29}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_2
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=562\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_31}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
MEP\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_33}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mk\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_35}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mo\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_37}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mo\_1
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=561\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_53_39}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[28]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{6}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}
\end{sphinxVerbatim}
}


\paragraph{5.3. Calculate netowrk score}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:5.3.-Calculate-netowrk-score}}
Next, we calculate several network score using some R libraries. Please make sure that R libraries are installed in your PC before running the command below.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[25]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Calculate network scores. It takes several minutes.}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{get\PYGZus{}score}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
processing{\ldots} batch 1/3
Ery\_0: finished.
Ery\_1: finished.
Ery\_2: finished.
Ery\_3: finished.
Ery\_4: finished.
Ery\_5: finished.
Ery\_6: finished.
Ery\_7: finished.
processing{\ldots} batch 2/3
Ery\_8: finished.
Ery\_9: finished.
GMP\_0: finished.
GMP\_1: finished.
GMPl\_0: finished.
Gran\_0: finished.
Gran\_1: finished.
Gran\_2: finished.
processing{\ldots} batch 3/3
MEP\_0: finished.
Mk\_0: finished.
Mo\_0: finished.
Mo\_1: finished.
\end{sphinxVerbatim}
}

The score is stored as a attribute called “merged\_score”, and the score will also be saved in a folder in your computer.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[82]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{merged\PYGZus{}score}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[82]:\,\hspace{\fboxrule}\hspace{\fboxsep}}       degree\_all  degree\_in  degree\_out  clustering\_coefficient  \textbackslash{}
Stat3          82          0          82                0.021981
Mycn           30          0          30                0.011494
E2f4          181          2         179                0.009822
Zbtb1          22          0          22                0.000000
Ybx1           69          9          60                0.028133

       clustering\_coefficient\_weighted  degree\_centrality\_all  \textbackslash{}
Stat3                         0.022055               0.151292
Mycn                          0.009986               0.055351
E2f4                          0.011874               0.333948
Zbtb1                         0.000000               0.040590
Ybx1                          0.027709               0.127306

       degree\_centrality\_in  degree\_centrality\_out  betweenness\_centrality  \textbackslash{}
Stat3              0.000000               0.151292                       0
Mycn               0.000000               0.055351                       0
E2f4               0.003690               0.330258                    3158
Zbtb1              0.000000               0.040590                       0
Ybx1               0.016605               0.110701                    1051

       closeness\_centrality  eigenvector\_centrality  page\_rank  \textbackslash{}
Stat3              0.000012                0.487978   0.001633
Mycn               0.000010                0.245650   0.001633
E2f4               0.000010                1.000000   0.001724
Zbtb1              0.000004                0.113727   0.001633
Ybx1               0.000004                0.385376   0.002133

       assortative\_coefficient  average\_path\_length  community\_random\_walk  \textbackslash{}
Stat3                -0.161693             2.621324                      1
Mycn                 -0.161693             2.621324                      1
E2f4                 -0.161693             2.621324                      1
Zbtb1                -0.161693             2.621324                     18
Ybx1                 -0.161693             2.621324                      2

       module  connectivity  participation           role cluster
Stat3       0      3.573858       0.647531  Connector Hub   Ery\_0
Mycn        2      1.767680       0.595556     Peripheral   Ery\_0
E2f4        0      8.575037       0.632252  Connector Hub   Ery\_0
Zbtb1       2      1.317952       0.533058     Peripheral   Ery\_0
Ybx1        3      5.306800       0.687251  Connector Hub   Ery\_0
\end{sphinxVerbatim}
}


\paragraph{6.4. Save}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:6.4.-Save}}
Save processed GRN. We use this file in the next notebook; “in silico perturbation with GRNs”.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[31]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Save Links object.}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{to\PYGZus{}hdf5}\PYG{p}{(}\PYG{n}{file\PYGZus{}path}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{links.celloracle.links}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[70]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} You can load files with the following command.}
\PYG{n}{links} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{load\PYGZus{}hdf5}\PYG{p}{(}\PYG{n}{file\PYGZus{}path}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{links.celloracle.links}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\end{sphinxVerbatim}
}


\subsubsection{7. Network analysis; Network score for each gene}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:7.-Network-analysis;-Network-score-for-each-gene}}
The Links class has many functions to visualize network score. See the documentation for the details of the functions.


\paragraph{7.1. Network score in each cluster}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:7.1.-Network-score-in-each-cluster}}
We have calculated several network scores using different centrality metrics. We can use the centrality score to identify key regulatory genes because centrality is one of the important indicators of network structure (\sphinxurl{https://en.wikipedia.org/wiki/Centrality}).

Let’s visualize genes with high network centrality.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[83]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Check cluster name}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{cluster}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[83]:\,\hspace{\fboxrule}\hspace{\fboxsep}}['Ery\_0',
 'Ery\_1',
 'Ery\_2',
 'Ery\_3',
 'Ery\_4',
 'Ery\_5',
 'Ery\_6',
 'Ery\_7',
 'Ery\_8',
 'Ery\_9',
 'GMP\_0',
 'GMP\_1',
 'GMPl\_0',
 'Gran\_0',
 'Gran\_1',
 'Gran\_2',
 'MEP\_0',
 'Mk\_0',
 'Mo\_0',
 'Mo\_1']
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[53]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Visualize top n\PYGZhy{}th genes that have high scores.}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}scores\PYGZus{}as\PYGZus{}rank}\PYG{p}{(}\PYG{n}{cluster}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{n\PYGZus{}gene}\PYG{o}{=}\PYG{l+m+mi}{30}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{/ranked\PYGZus{}score}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=304\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_65_0}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=341\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_65_1}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=304\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_65_2}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=305\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_65_3}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=296\sphinxpxdimen,height=319\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_65_4}.png}

\end{nbsphinxfancyoutput}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=342\sphinxpxdimen,height=318\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_65_5}.png}

\end{nbsphinxfancyoutput}


\paragraph{7.2. Network score comparison between two clusters}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:7.2.-Network-score-comparison-between-two-clusters}}
By comparing network scores between two clusters, we can analyze differences in GRN structure.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[54]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ticklabel\PYGZus{}format}\PYG{p}{(}\PYG{n}{style}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sci}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{axis}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{y}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{scilimits}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}comparison\PYGZus{}2D}\PYG{p}{(}\PYG{n}{value}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{eigenvector\PYGZus{}centrality}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{cluster1}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{cluster2}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{GMPl\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{percentile}\PYG{o}{=}\PYG{l+m+mi}{98}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{/score\PYGZus{}comparison}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=407\sphinxpxdimen,height=304\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_68_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[55]:\,\hspace{\fboxrule}\hspace{\fboxsep}}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ticklabel\PYGZus{}format}\PYG{p}{(}\PYG{n}{style}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sci}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{axis}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{y}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{scilimits}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}comparison\PYGZus{}2D}\PYG{p}{(}\PYG{n}{value}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{betweenness\PYGZus{}centrality}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{cluster1}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{cluster2}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{GMPl\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{percentile}\PYG{o}{=}\PYG{l+m+mi}{98}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{/score\PYGZus{}comparison}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=416\sphinxpxdimen,height=304\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_69_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[56]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ticklabel\PYGZus{}format}\PYG{p}{(}\PYG{n}{style}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sci}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{axis}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{y}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{scilimits}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}comparison\PYGZus{}2D}\PYG{p}{(}\PYG{n}{value}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{degree\PYGZus{}centrality\PYGZus{}all}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{cluster1}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MEP\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{cluster2}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{GMPl\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                               \PYG{n}{percentile}\PYG{o}{=}\PYG{l+m+mi}{98}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{/score\PYGZus{}comparison}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=392\sphinxpxdimen,height=304\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_70_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{7.3. Network score dynamics}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:7.3.-Network-score-dynamics}}
In the following session, we focus on how a gene’s network score changes during the differentiation.

Using Gata2, we will demonstrate how you can visualize networks scores for a single gene.

Gata2 is known to play an essential role in the early MEP and GMP populations. .

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[30]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Visualize Gata2 network score dynamics}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}per\PYGZus{}cluster}\PYG{p}{(}\PYG{n}{goi}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{/network\PYGZus{}score\PYGZus{}per\PYGZus{}gene/}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gata2
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=575\sphinxpxdimen,height=301\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_72_1}.png}

\end{nbsphinxfancyoutput}

If a gene have no connections in a cluster, it is impossible to calculate network degree scores. Thus the scores will not be shown. For example, Cebpa have no connection in the erythloids clusters, and there is no degree scores for Cebpa in these clusters as follows.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[38]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}per\PYGZus{}cluster}\PYG{p}{(}\PYG{n}{goi}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Cebpa}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Cebpa
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=408\sphinxpxdimen,height=301\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_74_1}.png}

\end{nbsphinxfancyoutput}

You can check filtered network edge as follows.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[39]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{cluster\PYGZus{}name} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ery\PYGZus{}0}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{filtered\PYGZus{}links\PYGZus{}df} \PYG{o}{=} \PYG{n}{links}\PYG{o}{.}\PYG{n}{filtered\PYGZus{}links}\PYG{p}{[}\PYG{n}{cluster\PYGZus{}name}\PYG{p}{]}
\PYG{n}{filtered\PYGZus{}links\PYGZus{}df}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[39]:\,\hspace{\fboxrule}\hspace{\fboxsep}}      source  target  coef\_mean  coef\_abs             p      -logp
68775  Stat3   Top2a  -0.107635  0.107635  1.976987e-14  13.703996
51655   Mycn   Prdx6  -0.096651  0.096651  8.076169e-11  10.092795
41345   Mycn     Mt1  -0.093897  0.093897  8.228218e-15  14.084694
5136    Ybx1  Anp32b   0.089403  0.089403  4.498303e-14  13.346951
41326   E2f4     Mt1   0.089261  0.089261  7.447929e-10   9.127964
\end{sphinxVerbatim}
}

You can confirm that there is no Cebpa connection in Ery\_0 cluster.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[41]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{filtered\PYGZus{}links\PYGZus{}df}\PYG{p}{[}\PYG{n}{filtered\PYGZus{}links\PYGZus{}df}\PYG{o}{.}\PYG{n}{source} \PYG{o}{==} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Cebpa}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[41]:\,\hspace{\fboxrule}\hspace{\fboxsep}}Empty DataFrame
Columns: [source, target, coef\_mean, coef\_abs, p, -logp]
Index: []
\end{sphinxVerbatim}
}


\paragraph{7.4. Gene cartography analysis}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:7.4.-Gene-cartography-analysis}}
Gene cartography is a method for gene network analysis. The method classifies gene into several groups using the network module structure and connections. It provides us an insight about the role and regulatory mechanism for each gene. For more information on gene cartography, please refer to the following paper (\sphinxurl{https://www.nature.com/articles/nature03288}).

The gene cartography will be calculated for the GRN in each cluster. Thus we can know how the gene cartography change by comparing the the score between clusters.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[58]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Plot cartography as a scatter plot}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}cartography\PYGZus{}scatter\PYGZus{}per\PYGZus{}cluster}\PYG{p}{(}\PYG{n}{scatter}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,}
                                           \PYG{n}{kde}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,}
                                           \PYG{n}{gois}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Sfpi1}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
                                           \PYG{n}{auto\PYGZus{}gene\PYGZus{}annot}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,}
                                           \PYG{n}{args\PYGZus{}dot}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{n\PYGZus{}levels}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{l+m+mi}{105}\PYG{p}{\PYGZcb{}}\PYG{p}{,}
                                           \PYG{n}{args\PYGZus{}line}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{c}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{gray}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{\PYGZcb{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{/cartography}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_1}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_1
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_3}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_2
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_5}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_3
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_7}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_4
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_9}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_5
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_11}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_6
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_13}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_7
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_15}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_8
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_17}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Ery\_9
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_19}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMP\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_21}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMP\_1
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_23}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
GMPl\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_25}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_27}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_1
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_29}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gran\_2
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_31}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
MEP\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_33}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mk\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_35}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mo\_0
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=356\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_37}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mo\_1
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=362\sphinxpxdimen,height=264\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_80_39}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[66]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Plot the summary of cartography analysis}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}cartography\PYGZus{}term}\PYG{p}{(}\PYG{n}{goi}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{/cartography}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Gata2
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=408\sphinxpxdimen,height=346\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_81_1}.png}

\end{nbsphinxfancyoutput}


\subsubsection{8. Network analysis; network score distribution}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:8.-Network-analysis;-network-score-distribution}}
Next, we visualize the distribution of network score to get insight into the global trend of the GRNs.


\paragraph{8.1. Distribution of network degree}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:8.1.-Distribution-of-network-degree}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[60]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.15}\PYG{p}{,} \PYG{n}{bottom}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylim}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mf}{0.040}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}discributions}\PYG{p}{(}\PYG{n}{values}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{degree\PYGZus{}centrality\PYGZus{}all}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{method}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{boxplot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}


\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
degree\_centrality\_all
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=390\sphinxpxdimen,height=263\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_84_1}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[61]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.15}\PYG{p}{,} \PYG{n}{bottom}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylim}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{0.40}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}score\PYGZus{}discributions}\PYG{p}{(}\PYG{n}{values}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{eigenvector\PYGZus{}centrality}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{method}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{boxplot}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}



\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
eigenvector\_centrality
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=383\sphinxpxdimen,height=263\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_85_1}.png}

\end{nbsphinxfancyoutput}


\paragraph{8.2. Distribution of netowrk entropy}
\label{\detokenize{notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data:8.2.-Distribution-of-netowrk-entropy}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[62]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.15}\PYG{p}{,} \PYG{n}{bottom}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{)}
\PYG{n}{links}\PYG{o}{.}\PYG{n}{plot\PYGZus{}network\PYGZus{}entropy\PYGZus{}distributions}\PYG{p}{(}\PYG{n}{save}\PYG{o}{=}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{save\PYGZus{}folder}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}


\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/scipy/stats/\_distn\_infrastructure.py:2614: RuntimeWarning: invalid value encountered in true\_divide
  pk = 1.0*pk / np.sum(pk, axis=0)
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/celloracle/network\_analysis/links\_object.py:345: RuntimeWarning: divide by zero encountered in log
  ent\_norm.append(en/np.log(k[i]))
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/celloracle/network\_analysis/links\_object.py:345: RuntimeWarning: invalid value encountered in double\_scalars
  ent\_norm.append(en/np.log(k[i]))
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=393\sphinxpxdimen,height=263\sphinxpxdimen]{{notebooks_04_Network_analysis_Network_analysis_with_Paul_etal_2015_data_87_1}.png}

\end{nbsphinxfancyoutput}

Using the network scores, we could pick up cluster\sphinxhyphen{}specific key TFs. Gata2, Gata1, Klf1, E2f1, for example, are known to play an essential role in MEP, and these TFs showed high network score in our GRN.

However, it is important to note that network analysis alone cannot shed light on the specific functions or roles these TFs play in cell fate determination.

In the next section, we will begin to investigate each TF’s contribution to cell fate by running GRN simulations

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[ ]:\,\hspace{\fboxrule}\hspace{\fboxsep}}
\end{sphinxVerbatim}
}


\subsection{Simulation with GRNs}
\label{\detokenize{tutorials/simulation:simulation-with-grns}}\label{\detokenize{tutorials/simulation:simulation}}\label{\detokenize{tutorials/simulation::doc}}
\sphinxcode{\sphinxupquote{celloracle}} leverage GRNs to simulate signal propagation inside a cell.
We can estimate the effect of gene perturbation by the simulation with GRNs.

Additonally, we will combine the signal propagation simulation with a cell state transition simulation. The latter simulation is performed by a python library for RNA\sphinxhyphen{}velocity analysis, called \sphinxcode{\sphinxupquote{velocyto}} .
This analysis may provide an insight into a complex system how TF controls enormous target genes to determines cell fate.

Python notebook


\subsubsection{0. Import libraries}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:0.-Import-libraries}}\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data::doc}}

\paragraph{0.1. Import public libraries}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:0.1.-Import-public-libraries}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{sys}

\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{colors} \PYG{k}{as} \PYG{n+nn}{colors}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{scanpy} \PYG{k}{as} \PYG{n+nn}{sc}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[2]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{k+kn}{import} \PYG{n+nn}{celloracle} \PYG{k}{as} \PYG{n+nn}{co}
\end{sphinxVerbatim}
}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[3]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{font.family}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{arial}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figure.figsize}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{9}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{]}
\PYG{o}{\PYGZpc{}}\PYG{k}{config} InlineBackend.figure\PYGZus{}format = \PYGZsq{}retina\PYGZsq{}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{savefig.dpi}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{600}

\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline
\end{sphinxVerbatim}
}


\paragraph{0.1. Make a folder to save graph}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:0.1.-Make-a-folder-to-save-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[5]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Make folder to save plots}
\PYG{n}{save\PYGZus{}folder} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{figures}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{save\PYGZus{}folder}\PYG{p}{,} \PYG{n}{exist\PYGZus{}ok}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{1. Load data}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:1.-Load-data}}

\paragraph{1.1. Load processed oracle object}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:1.1.-Load-processed-oracle-object}}
Load the oracle object. See the previous notebook for the notes on how to prepare the oracle object.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[7]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{oracle} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{load\PYGZus{}hdf5}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{../04\PYGZus{}Network\PYGZus{}analysis/Paul\PYGZus{}15\PYGZus{}data.celloracle.oracle}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{1.2. Load inferred GRNs}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:1.2.-Load-inferred-GRNs}}
In the previous notebook, we calculated GRNs. Now, we will use these GRNs for simulation. We import GRNs which were saved in the Links object.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[8]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links} \PYG{o}{=} \PYG{n}{co}\PYG{o}{.}\PYG{n}{load\PYGZus{}hdf5}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{../04\PYGZus{}Network\PYGZus{}analysis/links.celloracle.links}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}


\subsubsection{2. Make predictive models for simulation}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:2.-Make-predictive-models-for-simulation}}
We will fit ridge regression models again. This process takes less time than the GRN inference in the previous notebook because we only use significant TFs to predict target gene instead of all regulatory candidate TFs.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[12]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{links}\PYG{o}{.}\PYG{n}{filter\PYGZus{}links}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{get\PYGZus{}cluster\PYGZus{}specific\PYGZus{}TFdict\PYGZus{}from\PYGZus{}Links}\PYG{p}{(}\PYG{n}{links\PYGZus{}object}\PYG{o}{=}\PYG{n}{links}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{fit\PYGZus{}GRN\PYGZus{}for\PYGZus{}simulation}\PYG{p}{(}\PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{n}{use\PYGZus{}cluster\PYGZus{}specific\PYGZus{}TFdict}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
calculating GRN using cluster specicif TF dict{\ldots}
calculating GRN in Ery\_0
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1074 genes
calculating GRN in Ery\_1
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1092 genes
calculating GRN in Ery\_2
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1064 genes
calculating GRN in Ery\_3
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1105 genes
calculating GRN in Ery\_4
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1102 genes
calculating GRN in Ery\_5
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1116 genes
calculating GRN in Ery\_6
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1097 genes
calculating GRN in Ery\_7
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1062 genes
calculating GRN in Ery\_8
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1117 genes
calculating GRN in Ery\_9
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1121 genes
calculating GRN in GMP\_0
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1107 genes
calculating GRN in GMP\_1
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1104 genes
calculating GRN in GMPl\_0
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1089 genes
calculating GRN in Gran\_0
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1067 genes
calculating GRN in Gran\_1
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1076 genes
calculating GRN in Gran\_2
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1105 genes
calculating GRN in MEP\_0
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1152 genes
calculating GRN in Mk\_0
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1114 genes
calculating GRN in Mo\_0
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1085 genes
calculating GRN in Mo\_1
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
HBox(children=(IntProgress(value=0, max=1999), HTML(value='')))
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]

genes\_in\_gem: 1999
models made for 1074 genes
\end{sphinxVerbatim}
}


\subsubsection{3. in silico Perturbation\sphinxhyphen{}simulation}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:3.-in-silico-Perturbation-simulation}}
Next, we will simulate the effects of perturbing a single TF to investigate its function and regulatory mechanism. See the celloracle paper for the details and scientific premise on the algorithm.

In this notebook, we’ll show an example of the simulation; we’ll simulate knock\sphinxhyphen{}out of Gata1 gene in the hematopoiesis.

Previous studies have shown that Gata1 is one of the TFs that regulates cell fate decisions in myeloid progenitors. Additionally, Gata1 has been shown to affect erythroid cell differentiation.

Here, we will analyze Gata1 for the demonstration of celloracle; Celloracle try to recapitulate the previous findings of Gata1 gene above.


\paragraph{3.1. Check gene expression pattern.}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:3.1.-Check-gene-expression-pattern.}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[26]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Check gene expression}
\PYG{n}{goi} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gata1}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{pl}\PYG{o}{.}\PYG{n}{draw\PYGZus{}graph}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{p}{[}\PYG{n}{goi}\PYG{p}{,} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{cluster\PYGZus{}column\PYGZus{}name}\PYG{p}{]}\PYG{p}{,}
                 \PYG{n}{layer}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{imputed\PYGZus{}count}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{use\PYGZus{}raw}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=935\sphinxpxdimen,height=265\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_15_0}.png}

\end{nbsphinxfancyoutput}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[33]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Plot gene expression in histogram}
\PYG{n}{sc}\PYG{o}{.}\PYG{n}{get}\PYG{o}{.}\PYG{n}{obs\PYGZus{}df}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{p}{,} \PYG{n}{keys}\PYG{o}{=}\PYG{p}{[}\PYG{n}{goi}\PYG{p}{]}\PYG{p}{,} \PYG{n}{layer}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{imputed\PYGZus{}count}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{o}{.}\PYG{n}{hist}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=383\sphinxpxdimen,height=263\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_16_0}.png}

\end{nbsphinxfancyoutput}


\paragraph{3.2. calculate future gene expression after perturbation.}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:3.2.-calculate-future-gene-expression-after-perturbation.}}
Although you can use any gene expression value for the input of in silico perturbation, we recommend avoiding extreme values which are far from natural gene expression ranges. If you set Gata1 gene expression to 100, for example, it may lead to biologically infeasible results.

Here we simulate Gata1 KO; we predict what happens to the cells if Gata1 gene expression changed into 0.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[34]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Enter perturbation conditions to simulate signal propagation after the perturbation.}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{simulate\PYGZus{}shift}\PYG{p}{(}\PYG{n}{perturb\PYGZus{}condition}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{n}{goi}\PYG{p}{:} \PYG{l+m+mf}{0.0}\PYG{p}{\PYGZcb{}}\PYG{p}{,}
                      \PYG{n}{n\PYGZus{}propagation}\PYG{o}{=}\PYG{l+m+mi}{3}\PYG{p}{)}
\end{sphinxVerbatim}
}


\paragraph{3.3. calculate transition probability between cells}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:3.3.-calculate-transition-probability-between-cells}}
In the step above, we simulated simulated future gene expression values after perturbation. This prediction is based on itelative calculations of signal propagations within the GRN.

Next step, we will calculate the probabilityof a cell state transition based on the simulated data. Using the transition probability between cells, we can predict how a cell changes after perturbation.

This transition probability will be used in two ways.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{(}{)}%
\item {} 
Visualization of directed trjectory graph.

\item {} 
Markof simulation.

\end{enumerate}

In Step 4.2 and 4.3, we usefunctions imported from the velocytoloom class in velocyto.py. Please see the documentation of VelocytoLoom for more information. \sphinxurl{http://velocyto.org/velocyto.py/fullapi/api\_analysis.html}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[35]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Get transition probability}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{estimate\PYGZus{}transition\PYGZus{}prob}\PYG{p}{(}\PYG{n}{n\PYGZus{}neighbors}\PYG{o}{=}\PYG{l+m+mi}{200}\PYG{p}{,} \PYG{n}{knn\PYGZus{}random}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{sampled\PYGZus{}fraction}\PYG{o}{=}\PYG{l+m+mf}{0.5}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Calculate embedding}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{calculate\PYGZus{}embedding\PYGZus{}shift}\PYG{p}{(}\PYG{n}{sigma\PYGZus{}corr} \PYG{o}{=} \PYG{l+m+mf}{0.05}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Calculate global trend of cell transition}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{calculate\PYGZus{}grid\PYGZus{}arrows}\PYG{p}{(}\PYG{n}{smooth}\PYG{o}{=}\PYG{l+m+mf}{0.8}\PYG{p}{,} \PYG{n}{steps}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{40}\PYG{p}{,} \PYG{l+m+mi}{40}\PYG{p}{)}\PYG{p}{,} \PYG{n}{n\PYGZus{}neighbors}\PYG{o}{=}\PYG{l+m+mi}{300}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{nbsphinx-stderr}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3326: FutureWarning: arrays to stack must be passed as a "sequence" type such as list or tuple. Support for non-sequence iterables such as generators is deprecated as of NumPy 1.16 and will raise an error in the future.
  exec(code\_obj, self.user\_global\_ns, self.user\_ns)
WARNING:root:Nans encountered in corrcoef and corrected to 1s. If not identical cells were present it is probably a small isolated cluster converging after imputation.
\end{sphinxVerbatim}
}


\subsubsection{4. Visualization}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:4.-Visualization}}

\paragraph{4.1. Detailed directed trajectory graph}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:4.1.-Detailed-directed-trajectory-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[36]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{k+kc}{None}\PYG{p}{,}\PYG{p}{(}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{quiver\PYGZus{}scale} \PYG{o}{=} \PYG{l+m+mi}{40}


\PYG{n}{ix\PYGZus{}choice} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{size}\PYG{o}{=}\PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{/}\PYG{l+m+mf}{1.}\PYG{p}{)}\PYG{p}{,} \PYG{n}{replace}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}

\PYG{n}{embedding} \PYG{o}{=} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obsm}\PYG{p}{[}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{embedding\PYGZus{}name}\PYG{p}{]}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,}
            \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{0.8}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{,} \PYG{n}{s}\PYG{o}{=}\PYG{l+m+mi}{38}\PYG{p}{,} \PYG{n}{edgecolor}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{,} \PYG{n}{rasterized}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}

\PYG{n}{quiver\PYGZus{}kwargs}\PYG{o}{=}\PYG{n+nb}{dict}\PYG{p}{(}\PYG{n}{headaxislength}\PYG{o}{=}\PYG{l+m+mi}{7}\PYG{p}{,} \PYG{n}{headlength}\PYG{o}{=}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{n}{headwidth}\PYG{o}{=}\PYG{l+m+mi}{8}\PYG{p}{,}
                   \PYG{n}{linewidths}\PYG{o}{=}\PYG{l+m+mf}{0.35}\PYG{p}{,} \PYG{n}{width}\PYG{o}{=}\PYG{l+m+mf}{0.0045}\PYG{p}{,}\PYG{n}{edgecolors}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{k}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
                   \PYG{n}{color}\PYG{o}{=}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{colorandum}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{quiver}\PYG{p}{(}\PYG{n}{embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,}
           \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{delta\PYGZus{}embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{delta\PYGZus{}embedding}\PYG{p}{[}\PYG{n}{ix\PYGZus{}choice}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,}
           \PYG{n}{scale}\PYG{o}{=}\PYG{n}{quiver\PYGZus{}scale}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{quiver\PYGZus{}kwargs}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{off}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}plt.savefig(f\PYGZdq{}\PYGZob{}save\PYGZus{}folder\PYGZcb{}/full\PYGZus{}arrows\PYGZob{}goi\PYGZcb{}.png\PYGZdq{},  transparent=True)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[36]:\,\hspace{\fboxrule}\hspace{\fboxsep}}(-10815.27020913708, 10950.84121716522, -10711.36365432337, 10949.477199695968)
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=398\sphinxpxdimen,height=357\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_22_1}.png}

\end{nbsphinxfancyoutput}


\paragraph{4.2. Grid graph}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:4.2.-Grid-graph}}
{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[37]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Plot whole graph}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{k+kc}{None}\PYG{p}{,}\PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}grid\PYGZus{}arrows}\PYG{p}{(}\PYG{n}{quiver\PYGZus{}scale}\PYG{o}{=}\PYG{l+m+mf}{2.0}\PYG{p}{,}
                        \PYG{n}{scatter\PYGZus{}kwargs\PYGZus{}dict}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{alpha}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+m+mf}{0.35}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{lw}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+m+mf}{0.35}\PYG{p}{,}
                                              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{edgecolor}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{0.4}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{s}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{l+m+mi}{38}\PYG{p}{,}
                                              \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{rasterized}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:}\PYG{k+kc}{True}\PYG{p}{\PYGZcb{}}\PYG{p}{,}
                        \PYG{n}{min\PYGZus{}mass}\PYG{o}{=}\PYG{l+m+mf}{0.015}\PYG{p}{,} \PYG{n}{angles}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{xy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{scale\PYGZus{}units}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{xy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                        \PYG{n}{headaxislength}\PYG{o}{=}\PYG{l+m+mf}{2.75}\PYG{p}{,}
                        \PYG{n}{headlength}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{n}{headwidth}\PYG{o}{=}\PYG{l+m+mf}{4.8}\PYG{p}{,} \PYG{n}{minlength}\PYG{o}{=}\PYG{l+m+mf}{1.5}\PYG{p}{,}
                        \PYG{n}{plot\PYGZus{}random}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{n}{scale\PYGZus{}type}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{relative}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}plt.savefig(f\PYGZdq{}\PYGZob{}save\PYGZus{}folder\PYGZcb{}/vectorfield\PYGZus{}\PYGZob{}goi\PYGZcb{}.png\PYGZdq{}, transparent=True)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=619\sphinxpxdimen,height=574\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_24_0}.png}

\end{nbsphinxfancyoutput}


\subsubsection{5. Markov simulation to analyze the effects of perturbation on cell fate transition}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:5.-Markov-simulation-to-analyze-the-effects-of-perturbation-on-cell-fate-transition}}
We can also simulate cell state transition using Markof simulation.


\subsubsection{5.1. Do Markov simulation}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:5.1.-Do-Markov-simulation}}
We will simulate using the parameters, “n\_steps=200” and “n\_duplication=5” in the following example.

To elaborate, this means:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{(}{)}%
\item {} 
We will do 200 times of iterative simulations to predict how the cell changes over time

\item {} 
We will repeat 5 rounds of simulations

\end{enumerate}

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[83]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{o}{\PYGZpc{}\PYGZpc{}time}
\PYG{c+c1}{\PYGZsh{} n\PYGZus{}steps is the number of steps in markov simulation.}
\PYG{c+c1}{\PYGZsh{} n\PYGZus{}duplication is the number of technical duplication for the simulation}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{run\PYGZus{}markov\PYGZus{}chain\PYGZus{}simulation}\PYG{p}{(}\PYG{n}{n\PYGZus{}steps}\PYG{o}{=}\PYG{l+m+mi}{200}\PYG{p}{,} \PYG{n}{n\PYGZus{}duplication}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
CPU times: user 1.33 s, sys: 0 ns, total: 1.33 s
Wall time: 1.33 s
\end{sphinxVerbatim}
}


\paragraph{5.2. Check the results of the simulation for specific cells}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:5.2.-Check-the-results-of-the-simulation-for-specific-cells}}
Check the results of simulation. Pick up some cells and visualize their transition trajectory.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[88]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Randomly pick up 3 cells}
\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{12}\PYG{p}{)}
\PYG{n}{cells} \PYG{o}{=} \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{adata}\PYG{o}{.}\PYG{n}{obs}\PYG{o}{.}\PYG{n}{index}\PYG{o}{.}\PYG{n}{values}\PYG{p}{[}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{ixs\PYGZus{}mcmc}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{)}\PYG{p}{]}

\PYG{c+c1}{\PYGZsh{} Visualize the simulated results of cell transition after perturbation}
\PYG{k}{for} \PYG{n}{k} \PYG{o+ow}{in} \PYG{n}{cells}\PYG{p}{:}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cell }\PYG{l+s+si}{\PYGZob{}}\PYG{n}{k}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{9}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{]}\PYG{p}{)}
    \PYG{k}{for} \PYG{n}{j}\PYG{p}{,} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{20}\PYG{p}{,} \PYG{l+m+mi}{50}\PYG{p}{]}\PYG{p}{)}\PYG{p}{:} \PYG{c+c1}{\PYGZsh{} time points}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{p}{(}\PYG{n}{j}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
        \PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}mc\PYGZus{}result\PYGZus{}as\PYGZus{}trajectory}\PYG{p}{(}\PYG{n}{k}\PYG{p}{,} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{i}\PYG{p}{)}\PYG{p}{)}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{f}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{simulation step: 0\PYGZti{}}\PYG{l+s+si}{\PYGZob{}}\PYG{n}{i}\PYG{l+s+si}{\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{off}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}
}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
cell 1961
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=573\sphinxpxdimen,height=209\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_32_1}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
cell 43
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=573\sphinxpxdimen,height=209\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_32_3}.png}

\end{nbsphinxfancyoutput}

{

\kern-\sphinxverbatimsmallskipamount\kern-\baselineskip
\kern+\FrameHeightAdjust\kern-\fboxrule
\vspace{\nbsphinxcodecellspacing}

\sphinxsetup{VerbatimColor={named}{white}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
cell 1567
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=573\sphinxpxdimen,height=209\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_32_5}.png}

\end{nbsphinxfancyoutput}


\paragraph{5.3. Summarize the results of simulation by plotting sankey diagram}
\label{\detokenize{notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data:5.3.-Summarize-the-results-of-simulation-by-plotting-sankey-diagram}}
Sankey diagrams are useful when you want to visualize proportional cell transitions between some groups.

For the grouping of cells, you can use arbitrary cluster unit.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[89]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{c+c1}{\PYGZsh{} Plot sankey diagram}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{cl} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}mc\PYGZus{}resutls\PYGZus{}as\PYGZus{}sankey}\PYG{p}{(}\PYG{n}{cluster\PYGZus{}use}\PYG{o}{=}\PYG{n}{cl}\PYG{p}{,} \PYG{n}{start}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{end}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=357\sphinxpxdimen,height=357\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_35_0}.png}

\end{nbsphinxfancyoutput}

The Sankey diagram above looks messy because the cluster order is random.

Let’s change the cluster order and make the plot again

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[90]:\,\hspace{\fboxrule}\hspace{\fboxsep}}
\PYG{n}{cl} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{louvain\PYGZus{}annot}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{order} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mk\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}4}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}6}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}7}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}8}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ery\PYGZus{}9}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMPl\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMPl\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
         \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mo\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}0}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gran\PYGZus{}3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.3}\PYG{p}{,} \PYG{n}{right}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}mc\PYGZus{}resutls\PYGZus{}as\PYGZus{}sankey}\PYG{p}{(}\PYG{n}{cluster\PYGZus{}use}\PYG{o}{=}\PYG{n}{cl}\PYG{p}{,}  \PYG{n}{start}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{end}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,} \PYG{n}{order}\PYG{o}{=}\PYG{n}{order}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}plt.savefig(f\PYGZdq{}\PYGZob{}save\PYGZus{}folder\PYGZcb{}/mcmc\PYGZus{}\PYGZob{}cl\PYGZcb{}.png\PYGZdq{})}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=226\sphinxpxdimen,height=357\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_37_0}.png}

\end{nbsphinxfancyoutput}

Make another Saneky diagram with different cluster units.

{
\sphinxsetup{VerbatimColor={named}{nbsphinx-code-bg}}
\sphinxsetup{VerbatimBorderColor={named}{nbsphinx-code-border}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxin}[92]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{n}{order} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Megakaryocytes}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEP}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Erythroids}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{GMP}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{late\PYGZus{}GMP}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Monocytes}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Granulocytes}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{cl} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{cell\PYGZus{}type}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.35}\PYG{p}{,} \PYG{n}{right}\PYG{o}{=}\PYG{l+m+mf}{0.65}\PYG{p}{)}
\PYG{n}{oracle}\PYG{o}{.}\PYG{n}{plot\PYGZus{}mc\PYGZus{}resutls\PYGZus{}as\PYGZus{}sankey}\PYG{p}{(}\PYG{n}{cluster\PYGZus{}use}\PYG{o}{=}\PYG{n}{cl}\PYG{p}{,} \PYG{n}{start}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{end}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,} \PYG{n}{order}\PYG{o}{=}\PYG{n}{order}\PYG{p}{,} \PYG{n}{font\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mi}{14}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}plt.savefig(f\PYGZdq{}\PYGZob{}save\PYGZus{}folder\PYGZcb{}/mcmc\PYGZus{}\PYGZob{}cl\PYGZcb{}\PYGZob{}goi\PYGZcb{}.png\PYGZdq{}, transparent=True)}
\end{sphinxVerbatim}
}

\hrule height -\fboxrule\relax
\vspace{\nbsphinxcodecellspacing}

\makeatletter\setbox\nbsphinxpromptbox\box\voidb@x\makeatother

\begin{nbsphinxfancyoutput}

\noindent\sphinxincludegraphics[width=346\sphinxpxdimen,height=360\sphinxpxdimen]{{notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_39_0}.png}

\end{nbsphinxfancyoutput}

Based on the results, we may conclude several things as follows.

Gata1 KO induced both cell state transitions from Erythroids to MEP, and from MEP to GMP.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{(}{)}%
\item {} 
These results suggest that Gata1 may play a role in the progression of Erythroid differentiation and cell state determination between the MEP and GMP lineages.

\item {} 
Gata1 KO also induced cell state transitions from granulocytes to late GMP, suggesting Gata1’s involvement in Granulocytes differentiation.

\end{enumerate}

These results agree with previous reports about Gata1 and recapitulate Gata1’s cell\sphinxhyphen{}type\sphinxhyphen{}specific function regarding the cell fate decisions in hematopoiesis.


\section{API}
\label{\detokenize{modules/index:api}}\label{\detokenize{modules/index::doc}}

\subsection{Command Line API}
\label{\detokenize{modules/index:command-line-api}}
CellOracle has a command line API.
This command can be used to convert scRNA\sphinxhyphen{}seq data.
If you have a scRNA\sphinxhyphen{}seq data which was processed with Seurat and saved as Rds file, you can use the following command to make anndata from Seurat object.
The anndata object produced by this command can be used for input of celloracle.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{seuratToAnndata} \PYG{n}{YOUR\PYGZus{}SEURAT\PYGZus{}OBJECT}\PYG{o}{.}\PYG{n}{Rds} \PYG{n}{OUTPUT\PYGZus{}PATH}
\end{sphinxVerbatim}


\subsection{Python API}
\label{\detokenize{modules/index:python-api}}

\subsubsection{Custom class in celloracle}
\label{\detokenize{modules/celloracle:custom-class-in-celloracle}}\label{\detokenize{modules/celloracle::doc}}
We define some custom classes in celloracle.

\phantomsection\label{\detokenize{modules/celloracle:module-celloracle}}\index{module@\spxentry{module}!celloracle@\spxentry{celloracle}}\index{celloracle@\spxentry{celloracle}!module@\spxentry{module}}\index{Links (class in celloracle)@\spxentry{Links}\spxextra{class in celloracle}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.}}\sphinxbfcode{\sphinxupquote{Links}}}{\emph{\DUrole{n}{name}}, \emph{\DUrole{n}{links\_dict}\DUrole{o}{=}\DUrole{default_value}{\{\}}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

This is a class for the processing and visualization of GRNs.
Links object stores cluster\sphinxhyphen{}specific GRNs and metadata.
Please use “get\_links” function in Oracle object to generate Links object.
\index{links\_dict (celloracle.Links attribute)@\spxentry{links\_dict}\spxextra{celloracle.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.links_dict}}\pysigline{\sphinxbfcode{\sphinxupquote{links\_dict}}}
Dictionary that store unprocessed network data.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{filtered\_links (celloracle.Links attribute)@\spxentry{filtered\_links}\spxextra{celloracle.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.filtered_links}}\pysigline{\sphinxbfcode{\sphinxupquote{filtered\_links}}}
Dictionary that store filtered network data.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{merged\_score (celloracle.Links attribute)@\spxentry{merged\_score}\spxextra{celloracle.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.merged_score}}\pysigline{\sphinxbfcode{\sphinxupquote{merged\_score}}}
Network scores.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{cluster (celloracle.Links attribute)@\spxentry{cluster}\spxextra{celloracle.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.cluster}}\pysigline{\sphinxbfcode{\sphinxupquote{cluster}}}
List of cluster name.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
list of str

\end{description}\end{quote}

\end{fulllineitems}

\index{name (celloracle.Links attribute)@\spxentry{name}\spxextra{celloracle.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.name}}\pysigline{\sphinxbfcode{\sphinxupquote{name}}}
Name of clustering unit.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{palette (celloracle.Links attribute)@\spxentry{palette}\spxextra{celloracle.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.palette}}\pysigline{\sphinxbfcode{\sphinxupquote{palette}}}
DataFrame that store color information.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{filter\_links() (celloracle.Links method)@\spxentry{filter\_links()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.filter_links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{filter\_links}}}{\emph{\DUrole{n}{p}\DUrole{o}{=}\DUrole{default_value}{0.001}}, \emph{\DUrole{n}{weight}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}coef\_abs\textquotesingle{}}}, \emph{\DUrole{n}{thread\_number}\DUrole{o}{=}\DUrole{default_value}{10000}}, \emph{\DUrole{n}{genelist\_source}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{genelist\_target}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Filter network edges.
In most cases, inferred GRN has non\sphinxhyphen{}significant random edges.
We have to remove these edges before analyzing the network structure.
You can do the filtering in any of the following ways.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{(}{)}%
\item {} 
Filter based on the p\sphinxhyphen{}value of the network edge.
Please enter p\sphinxhyphen{}value for thresholding.

\item {} 
Filter based on network edge number.
If you set the number, network edges will be filtered based on the order of a network score. The top n\sphinxhyphen{}th network edges with network weight will remain, and the other edges will be removed.
The network data has several types of network weight, so you have to select which network weight do you want to use.

\item {} 
Filter based on an arbitrary gene list. You can set a gene list for source nodes or target nodes.

\end{enumerate}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{p}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} threshold for p\sphinxhyphen{}value of the network edge.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{weight}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Please select network weight name for the filtering

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{genelist\_source}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list to remain in regulatory gene nodes. Default is None.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{genelist\_target}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list to remain in target gene nodes. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_network\_entropy() (celloracle.Links method)@\spxentry{get\_network\_entropy()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.get_network_entropy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_network\_entropy}}}{\emph{\DUrole{n}{value}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}coef\_abs\textquotesingle{}}}}{}
Calculate network entropy scores.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{value}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Default is “coef\_abs”.

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_score() (celloracle.Links method)@\spxentry{get\_score()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.get_score}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_score}}}{\emph{\DUrole{n}{test\_mode}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
Get several network sores using R libraries.
Make sure all dependent R libraries are installed in your environment before running this function.
You can check the installation for the R libraries by running test\_installation() in network\_analysis module.

\end{fulllineitems}

\index{plot\_cartography\_scatter\_per\_cluster() (celloracle.Links method)@\spxentry{plot\_cartography\_scatter\_per\_cluster()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_cartography_scatter_per_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cartography\_scatter\_per\_cluster}}}{\emph{\DUrole{n}{gois}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{clusters}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{scatter}\DUrole{o}{=}\DUrole{default_value}{True}}, \emph{\DUrole{n}{kde}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{auto\_gene\_annot}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{percentile}\DUrole{o}{=}\DUrole{default_value}{98}}, \emph{\DUrole{n}{args\_dot}\DUrole{o}{=}\DUrole{default_value}{\{\textquotesingle{}n\_levels\textquotesingle{}: 105\}}}, \emph{\DUrole{n}{args\_line}\DUrole{o}{=}\DUrole{default_value}{\{\textquotesingle{}c\textquotesingle{}: \textquotesingle{}gray\textquotesingle{}\}}}, \emph{\DUrole{n}{args\_annot}\DUrole{o}{=}\DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Make a gene network cartography plot.
Please read the original paper describing gene network cartography for more information.
\sphinxurl{https://www.nature.com/articles/nature03288}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gois}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of srt}}) \textendash{} List of gene name to highlight.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{clusters}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} List of cluster name to analyze. If None, all clusters in Links object will be analyzed.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scatter}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to make a scatter plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{auto\_gene\_annot}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to pick up genes to make an annotation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{percentile}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Genes with a network score above the percentile will be shown with annotation. Default is 98.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_dot}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for scatter plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_line}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for lines in cartography plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_annot}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for annotation in plots.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_cartography\_term() (celloracle.Links method)@\spxentry{plot\_cartography\_term()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_cartography_term}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cartography\_term}}}{\emph{\DUrole{n}{goi}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the gene network cartography term like a heatmap.
Please read the original paper of gene network cartography for the principle of gene network cartography.
\sphinxurl{https://www.nature.com/articles/nature03288}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gois}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of srt}}) \textendash{} List of gene name to highlight.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_degree\_distributions() (celloracle.Links method)@\spxentry{plot\_degree\_distributions()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_degree_distributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_degree\_distributions}}}{\emph{\DUrole{n}{plot\_model}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the network degree distributions (the number of edge per gene).
The network degree will be visualized in both linear scale and log scale.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plot\_model}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to plot linear approximation line.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_network\_entropy\_distributions() (celloracle.Links method)@\spxentry{plot\_network\_entropy\_distributions()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_network_entropy_distributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_network\_entropy\_distributions}}}{\emph{\DUrole{n}{update\_network\_entropy}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the distribution for network entropy.
See the CellOracle paper for more detail.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} (\sphinxstyleliteralemphasis{\sphinxupquote{Links object}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{values}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The list of score to visualize. If it is None, all network score (listed above) will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{update\_network\_entropy}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to recalculate network entropy.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_comparison\_2D() (celloracle.Links method)@\spxentry{plot\_score\_comparison\_2D()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_score_comparison_2D}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_comparison\_2D}}}{\emph{\DUrole{n}{value}}, \emph{\DUrole{n}{cluster1}}, \emph{\DUrole{n}{cluster2}}, \emph{\DUrole{n}{percentile}\DUrole{o}{=}\DUrole{default_value}{99}}, \emph{\DUrole{n}{annot\_shifts}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Make a scatter plot that compares specific network scores in two groups.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{value}} (\sphinxstyleliteralemphasis{\sphinxupquote{srt}}) \textendash{} The network score type.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster1}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name. Network scores in cluster1 will be visualized in the x\sphinxhyphen{}axis.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster2}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name. Network scores in cluster2 will be visualized in the y\sphinxhyphen{}axis.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{percentile}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Genes with a network score above the percentile will be shown with annotation. Default is 99.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{annot\_shifts}} (\sphinxstyleliteralemphasis{\sphinxupquote{(}}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{)}}) \textendash{} Annotation visualization setting.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_discributions() (celloracle.Links method)@\spxentry{plot\_score\_discributions()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_score_discributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_discributions}}}{\emph{\DUrole{n}{values}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{method}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}boxplot\textquotesingle{}}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the distribution of network scores.
An individual data point is a network edge (gene).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See Links class for details.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{values}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The list of score to visualize. If it is None, all of the network score will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Plotting method. Select either “boxplot” or “barplot”.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_per\_cluster() (celloracle.Links method)@\spxentry{plot\_score\_per\_cluster()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_score_per_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_per\_cluster}}}{\emph{\DUrole{n}{goi}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot network score for a gene.
This function visualizes the network score for a specific gene between clusters to get an insight into the dynamics of the gene.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{goi}} (\sphinxstyleliteralemphasis{\sphinxupquote{srt}}) \textendash{} Gene name.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_scores\_as\_rank() (celloracle.Links method)@\spxentry{plot\_scores\_as\_rank()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.plot_scores_as_rank}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_scores\_as\_rank}}}{\emph{\DUrole{n}{cluster}}, \emph{\DUrole{n}{n\_gene}\DUrole{o}{=}\DUrole{default_value}{50}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Pick up top n\sphinxhyphen{}th genes wich high\sphinxhyphen{}network scores and make plots.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name to analyze.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_gene}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Number of genes to plot. Default is 50.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.Links method)@\spxentry{to\_hdf5()}\spxextra{celloracle.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Links.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{\DUrole{n}{file\_path}}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.links’

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{Net (class in celloracle)@\spxentry{Net}\spxextra{class in celloracle}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.}}\sphinxbfcode{\sphinxupquote{Net}}}{\emph{\DUrole{n}{gene\_expression\_matrix}}, \emph{\DUrole{n}{gem\_standerdized}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{TFinfo\_matrix}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{cellstate}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{TFinfo\_dic}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{annotation}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

Net is a custom class for inferring sample\sphinxhyphen{}specific GRN from scRNA\sphinxhyphen{}seq data.
This class is used inside the Oracle class for GRN inference.
This class requires two types of information below.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{(}{)}%
\item {} 
Single\sphinxhyphen{}cell RNA\sphinxhyphen{}seq data:
The Net class needs processed scRNA\sphinxhyphen{}seq data.
Gene and cell filtering, quality check, normalization, log\sphinxhyphen{}transformation (but not scaling and centering) have to be done before starting the GRN calculation with this class.
You can also use any arbitrary metadata (i.e., mRNA count, cell\sphinxhyphen{}cycle phase) for GRN input.

\item {} 
Potential regulatory connection (or base GRN):
This method uses the list of potential regulatory TFs as input.
This information can be calculated from ATAC\sphinxhyphen{}seq data using the motif\sphinxhyphen{}analysis module.
If sample\sphinxhyphen{}specific ATAC\sphinxhyphen{}seq data is not available,
you can use general TF\sphinxhyphen{}binding info derived from public ATAC\sphinxhyphen{}seq dataset of various tissue/cell type.

\end{enumerate}
\index{linkList (celloracle.Net attribute)@\spxentry{linkList}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.linkList}}\pysigline{\sphinxbfcode{\sphinxupquote{linkList}}}
The results of the GRN inference.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{all\_genes (celloracle.Net attribute)@\spxentry{all\_genes}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.all_genes}}\pysigline{\sphinxbfcode{\sphinxupquote{all\_genes}}}
An array of all genes that exist in the input gene expression matrix
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
numpy.array

\end{description}\end{quote}

\end{fulllineitems}

\index{embedding\_name (celloracle.Net attribute)@\spxentry{embedding\_name}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.embedding_name}}\pysigline{\sphinxbfcode{\sphinxupquote{embedding\_name}}}
The key name name in adata.obsm containing dimensional reduction coordinates
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{annotation (celloracle.Net attribute)@\spxentry{annotation}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.annotation}}\pysigline{\sphinxbfcode{\sphinxupquote{annotation}}}
Annotation. you can add custom annotation.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{coefs\_dict (celloracle.Net attribute)@\spxentry{coefs\_dict}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.coefs_dict}}\pysigline{\sphinxbfcode{\sphinxupquote{coefs\_dict}}}
Coefs of linear regression.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{stats\_dict (celloracle.Net attribute)@\spxentry{stats\_dict}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.stats_dict}}\pysigline{\sphinxbfcode{\sphinxupquote{stats\_dict}}}
Statistic values about coefs.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{fitted\_genes (celloracle.Net attribute)@\spxentry{fitted\_genes}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.fitted_genes}}\pysigline{\sphinxbfcode{\sphinxupquote{fitted\_genes}}}
List of genes where the regression model was successfully calculated.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
list of str

\end{description}\end{quote}

\end{fulllineitems}

\index{failed\_genes (celloracle.Net attribute)@\spxentry{failed\_genes}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.failed_genes}}\pysigline{\sphinxbfcode{\sphinxupquote{failed\_genes}}}
List of genes that were not assigned coefs
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
list of str

\end{description}\end{quote}

\end{fulllineitems}

\index{cellstate (celloracle.Net attribute)@\spxentry{cellstate}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.cellstate}}\pysigline{\sphinxbfcode{\sphinxupquote{cellstate}}}
A metadata for GRN input
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{TFinfo (celloracle.Net attribute)@\spxentry{TFinfo}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.TFinfo}}\pysigline{\sphinxbfcode{\sphinxupquote{TFinfo}}}
Information about potential regulatory TFs.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{gem (celloracle.Net attribute)@\spxentry{gem}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.gem}}\pysigline{\sphinxbfcode{\sphinxupquote{gem}}}
Merged matrix made with gene\_expression\_matrix and cellstate matrix.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{gem\_standerdized (celloracle.Net attribute)@\spxentry{gem\_standerdized}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.gem_standerdized}}\pysigline{\sphinxbfcode{\sphinxupquote{gem\_standerdized}}}
Almost the same as gem, but the gene\_expression\_matrix was standardized.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{library\_last\_update\_date (celloracle.Net attribute)@\spxentry{library\_last\_update\_date}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.library_last_update_date}}\pysigline{\sphinxbfcode{\sphinxupquote{library\_last\_update\_date}}}
Last update date of this code. This info is for code development. It can be deprecated in the future
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{object\_initiation\_date (celloracle.Net attribute)@\spxentry{object\_initiation\_date}\spxextra{celloracle.Net attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.object_initiation_date}}\pysigline{\sphinxbfcode{\sphinxupquote{object\_initiation\_date}}}
The date when this object was made.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{addAnnotation() (celloracle.Net method)@\spxentry{addAnnotation()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.addAnnotation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addAnnotation}}}{\emph{\DUrole{n}{annotation\_dictionary}}}{}
Add a new annotation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{annotation\_dictionary}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} e.g. \{“sample\_name”: “NIH 3T3 cell”\}

\end{description}\end{quote}

\end{fulllineitems}

\index{addTFinfo\_dictionary() (celloracle.Net method)@\spxentry{addTFinfo\_dictionary()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.addTFinfo_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addTFinfo\_dictionary}}}{\emph{\DUrole{n}{TFdict}}}{}
Add a new TF info to pre\sphinxhyphen{}exiting TFdict.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} python dictionary of TF info.

\end{description}\end{quote}

\end{fulllineitems}

\index{addTFinfo\_matrix() (celloracle.Net method)@\spxentry{addTFinfo\_matrix()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.addTFinfo_matrix}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addTFinfo\_matrix}}}{\emph{\DUrole{n}{TFinfo\_matrix}}}{}
Load TF info dataframe.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFinfo}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.DataFrame}}) \textendash{} information about potential regulatory TFs.

\end{description}\end{quote}

\end{fulllineitems}

\index{copy() (celloracle.Net method)@\spxentry{copy()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.copy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{copy}}}{}{}
Deepcopy itself

\end{fulllineitems}

\index{fit\_All\_genes() (celloracle.Net method)@\spxentry{fit\_All\_genes()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.fit_All_genes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_All\_genes}}}{\emph{\DUrole{n}{bagging\_number}\DUrole{o}{=}\DUrole{default_value}{200}}, \emph{\DUrole{n}{scaling}\DUrole{o}{=}\DUrole{default_value}{True}}, \emph{\DUrole{n}{model\_method}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}bagging\_ridge\textquotesingle{}}}, \emph{\DUrole{n}{command\_line\_mode}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{log}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{alpha}\DUrole{o}{=}\DUrole{default_value}{1}}, \emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Make ML models for all genes.
The calculation will be performed in parallel using scikit\sphinxhyphen{}learn bagging function.
You can select a modeling method (bagging\_ridge or bayesian\_ridge).  This calculation usually takes a long time.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of estimators for bagging.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scaling}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to scale regulatory gene expression values.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{model\_method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} ML model name. Please select either “bagging\_ridge” or “bayesian\_ridge”

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{command\_line\_mode}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Please select False if the calculation is performed on jupyter notebook.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log}} (\sphinxstyleliteralemphasis{\sphinxupquote{logging object}}) \textendash{} log object to output log

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Strength of regularization.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to show a progress bar.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{fit\_All\_genes\_parallel() (celloracle.Net method)@\spxentry{fit\_All\_genes\_parallel()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.fit_All_genes_parallel}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_All\_genes\_parallel}}}{\emph{\DUrole{n}{bagging\_number}\DUrole{o}{=}\DUrole{default_value}{200}}, \emph{\DUrole{n}{scaling}\DUrole{o}{=}\DUrole{default_value}{True}}, \emph{\DUrole{n}{log}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{10}}}{}
IMPORTANT: this function being debugged and is currently unavailable.

Make ML models for all genes.
The calculation will be performed in parallel using joblib parallel module.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of estimators for bagging.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scaling}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to scale regulatory gene expression values.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log}} (\sphinxstyleliteralemphasis{\sphinxupquote{logging object}}) \textendash{} log object to output log

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} verbose for joblib parallel

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{fit\_genes() (celloracle.Net method)@\spxentry{fit\_genes()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.fit_genes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_genes}}}{\emph{\DUrole{n}{target\_genes}}, \emph{\DUrole{n}{bagging\_number}\DUrole{o}{=}\DUrole{default_value}{200}}, \emph{\DUrole{n}{scaling}\DUrole{o}{=}\DUrole{default_value}{True}}, \emph{\DUrole{n}{model\_method}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}bagging\_ridge\textquotesingle{}}}, \emph{\DUrole{n}{save\_coefs}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{command\_line\_mode}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{log}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{alpha}\DUrole{o}{=}\DUrole{default_value}{1}}, \emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Make ML models for genes of interest.
This calculation will be performed in parallel using scikit\sphinxhyphen{}learn’s bagging function.
You can select a modeling method; Please chose either bagging\_ridge or bayesian\_ridge.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{target\_genes}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of estimators for bagging.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scaling}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to scale regulatory gene expression values.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{model\_method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} ML model name. Please select either “bagging\_ridge” or “bayesian\_ridge”

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save\_coefs}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to store details of coef values in bagging model.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{command\_line\_mode}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Please select False if the calculation is performed on jupyter notebook.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log}} (\sphinxstyleliteralemphasis{\sphinxupquote{logging object}}) \textendash{} log object to output log

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Strength of regularization.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to show a progress bar.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plotCoefs() (celloracle.Net method)@\spxentry{plotCoefs()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.plotCoefs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plotCoefs}}}{\emph{\DUrole{n}{target\_gene}}, \emph{\DUrole{n}{sort}\DUrole{o}{=}\DUrole{default_value}{True}}, \emph{\DUrole{n}{threshold\_p}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the distribution of Coef values (network edge weights).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{target\_gene}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} gene name

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sort}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to sort genes by its strength

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of estimators for bagging.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{threshold\_p}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} the threshold for p\sphinxhyphen{}values. TFs will be filtered based on the p\sphinxhyphen{}value.
if None, no filtering is applied.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.Net method)@\spxentry{to\_hdf5()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{\DUrole{n}{file\_path}}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.net’

\end{description}\end{quote}

\end{fulllineitems}

\index{updateLinkList() (celloracle.Net method)@\spxentry{updateLinkList()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.updateLinkList}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{updateLinkList}}}{\emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Update LinkList.
LinkList is a data frame that store information about inferred GRNs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to show a progress bar

\end{description}\end{quote}

\end{fulllineitems}

\index{updateTFinfo\_dictionary() (celloracle.Net method)@\spxentry{updateTFinfo\_dictionary()}\spxextra{celloracle.Net method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Net.updateTFinfo_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{updateTFinfo\_dictionary}}}{\emph{\DUrole{n}{TFdict}}}{}
Update TF info matrix
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} A python dictionary in which a key is Target gene, value are potential regulatory genes for the target gene.

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{Oracle (class in celloracle)@\spxentry{Oracle}\spxextra{class in celloracle}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle}}\pysigline{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.}}\sphinxbfcode{\sphinxupquote{Oracle}}}
Bases: \sphinxcode{\sphinxupquote{celloracle.trajectory.modified\_VelocytoLoom\_class.modified\_VelocytoLoom}}

Oracle is the main class in CellOracle. Oracle object imports scRNA\sphinxhyphen{}seq data (anndata) and TF information to infer cluster\sphinxhyphen{}specific GRNs. It can predict the future gene expression patterns and cell state transitions in response to  the perturbation of TFs. Please see the CellOracle paper for details.
The code of the Oracle class was made of the three components below.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{(}{)}%
\item {} 
Anndata: Gene expression matrix and metadata from single\sphinxhyphen{}cell RNA\sphinxhyphen{}seq are stored in the anndata object. Processed values, such as normalized counts and simulated values, are stored as layers of anndata. Metadata (i.e., Cluster info) are saved in anndata.obs. Refer to scanpy/anndata documentation for detail.

\item {} 
Net: Net is a custom class in celloracle. Net object processes several data to infer GRN. See the Net class documentation for details.

\item {} 
VelycytoLoom: Calculation of transition probability and visualization of directed trajectory graph will be performed in the same way as velocytoloom. VelocytoLoom is class from Velocyto, a python library for RNA\sphinxhyphen{}velocity analysis. In celloracle, we use some functions in velocytoloom for the visualization.

\end{enumerate}
\index{adata (celloracle.Oracle attribute)@\spxentry{adata}\spxextra{celloracle.Oracle attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.adata}}\pysigline{\sphinxbfcode{\sphinxupquote{adata}}}
Imported anndata object
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
anndata

\end{description}\end{quote}

\end{fulllineitems}

\index{cluster\_column\_name (celloracle.Oracle attribute)@\spxentry{cluster\_column\_name}\spxextra{celloracle.Oracle attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.cluster_column_name}}\pysigline{\sphinxbfcode{\sphinxupquote{cluster\_column\_name}}}
The column name in adata.obs containing cluster info
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{embedding\_name (celloracle.Oracle attribute)@\spxentry{embedding\_name}\spxextra{celloracle.Oracle attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.embedding_name}}\pysigline{\sphinxbfcode{\sphinxupquote{embedding\_name}}}
The key name in adata.obsm containing dimensional reduction cordinates
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{addTFinfo\_dictionary() (celloracle.Oracle method)@\spxentry{addTFinfo\_dictionary()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.addTFinfo_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addTFinfo\_dictionary}}}{\emph{\DUrole{n}{TFdict}}}{}
Add new TF info to pre\sphinxhyphen{}existing TFdict.
Values in the old TF dictionary will remain.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Python dictionary of TF info.

\end{description}\end{quote}

\end{fulllineitems}

\index{copy() (celloracle.Oracle method)@\spxentry{copy()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.copy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{copy}}}{}{}
Deepcopy itself.

\end{fulllineitems}

\index{count\_cells\_in\_mc\_resutls() (celloracle.Oracle method)@\spxentry{count\_cells\_in\_mc\_resutls()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.count_cells_in_mc_resutls}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{count\_cells\_in\_mc\_resutls}}}{\emph{\DUrole{n}{cluster\_use}}, \emph{\DUrole{n}{end}\DUrole{o}{=}\DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{order}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Count the simulated cell by the cluster.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_use}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} cluster information name in anndata.obs.
You can use any cluster information in anndata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{end}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The end point of Sankey\sphinxhyphen{}diagram. Please select a  step in the Markov simulation.
if you set {[}end=\sphinxhyphen{}1{]}, the final step of Markov simulation will be used.

\end{itemize}

\item[{Returns}] \leavevmode
Number of cells before / after simulation

\item[{Return type}] \leavevmode
pandas.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{fit\_GRN\_for\_simulation() (celloracle.Oracle method)@\spxentry{fit\_GRN\_for\_simulation()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.fit_GRN_for_simulation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_GRN\_for\_simulation}}}{\emph{\DUrole{n}{GRN\_unit}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}cluster\textquotesingle{}}}, \emph{\DUrole{n}{alpha}\DUrole{o}{=}\DUrole{default_value}{1}}, \emph{\DUrole{n}{use\_cluster\_specific\_TFdict}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
Do GRN inference.
Please see the paper of CellOracle paper for details.

GRN can be constructed for the entire population or each clusters.
If you want to infer cluster\sphinxhyphen{}specific GRN, please set {[}GRN\_unit=”cluster”{]}.
You can select cluster information when you import data.

If you set {[}GRN\_unit=”whole”{]}, GRN will be made using all cells.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{GRN\_unit}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Select “cluster” or “whole”

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{ or }}\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The strength of regularization.
If you set a lower value, the sensitivity increases, and you can detect weaker network connections. However, there may be more noise.
If you select a higher value, it will reduce the chance of overfitting.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_cluster\_specific\_TFdict\_from\_Links() (celloracle.Oracle method)@\spxentry{get\_cluster\_specific\_TFdict\_from\_Links()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.get_cluster_specific_TFdict_from_Links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_cluster\_specific\_TFdict\_from\_Links}}}{\emph{\DUrole{n}{links\_object}}}{}
Extract TF and its target gene information from Links object.
This function can be used to reconstruct GRNs based on pre\sphinxhyphen{}existing GRNs saved in Links object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{links\_object}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} Please see the explanation of Links class.

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_links() (celloracle.Oracle method)@\spxentry{get\_links()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.get_links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_links}}}{\emph{\DUrole{n}{cluster\_name\_for\_GRN\_unit}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{alpha}\DUrole{o}{=}\DUrole{default_value}{10}}, \emph{\DUrole{n}{bagging\_number}\DUrole{o}{=}\DUrole{default_value}{20}}, \emph{\DUrole{n}{verbose\_level}\DUrole{o}{=}\DUrole{default_value}{1}}, \emph{\DUrole{n}{test\_mode}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
Makes GRN for each cluster and returns results as a Links object.
Several preprocessing should be done before using this function.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_name\_for\_GRN\_unit}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name for GRN calculation. The cluster information should be stored in Oracle.adata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{ or }}\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The strength of regularization.
If you set a lower value, the sensitivity increases, and you can detect weaker network connections. However, there may be more noise.
If you select a higher value, it will reduce the chance of overfitting.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number used in bagging calculation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose\_level}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} if {[}verbose\_level\textgreater{}1{]}, most detailed progress information will be shown.
if {[}verbose\_level \textgreater{} 0{]}, one progress bar will be shown.
if {[}verbose\_level == 0{]}, no progress bar will be shown.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{test\_mode}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} If test\_mode is True, GRN calculation will be done for only one cluster rather than all clusters.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{import\_TF\_data() (celloracle.Oracle method)@\spxentry{import\_TF\_data()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.import_TF_data}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{import\_TF\_data}}}{\emph{\DUrole{n}{TF\_info\_matrix}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{TF\_info\_matrix\_path}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{TFdict}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Load data about potential\sphinxhyphen{}regulatory TFs.
You can import either TF\_info\_matrix or TFdict.
For more information on how to make these files, please see the motif analysis module within the celloracle tutorial.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{TF\_info\_matrix}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.DataFrame}}) \textendash{} TF\_info\_matrix.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{TF\_info\_matrix\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} File path for TF\_info\_matrix (pandas.DataFrame).

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Python dictionary of TF info.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{import\_anndata\_as\_normalized\_count() (celloracle.Oracle method)@\spxentry{import\_anndata\_as\_normalized\_count()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.import_anndata_as_normalized_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{import\_anndata\_as\_normalized\_count}}}{\emph{\DUrole{n}{adata}}, \emph{\DUrole{n}{cluster\_column\_name}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{embedding\_name}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Load scRNA\sphinxhyphen{}seq data. scRNA\sphinxhyphen{}seq data should be prepared as an anndata object.
Preprocessing (cell and gene filtering, dimensional reduction, clustering, etc.) should be done before loading data.
The method will import NORMALIZED and LOG TRANSFORMED data but NOT SCALED and NOT CENTERED data.
See the tutorial for more details on how to process scRNA\sphinxhyphen{}seq data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{adata}} (\sphinxstyleliteralemphasis{\sphinxupquote{anndata}}) \textendash{} anndata object containing scRNA\sphinxhyphen{}seq data.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_column\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of column containing cluster information in anndata.obs.
Clustering data should be in anndata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{embedding\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the key name for dimensional reduction information in anndata.obsm.
Dimensional reduction (or 2D trajectory graph) should be in anndata.obsm.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{transform}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The method for log\sphinxhyphen{}transformation. Chose one from “natural\_log” or “log2”.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{import\_anndata\_as\_raw\_count() (celloracle.Oracle method)@\spxentry{import\_anndata\_as\_raw\_count()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.import_anndata_as_raw_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{import\_anndata\_as\_raw\_count}}}{\emph{\DUrole{n}{adata}}, \emph{\DUrole{n}{cluster\_column\_name}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{embedding\_name}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{transform}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}natural\_log\textquotesingle{}}}}{}
Load scRNA\sphinxhyphen{}seq data. scRNA\sphinxhyphen{}seq data should be prepared as an anndata object.
Preprocessing (cell and gene filtering, dimensional reduction, clustering, etc.) should be done before loading data.
The method imports RAW GENE COUNTS because unscaled and uncentered gene expression data are required for the GRN inference and simulation.
See tutorial notebook for the details about how to process scRNA\sphinxhyphen{}seq data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{adata}} (\sphinxstyleliteralemphasis{\sphinxupquote{anndata}}) \textendash{} anndata object that stores scRNA\sphinxhyphen{}seq data.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_column\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of column containing cluster information in anndata.obs.
Clustering data should be in anndata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{embedding\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the key name for dimensional reduction information in anndata.obsm.
Dimensional reduction (or 2D trajectory graph) should be in anndata.obsm.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{transform}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The method for log\sphinxhyphen{}transformation. Chose one from “natural\_log” or “log2”.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_mc\_result\_as\_kde() (celloracle.Oracle method)@\spxentry{plot\_mc\_result\_as\_kde()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.plot_mc_result_as_kde}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_mc\_result\_as\_kde}}}{\emph{\DUrole{n}{n\_time}}, \emph{\DUrole{n}{args}\DUrole{o}{=}\DUrole{default_value}{\{\}}}}{}
Pick up one timepoint in the cell state\sphinxhyphen{}transition simulation and plot as a kde plot.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_time}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the number in Markov simulation

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} An argument for seaborn.kdeplot.
See seaborn documentation for details (\sphinxurl{https://seaborn.pydata.org/generated/seaborn.kdeplot.html\#seaborn.kdeplot}).

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_mc\_result\_as\_trajectory() (celloracle.Oracle method)@\spxentry{plot\_mc\_result\_as\_trajectory()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.plot_mc_result_as_trajectory}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_mc\_result\_as\_trajectory}}}{\emph{\DUrole{n}{cell\_name}}, \emph{\DUrole{n}{time\_range}}, \emph{\DUrole{n}{args}\DUrole{o}{=}\DUrole{default_value}{\{\}}}}{}
Pick up several timepoints in the cell state\sphinxhyphen{}transition simulation and plot as a line plot.
This function can be used to visualize how cell\sphinxhyphen{}state changes after perturbation focusing on a specific cell.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cell\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} cell name. chose from adata.obs.index

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{time\_range}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of int}}) \textendash{} the list of index in Markov simulation

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} dictionary for the arguments for matplotlib.pyplit.plot.
See matplotlib documentation for details (\sphinxurl{https://matplotlib.org/api/\_as\_gen/matplotlib.pyplot.plot.html\#matplotlib.pyplot.plot}).

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_mc\_resutls\_as\_sankey() (celloracle.Oracle method)@\spxentry{plot\_mc\_resutls\_as\_sankey()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.plot_mc_resutls_as_sankey}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_mc\_resutls\_as\_sankey}}}{\emph{\DUrole{n}{cluster\_use}}, \emph{\DUrole{n}{start}\DUrole{o}{=}\DUrole{default_value}{0}}, \emph{\DUrole{n}{end}\DUrole{o}{=}\DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{order}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{font\_size}\DUrole{o}{=}\DUrole{default_value}{10}}}{}
Plot the simulated cell state\sphinxhyphen{}transition as a Sankey\sphinxhyphen{}diagram after groping by the cluster.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_use}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} cluster information name in anndata.obs.
You can use any cluster information in anndata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{start}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The starting point of Sankey\sphinxhyphen{}diagram. Please select a  step in the Markov simulation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{end}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The end point of Sankey\sphinxhyphen{}diagram. Please select a  step in the Markov simulation.
if you set {[}end=\sphinxhyphen{}1{]}, the final step of Markov simulation will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{order}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The order of cluster name in the Sankey\sphinxhyphen{}diagram.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{font\_size}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Font size for cluster name label in the Sankey diagram.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{prepare\_markov\_simulation() (celloracle.Oracle method)@\spxentry{prepare\_markov\_simulation()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.prepare_markov_simulation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{prepare\_markov\_simulation}}}{\emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
Pick up cells for Markov simulation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} If True, it plots selected cells.

\end{description}\end{quote}

\end{fulllineitems}

\index{run\_markov\_chain\_simulation() (celloracle.Oracle method)@\spxentry{run\_markov\_chain\_simulation()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.run_markov_chain_simulation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{run\_markov\_chain\_simulation}}}{\emph{\DUrole{n}{n\_steps}\DUrole{o}{=}\DUrole{default_value}{500}}, \emph{\DUrole{n}{n\_duplication}\DUrole{o}{=}\DUrole{default_value}{5}}, \emph{\DUrole{n}{seed}\DUrole{o}{=}\DUrole{default_value}{123}}, \emph{\DUrole{n}{calculate\_randomized}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Do Markov simlations to predict cell transition after perturbation.
The transition probability between cells has been calculated
based on simulated gene expression values in the signal propagation process.
The cell state transition will be simulated based on the probability.
You can simulate the process multiple times to get a robust outcome.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_steps}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} steps for Markov simulation. This value is equivalent to the amount of time after perturbation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_duplication}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the number for multiple calculations.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{simulate\_shift() (celloracle.Oracle method)@\spxentry{simulate\_shift()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.simulate_shift}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{simulate\_shift}}}{\emph{\DUrole{n}{perturb\_condition}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{GRN\_unit}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}cluster\textquotesingle{}}}, \emph{\DUrole{n}{n\_propagation}\DUrole{o}{=}\DUrole{default_value}{3}}, \emph{\DUrole{n}{ignore\_warning}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
Simulate signal propagation with GRNs. Please see the CellOracle paper for details.
This function simulates a gene expression pattern in the near future.
Simulated values will be stored in anndata.layers: {[}“simulated\_count”{]}

The simulation use three types of data.
(1) GRN inference results (coef\_matrix).
(2) Perturb\_condition: You can set arbitrary perturbation condition.
(3) Gene expression matrix: The simulation starts from imputed gene expression data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{perturb\_condition}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} condition for perturbation.
if you want to simulate knockout for GeneX, please set {[}perturb\_condition=\{“GeneX”: 0.0\}{]}
Although you can set any non\sphinxhyphen{}negative values for the gene condition, avoid setting biologically infeasible values for the perturb condition.
It is strongly recommended to check gene expression values in your data before selecting the perturb condition.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{GRN\_unit}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} GRN type. Please select either “whole” or “cluster”. See the documentation of “fit\_GRN\_for\_simulation” for the detailed explanation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_propagation}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Calculation will be performed iteratively to simulate signal propagation in GRN.
You can set the number of steps for this calculation.
With a higher number, the results may recapitulate signal propagation for many genes.
However, a higher number of propagation may cause more error/noise.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{summarize\_mc\_results\_by\_cluster() (celloracle.Oracle method)@\spxentry{summarize\_mc\_results\_by\_cluster()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.summarize_mc_results_by_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{summarize\_mc\_results\_by\_cluster}}}{\emph{\DUrole{n}{cluster\_use}}, \emph{\DUrole{n}{random}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
This function summarizes the simulated cell state\sphinxhyphen{}transition by groping the results into each cluster.
It returns sumarized results as a pandas.DataFrame.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_use}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} cluster information name in anndata.obs.
You can use any arbitrary cluster information in anndata.obs.

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.Oracle method)@\spxentry{to\_hdf5()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{\DUrole{n}{file\_path}}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.oracle’

\end{description}\end{quote}

\end{fulllineitems}

\index{updateTFinfo\_dictionary() (celloracle.Oracle method)@\spxentry{updateTFinfo\_dictionary()}\spxextra{celloracle.Oracle method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.Oracle.updateTFinfo_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{updateTFinfo\_dictionary}}}{\emph{\DUrole{n}{TFdict}}}{}
Update a TF dictionary.
If a key in the new TF dictionary already exists in the old TF dictionary, old values will be replaced with a new one.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{TFdict}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Python dictionary of TF info.

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{load\_hdf5() (in module celloracle)@\spxentry{load\_hdf5()}\spxextra{in module celloracle}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle:celloracle.load_hdf5}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.}}\sphinxbfcode{\sphinxupquote{load\_hdf5}}}{\emph{\DUrole{n}{file\_path}}, \emph{\DUrole{n}{object\_class\_name}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Load an object of celloracle’s custom class that was saved as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file\_path.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{object\_class\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Types of object.
If it is None, object class will be identified from the extension of file\_name.
Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}



\subsubsection{Modules for ATAC\sphinxhyphen{}seq analysis}
\label{\detokenize{modules/celloracle:modules-for-atac-seq-analysis}}\begin{quote}


\paragraph{celloracle.motif\_analysis module}
\label{\detokenize{modules/celloracle.motif_analysis:module-celloracle.motif_analysis}}\label{\detokenize{modules/celloracle.motif_analysis:celloracle-motif-analysis-module}}\label{\detokenize{modules/celloracle.motif_analysis::doc}}\index{module@\spxentry{module}!celloracle.motif\_analysis@\spxentry{celloracle.motif\_analysis}}\index{celloracle.motif\_analysis@\spxentry{celloracle.motif\_analysis}!module@\spxentry{module}}
The {\hyperref[\detokenize{modules/celloracle.motif_analysis:module-celloracle.motif_analysis}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{motif\_analysis}}}}} module implements transcription factor motif scan.

Genomic activity information (peak of ATAC\sphinxhyphen{}seq or Chip\sphinxhyphen{}seq) is extracted first.
Then the peak DNA sequence will be subjected to TF motif scan.
Finally we will get list of TFs that potentially binds to a specific gene.
\index{TFinfo (class in celloracle.motif\_analysis)@\spxentry{TFinfo}\spxextra{class in celloracle.motif\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{TFinfo}}}{\emph{\DUrole{n}{peak\_data\_frame}}, \emph{\DUrole{n}{ref\_genome}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

This is a custom class for motif analysis in celloracle.
TFinfo object performs motif scan using the TF motif database in gimmemotifs and several functions of genomepy.
Analysis results can be exported as a python dictionary or dataframe.
These files; python dictionary of dataframe of TF binding information, are needed during GRN inference.
\index{peak\_df (celloracle.motif\_analysis.TFinfo attribute)@\spxentry{peak\_df}\spxextra{celloracle.motif\_analysis.TFinfo attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.peak_df}}\pysigline{\sphinxbfcode{\sphinxupquote{peak\_df}}}
dataframe about DNA peak and target gene data.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{all\_target\_gene (celloracle.motif\_analysis.TFinfo attribute)@\spxentry{all\_target\_gene}\spxextra{celloracle.motif\_analysis.TFinfo attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.all_target_gene}}\pysigline{\sphinxbfcode{\sphinxupquote{all\_target\_gene}}}
target genes.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
array of str

\end{description}\end{quote}

\end{fulllineitems}

\index{ref\_genome (celloracle.motif\_analysis.TFinfo attribute)@\spxentry{ref\_genome}\spxextra{celloracle.motif\_analysis.TFinfo attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.ref_genome}}\pysigline{\sphinxbfcode{\sphinxupquote{ref\_genome}}}
reference genome name that was used in DNA peak generation.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{scanned\_df (celloracle.motif\_analysis.TFinfo attribute)@\spxentry{scanned\_df}\spxextra{celloracle.motif\_analysis.TFinfo attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.scanned_df}}\pysigline{\sphinxbfcode{\sphinxupquote{scanned\_df}}}
Results of motif scan. Key is a peak name. Value is a dataframe of motif scan.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{dic\_targetgene2TFs (celloracle.motif\_analysis.TFinfo attribute)@\spxentry{dic\_targetgene2TFs}\spxextra{celloracle.motif\_analysis.TFinfo attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.dic_targetgene2TFs}}\pysigline{\sphinxbfcode{\sphinxupquote{dic\_targetgene2TFs}}}
Final product of motif scan. Key is a target gene. Value is a list of regulatory candidate genes.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{dic\_peak2Targetgene (celloracle.motif\_analysis.TFinfo attribute)@\spxentry{dic\_peak2Targetgene}\spxextra{celloracle.motif\_analysis.TFinfo attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.dic_peak2Targetgene}}\pysigline{\sphinxbfcode{\sphinxupquote{dic\_peak2Targetgene}}}
Dictionary. Key is a peak name. Value is a list of the target gene.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{dic\_TF2targetgenes (celloracle.motif\_analysis.TFinfo attribute)@\spxentry{dic\_TF2targetgenes}\spxextra{celloracle.motif\_analysis.TFinfo attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.dic_TF2targetgenes}}\pysigline{\sphinxbfcode{\sphinxupquote{dic\_TF2targetgenes}}}
Final product of motif scan. Key is a TF. Value is a list of potential target genes of the TF.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{copy() (celloracle.motif\_analysis.TFinfo method)@\spxentry{copy()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.copy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{copy}}}{}{}
Deepcoty itself.

\end{fulllineitems}

\index{filter\_motifs\_by\_score() (celloracle.motif\_analysis.TFinfo method)@\spxentry{filter\_motifs\_by\_score()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.filter_motifs_by_score}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{filter\_motifs\_by\_score}}}{\emph{\DUrole{n}{threshold}}, \emph{\DUrole{n}{method}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}cumlative\_score\textquotesingle{}}}}{}
Remove motifs with low binding scores.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} thresholding method. Select either of {[}“indivisual\_score”, “cumlative\_score”{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{filter\_peaks() (celloracle.motif\_analysis.TFinfo method)@\spxentry{filter\_peaks()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.filter_peaks}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{filter\_peaks}}}{\emph{\DUrole{n}{peaks\_to\_be\_remained}}}{}
Filter peaks.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{peaks\_to\_be\_remained}} (\sphinxstyleliteralemphasis{\sphinxupquote{array of str}}) \textendash{} list of peaks. Peaks that are NOT in the list will be removed.

\end{description}\end{quote}

\end{fulllineitems}

\index{make\_TFinfo\_dataframe\_and\_dictionary() (celloracle.motif\_analysis.TFinfo method)@\spxentry{make\_TFinfo\_dataframe\_and\_dictionary()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.make_TFinfo_dataframe_and_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{make\_TFinfo\_dataframe\_and\_dictionary}}}{\emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
This is the final step of motif\_analysis.
Convert scanned results into a data frame and dictionaries.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar.

\end{description}\end{quote}

\end{fulllineitems}

\index{reset\_dictionary\_and\_df() (celloracle.motif\_analysis.TFinfo method)@\spxentry{reset\_dictionary\_and\_df()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.reset_dictionary_and_df}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{reset\_dictionary\_and\_df}}}{}{}
Reset TF dictionary and TF dataframe.
The following attributes will be erased: TF\_onehot, dic\_targetgene2TFs, dic\_peak2Targetgene, dic\_TF2targetgenes.

\end{fulllineitems}

\index{reset\_filtering() (celloracle.motif\_analysis.TFinfo method)@\spxentry{reset\_filtering()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.reset_filtering}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{reset\_filtering}}}{}{}
Reset filtering information.
You can use this function to stat over the filtering step with new conditions.
The following attributes will be erased: TF\_onehot, dic\_targetgene2TFs, dic\_peak2Targetgene, dic\_TF2targetgenes.

\end{fulllineitems}

\index{save\_as\_parquet() (celloracle.motif\_analysis.TFinfo method)@\spxentry{save\_as\_parquet()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.save_as_parquet}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{save\_as\_parquet}}}{\emph{\DUrole{n}{folder\_path}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Save itself. Some attributes are saved as parquet file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{folder\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} folder path

\end{description}\end{quote}

\end{fulllineitems}

\index{scan() (celloracle.motif\_analysis.TFinfo method)@\spxentry{scan()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.scan}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{scan}}}{\emph{\DUrole{n}{background\_length}\DUrole{o}{=}\DUrole{default_value}{200}}, \emph{\DUrole{n}{fpr}\DUrole{o}{=}\DUrole{default_value}{0.02}}, \emph{\DUrole{n}{n\_cpus}\DUrole{o}{=}\DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Scan DNA sequences searching for TF binding motifs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{background\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} background length. This is used for the calculation of the binding score.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{fpr}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} False positive rate for motif identification.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_cpus}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} number of CPUs for parallel calculation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_dataframe() (celloracle.motif\_analysis.TFinfo method)@\spxentry{to\_dataframe()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.to_dataframe}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_dataframe}}}{\emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Return results as a dataframe.
Rows are peak\_id, and columns are TFs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to show a progress bar.

\item[{Returns}] \leavevmode
TFinfo matrix.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_dictionary() (celloracle.motif\_analysis.TFinfo method)@\spxentry{to\_dictionary()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.to_dictionary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_dictionary}}}{\emph{\DUrole{n}{dictionary\_type}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}targetgene2TFs\textquotesingle{}}}, \emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Return TF information as a python dictionary.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{dictionary\_type}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Type of dictionary. Select from {[}“targetgene2TFs”, “TF2targetgenes”{]}.
If you chose “targetgene2TFs”, it returns a dictionary in which a key is a target gene, and a value is a list of regulatory candidate genes (TFs) of the target.
If you chose “TF2targetgenes”, it returns a dictionary in which a key is a TF and a value is a list of potential target genes of the TF.

\item[{Returns}] \leavevmode
dictionary.

\item[{Return type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.motif\_analysis.TFinfo method)@\spxentry{to\_hdf5()}\spxextra{celloracle.motif\_analysis.TFinfo method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{\DUrole{n}{file\_path}}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.tfinfo’

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{get\_tss\_info() (in module celloracle.motif\_analysis)@\spxentry{get\_tss\_info()}\spxextra{in module celloracle.motif\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.get_tss_info}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{get\_tss\_info}}}{\emph{\DUrole{n}{peak\_str\_list}}, \emph{\DUrole{n}{ref\_genome}}, \emph{\DUrole{n}{verbose}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Get annotation about Transcription Starting Site (TSS).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peak\_str\_list}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} list of peak\_id. e.g., {[}“chr5\_0930303\_9499409”, “chr11\_123445555\_123445577”{]}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ref\_genome}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} reference genome name.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} verbosity.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{integrate\_tss\_peak\_with\_cicero() (in module celloracle.motif\_analysis)@\spxentry{integrate\_tss\_peak\_with\_cicero()}\spxextra{in module celloracle.motif\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.integrate_tss_peak_with_cicero}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{integrate\_tss\_peak\_with\_cicero}}}{\emph{\DUrole{n}{tss\_peak}}, \emph{\DUrole{n}{cicero\_connections}}}{}
Process output of cicero data and returns DNA peak information for motif analysis in celloracle.
Please see the celloracle tutorial for more information.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tss\_peak}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.dataframe}}) \textendash{} dataframe about TSS information. Please use the function, “get\_tss\_info” to get this dataframe.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cicero\_connections}} (\sphinxstyleliteralemphasis{\sphinxupquote{dataframe}}) \textendash{} dataframe that stores the results of cicero analysis.

\end{itemize}

\item[{Returns}] \leavevmode
DNA peak about promoter/enhancer and its annotation about target gene.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{is\_genome\_installed() (in module celloracle.motif\_analysis)@\spxentry{is\_genome\_installed()}\spxextra{in module celloracle.motif\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.is_genome_installed}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{is\_genome\_installed}}}{\emph{\DUrole{n}{ref\_genome}}}{}
Celloracle motif\_analysis module uses gimmemotifs and genomepy internally.
Reference genome files should be installed in the PC to use gimmemotifs and genomepy.
This function checks the installation status of the reference genome.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ref\_genome}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} names of reference genome. i.e., “mm10”, “hg19”

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_TFinfo\_from\_parquets() (in module celloracle.motif\_analysis)@\spxentry{load\_TFinfo\_from\_parquets()}\spxextra{in module celloracle.motif\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.load_TFinfo_from_parquets}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{load\_TFinfo\_from\_parquets}}}{\emph{\DUrole{n}{folder\_path}}}{}
Load TFinfo object which was saved with the function; “save\_as\_parquet”.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{folder\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} folder path

\item[{Returns}] \leavevmode
Loaded TFinfo object.

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.TFinfo}]{\sphinxcrossref{TFinfo}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{make\_TFinfo\_from\_scanned\_file() (in module celloracle.motif\_analysis)@\spxentry{make\_TFinfo\_from\_scanned\_file()}\spxextra{in module celloracle.motif\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.make_TFinfo_from_scanned_file}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{make\_TFinfo\_from\_scanned\_file}}}{\emph{\DUrole{n}{path\_to\_raw\_bed}}, \emph{\DUrole{n}{path\_to\_scanned\_result\_bed}}, \emph{\DUrole{n}{ref\_genome}}}{}
This function is currently an available.

\end{fulllineitems}

\index{peak2fasta() (in module celloracle.motif\_analysis)@\spxentry{peak2fasta()}\spxextra{in module celloracle.motif\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.peak2fasta}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{peak2fasta}}}{\emph{\DUrole{n}{peak\_ids}}, \emph{\DUrole{n}{ref\_genome}}}{}
Convert peak\_id into fasta object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peak\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{ or }}\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} Peak\_id.  e.g. “chr5\_0930303\_9499409”
or it can be a list of peak\_id.  e.g. {[}“chr5\_0930303\_9499409”, “chr11\_123445555\_123445577”{]}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ref\_genome}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Reference genome name.   e.g. “mm9”, “mm10”, “hg19” etc

\end{itemize}

\item[{Returns}] \leavevmode
DNA sequence in fasta format

\item[{Return type}] \leavevmode
gimmemotifs fasta object

\end{description}\end{quote}

\end{fulllineitems}

\index{read\_bed() (in module celloracle.motif\_analysis)@\spxentry{read\_bed()}\spxextra{in module celloracle.motif\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.motif_analysis:celloracle.motif_analysis.read_bed}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.motif\_analysis.}}\sphinxbfcode{\sphinxupquote{read\_bed}}}{\emph{\DUrole{n}{bed\_path}}}{}
Load bed file and return as dataframe.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{bed\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} File path.

\item[{Returns}] \leavevmode
bed file in dataframe.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}


\subsubsection{Modules for Network analysis}
\label{\detokenize{modules/celloracle:modules-for-network-analysis}}\begin{quote}


\paragraph{celloracle.network\_analysis module}
\label{\detokenize{modules/celloracle.network_analysis:module-celloracle.network_analysis}}\label{\detokenize{modules/celloracle.network_analysis:celloracle-network-analysis-module}}\label{\detokenize{modules/celloracle.network_analysis::doc}}\index{module@\spxentry{module}!celloracle.network\_analysis@\spxentry{celloracle.network\_analysis}}\index{celloracle.network\_analysis@\spxentry{celloracle.network\_analysis}!module@\spxentry{module}}
The {\hyperref[\detokenize{modules/celloracle.network_analysis:module-celloracle.network_analysis}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{network\_analysis}}}}} module implements Network analysis.
\index{Links (class in celloracle.network\_analysis)@\spxentry{Links}\spxextra{class in celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{Links}}}{\emph{\DUrole{n}{name}}, \emph{\DUrole{n}{links\_dict}\DUrole{o}{=}\DUrole{default_value}{\{\}}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

This is a class for the processing and visualization of GRNs.
Links object stores cluster\sphinxhyphen{}specific GRNs and metadata.
Please use “get\_links” function in Oracle object to generate Links object.
\index{links\_dict (celloracle.network\_analysis.Links attribute)@\spxentry{links\_dict}\spxextra{celloracle.network\_analysis.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.links_dict}}\pysigline{\sphinxbfcode{\sphinxupquote{links\_dict}}}
Dictionary that store unprocessed network data.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{filtered\_links (celloracle.network\_analysis.Links attribute)@\spxentry{filtered\_links}\spxextra{celloracle.network\_analysis.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.filtered_links}}\pysigline{\sphinxbfcode{\sphinxupquote{filtered\_links}}}
Dictionary that store filtered network data.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
dictionary

\end{description}\end{quote}

\end{fulllineitems}

\index{merged\_score (celloracle.network\_analysis.Links attribute)@\spxentry{merged\_score}\spxextra{celloracle.network\_analysis.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.merged_score}}\pysigline{\sphinxbfcode{\sphinxupquote{merged\_score}}}
Network scores.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{cluster (celloracle.network\_analysis.Links attribute)@\spxentry{cluster}\spxextra{celloracle.network\_analysis.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.cluster}}\pysigline{\sphinxbfcode{\sphinxupquote{cluster}}}
List of cluster name.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
list of str

\end{description}\end{quote}

\end{fulllineitems}

\index{name (celloracle.network\_analysis.Links attribute)@\spxentry{name}\spxextra{celloracle.network\_analysis.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.name}}\pysigline{\sphinxbfcode{\sphinxupquote{name}}}
Name of clustering unit.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{palette (celloracle.network\_analysis.Links attribute)@\spxentry{palette}\spxextra{celloracle.network\_analysis.Links attribute}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.palette}}\pysigline{\sphinxbfcode{\sphinxupquote{palette}}}
DataFrame that store color information.
\begin{quote}\begin{description}
\item[{Type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{filter\_links() (celloracle.network\_analysis.Links method)@\spxentry{filter\_links()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.filter_links}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{filter\_links}}}{\emph{\DUrole{n}{p}\DUrole{o}{=}\DUrole{default_value}{0.001}}, \emph{\DUrole{n}{weight}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}coef\_abs\textquotesingle{}}}, \emph{\DUrole{n}{thread\_number}\DUrole{o}{=}\DUrole{default_value}{10000}}, \emph{\DUrole{n}{genelist\_source}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{genelist\_target}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Filter network edges.
In most cases, inferred GRN has non\sphinxhyphen{}significant random edges.
We have to remove these edges before analyzing the network structure.
You can do the filtering in any of the following ways.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{(}{)}%
\item {} 
Filter based on the p\sphinxhyphen{}value of the network edge.
Please enter p\sphinxhyphen{}value for thresholding.

\item {} 
Filter based on network edge number.
If you set the number, network edges will be filtered based on the order of a network score. The top n\sphinxhyphen{}th network edges with network weight will remain, and the other edges will be removed.
The network data has several types of network weight, so you have to select which network weight do you want to use.

\item {} 
Filter based on an arbitrary gene list. You can set a gene list for source nodes or target nodes.

\end{enumerate}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{p}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} threshold for p\sphinxhyphen{}value of the network edge.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{weight}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Please select network weight name for the filtering

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{genelist\_source}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list to remain in regulatory gene nodes. Default is None.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{genelist\_target}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} gene list to remain in target gene nodes. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_network\_entropy() (celloracle.network\_analysis.Links method)@\spxentry{get\_network\_entropy()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.get_network_entropy}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_network\_entropy}}}{\emph{\DUrole{n}{value}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}coef\_abs\textquotesingle{}}}}{}
Calculate network entropy scores.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{value}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Default is “coef\_abs”.

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_score() (celloracle.network\_analysis.Links method)@\spxentry{get\_score()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.get_score}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_score}}}{\emph{\DUrole{n}{test\_mode}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
Get several network sores using R libraries.
Make sure all dependent R libraries are installed in your environment before running this function.
You can check the installation for the R libraries by running test\_installation() in network\_analysis module.

\end{fulllineitems}

\index{plot\_cartography\_scatter\_per\_cluster() (celloracle.network\_analysis.Links method)@\spxentry{plot\_cartography\_scatter\_per\_cluster()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_cartography_scatter_per_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cartography\_scatter\_per\_cluster}}}{\emph{\DUrole{n}{gois}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{clusters}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{scatter}\DUrole{o}{=}\DUrole{default_value}{True}}, \emph{\DUrole{n}{kde}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{auto\_gene\_annot}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{percentile}\DUrole{o}{=}\DUrole{default_value}{98}}, \emph{\DUrole{n}{args\_dot}\DUrole{o}{=}\DUrole{default_value}{\{\textquotesingle{}n\_levels\textquotesingle{}: 105\}}}, \emph{\DUrole{n}{args\_line}\DUrole{o}{=}\DUrole{default_value}{\{\textquotesingle{}c\textquotesingle{}: \textquotesingle{}gray\textquotesingle{}\}}}, \emph{\DUrole{n}{args\_annot}\DUrole{o}{=}\DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Make a gene network cartography plot.
Please read the original paper describing gene network cartography for more information.
\sphinxurl{https://www.nature.com/articles/nature03288}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gois}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of srt}}) \textendash{} List of gene name to highlight.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{clusters}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} List of cluster name to analyze. If None, all clusters in Links object will be analyzed.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{scatter}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to make a scatter plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{auto\_gene\_annot}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to pick up genes to make an annotation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{percentile}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Genes with a network score above the percentile will be shown with annotation. Default is 98.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_dot}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for scatter plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_line}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for lines in cartography plot.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{args\_annot}} (\sphinxstyleliteralemphasis{\sphinxupquote{dictionary}}) \textendash{} Arguments for annotation in plots.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_cartography\_term() (celloracle.network\_analysis.Links method)@\spxentry{plot\_cartography\_term()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_cartography_term}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cartography\_term}}}{\emph{\DUrole{n}{goi}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the gene network cartography term like a heatmap.
Please read the original paper of gene network cartography for the principle of gene network cartography.
\sphinxurl{https://www.nature.com/articles/nature03288}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gois}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of srt}}) \textendash{} List of gene name to highlight.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_degree\_distributions() (celloracle.network\_analysis.Links method)@\spxentry{plot\_degree\_distributions()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_degree_distributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_degree\_distributions}}}{\emph{\DUrole{n}{plot\_model}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the network degree distributions (the number of edge per gene).
The network degree will be visualized in both linear scale and log scale.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plot\_model}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to plot linear approximation line.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_network\_entropy\_distributions() (celloracle.network\_analysis.Links method)@\spxentry{plot\_network\_entropy\_distributions()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_network_entropy_distributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_network\_entropy\_distributions}}}{\emph{\DUrole{n}{update\_network\_entropy}\DUrole{o}{=}\DUrole{default_value}{False}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the distribution for network entropy.
See the CellOracle paper for more detail.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} (\sphinxstyleliteralemphasis{\sphinxupquote{Links object}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{values}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The list of score to visualize. If it is None, all network score (listed above) will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{update\_network\_entropy}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to recalculate network entropy.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_comparison\_2D() (celloracle.network\_analysis.Links method)@\spxentry{plot\_score\_comparison\_2D()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_score_comparison_2D}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_comparison\_2D}}}{\emph{\DUrole{n}{value}}, \emph{\DUrole{n}{cluster1}}, \emph{\DUrole{n}{cluster2}}, \emph{\DUrole{n}{percentile}\DUrole{o}{=}\DUrole{default_value}{99}}, \emph{\DUrole{n}{annot\_shifts}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Make a scatter plot that compares specific network scores in two groups.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{value}} (\sphinxstyleliteralemphasis{\sphinxupquote{srt}}) \textendash{} The network score type.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster1}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name. Network scores in cluster1 will be visualized in the x\sphinxhyphen{}axis.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster2}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name. Network scores in cluster2 will be visualized in the y\sphinxhyphen{}axis.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{percentile}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Genes with a network score above the percentile will be shown with annotation. Default is 99.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{annot\_shifts}} (\sphinxstyleliteralemphasis{\sphinxupquote{(}}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{)}}) \textendash{} Annotation visualization setting.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_discributions() (celloracle.network\_analysis.Links method)@\spxentry{plot\_score\_discributions()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_score_discributions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_discributions}}}{\emph{\DUrole{n}{values}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{method}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}boxplot\textquotesingle{}}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot the distribution of network scores.
An individual data point is a network edge (gene).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See Links class for details.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{values}} (\sphinxstyleliteralemphasis{\sphinxupquote{list of str}}) \textendash{} The list of score to visualize. If it is None, all of the network score will be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Plotting method. Select either “boxplot” or “barplot”.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_score\_per\_cluster() (celloracle.network\_analysis.Links method)@\spxentry{plot\_score\_per\_cluster()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_score_per_cluster}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_score\_per\_cluster}}}{\emph{\DUrole{n}{goi}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Plot network score for a gene.
This function visualizes the network score for a specific gene between clusters to get an insight into the dynamics of the gene.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{goi}} (\sphinxstyleliteralemphasis{\sphinxupquote{srt}}) \textendash{} Gene name.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_scores\_as\_rank() (celloracle.network\_analysis.Links method)@\spxentry{plot\_scores\_as\_rank()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.plot_scores_as_rank}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_scores\_as\_rank}}}{\emph{\DUrole{n}{cluster}}, \emph{\DUrole{n}{n\_gene}\DUrole{o}{=}\DUrole{default_value}{50}}, \emph{\DUrole{n}{save}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Pick up top n\sphinxhyphen{}th genes wich high\sphinxhyphen{}network scores and make plots.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} See network\_analysis.Links class for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name to analyze.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{n\_gene}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Number of genes to plot. Default is 50.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{save}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if {[}save=None{]}. Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_hdf5() (celloracle.network\_analysis.Links method)@\spxentry{to\_hdf5()}\spxextra{celloracle.network\_analysis.Links method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.Links.to_hdf5}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_hdf5}}}{\emph{\DUrole{n}{file\_path}}}{}
Save object as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path to save file. Filename needs to end with ‘.celloracle.links’

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{draw\_network() (in module celloracle.network\_analysis)@\spxentry{draw\_network()}\spxextra{in module celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.draw_network}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{draw\_network}}}{\emph{\DUrole{n}{linkList}}, \emph{\DUrole{n}{return\_graph}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
Plot network graph.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{linkList}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.DataFrame}}) \textendash{} GRN saved as linkList.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{return\_graph}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to return graph object.

\end{itemize}

\item[{Returns}] \leavevmode
Network X graph objenct.

\item[{Return type}] \leavevmode
Graph object

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_R\_path() (in module celloracle.network\_analysis)@\spxentry{get\_R\_path()}\spxextra{in module celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.get_R_path}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{get\_R\_path}}}{}{}
\end{fulllineitems}

\index{get\_links() (in module celloracle.network\_analysis)@\spxentry{get\_links()}\spxextra{in module celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.get_links}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{get\_links}}}{\emph{\DUrole{n}{oracle\_object}}, \emph{\DUrole{n}{cluster\_name\_for\_GRN\_unit}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{alpha}\DUrole{o}{=}\DUrole{default_value}{10}}, \emph{\DUrole{n}{bagging\_number}\DUrole{o}{=}\DUrole{default_value}{20}}, \emph{\DUrole{n}{verbose\_level}\DUrole{o}{=}\DUrole{default_value}{1}}, \emph{\DUrole{n}{test\_mode}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
Make GRN for each cluster and returns results as a Links object.
Several preprocessing should be done before using this function.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{oracle\_object}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Oracle}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Oracle}}}}}) \textendash{} See Oracle module for detail.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cluster\_name\_for\_GRN\_unit}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Cluster name for GRN calculation. The cluster information should be stored in Oracle.adata.obs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{ or }}\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The strength of regularization.
If you set a lower value, the sensitivity increases, and you can detect weaker network connections. However, there may be more noise.
If you select a higher value, it will reduce the chance of overfitting.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bagging\_number}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number used in bagging calculation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose\_level}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} if {[}verbose\_level\textgreater{}1{]}, most detailed progress information will be shown.
if {[}verbose\_level \textgreater{} 0{]}, one progress bar will be shown.
if {[}verbose\_level == 0{]}, no progress bar will be shown.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{test\_mode}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} If test\_mode is True, GRN calculation will be done for only one cluster rather than all clusters.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{linkList\_to\_networkgraph() (in module celloracle.network\_analysis)@\spxentry{linkList\_to\_networkgraph()}\spxextra{in module celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.linkList_to_networkgraph}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{linkList\_to\_networkgraph}}}{\emph{\DUrole{n}{filteredlinkList}}}{}
Convert linkList into Graph object in NetworkX.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{filteredlinkList}} (\sphinxstyleliteralemphasis{\sphinxupquote{pandas.DataFrame}}) \textendash{} GRN saved as linkList.

\item[{Returns}] \leavevmode
Network X graph objenct.

\item[{Return type}] \leavevmode
Graph object

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_links() (in module celloracle.network\_analysis)@\spxentry{load\_links()}\spxextra{in module celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.load_links}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{load\_links}}}{\emph{\DUrole{n}{file\_path}}}{}
Load links object saved as a hdf5 file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path.

\item[{Returns}] \leavevmode
loaded links object.

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{Links}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{set\_R\_path() (in module celloracle.network\_analysis)@\spxentry{set\_R\_path()}\spxextra{in module celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.set_R_path}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{set\_R\_path}}}{\emph{\DUrole{n}{R\_path}}}{}
\end{fulllineitems}

\index{test\_R\_libraries\_installation() (in module celloracle.network\_analysis)@\spxentry{test\_R\_libraries\_installation()}\spxextra{in module celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.test_R_libraries_installation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{test\_R\_libraries\_installation}}}{\emph{\DUrole{n}{show\_all\_stdout}\DUrole{o}{=}\DUrole{default_value}{False}}}{}
CellOracle.network\_analysis use several R libraries for network analysis.
This is a test function to check for instalation of the necessary R libraries.

\end{fulllineitems}

\index{transfer\_scores\_from\_links\_to\_adata() (in module celloracle.network\_analysis)@\spxentry{transfer\_scores\_from\_links\_to\_adata()}\spxextra{in module celloracle.network\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.network_analysis:celloracle.network_analysis.transfer_scores_from_links_to_adata}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.network\_analysis.}}\sphinxbfcode{\sphinxupquote{transfer\_scores\_from\_links\_to\_adata}}}{\emph{\DUrole{n}{adata}}, \emph{\DUrole{n}{links}}, \emph{\DUrole{n}{method}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}median\textquotesingle{}}}}{}
Transfer the summary of network scores (median or mean) per group from Links object into adata.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{adata}} (\sphinxstyleliteralemphasis{\sphinxupquote{anndata}}) \textendash{} anndata

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{links}} ({\hyperref[\detokenize{modules/celloracle:celloracle.Links}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Links}}}}}) \textendash{} Likns object

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The method to summarize data.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}


\subsubsection{Other modules}
\label{\detokenize{modules/celloracle:other-modules}}\begin{quote}


\paragraph{celloracle.go\_analysis module}
\label{\detokenize{modules/celloracle.go_analysis:celloracle-go-analysis-module}}\label{\detokenize{modules/celloracle.go_analysis::doc}}\begin{quote}
\phantomsection\label{\detokenize{modules/celloracle.go_analysis:module-celloracle.go_analysis}}\index{module@\spxentry{module}!celloracle.go\_analysis@\spxentry{celloracle.go\_analysis}}\index{celloracle.go\_analysis@\spxentry{celloracle.go\_analysis}!module@\spxentry{module}}
The {\hyperref[\detokenize{modules/celloracle.go_analysis:module-celloracle.go_analysis}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{go\_analysis}}}}} module implements Gene Ontology analysis.
This module use goatools internally.
\index{geneID2Symbol() (in module celloracle.go\_analysis)@\spxentry{geneID2Symbol()}\spxextra{in module celloracle.go\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.go_analysis:celloracle.go_analysis.geneID2Symbol}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.go\_analysis.}}\sphinxbfcode{\sphinxupquote{geneID2Symbol}}}{\emph{\DUrole{n}{IDs}}, \emph{\DUrole{n}{species}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}mouse\textquotesingle{}}}}{}
Convert Entrez gene id into gene symbol.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IDs}} (\sphinxstyleliteralemphasis{\sphinxupquote{array of str}}) \textendash{} Entrez gene id.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{species}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Select species. Either “mouse” or “human”.

\end{itemize}

\item[{Returns}] \leavevmode
Gene symbol

\item[{Return type}] \leavevmode
list of str

\end{description}\end{quote}

\end{fulllineitems}

\index{geneSymbol2ID() (in module celloracle.go\_analysis)@\spxentry{geneSymbol2ID()}\spxextra{in module celloracle.go\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.go_analysis:celloracle.go_analysis.geneSymbol2ID}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.go\_analysis.}}\sphinxbfcode{\sphinxupquote{geneSymbol2ID}}}{\emph{\DUrole{n}{symbols}}, \emph{\DUrole{n}{species}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}mouse\textquotesingle{}}}}{}
Convert gene symbol into Entrez gene id.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{symbols}} (\sphinxstyleliteralemphasis{\sphinxupquote{array of str}}) \textendash{} gene symbol

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{species}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Select species. Either “mouse” or “human”

\end{itemize}

\item[{Returns}] \leavevmode
Entrez gene id

\item[{Return type}] \leavevmode
list of str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_GO() (in module celloracle.go\_analysis)@\spxentry{get\_GO()}\spxextra{in module celloracle.go\_analysis}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.go_analysis:celloracle.go_analysis.get_GO}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.go\_analysis.}}\sphinxbfcode{\sphinxupquote{get\_GO}}}{\emph{\DUrole{n}{gene\_query}}, \emph{\DUrole{n}{species}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}mouse\textquotesingle{}}}}{}
Get Gene Ontologies (GOs).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_query}} (\sphinxstyleliteralemphasis{\sphinxupquote{array of str}}) \textendash{} gene list.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{species}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Select species. Either “mouse” or “human”

\end{itemize}

\item[{Returns}] \leavevmode
GO analysis results as dataframe.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}


\paragraph{celloracle.utility module}
\label{\detokenize{modules/celloracle.utility:celloracle-utility-module}}\label{\detokenize{modules/celloracle.utility::doc}}\begin{quote}
\phantomsection\label{\detokenize{modules/celloracle.utility:module-celloracle.utility}}\index{module@\spxentry{module}!celloracle.utility@\spxentry{celloracle.utility}}\index{celloracle.utility@\spxentry{celloracle.utility}!module@\spxentry{module}}
The {\hyperref[\detokenize{modules/celloracle.utility:module-celloracle.utility}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{utility}}}}} module has several functions that support celloracle.
\index{exec\_process() (in module celloracle.utility)@\spxentry{exec\_process()}\spxextra{in module celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.exec_process}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{exec\_process}}}{\emph{\DUrole{n}{commands}}, \emph{\DUrole{n}{message}\DUrole{o}{=}\DUrole{default_value}{True}}, \emph{\DUrole{n}{wait\_finished}\DUrole{o}{=}\DUrole{default_value}{True}}, \emph{\DUrole{n}{return\_process}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Excute a command. This is a wrapper of “subprocess.Popen”
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{commands}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} command.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{message}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to return a message or not.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{wait\_finished}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether or not to wait for the process to finish. If false, the process will be perfomed in background and the function will finish immediately

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{return\_process}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to return “process”.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{intersect() (in module celloracle.utility)@\spxentry{intersect()}\spxextra{in module celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.intersect}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{intersect}}}{\emph{\DUrole{n}{list1}}, \emph{\DUrole{n}{list2}}}{}
Intersect two list and get components that exists in both list.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{list1}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} input list.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{list2}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} input list.

\end{itemize}

\item[{Returns}] \leavevmode
intersected list.

\item[{Return type}] \leavevmode
list

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_hdf5() (in module celloracle.utility)@\spxentry{load\_hdf5()}\spxextra{in module celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.load_hdf5}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{load\_hdf5}}}{\emph{\DUrole{n}{file\_path}}, \emph{\DUrole{n}{object\_class\_name}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Load an object of celloracle’s custom class that was saved as hdf5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file\_path.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{object\_class\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Types of object.
If it is None, object class will be identified from the extension of file\_name.
Default is None.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_pickled\_object() (in module celloracle.utility)@\spxentry{load\_pickled\_object()}\spxextra{in module celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.load_pickled_object}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{load\_pickled\_object}}}{\emph{\DUrole{n}{filepath}}}{}
Load pickled object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{filepath}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path.

\item[{Returns}] \leavevmode
loaded object.

\item[{Return type}] \leavevmode
python object

\end{description}\end{quote}

\end{fulllineitems}

\index{makelog (class in celloracle.utility)@\spxentry{makelog}\spxextra{class in celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.makelog}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{makelog}}}{\emph{\DUrole{n}{file\_name}\DUrole{o}{=}\DUrole{default_value}{None}}, \emph{\DUrole{n}{directory}\DUrole{o}{=}\DUrole{default_value}{None}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

This is a class for making log.
\index{info() (celloracle.utility.makelog method)@\spxentry{info()}\spxextra{celloracle.utility.makelog method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.makelog.info}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{info}}}{\emph{\DUrole{n}{comment}}}{}
Add comment into the log file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{comment}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} comment.

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{save\_as\_pickled\_object() (in module celloracle.utility)@\spxentry{save\_as\_pickled\_object()}\spxextra{in module celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.save_as_pickled_object}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{save\_as\_pickled\_object}}}{\emph{\DUrole{n}{obj}}, \emph{\DUrole{n}{filepath}}}{}
Save any object using pickle.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{obj}} (\sphinxstyleliteralemphasis{\sphinxupquote{any python object}}) \textendash{} python object.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{filepath}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} file path.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{standard() (in module celloracle.utility)@\spxentry{standard()}\spxextra{in module celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.standard}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{standard}}}{\emph{\DUrole{n}{df}}}{}
Standardize value.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{df}} (\sphinxstyleliteralemphasis{\sphinxupquote{padas.dataframe}}) \textendash{} dataframe.

\item[{Returns}] \leavevmode
Data after standardization.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\index{transfer\_all\_colors\_between\_anndata() (in module celloracle.utility)@\spxentry{transfer\_all\_colors\_between\_anndata()}\spxextra{in module celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.transfer_all_colors_between_anndata}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{transfer\_all\_colors\_between\_anndata}}}{\emph{\DUrole{n}{anndata\_ref}}, \emph{\DUrole{n}{anndata\_que}}}{}
\end{fulllineitems}

\index{transfer\_color\_between\_anndata() (in module celloracle.utility)@\spxentry{transfer\_color\_between\_anndata()}\spxextra{in module celloracle.utility}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.utility:celloracle.utility.transfer_color_between_anndata}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.utility.}}\sphinxbfcode{\sphinxupquote{transfer\_color\_between\_anndata}}}{\emph{\DUrole{n}{anndata\_ref}}, \emph{\DUrole{n}{anndata\_que}}, \emph{\DUrole{n}{cluster\_name}}}{}
\end{fulllineitems}

\end{quote}


\paragraph{celloracle.data module}
\label{\detokenize{modules/celloracle.data:celloracle-data-module}}\label{\detokenize{modules/celloracle.data::doc}}\begin{quote}
\phantomsection\label{\detokenize{modules/celloracle.data:module-celloracle.data}}\index{module@\spxentry{module}!celloracle.data@\spxentry{celloracle.data}}\index{celloracle.data@\spxentry{celloracle.data}!module@\spxentry{module}}
The {\hyperref[\detokenize{modules/celloracle.data:module-celloracle.data}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{data}}}}} module implements data download and loading.
\index{load\_TFinfo\_df\_mm9\_mouse\_atac\_atlas() (in module celloracle.data)@\spxentry{load\_TFinfo\_df\_mm9\_mouse\_atac\_atlas()}\spxextra{in module celloracle.data}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.data:celloracle.data.load_TFinfo_df_mm9_mouse_atac_atlas}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.data.}}\sphinxbfcode{\sphinxupquote{load\_TFinfo\_df\_mm9\_mouse\_atac\_atlas}}}{}{}
Load Transcription factor binding information made from mouse scATAC\sphinxhyphen{}seq atlas dataset.
mm9 genome was used for the reference genome.

Args:
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
TF binding info.

\item[{Return type}] \leavevmode
pandas.dataframe

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}


\paragraph{celloracle.data\_conversion module}
\label{\detokenize{modules/celloracle.data_conversion:celloracle-data-conversion-module}}\label{\detokenize{modules/celloracle.data_conversion::doc}}\begin{quote}
\phantomsection\label{\detokenize{modules/celloracle.data_conversion:module-celloracle.data_conversion}}\index{module@\spxentry{module}!celloracle.data\_conversion@\spxentry{celloracle.data\_conversion}}\index{celloracle.data\_conversion@\spxentry{celloracle.data\_conversion}!module@\spxentry{module}}
The {\hyperref[\detokenize{modules/celloracle.data_conversion:module-celloracle.data_conversion}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{data\_conversion}}}}} module implements data conversion between different platform.
\index{seurat\_object\_to\_anndata() (in module celloracle.data\_conversion)@\spxentry{seurat\_object\_to\_anndata()}\spxextra{in module celloracle.data\_conversion}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules/celloracle.data_conversion:celloracle.data_conversion.seurat_object_to_anndata}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{celloracle.data\_conversion.}}\sphinxbfcode{\sphinxupquote{seurat\_object\_to\_anndata}}}{\emph{\DUrole{n}{file\_path\_seurat\_object}}, \emph{\DUrole{n}{delete\_tmp\_file}\DUrole{o}{=}\DUrole{default_value}{True}}}{}
Convert seurat object into anndata.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{file\_path\_seurat\_object}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} File path of seurat object. Seurat object should be saved as Rds format.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{delete\_tmp\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether to delete temporary file.

\end{itemize}

\item[{Returns}] \leavevmode
anndata object.

\item[{Return type}] \leavevmode
anndata

\end{description}\end{quote}

\end{fulllineitems}

\end{quote}
\end{quote}


\section{Changelog}
\label{\detokenize{changelog/index:changelog}}\label{\detokenize{changelog/index:id1}}\label{\detokenize{changelog/index::doc}}\begin{itemize}
\item {} 
\sphinxtitleref{0.3.1 \textless{}2020\sphinxhyphen{}03\sphinxhyphen{}23\textgreater{}}

\end{itemize}
\begin{quote}

\sphinxtitleref{\sphinxhyphen{}} Fix an error when try to save file larger than 4GB file
\end{quote}
\begin{itemize}
\item {} 
\sphinxtitleref{0.3.0 \textless{}2020\sphinxhyphen{}2\sphinxhyphen{}17\textgreater{}}

\end{itemize}
\begin{quote}

\sphinxtitleref{\sphinxhyphen{}} Release beta version
\end{quote}


\section{License}
\label{\detokenize{license/index:license}}\label{\detokenize{license/index:id1}}\label{\detokenize{license/index::doc}}
The software is provided under a modified Apache License Version 2.0.
The software may be used for non\sphinxhyphen{}commercial academic purposes only.
For any other use of the Work, including commercial use, please contact Morris lab.

Copyright 2020 Kenji Kamimoto, Christy Hoffmann, Samantha Morris

Apache License
Version 2.0, January 2004
\sphinxurl{http://www.apache.org/licenses/}

TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Definitions.

\end{enumerate}

“License” shall mean the terms and conditions for use, reproduction,
and distribution as defined by Sections 1 through 9 of this document.

“Licensor” shall mean the copyright owner or entity authorized by
the copyright owner that is granting the License.

“Legal Entity” shall mean the union of the acting entity and all
other entities that control, are controlled by, or are under common
control with that entity. For the purposes of this definition,
“control” means (i) the power, direct or indirect, to cause the
direction or management of such entity, whether by contract or
otherwise, or (ii) ownership of fifty percent (50\%) or more of the
outstanding shares, or (iii) beneficial ownership of such entity.

“You” (or “Your”) shall mean an individual or Legal Entity
exercising permissions granted by this License.

“Source” form shall mean the preferred form for making modifications,
including but not limited to software source code, documentation
source, and configuration files.

“Object” form shall mean any form resulting from mechanical
transformation or translation of a Source form, including but
not limited to compiled object code, generated documentation,
and conversions to other media types.

“Work” shall mean the work of authorship, whether in Source or
Object form, made available under the License, as indicated by a
copyright notice that is included in or attached to the work
(an example is provided in the Appendix below).

“Derivative Works” shall mean any work, whether in Source or Object
form, that is based on (or derived from) the Work and for which the
editorial revisions, annotations, elaborations, or other modifications
represent, as a whole, an original work of authorship. For the purposes
of this License, Derivative Works shall not include works that remain
separable from, or merely link (or bind by name) to the interfaces of,
the Work and Derivative Works thereof.

“Contribution” shall mean any work of authorship, including
the original version of the Work and any modifications or additions
to that Work or Derivative Works thereof, that is intentionally
submitted to Licensor for inclusion in the Work by the copyright owner
or by an individual or Legal Entity authorized to submit on behalf of
the copyright owner. For the purposes of this definition, “submitted”
means any form of electronic, verbal, or written communication sent
to the Licensor or its representatives, including but not limited to
communication on electronic mailing lists, source code control systems,
and issue tracking systems that are managed by, or on behalf of, the
Licensor for the purpose of discussing and improving the Work, but
excluding communication that is conspicuously marked or otherwise
designated in writing by the copyright owner as “Not a Contribution.”

“Contributor” shall mean Licensor and any individual or Legal Entity
on behalf of whom a Contribution has been received by Licensor and
subsequently incorporated within the Work.

2. Grant of Copyright License. Subject to the terms and conditions of
this License, each Contributor hereby grants to You a perpetual,
worldwide, non\sphinxhyphen{}exclusive, no\sphinxhyphen{}charge, royalty\sphinxhyphen{}free, irrevocable
copyright license to reproduce, prepare Derivative Works of,
publicly display, publicly perform, sublicense, and distribute the
Work and such Derivative Works in Source or Object form.

3. Grant of Patent License. Subject to the terms and conditions of
this License, each Contributor hereby grants to You a perpetual,
worldwide, non\sphinxhyphen{}exclusive, no\sphinxhyphen{}charge, royalty\sphinxhyphen{}free, irrevocable
(except as stated in this section) patent license to make, have made,
use, offer to sell, sell, import, and otherwise transfer the Work,
where such license applies only to those patent claims licensable
by such Contributor that are necessarily infringed by their
Contribution(s) alone or by combination of their Contribution(s)
with the Work to which such Contribution(s) was submitted. If You
institute patent litigation against any entity (including a
cross\sphinxhyphen{}claim or counterclaim in a lawsuit) alleging that the Work
or a Contribution incorporated within the Work constitutes direct
or contributory patent infringement, then any patent licenses
granted to You under this License for that Work shall terminate
as of the date such litigation is filed.

4. Redistribution. You may reproduce and distribute copies of the
Work or Derivative Works thereof in any medium, with or without
modifications, and in Source or Object form, provided that You
meet the following conditions:

(a) You must give any other recipients of the Work or
Derivative Works a copy of this License; and

(b) You must cause any modified files to carry prominent notices
stating that You changed the files; and

(c) You must retain, in the Source form of any Derivative Works
that You distribute, all copyright, patent, trademark, and
attribution notices from the Source form of the Work,
excluding those notices that do not pertain to any part of
the Derivative Works; and

(d) If the Work includes a “NOTICE” text file as part of its
distribution, then any Derivative Works that You distribute must
include a readable copy of the attribution notices contained
within such NOTICE file, excluding those notices that do not
pertain to any part of the Derivative Works, in at least one
of the following places: within a NOTICE text file distributed
as part of the Derivative Works; within the Source form or
documentation, if provided along with the Derivative Works; or,
within a display generated by the Derivative Works, if and
wherever such third\sphinxhyphen{}party notices normally appear. The contents
of the NOTICE file are for informational purposes only and
do not modify the License. You may add Your own attribution
notices within Derivative Works that You distribute, alongside
or as an addendum to the NOTICE text from the Work, provided
that such additional attribution notices cannot be construed
as modifying the License.

You may add Your own copyright statement to Your modifications and
may provide additional or different license terms and conditions
for use, reproduction, or distribution of Your modifications, or
for any such Derivative Works as a whole, provided Your use,
reproduction, and distribution of the Work otherwise complies with
the conditions stated in this License.

5. Submission of Contributions. Unless You explicitly state otherwise,
any Contribution intentionally submitted for inclusion in the Work
by You to the Licensor shall be under the terms and conditions of
this License, without any additional terms or conditions.
Notwithstanding the above, nothing herein shall supersede or modify
the terms of any separate license agreement you may have executed
with Licensor regarding such Contributions.

6. Trademarks. This License does not grant permission to use the trade
names, trademarks, service marks, or product names of the Licensor,
except as required for reasonable and customary use in describing the
origin of the Work and reproducing the content of the NOTICE file.

7. Disclaimer of Warranty. Unless required by applicable law or
agreed to in writing, Licensor provides the Work (and each
Contributor provides its Contributions) on an “AS IS” BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
implied, including, without limitation, any warranties or conditions
of TITLE, NON\sphinxhyphen{}INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A
PARTICULAR PURPOSE. You are solely responsible for determining the
appropriateness of using or redistributing the Work and assume any
risks associated with Your exercise of permissions under this License.

8. Limitation of Liability. In no event and under no legal theory,
whether in tort (including negligence), contract, or otherwise,
unless required by applicable law (such as deliberate and grossly
negligent acts) or agreed to in writing, shall any Contributor be
liable to You for damages, including any direct, indirect, special,
incidental, or consequential damages of any character arising as a
result of this License or out of the use or inability to use the
Work (including but not limited to damages for loss of goodwill,
work stoppage, computer failure or malfunction, or any and all
other commercial damages or losses), even if such Contributor
has been advised of the possibility of such damages.

9. Accepting Warranty or Additional Liability. While redistributing
the Work or Derivative Works thereof, You may choose to offer,
and charge a fee for, acceptance of support, warranty, indemnity,
or other liability obligations and/or rights consistent with this
License. However, in accepting such obligations, You may act only
on Your own behalf and on Your sole responsibility, not on behalf
of any other Contributor, and only if You agree to indemnify,
defend, and hold each Contributor harmless for any liability
incurred by, or claims asserted against, such Contributor by reason
of your accepting any such warranty or additional liability.

END OF TERMS AND CONDITIONS

APPENDIX: How to apply the Apache License to your work.

To apply the Apache License to your work, attach the following
boilerplate notice, with the fields enclosed by brackets “{[}{]}”
replaced with your own identifying information. (Don’t include
the brackets!)  The text should be enclosed in the appropriate
comment syntax for the file format. We also recommend that a
file or class name and description of purpose be included on the
same “printed page” as the copyright notice for easier
identification within third\sphinxhyphen{}party archives.

Copyright 2020 Kenji Kamimoto, Christy Hoffmann, Samantha Morris

Licensed under the Apache License, Version 2.0 (the “License”);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

\sphinxurl{http://www.apache.org/licenses/LICENSE-2.0}

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an “AS IS” BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


\section{Authors and citations}
\label{\detokenize{citation/index:authors-and-citations}}\label{\detokenize{citation/index:citing}}\label{\detokenize{citation/index::doc}}

\subsection{Cite celloracle}
\label{\detokenize{citation/index:cite-celloracle}}
If you use celloracle please cite our bioarxiv preprint \sphinxhref{https://www.biorxiv.org/content/10.1101/2020.02.17.947416v3}{CellOracle: Dissecting cell identity via network inference and in silico gene perturbation}.


\subsection{celloracle software development}
\label{\detokenize{citation/index:celloracle-software-development}}
celloracle is developed and maintained by \sphinxhref{http://morrislab.wustl.edu/lab-members/}{Kenji Kamimoto and members of Samantha Morris Lab}.
Please post troubles or questions on \sphinxhref{https://github.com/morris-lab/CellOracle/issues}{the Github repository}.


\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\let\bigletter\sphinxstyleindexlettergroup
\bigletter{c}
\item\relax\sphinxstyleindexentry{celloracle}\sphinxstyleindexpageref{modules/celloracle:\detokenize{module-celloracle}}
\item\relax\sphinxstyleindexentry{celloracle.data}\sphinxstyleindexpageref{modules/celloracle.data:\detokenize{module-celloracle.data}}
\item\relax\sphinxstyleindexentry{celloracle.data\_conversion}\sphinxstyleindexpageref{modules/celloracle.data_conversion:\detokenize{module-celloracle.data_conversion}}
\item\relax\sphinxstyleindexentry{celloracle.go\_analysis}\sphinxstyleindexpageref{modules/celloracle.go_analysis:\detokenize{module-celloracle.go_analysis}}
\item\relax\sphinxstyleindexentry{celloracle.motif\_analysis}\sphinxstyleindexpageref{modules/celloracle.motif_analysis:\detokenize{module-celloracle.motif_analysis}}
\item\relax\sphinxstyleindexentry{celloracle.network\_analysis}\sphinxstyleindexpageref{modules/celloracle.network_analysis:\detokenize{module-celloracle.network_analysis}}
\item\relax\sphinxstyleindexentry{celloracle.utility}\sphinxstyleindexpageref{modules/celloracle.utility:\detokenize{module-celloracle.utility}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}