

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>celloracle 0.1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'./',
              VERSION:'0.1.0',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html#document-index" class="icon icon-home"> celloracle
          

          
          </a>

          
            
            
          

          

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="index.html#document-index">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-installation/index">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#python-requirements">Python Requirements</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#optional-make-a-new-environment">0. (Optional) Make a new environment</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#add-conda-channels">1. Add conda channels</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#install-velocyto">2. Install velocyto</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#install-scanpy">3. Install scanpy</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#install-gimmemotifs">4. Install gimmemotifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#install-other-python-libraries">5. Install other python libraries</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#install-celloracle-from-github">6. install celloracle from github</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#r-requirements">R requirements</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#id3">Seurat</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#id5">Cicero</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#id7">igraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#id9">linkcomm</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#id11">rnetcarto</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#check-installation">Check installation</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-tutorials/index">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-tutorials/atac">1. ATAC-seq data preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#a-extract-tf-binding-information-from-scatac-seq-data">A. Extract TF binding information from scATAC-seq data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero">0. Import library</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#1.-Prepare-data">1. Prepare data</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#2.-Load-data-and-make-Cell-Data-Set-(CDS)-object">2. Load data and make Cell Data Set (CDS) object</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#2.1.-Process-data-to-make-CDS-object">2.1. Process data to make CDS object</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="index.html#3.-Qauality-check-and-Filtering">3. Qauality check and Filtering</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.-Process-cicero-CDS-object">4. Process cicero-CDS object</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#5.-Run-cicero-to-get-cis-regulatory-connection-scores">5. Run cicero to get cis-regulatory connection scores</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#6.-Save-results-for-next-step">6. Save results for next step</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data">0. Import libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#1.-Load-data-made-with-cicero">1. Load data made with cicero</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#2.-Make-TSS-annotation">2. Make TSS annotation</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#!!-Please-make-sure-that-you-are-setting-correct-reference-genoms.">!! Please make sure that you are setting correct reference genoms.</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="index.html#3.-Integrate-TSS-info-and-cicero-connections">3. Integrate TSS info and cicero connections</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.-Filter-peaks">4. Filter peaks</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#5.-Save-data">5. Save data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#b-extract-tf-binding-information-from-bulk-atac-seq-data-or-chip-seq-data">B. Extract TF binding information from bulk ATAC-seq data or Chip-seq data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data">0. Import libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#1.-Load-bed-file">1. Load bed file</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#2.-Make-TSS-annotation">2. Make TSS annotation</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#3.-Save-data">3. Save data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-tutorials/motifscan">2. Transcription factor binding motif scan</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#document-notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901">0. Import libraries</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#1.-Load-data">1. Load data</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#2.-Check-data">2. Check data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#2.1.-Remove-short-peaks">2.1. Remove short peaks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs">3. Instantiate TFinfo object and search for TF binding motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#3.1-check-reference-genome-installation">3.1 check reference genome installation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#3.2.-Install-reference-genome-(if-refgenome-is-not-installed)">3.2. Install reference genome (if refgenome is not installed)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#4.-Scan-motifs-and-save-object">4. Scan motifs and save object</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#5.-Filtering-motifs">5. Filtering motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#6.-Get-Final-results">6. Get Final results</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#6.1.-Get-resutls-as-a-dictionary">6.1. Get resutls as a dictionary</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#6.2.-Get-results-as-a-dataframe">6.2. Get results as a dataframe</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#7.-Save-TFinfo-as-dictionary-and-dataframe">7. Save TFinfo as dictionary and dataframe</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-tutorials/scrnaprocess">3. Single-cell RNA-seq data preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#a-scrna-seq-data-preprocessing-with-scanpy">A. scRNA-seq data preprocessing with scanpy</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data">0. Import libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#1.-Load-data">1. Load data</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#2.-Filtering">2. Filtering</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#3.-Normalization">3. Normalization</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.-Identification-of-highly-variable-genes">4. Identification of highly variable genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#5.-Log-transformation">5. Log transformation</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#6.-Dimensional-reduction">6. Dimensional reduction</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#7.-Clustering">7. Clustering</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#(Optional)-Re-calculate-Dimensional-reduction-graph">(Optional) Re-calculate Dimensional reduction graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#8.-Check-data">8. Check data</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#8.1.-Visualize-marker-gene-expression">8.1. Visualize marker gene expression</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="index.html#8.-Make-annotation-for-cluster">8. Make annotation for cluster</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#8.1.-Make-annotation-(1)">8.1. Make annotation (1)</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#8.2.-Make-annotation-(2)">8.2. Make annotation (2)</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="index.html#9.-(Option)-Subset-cells">9. (Option) Subset cells</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#10.-Save-data">10. Save data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#b-scrna-seq-data-preprocessing-with-seurat">B. scRNA-seq data preprocessing with Seurat</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-tutorials/networkanalysis">4. Network analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#document-notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data">0. Import libraries</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#0.1.-Check-installation">0.1. Check installation</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#0.2.-Make-a-folder-to-save-graph">0.2. Make a folder to save graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#1.-Load-data">1. Load data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#1.1.-Load-processed-gene-expression-data-(anndata)">1.1. Load processed gene expression data (anndata)</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#1.2.-Load-TF-data.">1.2. Load TF data.</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#2.-Initiate-Oracle-object">2. Initiate Oracle object</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#2.1.-load-gene-expression-data-into-oracle-object.">2.1. load gene expression data into oracle object.</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#2.2.-load-TFinfo-into-oracle-object">2.2. load TFinfo into oracle object</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#2.3.-(Optional)-Add-TF-info-manually">2.3. (Optional) Add TF info manually</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#2.3.1.-Make-TF-info-dictionary-manually">2.3.1. Make TF info dictionary manually</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#2.3.2.-Add-TF-informatio-dictionary-into-the-oracle-object">2.3.2. Add TF informatio dictionary into the oracle object</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#3.-Knn-imputation">3. Knn imputation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#3.1.-PCA">3.1. PCA</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#3.2.-Knn-imputation">3.2. Knn imputation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#4.-Save-and-Load.">4. Save and Load.</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#4.-GRN-calculation">4. GRN calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.1.-Get-GRNs">4.1. Get GRNs</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.2.-(Optional)-Export-GRNs">4.2. (Optional) Export GRNs</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.3.-(Optional)-Change-order">4.3. (Optional) Change order</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#5.-Network-preprocessing">5. Network preprocessing</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#5.1.-Filter-network-edges">5.1. Filter network edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#5.2.-Degree-distribution">5.2. Degree distribution</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#5.3.-Calculate-netowrk-score">5.3. Calculate netowrk score</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#5.4.-Save">5.4. Save</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#6.-Network-analysis;-Network-score-for-each-gene">6. Network analysis; Network score for each gene</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#6.1.-Network-score-in-each-cluster">6.1. Network score in each cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#6.2.-Network-score-comparison-between-two-clusters">6.2. Network score comparison between two clusters</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#6.3.-Network-score-dynamics">6.3. Network score dynamics</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#6.4.-Gene-cartography-analysis">6.4. Gene cartography analysis</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#7.-Network-analysis;-Network-score-distribution">7. Network analysis; Network score distribution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#7.1.-Distribution-of-network-degree">7.1. Distribution of network degree</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#7.2.-Distribution-of-netowrk-entolopy">7.2. Distribution of netowrk entolopy</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-tutorials/simulation">5. Simulation with GRNs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#document-notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data">0. Import libraries</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#0.1.-Import-public-libraries">0.1. Import public libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#0.1.-Make-a-folder-to-save-graph">0.1. Make a folder to save graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#1.-Load-data">1. Load data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#1.1.-Load-processed-oracle-object">1.1. Load processed oracle object</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#1.2.-Load-inferred-GRNs">1.2. Load inferred GRNs</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#3.-Make-predictive-models-for-simulation">3. Make predictive models for simulation</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#4.-in-silico-Perturbation-simulation">4. in silico Perturbation-simulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.1.-Check-gene-expression-pattern.">4.1. Check gene expression pattern.</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.1.-calculate-future-gene-expression-after-perturbation.">4.1. calculate future gene expression after perturbation.</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.2.-calculate-transition-probability-between-cells">4.2. calculate transition probability between cells</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.3.-Visualization">4.3. Visualization</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#4.3.1.-Detailed-directed-trajectory-graph">4.3.1. Detailed directed trajectory graph</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#4.3.2.-Grid-graph">4.3.2. Grid graph</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="index.html#4.4.-Markov-simulation-to-analyze-the-effects-of-perturbation-on-cell-fate-transition">4.4. Markov simulation to analyze the effects of perturbation on cell fate transition</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#4.4.1.-Do-simulation">4.4.1. Do simulation</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#4.4.2.-Check-the-results-of-simulation-focusing-on-a-specific-cells">4.4.2. Check the results of simulation focusing on a specific cells</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#4.4.3.-Summarize-the-results-of-simulation-by-plotting-sankey-diagram">4.4.3. Summarize the results of simulation by plotting sankey diagram</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-modules/index">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#command-line-api">Command Line API</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#python-api">Python API</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#document-modules/celloracle">Custom class in celloracle</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#modules-for-atac-seq-analysis">Modules for ATAC-seq analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-modules/celloracle.motif_analysis">celloracle.motif_analysis module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#modules-for-network-analysis">Modules for Network analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-modules/celloracle.network_analysis">celloracle.network_analysis module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#other-modules">Other modules</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-modules/celloracle.go_analysis">celloracle.go_analysis module</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-modules/celloracle.utility">celloracle.utility module</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-modules/celloracle.data">celloracle.data module</a></li>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-modules/celloracle.data_conversion">celloracle.data_conversion module</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-changelog">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-license/index">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-citation/index">Authors and citations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#cite-celloracle">Cite celloracle</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#celloracle-software-development">celloracle software development</a></li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html#document-index">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html#document-index">Docs</a> &raquo;</li>
        
      <li>celloracle 0.1.0 documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="welcome-to-celloracle-s-documentation">
<h1>Welcome to celloracle’s documentation!<a class="headerlink" href="#welcome-to-celloracle-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>CellOracle is a python library for the analysis of Gene Regulatory Network with single cell data.</p>
<p>Source code are available at <a class="reference external" href="https://github.com/KenjiKamimoto-wustl122/CellOracle">celloracle  git hub repository</a></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<div class="line-block">
<div class="line"><strong>For colleagues in Morris Lab</strong></div>
<div class="line"><strong>Thank you for your kind help!!</strong></div>
</div>
<div class="line-block">
<div class="line">Documentation is also available as a word and pdf file.</div>
</div>
<p><a class="reference download internal" href="_downloads/celloracle.pdf" download=""><code class="xref download docutils literal notranslate"><span class="pre">pdf</span> <span class="pre">documentation</span></code></a></p>
<p><a class="reference download internal" href="_downloads/documentation_microsoft_word.docx" download=""><code class="xref download docutils literal notranslate"><span class="pre">word</span> <span class="pre">documentation</span></code></a></p>
<div class="line-block">
<div class="line"><strong>Could you please add comments on the file and send it back to Kenji?</strong></div>
<div class="line">I really appreciate your support!</div>
<div class="line"><br /></div>
<div class="line">Thanks,</div>
<div class="line">Kenji</div>
</div>
<blockquote class="last">
<div>(This message will be deleted when we launch celloracle.)</div></blockquote>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">CellOracle is still under development. It is alpha version and functions in this package may change in the future release.</p>
</div>
</div>
<div class="section" id="contents">
<h1>Contents<a class="headerlink" href="#contents" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
<span id="document-installation/index"></span><div class="section" id="installation">
<span id="install"></span><h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">celloracle</span></code> uses several python libraries and R library. Please follow this guide below to install the dependent software of celloracle.</p>
<div class="section" id="python-requirements">
<span id="require"></span><h3>Python Requirements<a class="headerlink" href="#python-requirements" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">celloracle</span></code> was developed with python 3.6. We do not support python 2.7x or python &lt;=3.5.</li>
<li><code class="docutils literal notranslate"><span class="pre">celloracle</span></code> was developed in Linux and macOS. We do not guarantee that <code class="docutils literal notranslate"><span class="pre">celloracle</span></code> works in Windows OS.</li>
<li>We highly recommend using <a class="reference external" href="https://www.continuum.io/downloads">anaconda</a> to setup python environment.</li>
<li>Please install all dependent libraries before installing <code class="docutils literal notranslate"><span class="pre">celloracle</span></code> according to the instructions below.</li>
<li><code class="docutils literal notranslate"><span class="pre">celloracle</span></code>  is still beta version and it is not available through PyPI or anaconda distribution yet. Please install <code class="docutils literal notranslate"><span class="pre">celloracle</span></code> from GitHub repository according to the instruction below.</li>
</ul>
<div class="section" id="optional-make-a-new-environment">
<h4>0. (Optional) Make a new environment<a class="headerlink" href="#optional-make-a-new-environment" title="Permalink to this headline">¶</a></h4>
<p>This step is optional. Please make a new python environment for celloracle and install dependent libraries in it if you get some software conflicts.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">make</span> <span class="o">-</span><span class="n">n</span> <span class="n">celloracle_env</span> <span class="n">python</span><span class="o">=</span><span class="mf">3.6</span>
<span class="n">conda</span> <span class="n">activate</span> <span class="n">celloracle_env</span>
</pre></div>
</div>
</div>
<div class="section" id="add-conda-channels">
<h4>1. Add conda channels<a class="headerlink" href="#add-conda-channels" title="Permalink to this headline">¶</a></h4>
<p>Installation of some libraries below requires non-default anaconda channels. Please add the channels below. Instead, you can explicitly enter the channel when you install a library.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">defaults</span>
<span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">bioconda</span>
<span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span>
</pre></div>
</div>
</div>
<div class="section" id="install-velocyto">
<h4>2. Install <a class="reference external" href="http://velocyto.org/velocyto.py/install/index.html">velocyto</a><a class="headerlink" href="#install-velocyto" title="Permalink to this headline">¶</a></h4>
<p>Please install velocyto with the following commands or <a class="reference external" href="http://velocyto.org/velocyto.py/install/index.html">the author’s instruction</a> .
On Mac OS, you may have a compile error during velocyto installation. I recommend you installing <a class="reference external" href="https://developer.apple.com/xcode/">Xcode</a> in that case.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="n">numpy</span> <span class="n">scipy</span> <span class="n">cython</span> <span class="n">numba</span> <span class="n">matplotlib</span> <span class="n">scikit</span><span class="o">-</span><span class="n">learn</span> <span class="n">h5py</span> <span class="n">click</span>
</pre></div>
</div>
<p>Then</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">velocyto</span>
</pre></div>
</div>
</div>
<div class="section" id="install-scanpy">
<h4>3. Install <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/installation.html">scanpy</a><a class="headerlink" href="#install-scanpy" title="Permalink to this headline">¶</a></h4>
<p>Please install scanpy with the following commands or <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/installation.html">the author’s instruction</a> .</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="n">seaborn</span> <span class="n">scikit</span><span class="o">-</span><span class="n">learn</span> <span class="n">statsmodels</span> <span class="n">numba</span> <span class="n">pytables</span> <span class="n">python</span><span class="o">-</span><span class="n">igraph</span> <span class="n">louvain</span>
</pre></div>
</div>
<p>Then</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">scanpy</span>
</pre></div>
</div>
</div>
<div class="section" id="install-gimmemotifs">
<h4>4. Install <a class="reference external" href="https://gimmemotifs.readthedocs.io/en/master/installation.html">gimmemotifs</a><a class="headerlink" href="#install-gimmemotifs" title="Permalink to this headline">¶</a></h4>
<p>Please install gimmemotifs with the following commands or <a class="reference external" href="https://gimmemotifs.readthedocs.io/en/master/installation.html">the author’s instruction</a> .</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="n">gimmemotifs</span> <span class="n">genomepy</span><span class="o">=</span><span class="mf">0.5</span><span class="o">.</span><span class="mi">5</span>
</pre></div>
</div>
</div>
<div class="section" id="install-other-python-libraries">
<h4>5. Install other python libraries<a class="headerlink" href="#install-other-python-libraries" title="Permalink to this headline">¶</a></h4>
<p>Please install other python libraries below with the following commands.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="n">goatools</span> <span class="n">pyarrow</span> <span class="n">tqdm</span> <span class="n">joblib</span> <span class="n">jupyter</span>
</pre></div>
</div>
</div>
<div class="section" id="install-celloracle-from-github">
<h4>6. install celloracle from github<a class="headerlink" href="#install-celloracle-from-github" title="Permalink to this headline">¶</a></h4>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">git</span><span class="o">+</span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">KenjiKamimoto</span><span class="o">-</span><span class="n">wustl122</span><span class="o">/</span><span class="n">CellOracle</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="r-requirements">
<h3>R requirements<a class="headerlink" href="#r-requirements" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">celloracle</span></code> use R library for the network analysis and scATAC-seq analysis.
Please install <a class="reference external" href="https://www.r-project.org">R</a> (&gt;=3.5) and R libraries below according to the author’s instruction.</p>
<div class="section" id="id3">
<h4><a class="reference external" href="https://satijalab.org/seurat/install.html">Seurat</a><a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h4>
<p>Please install <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> with the following r-script or <a class="reference external" href="https://satijalab.org/seurat/install.html">the author’s instruction</a> .
<code class="docutils literal notranslate"><span class="pre">celloracle</span></code> is compatible with both Seurat V2 and V3.
If you use only <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> for the scRNA-seq preprocessing and do not use <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> , you can skip installation of <code class="docutils literal notranslate"><span class="pre">Seurat</span></code>.</p>
<p>in R console,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&#39;Seurat&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id5">
<h4><a class="reference external" href="https://cole-trapnell-lab.github.io/cicero-release/docs/#installing-cicero">Cicero</a><a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h4>
<p>Please install <code class="docutils literal notranslate"><span class="pre">Cicero</span></code> with the following r-script or <a class="reference external" href="https://cole-trapnell-lab.github.io/cicero-release/docs/#installing-cicero">the author’s instruction</a> .
If you have no plan for scATAC-seq analysis and just want to use  <code class="docutils literal notranslate"><span class="pre">celloracle</span></code> with a default TF information which was supplied with celloracle, you can skip installation of <code class="docutils literal notranslate"><span class="pre">Cicero</span></code>.</p>
<p>in R console,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">requireNamespace</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;cicero&quot;</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="s">&quot;3.8&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id7">
<h4><a class="reference external" href="https://igraph.org/r/">igraph</a><a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h4>
<p>Please install <code class="docutils literal notranslate"><span class="pre">igraph</span></code> with the following r-script or <a class="reference external" href="https://igraph.org/r/">the author’s instruction</a> .</p>
<p>in R console,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;igraph&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id9">
<h4><a class="reference external" href="https://cran.r-project.org/web/packages/linkcomm/index.html">linkcomm</a><a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h4>
<p>Please install <code class="docutils literal notranslate"><span class="pre">linkcomm</span></code> with the following r-script or <a class="reference external" href="https://cran.r-project.org/web/packages/linkcomm/index.html">the author’s instruction</a> .</p>
<p>in R console,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;linkcomm&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id11">
<h4><a class="reference external" href="https://github.com/cran/rnetcarto/blob/master/src/rgraph/README.md">rnetcarto</a><a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">rnetcarto</span></code> installation has to be done with several steps. Please install rnetcarto with <a class="reference external" href="https://github.com/cran/rnetcarto/blob/master/src/rgraph/README.md">the author’s instruction</a> .
You need to install <a class="reference external" href="https://www.gnu.org/software/gsl/">the GNU Scientific Libraries</a> before installing rnetcarto. Detailed instruction can be found <a class="reference external" href="https://github.com/cran/rnetcarto/blob/master/src/rgraph/README.md">here</a> .</p>
</div>
<div class="section" id="check-installation">
<h4>Check installation<a class="headerlink" href="#check-installation" title="Permalink to this headline">¶</a></h4>
<p>These R libraries above are necessary for the network analysis in celloracle. You can check installation using celloracle’s function.</p>
<p>in python console,</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">celloracle</span> <span class="kn">as</span> <span class="nn">co</span>
<span class="n">co</span><span class="o">.</span><span class="n">network_analysis</span><span class="o">.</span><span class="n">test_R_libraries_installation</span><span class="p">()</span>
</pre></div>
</div>
<p>Please make sure that all R libraries are installed. The following message will be shown when all R libraries are appropriately installed.</p>
<div class="line-block">
<div class="line">checking R library installation: gProfileR -&gt; OK</div>
<div class="line">checking R library installation: igraph -&gt; OK</div>
<div class="line">checking R library installation: linkcomm -&gt; OK</div>
<div class="line">checking R library installation: rnetcarto -&gt; OK</div>
</div>
</div>
</div>
</div>
</div>
<div class="toctree-wrapper compound">
<span id="document-tutorials/index"></span><div class="section" id="tutorial">
<span id="id1"></span><h2>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h2>
<p>The analysis proceeds through multiple steps.
Please run the notebooks following the number of the notebook.
If you do not have ATAC-seq data and want to use the default TF binding information, you can skip the first and second step and start from the third step.</p>
<p>Please refer to the <code class="docutils literal notranslate"><span class="pre">celloracle</span></code> paper for scientific premise and the detail of the algorithm of celloracle.</p>
<p>The jupyter notebook files in this tutorial are available <a class="reference external" href="https://github.com/KenjiKamimoto-wustl122/CellOracle/tree/master/docs/notebooks">here</a> .</p>
<div class="toctree-wrapper compound">
<span id="document-tutorials/atac"></span><div class="section" id="atac-seq-data-preprocessing">
<span id="atac"></span><h3>1. ATAC-seq data preprocessing<a class="headerlink" href="#atac-seq-data-preprocessing" title="Permalink to this headline">¶</a></h3>
<p>In this step, we process scATAC-seq data (or bulk ATAC-seq data) to get open accessible promoter/enhancer DNA sequence.
We can get active proximal promoter/enhancer genome sequence by picking up ATAC-seq peaks that exist around the transcription starting site (TSS).
Distal cis-regulatory elements can be picked up using  <a class="reference external" href="https://cole-trapnell-lab.github.io/cicero-release/docs/#installing-cicero">Cicero</a> .
Cicero analyzes scATAC-seq data to calculate a co-accessible score between peaks.
We can identify cis-regulatory elements using Cicero co-access score and TSS information.</p>
<p>If you have bulk ATAC-seq data instead of scATAC-data, we’ll get only proximal promoter/enhancer genome sequence.</p>
<div class="section" id="a-extract-tf-binding-information-from-scatac-seq-data">
<h4>A. Extract TF binding information from scATAC-seq data<a class="headerlink" href="#a-extract-tf-binding-information-from-scatac-seq-data" title="Permalink to this headline">¶</a></h4>
<p>If you have a scATAC-seq data, you can get information of distal cis-regulatory elements.
This step uses Cicero and does not use celloracle. Please refer to <a class="reference external" href="https://cole-trapnell-lab.github.io/cicero-release/">the documentation of Cicero</a> for the detailed usage.</p>
<p>R notebook</p>
<div class="toctree-wrapper compound">
<span id="document-notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero"></span><div class="section" id="0.-Import-library">
<h5>0. Import library<a class="headerlink" href="#0.-Import-library" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">library</span><span class="p">(</span><span class="n">cicero</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.-Prepare-data">
<h5>1. Prepare data<a class="headerlink" href="#1.-Prepare-data" title="Permalink to this headline">¶</a></h5>
<p>In this tutorial we use acATAC-seq data in 10x genomics database. You do not need to download these data if you analyze your scATAC data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># create folder to store data</span>
<span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;data&quot;</span><span class="p">)</span>

<span class="c1"># download demo dataset from 10x genomics</span>
<span class="nf">system</span><span class="p">(</span><span class="s">&quot;wget -O data/matrix.tar.gz http://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_filtered_peak_bc_matrix.tar.gz&quot;</span><span class="p">)</span>

<span class="c1"># unzip data</span>
<span class="nf">system</span><span class="p">(</span><span class="s">&quot;tar -xvf data/matrix.tar.gz -C data&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># You can substitute the data path below with the data path of your scATAC data.</span>
<span class="n">data_folder</span> <span class="o">&lt;-</span> <span class="s">&quot;data/filtered_peak_bc_matrix&quot;</span>

<span class="c1"># Create a folder to save results</span>
<span class="n">output_folder</span> <span class="o">&lt;-</span> <span class="s">&quot;cicero_output&quot;</span>
<span class="nf">dir.create</span><span class="p">(</span><span class="n">output_folder</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.-Load-data-and-make-Cell-Data-Set-(CDS)-object">
<h5>2. Load data and make Cell Data Set (CDS) object<a class="headerlink" href="#2.-Load-data-and-make-Cell-Data-Set-(CDS)-object" title="Permalink to this headline">¶</a></h5>
<div class="section" id="2.1.-Process-data-to-make-CDS-object">
<h6>2.1. Process data to make CDS object<a class="headerlink" href="#2.1.-Process-data-to-make-CDS-object" title="Permalink to this headline">¶</a></h6>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># read in matrix data using the Matrix package</span>
<span class="n">indata</span> <span class="o">&lt;-</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">readMM</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">data_folder</span><span class="p">,</span> <span class="s">&quot;/matrix.mtx&quot;</span><span class="p">))</span>
<span class="c1"># binarize the matrix</span>
<span class="n">indata</span><span class="o">@</span><span class="n">x[indata</span><span class="o">@</span><span class="n">x</span> <span class="o">&gt;</span> <span class="m">0</span><span class="n">]</span> <span class="o">&lt;-</span> <span class="m">1</span>

<span class="c1"># format cell info</span>
<span class="n">cellinfo</span> <span class="o">&lt;-</span> <span class="nf">read.table</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">data_folder</span><span class="p">,</span> <span class="s">&quot;/barcodes.tsv&quot;</span><span class="p">))</span>
<span class="nf">row.names</span><span class="p">(</span><span class="n">cellinfo</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">cellinfo</span><span class="o">$</span><span class="n">V1</span>
<span class="nf">names</span><span class="p">(</span><span class="n">cellinfo</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&quot;cells&quot;</span>

<span class="c1"># format peak info</span>
<span class="n">peakinfo</span> <span class="o">&lt;-</span> <span class="nf">read.table</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">data_folder</span><span class="p">,</span> <span class="s">&quot;/peaks.bed&quot;</span><span class="p">))</span>
<span class="nf">names</span><span class="p">(</span><span class="n">peakinfo</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span> <span class="s">&quot;bp1&quot;</span><span class="p">,</span> <span class="s">&quot;bp2&quot;</span><span class="p">)</span>
<span class="n">peakinfo</span><span class="o">$</span><span class="n">site_name</span> <span class="o">&lt;-</span> <span class="nf">paste</span><span class="p">(</span><span class="n">peakinfo</span><span class="o">$</span><span class="n">chr</span><span class="p">,</span> <span class="n">peakinfo</span><span class="o">$</span><span class="n">bp1</span><span class="p">,</span> <span class="n">peakinfo</span><span class="o">$</span><span class="n">bp2</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;_&quot;</span><span class="p">)</span>
<span class="nf">row.names</span><span class="p">(</span><span class="n">peakinfo</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">peakinfo</span><span class="o">$</span><span class="n">site_name</span>

<span class="nf">row.names</span><span class="p">(</span><span class="n">indata</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">row.names</span><span class="p">(</span><span class="n">peakinfo</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">indata</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">row.names</span><span class="p">(</span><span class="n">cellinfo</span><span class="p">)</span>

<span class="c1"># make CDS</span>
<span class="n">input_cds</span> <span class="o">&lt;-</span>  <span class="nf">suppressWarnings</span><span class="p">(</span><span class="nf">newCellDataSet</span><span class="p">(</span><span class="n">indata</span><span class="p">,</span>
                            <span class="n">phenoData</span> <span class="o">=</span> <span class="n">methods</span><span class="o">::</span><span class="nf">new</span><span class="p">(</span><span class="s">&quot;AnnotatedDataFrame&quot;</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">cellinfo</span><span class="p">),</span>
                            <span class="n">featureData</span> <span class="o">=</span> <span class="n">methods</span><span class="o">::</span><span class="nf">new</span><span class="p">(</span><span class="s">&quot;AnnotatedDataFrame&quot;</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">peakinfo</span><span class="p">),</span>
                            <span class="n">expressionFamily</span><span class="o">=</span><span class="n">VGAM</span><span class="o">::</span><span class="nf">binomialff</span><span class="p">(),</span>
                            <span class="n">lowerDetectionLimit</span><span class="o">=</span><span class="m">0</span><span class="p">))</span>
<span class="n">input_cds</span><span class="o">@</span><span class="n">expressionFamily</span><span class="o">@</span><span class="n">vfamily</span> <span class="o">&lt;-</span> <span class="s">&quot;binomialff&quot;</span>
<span class="n">input_cds</span> <span class="o">&lt;-</span> <span class="n">monocle</span><span class="o">::</span><span class="nf">detectGenes</span><span class="p">(</span><span class="n">input_cds</span><span class="p">)</span>

<span class="c1">#Ensure there are no peaks included with zero reads</span>
<span class="n">input_cds</span> <span class="o">&lt;-</span> <span class="n">input_cds[Matrix</span><span class="o">::</span><span class="nf">rowSums</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">input_cds</span><span class="p">))</span> <span class="o">&gt;=</span> <span class="m">100</span><span class="p">,</span><span class="n">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="3.-Qauality-check-and-Filtering">
<h5>3. Qauality check and Filtering<a class="headerlink" href="#3.-Qauality-check-and-Filtering" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Visualize peak_count_per_cell</span>
<span class="nf">hist</span><span class="p">(</span><span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">input_cds</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_01_ATAC-seq_data_processing_option1_scATAC-seq_data_analysis_with_cicero_01_atacdata_to_cicero_8_0.png"><img alt="_images/notebooks_01_ATAC-seq_data_processing_option1_scATAC-seq_data_analysis_with_cicero_01_atacdata_to_cicero_8_0.png" src="_images/notebooks_01_ATAC-seq_data_processing_option1_scATAC-seq_data_analysis_with_cicero_01_atacdata_to_cicero_8_0.png" style="width: 420px; height: 420px;" /></a>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># filter cells by peak_count</span>
<span class="n">max_count</span> <span class="o">&lt;-</span>  <span class="m">15000</span> <span class="c1"># Please change the threshold value according to the distribution of the peak_count of your data</span>
<span class="n">min_count</span> <span class="o">&lt;-</span> <span class="m">2000</span> <span class="c1"># Please change the threshold value according to the distribution of the peak_count of your data</span>
<span class="n">input_cds</span> <span class="o">&lt;-</span> <span class="n">input_cds[</span><span class="p">,</span><span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">input_cds</span><span class="p">))</span> <span class="o">&gt;=</span> <span class="n">min_count]</span>
<span class="n">input_cds</span> <span class="o">&lt;-</span> <span class="n">input_cds[</span><span class="p">,</span><span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">input_cds</span><span class="p">))</span> <span class="o">&lt;=</span> <span class="n">max_count]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="4.-Process-cicero-CDS-object">
<h5>4. Process cicero-CDS object<a class="headerlink" href="#4.-Process-cicero-CDS-object" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Run cicero to get cis-regulatory networks</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">2017</span><span class="p">)</span>
<span class="n">input_cds</span> <span class="o">&lt;-</span> <span class="nf">detectGenes</span><span class="p">(</span><span class="n">input_cds</span><span class="p">)</span>
<span class="n">input_cds</span> <span class="o">&lt;-</span> <span class="nf">estimateSizeFactors</span><span class="p">(</span><span class="n">input_cds</span><span class="p">)</span>

<span class="n">input_cds</span> <span class="o">&lt;-</span> <span class="nf">reduceDimension</span><span class="p">(</span><span class="n">input_cds</span><span class="p">,</span> <span class="n">max_components</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span><span class="n">scaling</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">relative_expr</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span>
                      <span class="n">reduction_method</span> <span class="o">=</span> <span class="s">&#39;tSNE&#39;</span><span class="p">,</span> <span class="n">norm_method</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span>

<span class="n">tsne_coords</span> <span class="o">&lt;-</span> <span class="nf">t</span><span class="p">(</span><span class="nf">reducedDimA</span><span class="p">(</span><span class="n">input_cds</span><span class="p">))</span>
<span class="nf">row.names</span><span class="p">(</span><span class="n">tsne_coords</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">row.names</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">input_cds</span><span class="p">))</span>
<span class="n">cicero_cds</span> <span class="o">&lt;-</span> <span class="nf">make_cicero_cds</span><span class="p">(</span><span class="n">input_cds</span><span class="p">,</span> <span class="n">reduced_coordinates</span> <span class="o">=</span> <span class="n">tsne_coords</span><span class="p">)</span>

<span class="c1"># Save cicero-CDS object if you want.</span>
<span class="c1">#saveRDS(cicero_cds, paste0(output_folder, &quot;/cicero_cds.Rds&quot;))</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Remove noise by PCA ...

Reduce dimension by tSNE ...

Overlap QC metrics:
Cells per bin: 50
Maximum shared cells bin-bin: 44
Mean shared cells bin-bin: 0.76256263875674
Median shared cells bin-bin: 0

</pre></div></div>
</div>
</div>
<div class="section" id="5.-Run-cicero-to-get-cis-regulatory-connection-scores">
<h5>5. Run cicero to get cis-regulatory connection scores<a class="headerlink" href="#5.-Run-cicero-to-get-cis-regulatory-connection-scores" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># import genome length, which is needed for the function, run_cicero</span>
<span class="n">mm10_chromosome_length</span> <span class="o">&lt;-</span> <span class="nf">read.table</span><span class="p">(</span><span class="s">&quot;./mm10_chromosome_length.txt&quot;</span><span class="p">)</span>

<span class="c1"># runc the main function</span>
<span class="n">conns</span> <span class="o">&lt;-</span> <span class="nf">run_cicero</span><span class="p">(</span><span class="n">cicero_cds</span><span class="p">,</span> <span class="n">mm10_chromosome_length</span><span class="p">)</span> <span class="c1"># Takes a few minutes to run</span>

<span class="c1"># check results</span>
<span class="nf">head</span><span class="p">(</span><span class="n">conns</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[1] &#34;Starting Cicero&#34;
[1] &#34;Calculating distance_parameter value&#34;
[1] &#34;Running models&#34;
[1] &#34;Assembling connections&#34;
[1] &#34;Done&#34;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<table>
<caption>A data.frame: 6 × 3</caption>
<thead>
    <tr><th></th><th scope=col>Peak1</th><th scope=col>Peak2</th><th scope=col>coaccess</th></tr>
    <tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>2</th><td>chr1_3094484_3095479</td><td>chr1_3113499_3113979</td><td>-0.316289004</td></tr>
    <tr><th scope=row>3</th><td>chr1_3094484_3095479</td><td>chr1_3119478_3121690</td><td>-0.419240532</td></tr>
    <tr><th scope=row>4</th><td>chr1_3094484_3095479</td><td>chr1_3399730_3400368</td><td>-0.050867246</td></tr>
    <tr><th scope=row>5</th><td>chr1_3113499_3113979</td><td>chr1_3094484_3095479</td><td>-0.316289004</td></tr>
    <tr><th scope=row>7</th><td>chr1_3113499_3113979</td><td>chr1_3119478_3121690</td><td> 0.370342744</td></tr>
    <tr><th scope=row>8</th><td>chr1_3113499_3113979</td><td>chr1_3399730_3400368</td><td>-0.009276026</td></tr>
</tbody>
</table></div>
</div>
</div>
<div class="section" id="6.-Save-results-for-next-step">
<h5>6. Save results for next step<a class="headerlink" href="#6.-Save-results-for-next-step" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">all_peaks</span> <span class="o">&lt;-</span> <span class="nf">row.names</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">input_cds</span><span class="p">))</span>
<span class="nf">write.csv</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">all_peaks</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="n">output_folder</span><span class="p">,</span> <span class="s">&quot;/all_peaks.csv&quot;</span><span class="p">))</span>
<span class="nf">write.csv</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">conns</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="n">output_folder</span><span class="p">,</span> <span class="s">&quot;/cicero_connections.csv&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</div>
<p>Next, the results of Cicero analysis will be processed to make TSS annotations.</p>
<p>Python notebook</p>
<div class="toctree-wrapper compound">
<span id="document-notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data"></span><p>In this notebook, we process the results of cicero analysis to get active promoter/enhancer DNA peaks. First, we pick up peaks around the transcription starting site (TSS). Second, we merge cicero data with the peaks around TSS. Then we remove peaks that have a weak connection to TSS peak so that the final product includes TSS peaks and peaks that have a strong connection with the TSS peaks. We use this information as an active promoter/enhancer elements.</p>
<div class="section" id="0.-Import-libraries">
<h5>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="o">%</span><span class="k">matplotlib</span> inline

<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="c1">#import time</span>

<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">shutil</span><span class="o">,</span> <span class="nn">importlib</span><span class="o">,</span> <span class="nn">glob</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>

<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;

<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;savefig.dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">300</span>


</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">celloracle</span> <span class="k">import</span> <span class="n">motif_analysis</span> <span class="k">as</span> <span class="n">ma</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.-Load-data-made-with-cicero">
<h5>1. Load data made with cicero<a class="headerlink" href="#1.-Load-data-made-with-cicero" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># load all peaks</span>
<span class="n">peaks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;cicero_output/all_peaks.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">peaks</span> <span class="o">=</span> <span class="n">peaks</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">values</span>
<span class="n">peaks</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>array([&#39;chr1_3094484_3095479&#39;, &#39;chr1_3113499_3113979&#39;,
       &#39;chr1_3119478_3121690&#39;, ..., &#39;chrY_90804622_90805450&#39;,
       &#39;chrY_90808626_90809117&#39;, &#39;chrY_90810560_90811167&#39;], dtype=object)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># load cicero results</span>
<span class="n">cicero_connections</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;cicero_output/cicero_connections.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">cicero_connections</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Peak1</th>
      <th>Peak2</th>
      <th>coaccess</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>2</td>
      <td>chr1_3094484_3095479</td>
      <td>chr1_3113499_3113979</td>
      <td>-0.316289</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr1_3094484_3095479</td>
      <td>chr1_3119478_3121690</td>
      <td>-0.419241</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr1_3094484_3095479</td>
      <td>chr1_3399730_3400368</td>
      <td>-0.050867</td>
    </tr>
    <tr>
      <td>5</td>
      <td>chr1_3113499_3113979</td>
      <td>chr1_3094484_3095479</td>
      <td>-0.316289</td>
    </tr>
    <tr>
      <td>7</td>
      <td>chr1_3113499_3113979</td>
      <td>chr1_3119478_3121690</td>
      <td>0.370343</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="2.-Make-TSS-annotation">
<h5>2. Make TSS annotation<a class="headerlink" href="#2.-Make-TSS-annotation" title="Permalink to this headline">¶</a></h5>
<div class="section" id="!!-Please-make-sure-that-you-are-setting-correct-reference-genoms.">
<h6>!! Please make sure that you are setting correct reference genoms.<a class="headerlink" href="#!!-Please-make-sure-that-you-are-setting-correct-reference-genoms." title="Permalink to this headline">¶</a></h6>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">tss_annotated</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">get_tss_info</span><span class="p">(</span><span class="n">peak_str_list</span><span class="o">=</span><span class="n">peaks</span><span class="p">,</span> <span class="n">ref_genome</span><span class="o">=</span><span class="s2">&quot;mm10&quot;</span><span class="p">)</span>

<span class="c1"># check results</span>
<span class="n">tss_annotated</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
que bed peaks: 72402
tss peaks in que: 16987
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chr</th>
      <th>start</th>
      <th>end</th>
      <th>gene_short_name</th>
      <th>strand</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>16982</td>
      <td>chr1</td>
      <td>55130650</td>
      <td>55132118</td>
      <td>Mob4</td>
      <td>+</td>
    </tr>
    <tr>
      <td>16983</td>
      <td>chr6</td>
      <td>94499875</td>
      <td>94500767</td>
      <td>Slc25a26</td>
      <td>+</td>
    </tr>
    <tr>
      <td>16984</td>
      <td>chr19</td>
      <td>45659222</td>
      <td>45660823</td>
      <td>Fbxw4</td>
      <td>-</td>
    </tr>
    <tr>
      <td>16985</td>
      <td>chr12</td>
      <td>100898848</td>
      <td>100899597</td>
      <td>Gpr68</td>
      <td>-</td>
    </tr>
    <tr>
      <td>16986</td>
      <td>chr4</td>
      <td>129491262</td>
      <td>129492047</td>
      <td>Fam229a</td>
      <td>-</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
</div>
<div class="section" id="3.-Integrate-TSS-info-and-cicero-connections">
<h5>3. Integrate TSS info and cicero connections<a class="headerlink" href="#3.-Integrate-TSS-info-and-cicero-connections" title="Permalink to this headline">¶</a></h5>
<p>The output file after the integration process has three columns; “peak_id”, “gene_short_name”, and “coaccess”. “peak_id” is either of TSS peak or peaks that have a connection with a TSS peak. “gene_short_name” is a gene name of the TSS site. “coaccess” is a co-access score between a peak and TSS peak. If the score is 1, it means that the peak is TSS itself.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">integrated</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">integrate_tss_peak_with_cicero</span><span class="p">(</span><span class="n">tss_peak</span><span class="o">=</span><span class="n">tss_annotated</span><span class="p">,</span>
                                               <span class="n">cicero_connections</span><span class="o">=</span><span class="n">cicero_connections</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">integrated</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">integrated</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(263279, 3)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak_id</th>
      <th>gene_short_name</th>
      <th>coaccess</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr10_100015291_100017830</td>
      <td>Kitl</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr10_100018677_100020384</td>
      <td>Kitl</td>
      <td>0.086299</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr10_100050858_100051762</td>
      <td>Kitl</td>
      <td>0.034558</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr10_100052829_100053395</td>
      <td>Kitl</td>
      <td>0.167188</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr10_100128086_100128882</td>
      <td>Tmtc3</td>
      <td>0.022341</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="4.-Filter-peaks">
<h5>4. Filter peaks<a class="headerlink" href="#4.-Filter-peaks" title="Permalink to this headline">¶</a></h5>
<p>Remove peaks that have weak coaccess score.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">peak</span> <span class="o">=</span> <span class="n">integrated</span><span class="p">[</span><span class="n">integrated</span><span class="o">.</span><span class="n">coaccess</span> <span class="o">&gt;=</span> <span class="mf">0.8</span><span class="p">]</span>
<span class="n">peak</span> <span class="o">=</span> <span class="n">peak</span><span class="p">[[</span><span class="s2">&quot;peak_id&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_short_name&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">print</span><span class="p">(</span><span class="n">peak</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">peak</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(15680, 2)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak_id</th>
      <th>gene_short_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr10_100015291_100017830</td>
      <td>Kitl</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr10_100486534_100488209</td>
      <td>Tmtc3</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr10_100588641_100589556</td>
      <td>4930430F08Rik</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr10_100741247_100742505</td>
      <td>Gm35722</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr10_101681379_101682124</td>
      <td>Mgat4c</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="5.-Save-data">
<h5>5. Save data<a class="headerlink" href="#5.-Save-data" title="Permalink to this headline">¶</a></h5>
<p>Save the promoter/enhancer peak.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">peak</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="s2">&quot;peak_file.parquet&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>-&gt; go to next notebook</p>
</div>
</div>
</div>
<div class="section" id="b-extract-tf-binding-information-from-bulk-atac-seq-data-or-chip-seq-data">
<h4>B. Extract TF binding information from bulk ATAC-seq data or Chip-seq data<a class="headerlink" href="#b-extract-tf-binding-information-from-bulk-atac-seq-data-or-chip-seq-data" title="Permalink to this headline">¶</a></h4>
<p>Bulk DNA-seq data can be used to get open accessible promoter/enhancer sequence.</p>
<p>Python notebook</p>
<div class="toctree-wrapper compound">
<span id="document-notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data"></span><div class="section" id="0.-Import-libraries">
<h5>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="o">%</span><span class="k">matplotlib</span> inline

<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="c1">#import time</span>

<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">shutil</span><span class="o">,</span> <span class="nn">importlib</span><span class="o">,</span> <span class="nn">glob</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>

<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;

<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;savefig.dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">300</span>

</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># import celloracle function</span>
<span class="kn">from</span> <span class="nn">celloracle</span> <span class="k">import</span> <span class="n">motif_analysis</span> <span class="k">as</span> <span class="n">ma</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.-Load-bed-file">
<h5>1. Load bed file<a class="headerlink" href="#1.-Load-bed-file" title="Permalink to this headline">¶</a></h5>
<p>Import bed file of ATAC-seq data. This script can also be used DNase-seq or Chip-seq data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">file_path_of_bed_file</span> <span class="o">=</span> <span class="s2">&quot;data/all_peaks.bed&quot;</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># load bed_file</span>
<span class="n">bed</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="n">file_path_of_bed_file</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">bed</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(436206, 4)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>start</th>
      <th>end</th>
      <th>seqname</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr1</td>
      <td>3002478</td>
      <td>3002968</td>
      <td>chr1_3002478_3002968</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr1</td>
      <td>3084739</td>
      <td>3085712</td>
      <td>chr1_3084739_3085712</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr1</td>
      <td>3103576</td>
      <td>3104022</td>
      <td>chr1_3103576_3104022</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr1</td>
      <td>3106871</td>
      <td>3107210</td>
      <td>chr1_3106871_3107210</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr1</td>
      <td>3108932</td>
      <td>3109158</td>
      <td>chr1_3108932_3109158</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># convert bed file into peak name list</span>
<span class="n">peaks</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">process_bed_file</span><span class="o">.</span><span class="n">df_to_list_peakstr</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span>
<span class="n">peaks</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>array([&#39;chr1_3002478_3002968&#39;, &#39;chr1_3084739_3085712&#39;,
       &#39;chr1_3103576_3104022&#39;, ..., &#39;chrY_631222_631480&#39;,
       &#39;chrY_795887_796426&#39;, &#39;chrY_2397419_2397628&#39;], dtype=object)
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.-Make-TSS-annotation">
<h5>2. Make TSS annotation<a class="headerlink" href="#2.-Make-TSS-annotation" title="Permalink to this headline">¶</a></h5>
<p>!! Please make sure that you are setting correct ref genoms.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">tss_annotated</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">get_tss_info</span><span class="p">(</span><span class="n">peak_str_list</span><span class="o">=</span><span class="n">peaks</span><span class="p">,</span> <span class="n">ref_genome</span><span class="o">=</span><span class="s2">&quot;mm9&quot;</span><span class="p">)</span>

<span class="c1"># check results</span>
<span class="n">tss_annotated</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
que bed peaks: 436206
tss peaks in que: 24822
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chr</th>
      <th>start</th>
      <th>end</th>
      <th>gene_short_name</th>
      <th>strand</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>24817</td>
      <td>chr2</td>
      <td>60560211</td>
      <td>60561602</td>
      <td>Itgb6</td>
      <td>-</td>
    </tr>
    <tr>
      <td>24818</td>
      <td>chr15</td>
      <td>3975177</td>
      <td>3978654</td>
      <td>BC037032</td>
      <td>-</td>
    </tr>
    <tr>
      <td>24819</td>
      <td>chr14</td>
      <td>67690701</td>
      <td>67692101</td>
      <td>Ppp2r2a</td>
      <td>-</td>
    </tr>
    <tr>
      <td>24820</td>
      <td>chr17</td>
      <td>48455247</td>
      <td>48455773</td>
      <td>B430306N03Rik</td>
      <td>+</td>
    </tr>
    <tr>
      <td>24821</td>
      <td>chr10</td>
      <td>59861192</td>
      <td>59861608</td>
      <td>Gm17455</td>
      <td>+</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># change format</span>
<span class="n">peak_id_tss</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">process_bed_file</span><span class="o">.</span><span class="n">df_to_list_peakstr</span><span class="p">(</span><span class="n">tss_annotated</span><span class="p">)</span>
<span class="n">tss_annotated</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;peak_id&quot;</span><span class="p">:</span> <span class="n">peak_id_tss</span><span class="p">,</span>
                              <span class="s2">&quot;gene_short_name&quot;</span><span class="p">:</span> <span class="n">tss_annotated</span><span class="o">.</span><span class="n">gene_short_name</span><span class="o">.</span><span class="n">values</span><span class="p">})</span>
<span class="n">tss_annotated</span> <span class="o">=</span> <span class="n">tss_annotated</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tss_annotated</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">tss_annotated</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(24822, 2)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak_id</th>
      <th>gene_short_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr7_50691730_50692032</td>
      <td>Nkg7</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr7_50692077_50692785</td>
      <td>Nkg7</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr13_93564413_93564836</td>
      <td>Thbs4</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr13_14613429_14615645</td>
      <td>Hecw1</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr3_99688753_99689665</td>
      <td>Spag17</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="3.-Save-data">
<h5>3. Save data<a class="headerlink" href="#3.-Save-data" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">tss_annotated</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="s2">&quot;peak_file.parquet&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>-&gt; go to next notebook</p>
</div>
</div>
</div>
</div>
<span id="document-tutorials/motifscan"></span><div class="section" id="transcription-factor-binding-motif-scan">
<span id="motifscan"></span><h3>2. Transcription factor binding motif scan<a class="headerlink" href="#transcription-factor-binding-motif-scan" title="Permalink to this headline">¶</a></h3>
<p>We identified open-accessible Promoter/enhancer DNAs using ATAC-seq data.
Next, we scan the regulatory genomic sequence searching for the TF-binding motifs to get the list of TFs for each target gene.
In the later GRN inference process, this TF list per target gene will be used as a potential regulatory connection.</p>
<p>Python notebook</p>
<div class="toctree-wrapper compound">
<span id="document-notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901"></span><div class="section" id="0.-Import-libraries">
<h4>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">¶</a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="o">%</span><span class="k">matplotlib</span> inline

<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="c1">#import time</span>

<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">shutil</span><span class="o">,</span> <span class="nn">importlib</span><span class="o">,</span> <span class="nn">glob</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>

<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;

<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">7</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;savefig.dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">600</span>

</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">celloracle</span> <span class="k">import</span> <span class="n">motif_analysis</span> <span class="k">as</span> <span class="n">ma</span>
<span class="kn">from</span> <span class="nn">celloracle.utility</span> <span class="k">import</span> <span class="n">save_as_pickled_object</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.-Load-data">
<h4>1. Load data<a class="headerlink" href="#1.-Load-data" title="Permalink to this headline">¶</a></h4>
<p>load annotated peak data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">peaks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s2">&quot;../01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/peak_file.parquet&quot;</span><span class="p">)</span>
<span class="n">peaks</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak_id</th>
      <th>gene_short_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr10_100015291_100017830</td>
      <td>Kitl</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr10_100486534_100488209</td>
      <td>Tmtc3</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr10_100588641_100589556</td>
      <td>4930430F08Rik</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr10_100741247_100742505</td>
      <td>Gm35722</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr10_101681379_101682124</td>
      <td>Mgat4c</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="2.-Check-data">
<h4>2. Check data<a class="headerlink" href="#2.-Check-data" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check data</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;number of peak: {len(peaks.peak_id.unique())}&quot;</span><span class="p">)</span>
<span class="n">mean_</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">peaks</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;gene_short_name&quot;</span><span class="p">,</span> <span class="s2">&quot;peak_id&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">count</span><span class="p">())</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">peaks</span><span class="o">.</span><span class="n">gene_short_name</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;mean peaks per gene: </span><span class="si">{mean_}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">getLength</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;peak_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">b</span><span class="p">)</span>

<span class="c1">#</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">peaks</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">getLength</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;mean peak length: {df.values.mean()}&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
number of peak: 13919
mean peaks per gene: 1.0591731964333964
mean peak length: 1756.1744260204082
</pre></div></div>
</div>
<div class="section" id="2.1.-Remove-short-peaks">
<h5>2.1. Remove short peaks<a class="headerlink" href="#2.1.-Remove-short-peaks" title="Permalink to this headline">¶</a></h5>
<p>Short DNA fragment, such as less than 5 bases, cannot be used for motif scan. Remove short DNA fragment.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">peaks</span> <span class="o">=</span> <span class="n">peaks</span><span class="p">[</span><span class="n">df</span><span class="o">&gt;=</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs">
<h4>3. Instantiate TFinfo object and search for TF binding motifs<a class="headerlink" href="#3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs" title="Permalink to this headline">¶</a></h4>
<p>The motif analysis module has a custom class; TFinfo. TFinfo object converts a peak data into a DNA sequence and scans the DNA sequence searching for TF binding motifs. Then the results of motif scan will be filtered and converted into several files, such as python dictionary or dataframe. This TF information is necessary for GRN inference.</p>
</div>
<div class="section" id="3.1-check-reference-genome-installation">
<h4>3.1 check reference genome installation<a class="headerlink" href="#3.1-check-reference-genome-installation" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># PLEASE make sure that you are setting correct ref genome.</span>
<span class="n">ref_genome</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span>

<span class="n">ma</span><span class="o">.</span><span class="n">is_genome_installed</span><span class="p">(</span><span class="n">ref_genome</span><span class="o">=</span><span class="n">ref_genome</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
genome mm10 is not installed in this environment.
Please install genome using genomepy.
e.g.
    &gt;&gt;&gt; import genomepy
    &gt;&gt;&gt; genomepy.install_genome(&#34;mm9&#34;, &#34;UCSC&#34;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>False
</pre></div>
</div>
</div>
<div class="section" id="3.2.-Install-reference-genome-(if-refgenome-is-not-installed)">
<h5>3.2. Install reference genome (if refgenome is not installed)<a class="headerlink" href="#3.2.-Install-reference-genome-(if-refgenome-is-not-installed)" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">genomepy</span>
<span class="n">genomepy</span><span class="o">.</span><span class="n">install_genome</span><span class="p">(</span><span class="n">ref_genome</span><span class="p">,</span> <span class="s2">&quot;UCSC&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
downloading from http://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz...
done...
name: mm10
local name: mm10
fasta: /home/k/.local/share/genomes/mm10/mm10.fa
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check again</span>
<span class="n">ma</span><span class="o">.</span><span class="n">is_genome_installed</span><span class="p">(</span><span class="n">ref_genome</span><span class="o">=</span><span class="n">ref_genome</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>True
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Instantiate TFinfo object</span>
<span class="n">tfi</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">TFinfo</span><span class="p">(</span><span class="n">peak_data_frame</span><span class="o">=</span><span class="n">peaks</span><span class="p">,</span> <span class="c1"># peak info calculated from ATAC-seq data</span>
                <span class="n">ref_genome</span><span class="o">=</span><span class="n">ref_genome</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="4.-Scan-motifs-and-save-object">
<h4>4. Scan motifs and save object<a class="headerlink" href="#4.-Scan-motifs-and-save-object" title="Permalink to this headline">¶</a></h4>
<p>This step requires computational resource and may take long time</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%%time</span>
<span class="c1"># Scan motifs</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">scan</span><span class="p">(</span><span class="n">fpr</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># sace tfinfo object</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="s2">&quot;test.celloracle.tfinfo&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
initiating scanner ...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2019-09-22 23:00:18,604 - INFO - Using background: genome mm10 with length 200
2019-09-22 23:00:18,986 - INFO - Determining FPR-based threshold
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
getting DNA sequences ...
scanning motifs ...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "4a3dafd90b384a499ec1697f746b16c9", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

CPU times: user 52min 23s, sys: 36.8 s, total: 53min
Wall time: 52min 58s
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check motif scan results</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">scanned_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seqname</th>
      <th>motif_id</th>
      <th>factors_direct</th>
      <th>factors_indirect</th>
      <th>score</th>
      <th>pos</th>
      <th>strand</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Homeodomain.0001</td>
      <td>TGIF1</td>
      <td>ENSG00000234254, TGIF1</td>
      <td>10.311002</td>
      <td>1003</td>
      <td>1</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Mixed.0001</td>
      <td></td>
      <td>SRF, EGR1</td>
      <td>7.925873</td>
      <td>481</td>
      <td>1</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Mixed.0001</td>
      <td></td>
      <td>SRF, EGR1</td>
      <td>7.321375</td>
      <td>911</td>
      <td>-1</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Mixed.0001</td>
      <td></td>
      <td>SRF, EGR1</td>
      <td>7.276585</td>
      <td>811</td>
      <td>-1</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Nuclear_receptor.0002</td>
      <td>NR2C2</td>
      <td>NR2C2, Nr2c2</td>
      <td>9.067331</td>
      <td>449</td>
      <td>-1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We have the score for each sequence and motif_id pair. In the next step we will filter the motifs with low score.</p>
</div>
<div class="section" id="5.-Filtering-motifs">
<h4>5. Filtering motifs<a class="headerlink" href="#5.-Filtering-motifs" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># reset filtering</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">reset_filtering</span><span class="p">()</span>

<span class="c1"># do filtering</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">filter_motifs_by_score</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="mf">10.5</span><span class="p">)</span>

<span class="c1"># do post filtering process. Convert results into several file format.</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">make_TFinfo_dataframe_and_dictionary</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
peaks were filtered: 12934005 -&gt; 2285279
1. converting scanned results into one-hot encoded dataframe.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "666addf9ba6540fcb9762b2bbb40abbb", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

2. converting results into dictionaries.
converting scan results into dictionaries...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "92fac19056a7410fb88cf25038b4270c", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "afb7b3742cf246dca47e5bdfdfc0c5f9", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
</div>
<div class="section" id="6.-Get-Final-results">
<h4>6. Get Final results<a class="headerlink" href="#6.-Get-Final-results" title="Permalink to this headline">¶</a></h4>
<div class="section" id="6.1.-Get-resutls-as-a-dictionary">
<h5>6.1. Get resutls as a dictionary<a class="headerlink" href="#6.1.-Get-resutls-as-a-dictionary" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">td</span> <span class="o">=</span> <span class="n">tfi</span><span class="o">.</span><span class="n">to_dictionary</span><span class="p">(</span><span class="n">dictionary_type</span><span class="o">=</span><span class="s2">&quot;targetgene2TFs&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
</div>
<div class="section" id="6.2.-Get-results-as-a-dataframe">
<h5>6.2. Get results as a dataframe<a class="headerlink" href="#6.2.-Get-results-as-a-dataframe" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df</span> <span class="o">=</span> <span class="n">tfi</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak_id</th>
      <th>gene_short_name</th>
      <th>9430076c15rik</th>
      <th>Ac002126.6</th>
      <th>Ac012531.1</th>
      <th>Ac226150.2</th>
      <th>Afp</th>
      <th>Ahr</th>
      <th>Ahrr</th>
      <th>Aire</th>
      <th>...</th>
      <th>Znf784</th>
      <th>Znf8</th>
      <th>Znf816</th>
      <th>Znf85</th>
      <th>Zscan10</th>
      <th>Zscan16</th>
      <th>Zscan22</th>
      <th>Zscan26</th>
      <th>Zscan31</th>
      <th>Zscan4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr10_100015291_100017830</td>
      <td>Kitl</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr10_100486534_100488209</td>
      <td>Tmtc3</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr10_100588641_100589556</td>
      <td>4930430F08Rik</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr10_100741247_100742505</td>
      <td>Gm35722</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr10_101681379_101682124</td>
      <td>Mgat4c</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 1092 columns</p>
</div></div>
</div>
</div>
</div>
<div class="section" id="7.-Save-TFinfo-as-dictionary-and-dataframe">
<h4>7. Save TFinfo as dictionary and dataframe<a class="headerlink" href="#7.-Save-TFinfo-as-dictionary-and-dataframe" title="Permalink to this headline">¶</a></h4>
<p>We’ll use this information in the later step to make GRN. Save the results.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">folder</span> <span class="o">=</span> <span class="s2">&quot;TFinfo_outputs&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># save TFinfo as a dictionary</span>
<span class="n">td</span> <span class="o">=</span> <span class="n">tfi</span><span class="o">.</span><span class="n">to_dictionary</span><span class="p">(</span><span class="n">dictionary_type</span><span class="o">=</span><span class="s2">&quot;targetgene2TFs&quot;</span><span class="p">)</span>
<span class="n">save_as_pickled_object</span><span class="p">(</span><span class="n">td</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="s2">&quot;TFinfo_targetgene2TFs.pickled&quot;</span><span class="p">))</span>

<span class="c1"># save TFinfo as a dataframe</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">tfi</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="s2">&quot;TFinfo_dataframe.parquet&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<span id="document-tutorials/scrnaprocess"></span><div class="section" id="single-cell-rna-seq-data-preprocessing">
<span id="scrnaprocess"></span><h3>3. Single-cell RNA-seq data preprocessing<a class="headerlink" href="#single-cell-rna-seq-data-preprocessing" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt>Network analysis and simulation in celloracle will be performed using scRNA-seq data. The scRNA-seq data should include the compontnes below.</dt>
<dd><ul class="first last simple">
<li>Gene expression matrix; mRNA counts before scaling and transformation.</li>
<li>Clustering results.</li>
<li>Dimensional reduction results.</li>
</ul>
</dd>
<dt>In addition to these minimum requirements, we highly recommend doing these analyses below in the preprocessing step.</dt>
<dd><ul class="first last simple">
<li>Data quality check and Cell/Gene filtering.</li>
<li>Normalization</li>
<li>Identification of highly variable genes</li>
</ul>
</dd>
</dl>
<p>We recommend processing scRNA-seq data using either scanpy or Seurat.
If you are not familiar with the general workflow of scRNA-seq data processing, please go to <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">the documentation of scanpy</a> and <a class="reference external" href="https://satijalab.org/seurat/vignettes.html">the documentation of Seurat</a> .</p>
<p>If you already have preprocessed scRNA-seq data, which includes necessary information above, you can skip this part.</p>
<div class="section" id="a-scrna-seq-data-preprocessing-with-scanpy">
<h4>A. scRNA-seq data preprocessing with scanpy<a class="headerlink" href="#a-scrna-seq-data-preprocessing-with-scanpy" title="Permalink to this headline">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">scanpy</span></code> is a python library for the analysis of scRNA-seq data.</p>
<p>In this tutorial, we introduce an example of scRNA-seq preprocessing for celloracle with <code class="docutils literal notranslate"><span class="pre">scanpy</span></code>.
We wrote the notebook based on <a class="reference external" href="https://scanpy-tutorials.readthedocs.io/en/latest/paga-paul15.html">one of scanpy’s tutorials</a> with some modification.</p>
<p>Python notebook</p>
<div class="toctree-wrapper compound">
<span id="document-notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data"></span><div class="section" id="0.-Import-libraries">
<h5>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;savefig.dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">300</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.-Load-data">
<h5>1. Load data<a class="headerlink" href="#1.-Load-data" title="Permalink to this headline">¶</a></h5>
<p>In this notebook, we will show an example of how to process scRNA-seq data using a scRNA-seq data of hematopoiesis (Paul, F., Arkin, Y., Giladi, A., Jaitin, D. A., Kenigsberg, E., Keren-Shaul, H., et al. (2015). Transcriptional Heterogeneity and Lineage Commitment in Myeloid Progenitors. Cell, 163(7), 1663–1677. <a class="reference external" href="http://doi.org/10.1016/j.cell.2015.11.013">http://doi.org/10.1016/j.cell.2015.11.013</a>). You can easily download this scRNA-seq data with a scanpy’s function.</p>
<p>Please change the code below if you want to use your data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># download dataset. You can change the code blow if you use another data.</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">paul15</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: In Scanpy 0.*, this returned logarithmized data. Now it returns non-logarithmized data.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;paul15_clusters&#39; as categorical
Trying to set attribute `.uns` of view, making a copy.
</pre></div></div>
</div>
</div>
<div class="section" id="2.-Filtering">
<h5>2. Filtering<a class="headerlink" href="#2.-Filtering" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># only consider genes with more than 1 count</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_counts</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

</pre></div>
</div>
</div>
</div>
<div class="section" id="3.-Normalization">
<h5>3. Normalization<a class="headerlink" href="#3.-Normalization" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># normalize gene expression matrix with total UMI count per cell</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_per_cell</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">key_n_counts</span><span class="o">=</span><span class="s1">&#39;n_counts_all&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="4.-Identification-of-highly-variable-genes">
<h5>4. Identification of highly variable genes<a class="headerlink" href="#4.-Identification-of-highly-variable-genes" title="Permalink to this headline">¶</a></h5>
<p>Removing non-variable genes reduces calculation time in the GRN reconstruction and simulation. We recommend using top 2000~3000 variable genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># select top 2000 highly-variable genes</span>
<span class="n">filter_result</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes_dispersion</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span>
                                              <span class="n">flavor</span><span class="o">=</span><span class="s1">&#39;cell_ranger&#39;</span><span class="p">,</span>
                                              <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span>
                                              <span class="n">log</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># subset the genes</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">filter_result</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">]</span>

<span class="c1"># renormalize after filtering</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_per_cell</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Trying to set attribute `.obs` of view, making a copy.
</pre></div></div>
</div>
</div>
<div class="section" id="5.-Log-transformation">
<h5>5. Log transformation<a class="headerlink" href="#5.-Log-transformation" title="Permalink to this headline">¶</a></h5>
<p>We will do log transformation scaling because these are necessary for PCA, clustering, differential gene calculation. However, we also need non-transformed gene expression data in the celloracle analysis. Thus we keep raw count in anndata with the following command before log transformation.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># keep raw cont data before log transformation</span>
<span class="n">adata</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">adata</span>

<span class="c1"># Log transformation and scaling</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># (optional) Regressing out some values help reduce batch effects.</span>

<span class="c1">#sc.pp.regress_out(adata, [&#39;n_counts&#39;, &#39;percent_mito&#39;])</span>
<span class="c1">#sc.pp.regress_out(adata, [&#39;n_counts&#39;])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="6.-Dimensional-reduction">
<h5>6. Dimensional reduction<a class="headerlink" href="#6.-Dimensional-reduction" title="Permalink to this headline">¶</a></h5>
<p>Dimensional reduction is one of the most important parts of the scRNA-seq analysis. celloracle needs dimensional reduction embeddings to simulate cell transition.</p>
<p>Please choose a proper algorithm of dimensional reduction so that the embedding appropriately represents the data structure. We recommend using one of the dimensional reduction algorithms (or trajectory inference algorithms); UMAP, tSNE, diffusion map, Force-directed graph drawing, PAGA.</p>
<p>In this example, we use a combination of four algorithms; diffusion map, force-directed graph drawing, and PAGA.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># PCA</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># diffusion map</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">diffmap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="c1"># calculate neihbors again based on diffusionmap</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_diffmap&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="7.-Clustering">
<h5>7. Clustering<a class="headerlink" href="#7.-Clustering" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">louvain</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="(Optional)-Re-calculate-Dimensional-reduction-graph">
<h5>(Optional) Re-calculate Dimensional reduction graph<a class="headerlink" href="#(Optional)-Re-calculate-Dimensional-reduction-graph" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># PAGA graph construction</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="s1">&#39;louvain&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check current cluster name</span>
<span class="n">cluster_list</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">louvain</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="n">cluster_list</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>[5, 2, 12, 13, 0, ..., 6, 20, 14, 15, 21]
Length: 23
Categories (23, object): [5, 2, 12, 13, ..., 20, 14, 15, 21]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_23_0.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_23_0.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_23_0.png" style="width: 383px; height: 277px;" /></a>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">init_pos</span><span class="o">=</span><span class="s1">&#39;paga&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">123</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_25_0.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_25_0.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_25_0.png" style="width: 364px; height: 288px;" /></a>
</div>
</div>
</div>
<div class="section" id="8.-Check-data">
<h5>8. Check data<a class="headerlink" href="#8.-Check-data" title="Permalink to this headline">¶</a></h5>
<div class="section" id="8.1.-Visualize-marker-gene-expression">
<h6>8.1. Visualize marker gene expression<a class="headerlink" href="#8.1.-Visualize-marker-gene-expression" title="Permalink to this headline">¶</a></h6>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mf">4.5</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">markers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Erythroids&quot;</span><span class="p">:[</span><span class="s2">&quot;Gata1&quot;</span><span class="p">,</span> <span class="s2">&quot;Klf1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gypa&quot;</span><span class="p">,</span> <span class="s2">&quot;Hba-a2&quot;</span><span class="p">],</span>
           <span class="s2">&quot;Megakaryocytes&quot;</span><span class="p">:[</span><span class="s2">&quot;Itga2b&quot;</span><span class="p">,</span> <span class="s2">&quot;Pbx1&quot;</span><span class="p">,</span> <span class="s2">&quot;Sdpr&quot;</span><span class="p">,</span> <span class="s2">&quot;Vwf&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Granulocytes&quot;</span><span class="p">:[</span><span class="s2">&quot;Elane&quot;</span><span class="p">,</span> <span class="s2">&quot;Cebpe&quot;</span><span class="p">,</span> <span class="s2">&quot;Ctsg&quot;</span><span class="p">,</span> <span class="s2">&quot;Mpo&quot;</span><span class="p">,</span> <span class="s2">&quot;Gfi1&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Monocytes&quot;</span><span class="p">:[</span><span class="s2">&quot;Irf8&quot;</span><span class="p">,</span> <span class="s2">&quot;Csf1r&quot;</span><span class="p">,</span> <span class="s2">&quot;Ctsg&quot;</span><span class="p">,</span> <span class="s2">&quot;Mpo&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Mast_cells&quot;</span><span class="p">:[</span><span class="s2">&quot;Cma1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gzmb&quot;</span><span class="p">,</span> <span class="s2">&quot;Kit&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Basophils&quot;</span><span class="p">:[</span><span class="s2">&quot;Mcpt8&quot;</span><span class="p">,</span> <span class="s2">&quot;Prss34&quot;</span><span class="p">]</span>
            <span class="p">}</span>

<span class="k">for</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">genes</span> <span class="ow">in</span> <span class="n">markers</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;marker gene of </span><span class="si">{cell_type}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>


</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
marker gene of Erythroids
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_1.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_1.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_1.png" style="width: 656px; height: 574px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
marker gene of Megakaryocytes
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_3.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_3.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_3.png" style="width: 644px; height: 574px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
marker gene of Granulocytes
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_5.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_5.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_5.png" style="width: 655px; height: 856px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
marker gene of Monocytes
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_7.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_7.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_7.png" style="width: 656px; height: 574px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
marker gene of Mast_cells
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_9.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_9.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_9.png" style="width: 644px; height: 574px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
marker gene of Basophils
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_11.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_11.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_28_11.png" style="width: 644px; height: 292px;" /></a>
</div>
</div>
</div>
</div>
<div class="section" id="8.-Make-annotation-for-cluster">
<h5>8. Make annotation for cluster<a class="headerlink" href="#8.-Make-annotation-for-cluster" title="Permalink to this headline">¶</a></h5>
<p>Based on the marker gene expression and previous report, we will manually make annotation for each cluster.</p>
<div class="section" id="8.1.-Make-annotation-(1)">
<h6>8.1. Make annotation (1)<a class="headerlink" href="#8.1.-Make-annotation-(1)" title="Permalink to this headline">¶</a></h6>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;paul15_clusters&#39;</span><span class="p">],</span>
                 <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_32_0.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_32_0.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_32_0.png" style="width: 615px; height: 292px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check current cluster name</span>
<span class="n">cluster_list</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">louvain</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="n">cluster_list</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>[5, 2, 12, 13, 0, ..., 6, 20, 14, 15, 21]
Length: 23
Categories (23, object): [5, 2, 12, 13, ..., 20, 14, 15, 21]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># make anottation dictionary</span>
<span class="n">annotation</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;MEP&quot;</span><span class="p">:[</span><span class="mi">5</span><span class="p">],</span>
              <span class="s2">&quot;Erythroids&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">15</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">19</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">18</span><span class="p">],</span>
              <span class="s2">&quot;Megakaryocytes&quot;</span><span class="p">:[</span><span class="mi">17</span><span class="p">,</span> <span class="mi">22</span><span class="p">],</span>
              <span class="s2">&quot;GMP&quot;</span><span class="p">:[</span><span class="mi">11</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
              <span class="s2">&quot;late_GMP&quot;</span> <span class="p">:[</span><span class="mi">0</span><span class="p">],</span>
              <span class="s2">&quot;Granulocytes&quot;</span><span class="p">:[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
              <span class="s2">&quot;Monocytes&quot;</span><span class="p">:[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
              <span class="s2">&quot;DC&quot;</span><span class="p">:[</span><span class="mi">21</span><span class="p">],</span>
              <span class="s2">&quot;Lymphoid&quot;</span><span class="p">:[</span><span class="mi">20</span><span class="p">]}</span>

<span class="c1"># change dictionary format</span>
<span class="n">annotation_rev</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cluster_list</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">annotation</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="ow">in</span> <span class="n">annotation</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
            <span class="n">annotation_rev</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span>

<span class="c1"># check dictionary</span>
<span class="n">annotation_rev</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>{&#39;5&#39;: &#39;MEP&#39;,
 &#39;2&#39;: &#39;Monocytes&#39;,
 &#39;12&#39;: &#39;Erythroids&#39;,
 &#39;13&#39;: &#39;Granulocytes&#39;,
 &#39;0&#39;: &#39;late_GMP&#39;,
 &#39;10&#39;: &#39;Erythroids&#39;,
 &#39;3&#39;: &#39;Erythroids&#39;,
 &#39;18&#39;: &#39;Erythroids&#39;,
 &#39;11&#39;: &#39;GMP&#39;,
 &#39;7&#39;: &#39;Granulocytes&#39;,
 &#39;8&#39;: &#39;Erythroids&#39;,
 &#39;22&#39;: &#39;Megakaryocytes&#39;,
 &#39;16&#39;: &#39;Erythroids&#39;,
 &#39;1&#39;: &#39;GMP&#39;,
 &#39;17&#39;: &#39;Megakaryocytes&#39;,
 &#39;4&#39;: &#39;Granulocytes&#39;,
 &#39;19&#39;: &#39;Erythroids&#39;,
 &#39;9&#39;: &#39;Erythroids&#39;,
 &#39;6&#39;: &#39;Monocytes&#39;,
 &#39;20&#39;: &#39;Lymphoid&#39;,
 &#39;14&#39;: &#39;Erythroids&#39;,
 &#39;15&#39;: &#39;Erythroids&#39;,
 &#39;21&#39;: &#39;DC&#39;}
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">annotation_rev</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">louvain</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check results</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span> <span class="s1">&#39;paul15_clusters&#39;</span><span class="p">],</span>
                 <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;cell_type&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_36_1.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_36_1.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_36_1.png" style="width: 615px; height: 292px;" /></a>
</div>
</div>
</div>
<div class="section" id="8.2.-Make-annotation-(2)">
<h6>8.2. Make annotation (2)<a class="headerlink" href="#8.2.-Make-annotation-(2)" title="Permalink to this headline">¶</a></h6>
<p>We’ll make another annotation manually for each louvain clusters</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type&#39;</span><span class="p">],</span>
                 <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_38_0.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_38_0.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_38_0.png" style="width: 615px; height: 292px;" /></a>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">annotation_2</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;5&#39;</span><span class="p">:</span> <span class="s1">&#39;MEP_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;15&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;10&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_1&#39;</span><span class="p">,</span>
                <span class="s1">&#39;16&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_2&#39;</span><span class="p">,</span>
                <span class="s1">&#39;14&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_3&#39;</span><span class="p">,</span>
                <span class="s1">&#39;9&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_4&#39;</span><span class="p">,</span>
                <span class="s1">&#39;8&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_5&#39;</span><span class="p">,</span>
                <span class="s1">&#39;19&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_6&#39;</span><span class="p">,</span>
                <span class="s1">&#39;3&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_7&#39;</span><span class="p">,</span>
                <span class="s1">&#39;12&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_8&#39;</span><span class="p">,</span>
                <span class="s1">&#39;18&#39;</span><span class="p">:</span> <span class="s1">&#39;Ery_9&#39;</span><span class="p">,</span>
                <span class="s1">&#39;17&#39;</span><span class="p">:</span> <span class="s1">&#39;Mk_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;22&#39;</span><span class="p">:</span> <span class="s1">&#39;Mk_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;11&#39;</span><span class="p">:</span> <span class="s1">&#39;GMP_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;1&#39;</span><span class="p">:</span> <span class="s1">&#39;GMP_1&#39;</span><span class="p">,</span>
                <span class="s1">&#39;0&#39;</span><span class="p">:</span> <span class="s1">&#39;GMPl_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;7&#39;</span><span class="p">:</span> <span class="s1">&#39;Gran_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;13&#39;</span><span class="p">:</span> <span class="s1">&#39;Gran_1&#39;</span><span class="p">,</span>
                <span class="s1">&#39;4&#39;</span><span class="p">:</span> <span class="s1">&#39;Gran_2&#39;</span><span class="p">,</span>
                <span class="s1">&#39;6&#39;</span><span class="p">:</span> <span class="s1">&#39;Mo_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;2&#39;</span><span class="p">:</span> <span class="s1">&#39;Mo_1&#39;</span><span class="p">,</span>
                <span class="s1">&#39;21&#39;</span><span class="p">:</span> <span class="s1">&#39;DC_0&#39;</span><span class="p">,</span>
                <span class="s1">&#39;20&#39;</span><span class="p">:</span> <span class="s1">&#39;Lym_0&#39;</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;louvain_annot&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">annotation_2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">louvain</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check result</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;louvain_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type&#39;</span><span class="p">],</span>
                 <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;louvain_annot&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_41_1.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_41_1.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_41_1.png" style="width: 615px; height: 292px;" /></a>
</div>
</div>
<p>We’ve done secveral scRNA-preprocessing steps; filtering, normalization, clustering, dimensional reduction. In the next step, we’ll do GRN inference, network analysis, and in silico simulation based on these information.</p>
</div>
</div>
<div class="section" id="9.-(Option)-Subset-cells">
<h5>9. (Option) Subset cells<a class="headerlink" href="#9.-(Option)-Subset-cells" title="Permalink to this headline">¶</a></h5>
<p>In this tutorial, we are using scRNA-seq data of hematopoiesis. In the latter part, we will focus on the cell fate decision in the myeloid lineage. So we remove non-myeloid cell cluster, DC and Lymphoid now.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>[MEP, Monocytes, Erythroids, Granulocytes, late_GMP, GMP, Megakaryocytes, Lymphoid, DC]
Categories (9, object): [MEP, Monocytes, Erythroids, Granulocytes, ..., GMP, Megakaryocytes, Lymphoid, DC]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cell_of_interest</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="o">~</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;Lymphoid&quot;</span><span class="p">,</span> <span class="s2">&quot;DC&quot;</span><span class="p">])]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">cell_of_interest</span><span class="p">,</span> <span class="p">:]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check result</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;louvain_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type&#39;</span><span class="p">],</span>
                 <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_46_0.png"><img alt="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_46_0.png" src="_images/notebooks_03_scRNA-seq_data_preprocessing_scanpy_preprocessing_with_Paul_etal_2015_data_46_0.png" style="width: 615px; height: 292px;" /></a>
</div>
</div>
</div>
<div class="section" id="10.-Save-data">
<h5>10. Save data<a class="headerlink" href="#10.-Save-data" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#adata.write_h5ad(&quot;data/Paul_etal_15.h5ad&quot;)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="b-scrna-seq-data-preprocessing-with-seurat">
<h4>B. scRNA-seq data preprocessing with Seurat<a class="headerlink" href="#b-scrna-seq-data-preprocessing-with-seurat" title="Permalink to this headline">¶</a></h4>
<p>R notebook … comming soon</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If you use <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> for preprocessing, you need to convert the scRNA-seq data (Seurat object) into anndata to analyze the data with <code class="docutils literal notranslate"><span class="pre">celloracle</span></code>.
<code class="docutils literal notranslate"><span class="pre">celloracle</span></code> have python API and command-line API to convert Seurat object into anndata.
Please go to the documentation of celloracle API for more information.</p>
</div>
</div>
</div>
<span id="document-tutorials/networkanalysis"></span><div class="section" id="network-analysis">
<span id="networkanalysis"></span><h3>4. Network analysis<a class="headerlink" href="#network-analysis" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">celloracle</span></code> import scRNA-seq dataset and TF binding information to find active regulatory connections for all genes, generating sample-specific GRNs.</p>
<p>The inferred GRN is analyzed with several network algorithms to get various network scores. The network score is useful to identify key regulatory genes.</p>
<p>celloracle reconstructs GRN for each cluster enabling us to compare GRNs each other. It is also possible to analyze how the GRN change over differentiation.
The dynamics of the GRN structure provide us insight into the
context-dependent regulatory mechanism.</p>
<p>Python notebook</p>
<div class="toctree-wrapper compound">
<span id="document-notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data"></span><div class="section" id="0.-Import-libraries">
<h4>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">¶</a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># 0. Import</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">celloracle</span> <span class="k">as</span> <span class="nn">co</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># visualization settings</span>
<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;
<span class="o">%</span><span class="k">matplotlib</span> inline

<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;savefig.dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">300</span>

</pre></div>
</div>
</div>
<div class="section" id="0.1.-Check-installation">
<h5>0.1. Check installation<a class="headerlink" href="#0.1.-Check-installation" title="Permalink to this headline">¶</a></h5>
<p>Celloracle uses some R libraries in network analysis. Please make sure that all dependent R libraries are installed on your computer. You can test the installation with the following command.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">co</span><span class="o">.</span><span class="n">network_analysis</span><span class="o">.</span><span class="n">test_R_libraries_installation</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
checking R library installation: igraph -&gt; OK
checking R library installation: linkcomm -&gt; OK
checking R library installation: rnetcarto -&gt; OK
</pre></div></div>
</div>
</div>
<div class="section" id="0.2.-Make-a-folder-to-save-graph">
<h5>0.2. Make a folder to save graph<a class="headerlink" href="#0.2.-Make-a-folder-to-save-graph" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">save_folder</span> <span class="o">=</span> <span class="s2">&quot;figures&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">save_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="1.-Load-data">
<h4>1. Load data<a class="headerlink" href="#1.-Load-data" title="Permalink to this headline">¶</a></h4>
<div class="section" id="1.1.-Load-processed-gene-expression-data-(anndata)">
<h5>1.1. Load processed gene expression data (anndata)<a class="headerlink" href="#1.1.-Load-processed-gene-expression-data-(anndata)" title="Permalink to this headline">¶</a></h5>
<p>Please refer to the previous notebook in the tutorial for an example of how to process scRNA-seq data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># load data. !!Replace the data path below when you use another data.</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;../03_scRNA-seq_data_preprocessing/data/Paul_etal_15.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.2.-Load-TF-data.">
<h5>1.2. Load TF data.<a class="headerlink" href="#1.2.-Load-TF-data." title="Permalink to this headline">¶</a></h5>
<p>For the GRN inference, celloracle needs TF information, which contains lists of the regulatory candidate gene. There are several ways to make such TF information. We can generate TF information from scATAC-seq data or Bulk ATAC-seq data. Please refer to the first step of the tutorial for the details of how to make TF information.</p>
<p>If you do not have your scATAC-seq data, you can use some built-in data in celloracle. celloracle have a TFinfo made with various kind of tissue/cell-types of mouse ATAC-seq atlas dataset (<a class="reference external" href="http://atlas.gs.washington.edu/mouse-atac/">http://atlas.gs.washington.edu/mouse-atac/</a>).</p>
<p>You can load and use the data with the following command.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Load TF info which was made from mouse cell atlas dataset.</span>
<span class="n">TFinfo_df</span> <span class="o">=</span> <span class="n">co</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">load_TFinfo_df_mm9_mouse_atac_atlas</span><span class="p">()</span>

<span class="c1"># check data</span>
<span class="n">TFinfo_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak_id</th>
      <th>gene_short_name</th>
      <th>9430076c15rik</th>
      <th>Ac002126.6</th>
      <th>Ac012531.1</th>
      <th>Ac226150.2</th>
      <th>Afp</th>
      <th>Ahr</th>
      <th>Ahrr</th>
      <th>Aire</th>
      <th>...</th>
      <th>Znf784</th>
      <th>Znf8</th>
      <th>Znf816</th>
      <th>Znf85</th>
      <th>Zscan10</th>
      <th>Zscan16</th>
      <th>Zscan22</th>
      <th>Zscan26</th>
      <th>Zscan31</th>
      <th>Zscan4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr10_100050979_100052296</td>
      <td>4930430F08Rik</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr10_101006922_101007748</td>
      <td>SNORA17</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr10_101144061_101145000</td>
      <td>Mgat4c</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr10_10148873_10149183</td>
      <td>9130014G24Rik</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr10_10149425_10149815</td>
      <td>9130014G24Rik</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 1095 columns</p>
</div></div>
</div>
</div>
</div>
<div class="section" id="2.-Initiate-Oracle-object">
<h4>2. Initiate Oracle object<a class="headerlink" href="#2.-Initiate-Oracle-object" title="Permalink to this headline">¶</a></h4>
<p>celloracle have a custom class, Oracle. We can use Oracle for the data preprocessing and GRN inference. Oracle object stores all information and do the calculation with its internal functions. We instantiate an Oracle object, then input a gene expression data (anndata) and a TFinfo into the Oracle object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># make Oracle object</span>
<span class="n">oracle</span> <span class="o">=</span> <span class="n">co</span><span class="o">.</span><span class="n">Oracle</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="2.1.-load-gene-expression-data-into-oracle-object.">
<h5>2.1. load gene expression data into oracle object.<a class="headerlink" href="#2.1.-load-gene-expression-data-into-oracle-object." title="Permalink to this headline">¶</a></h5>
<p>When you load scRNA-seq data, please enter the name of clustering data and dimensional reduction data. Clustering data suppose to be stored in the attribute of “obs” in the anndata. Dimensional reduction data suppose to be stored in the attribute of “obsm” in the anndata. You can check these data by the following command.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># show data name in anndata</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;metadata columns :&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;dimensional reduction: &quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
metadata columns : [&#39;paul15_clusters&#39;, &#39;n_counts_all&#39;, &#39;n_counts&#39;, &#39;louvain&#39;, &#39;cell_type&#39;, &#39;louvain_annot&#39;]
dimensional reduction:  [&#39;X_diffmap&#39;, &#39;X_draw_graph_fa&#39;, &#39;X_pca&#39;]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The anndata shoud include (1) gene expression count, (2) clustering information, (3) trajectory (dimensional reduction embeddings) data.</span>
<span class="c1"># Please refer to another notebook for the detail of anndata preprocessing.</span>

<span class="c1"># In this notebook, we use raw mRNA count as an input of Oracle object.</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># Instantiate Oracle object. All calculation will be done in this object.</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">import_anndata_as_raw_count</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                                   <span class="n">cluster_column_name</span><span class="o">=</span><span class="s2">&quot;louvain_annot&quot;</span><span class="p">,</span>
                                   <span class="n">embedding_name</span><span class="o">=</span><span class="s2">&quot;X_draw_graph_fa&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.2.-load-TFinfo-into-oracle-object">
<h5>2.2. load TFinfo into oracle object<a class="headerlink" href="#2.2.-load-TFinfo-into-oracle-object" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># If you can load TF info dataframe with the following code.</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">import_TF_data</span><span class="p">(</span><span class="n">TF_info_matrix</span><span class="o">=</span><span class="n">TFinfo_df</span><span class="p">)</span>

<span class="c1"># Instead of the way avove, you can use TF &quot;dictionary&quot; with the following code.</span>
<span class="c1"># oracle.import_TF_data(TFdict=TFinfo_dictionary)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.3.-(Optional)-Add-TF-info-manually">
<h5>2.3. (Optional) Add TF info manually<a class="headerlink" href="#2.3.-(Optional)-Add-TF-info-manually" title="Permalink to this headline">¶</a></h5>
<p>While we mainly use TF info data made from scATAC-seq data, we can also add arbitrary information about the TF-target gene pair by the manual way.</p>
<p>For example, if there is a study or database about TF-target pair, you can use such information in the following way.</p>
<div class="section" id="2.3.1.-Make-TF-info-dictionary-manually">
<h6>2.3.1. Make TF info dictionary manually<a class="headerlink" href="#2.3.1.-Make-TF-info-dictionary-manually" title="Permalink to this headline">¶</a></h6>
<p>Here, we introduce a way to add TF binding information.</p>
<p>We have a TF binding data in about some TFs. The information is the supplemental table 4 in (<a class="reference external" href="http://doi.org/10.1016/j.cell.2015.11.013">http://doi.org/10.1016/j.cell.2015.11.013</a>).</p>
<p>In order to import TF data into Oracle object, we need to convert them into a python dictionary, in which dictionary keys are target gene, and values are regulatory candidate TFs.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># We have TF and its target gene information. This is from a supplemental Fig of Paul et. al, (2015).</span>
<span class="n">Paul_15_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;TF_data_in_Paul15.csv&quot;</span><span class="p">)</span>
<span class="n">Paul_15_data</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TF</th>
      <th>Target_genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>Cebpa</td>
      <td>Abcb1b, Acot1, C3, Cnpy3, Dhrs7, Dtx4, Edem2, ...</td>
    </tr>
    <tr>
      <td>1</td>
      <td>Irf8</td>
      <td>Abcd1, Aif1, BC017643, Cbl, Ccdc109b, Ccl6, d6...</td>
    </tr>
    <tr>
      <td>2</td>
      <td>Irf8</td>
      <td>1100001G20Rik, 4732418C07Rik, 9230105E10Rik, A...</td>
    </tr>
    <tr>
      <td>3</td>
      <td>Klf1</td>
      <td>2010011I20Rik, 5730469M10Rik, Acsl6, Add2, Ank...</td>
    </tr>
    <tr>
      <td>4</td>
      <td>Sfpi1</td>
      <td>0910001L09Rik, 2310014H01Rik, 4632428N05Rik, A...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Make dictionary: Key is TF, Value is list of Target gene</span>
<span class="n">TF_to_TG_dictionary</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">TF</span><span class="p">,</span> <span class="n">TGs</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">Paul_15_data</span><span class="o">.</span><span class="n">TF</span><span class="p">,</span> <span class="n">Paul_15_data</span><span class="o">.</span><span class="n">Target_genes</span><span class="p">):</span>
    <span class="c1"># convert target gene to list</span>
    <span class="n">TG_list</span> <span class="o">=</span> <span class="n">TGs</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
    <span class="c1"># store target gene list in a dictionary</span>
    <span class="n">TF_to_TG_dictionary</span><span class="p">[</span><span class="n">TF</span><span class="p">]</span> <span class="o">=</span> <span class="n">TG_list</span>

<span class="c1"># We have to make a dictionary, in which a Key is Target gene and value is TF.</span>
<span class="c1"># We inverse the dictionary above using a utility function in celloracle.</span>
<span class="n">TG_to_TF_dictionary</span> <span class="o">=</span> <span class="n">co</span><span class="o">.</span><span class="n">utility</span><span class="o">.</span><span class="n">inverse_dictionary</span><span class="p">(</span><span class="n">TF_to_TG_dictionary</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "6cdd0fba94fe41caae9aace9a609f5f9", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
</div>
<div class="section" id="2.3.2.-Add-TF-informatio-dictionary-into-the-oracle-object">
<h6>2.3.2. Add TF informatio dictionary into the oracle object<a class="headerlink" href="#2.3.2.-Add-TF-informatio-dictionary-into-the-oracle-object" title="Permalink to this headline">¶</a></h6>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Add TF information</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">addTFinfo_dictionary</span><span class="p">(</span><span class="n">TG_to_TF_dictionary</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="3.-Knn-imputation">
<h4>3. Knn imputation<a class="headerlink" href="#3.-Knn-imputation" title="Permalink to this headline">¶</a></h4>
<p>Celloracle uses almost the same strategy as velocyto for visualization of cell transition. This process requires knn imputation.</p>
<p>For the Knn imputation, we need PCA and PC selection first.</p>
<div class="section" id="3.1.-PCA">
<h5>3.1. PCA<a class="headerlink" href="#3.1.-PCA" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># perform PCA</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">perform_PCA</span><span class="p">()</span>

<span class="c1"># select important PCs</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">)[:</span><span class="mi">100</span><span class="p">])</span>
<span class="n">n_comps</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">))</span><span class="o">&gt;</span><span class="mf">0.002</span><span class="p">))[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">n_comps</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">n_comps</span><span class="p">)</span>
<span class="n">n_comps</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">n_comps</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
45
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_27_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_27_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_27_1.png" style="width: 380px; height: 278px;" /></a>
</div>
</div>
</div>
<div class="section" id="3.2.-Knn-imputation">
<h5>3.2. Knn imputation<a class="headerlink" href="#3.2.-Knn-imputation" title="Permalink to this headline">¶</a></h5>
<p>Estimate number of nearest neighbor for k-nn imputation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">n_cell</span> <span class="o">=</span> <span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;cell number is :</span><span class="si">{n_cell}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell number is :2671
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">k</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.025</span><span class="o">*</span><span class="n">n_cell</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Auto-selected k is :</span><span class="si">{k}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Auto-selected k is :66
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">oracle</span><span class="o">.</span><span class="n">knn_imputation</span><span class="p">(</span><span class="n">n_pca_dims</span><span class="o">=</span><span class="n">n_comps</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">balanced</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">b_sight</span><span class="o">=</span><span class="n">k</span><span class="o">*</span><span class="mi">8</span><span class="p">,</span>
                      <span class="n">b_maxl</span><span class="o">=</span><span class="n">k</span><span class="o">*</span><span class="mi">4</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="4.-Save-and-Load.">
<h4>4. Save and Load.<a class="headerlink" href="#4.-Save-and-Load." title="Permalink to this headline">¶</a></h4>
<p>Celloracle has some custom-class: Links, Oracle, TFinfo. You can save such an object using “to_hdf5”.</p>
<p>Pleasae use “load_hdf5” function to load the file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># save oracle object.</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="s2">&quot;Paul_15_data.celloracle.oracle&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># load file.</span>
<span class="c1"># oracle = co.load_hdf5(&quot;Paul_15_data.celloracle.oracle&quot;)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="4.-GRN-calculation">
<h4>4. GRN calculation<a class="headerlink" href="#4.-GRN-calculation" title="Permalink to this headline">¶</a></h4>
<p>Next step is constructing a cluster-specific GRN for all clusters.</p>
<p>You can calculate GRNs with “get_links” function, and the function returns GRNs as a Links object. Links object stores inferred GRNs and its metadata. You can do network analysis with Links object.</p>
<p>In the example below, we construct GRNs based on “louvain_annot” clustering unit.</p>
<p>GRN can be calculated at any arbitrary unit as long as the clustering information is stored in anndata.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check data</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;louvain_annot&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_37_0.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_37_0.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_37_0.png" style="width: 516px; height: 288px;" /></a>
</div>
</div>
<div class="section" id="4.1.-Get-GRNs">
<h5>4.1. Get GRNs<a class="headerlink" href="#4.1.-Get-GRNs" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># calculate GRN for each population in &quot;louvain_annot&quot; clustering unit.</span>
<span class="c1"># This step may take long time.</span>
<span class="n">links</span> <span class="o">=</span> <span class="n">oracle</span><span class="o">.</span><span class="n">get_links</span><span class="p">(</span><span class="n">cluster_name_for_GRN_unit</span><span class="o">=</span><span class="s2">&quot;louvain_annot&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                         <span class="n">verbose_level</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">test_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "362e1273b3b84cf79d50613bbf467d60", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
inferring GRN for Ery_0...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d0ece9e3ab72419f83481d238ca2947b", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_1...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "41484a6db6fd4bc49d17a27fd0b7697e", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_2...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "7afb8e0c7cf049d2b19ad654229cf4a8", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_3...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "5e85f43705754c8eba9941a34d2d02ab", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_4...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "c8053b749beb468da7e86ab193a57e8d", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_5...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "47360980e849429a8c25039ac7d9c209", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_6...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8a7b7018fd3c4a61aa4b78231bc6e665", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_7...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "fc9bd6862a0b4f7a9e8ff085590cfccd", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_8...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "9dac5d8dd3ef40f2ab82622c359a3b68", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Ery_9...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "4334aea238b04a0db3cfa55b55e09433", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for GMP_0...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "1de9fb3dae394349b614156b2766b86b", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for GMP_1...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "c26593bca9714d6e9743f3903d837768", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for GMPl_0...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "f45bda7e9c3349a2a0cedc661c416731", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Gran_0...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "adbdfde5e6a34407899ec129e356be5a", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Gran_1...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "a2aa7f5e52734c2784b62927c337986d", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Gran_2...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "5742657e64824d89ba89ded01ea620ca", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for MEP_0...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "5e8c644639694cb6a96889ae26310e20", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Mk_0...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "236d2ac06c90488db844a0cf0ba53b4d", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Mo_0...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "5fa9229a59b74b708b490f9186e5cec1", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

inferring GRN for Mo_1...
method: bagging_ridge
alpha: 10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "9f85c4a46d5c408a8bc7b17cd534ec75", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>


</pre></div></div>
</div>
</div>
<div class="section" id="4.2.-(Optional)-Export-GRNs">
<h5>4.2. (Optional) Export GRNs<a class="headerlink" href="#4.2.-(Optional)-Export-GRNs" title="Permalink to this headline">¶</a></h5>
<p>Although celloracle has many functions for network analysis, you can analyze GRNs by yourself. The raw GRN data is stored in the attribute of “links_dict”.</p>
<p>For example, you can get GRNs for “Ery_0” cluster with the following commands.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">links</span><span class="o">.</span><span class="n">links_dict</span><span class="p">[</span><span class="s2">&quot;Ery_0&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>target</th>
      <th>coef_mean</th>
      <th>coef_abs</th>
      <th>p</th>
      <th>-logp</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>Id2</td>
      <td>0610007L01Rik</td>
      <td>0.000552</td>
      <td>0.000552</td>
      <td>5.724488e-01</td>
      <td>0.242263</td>
    </tr>
    <tr>
      <td>1</td>
      <td>Klf2</td>
      <td>0610007L01Rik</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>NaN</td>
      <td>-0.000000</td>
    </tr>
    <tr>
      <td>2</td>
      <td>Stat5a</td>
      <td>0610007L01Rik</td>
      <td>-0.005179</td>
      <td>0.005179</td>
      <td>5.294497e-04</td>
      <td>3.276175</td>
    </tr>
    <tr>
      <td>3</td>
      <td>Elf1</td>
      <td>0610007L01Rik</td>
      <td>0.002107</td>
      <td>0.002107</td>
      <td>1.498884e-01</td>
      <td>0.824232</td>
    </tr>
    <tr>
      <td>4</td>
      <td>Gata1</td>
      <td>0610007L01Rik</td>
      <td>-0.000700</td>
      <td>0.000700</td>
      <td>5.731063e-01</td>
      <td>0.241765</td>
    </tr>
    <tr>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <td>74460</td>
      <td>Cxxc1</td>
      <td>Zyx</td>
      <td>-0.004999</td>
      <td>0.004999</td>
      <td>1.578852e-02</td>
      <td>1.801659</td>
    </tr>
    <tr>
      <td>74461</td>
      <td>Mef2c</td>
      <td>Zyx</td>
      <td>0.017708</td>
      <td>0.017708</td>
      <td>3.011616e-07</td>
      <td>6.521200</td>
    </tr>
    <tr>
      <td>74462</td>
      <td>Nfe2</td>
      <td>Zyx</td>
      <td>0.034433</td>
      <td>0.034433</td>
      <td>2.548244e-12</td>
      <td>11.593759</td>
    </tr>
    <tr>
      <td>74463</td>
      <td>Nr3c1</td>
      <td>Zyx</td>
      <td>-0.022663</td>
      <td>0.022663</td>
      <td>2.265408e-08</td>
      <td>7.644854</td>
    </tr>
    <tr>
      <td>74464</td>
      <td>Ets1</td>
      <td>Zyx</td>
      <td>0.012826</td>
      <td>0.012826</td>
      <td>4.813285e-09</td>
      <td>8.317558</td>
    </tr>
  </tbody>
</table>
<p>74465 rows × 6 columns</p>
</div></div>
</div>
<p>You can export the file as follows.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># set cluster name</span>
<span class="n">cluster</span> <span class="o">=</span> <span class="s2">&quot;Ery_0&quot;</span>

<span class="c1"># save as csv</span>
<span class="n">links</span><span class="o">.</span><span class="n">links_dict</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;raw_GRN_for_</span><span class="si">{cluster}</span><span class="s2">.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="4.3.-(Optional)-Change-order">
<h5>4.3. (Optional) Change order<a class="headerlink" href="#4.3.-(Optional)-Change-order" title="Permalink to this headline">¶</a></h5>
<p>Links object have a color information in a attribute, “palette”. This information is used for the visualization</p>
<p>The sample will be visualized in the order. Here we change the order.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Show the contents of pallete</span>
<span class="n">links</span><span class="o">.</span><span class="n">palette</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>palette</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Ery_0</td>
      <td>#7D87B9</td>
    </tr>
    <tr>
      <td>Ery_1</td>
      <td>#BEC1D4</td>
    </tr>
    <tr>
      <td>Ery_2</td>
      <td>#D6BCC0</td>
    </tr>
    <tr>
      <td>Ery_3</td>
      <td>#BB7784</td>
    </tr>
    <tr>
      <td>Ery_4</td>
      <td>#8E063B</td>
    </tr>
    <tr>
      <td>Ery_5</td>
      <td>#4A6FE3</td>
    </tr>
    <tr>
      <td>Ery_6</td>
      <td>#8595E1</td>
    </tr>
    <tr>
      <td>Ery_7</td>
      <td>#B5BBE3</td>
    </tr>
    <tr>
      <td>Ery_8</td>
      <td>#E6AFB9</td>
    </tr>
    <tr>
      <td>Ery_9</td>
      <td>#E07B91</td>
    </tr>
    <tr>
      <td>GMP_0</td>
      <td>#D33F6A</td>
    </tr>
    <tr>
      <td>GMP_1</td>
      <td>#11C638</td>
    </tr>
    <tr>
      <td>GMPl_0</td>
      <td>#8DD593</td>
    </tr>
    <tr>
      <td>Gran_0</td>
      <td>#C6DEC7</td>
    </tr>
    <tr>
      <td>Gran_1</td>
      <td>#EAD3C6</td>
    </tr>
    <tr>
      <td>Gran_2</td>
      <td>#F0B98D</td>
    </tr>
    <tr>
      <td>MEP_0</td>
      <td>#0FCFC0</td>
    </tr>
    <tr>
      <td>Mk_0</td>
      <td>#9CDED6</td>
    </tr>
    <tr>
      <td>Mo_0</td>
      <td>#D5EAE7</td>
    </tr>
    <tr>
      <td>Mo_1</td>
      <td>#F3E1EB</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># change the order of pallete</span>
<span class="n">order</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;MEP_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Mk_0&#39;</span><span class="p">,</span><span class="s1">&#39;Ery_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_1&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_2&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_3&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_4&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_5&#39;</span><span class="p">,</span>
         <span class="s1">&#39;Ery_6&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_7&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_8&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_9&#39;</span><span class="p">,</span><span class="s1">&#39;GMP_0&#39;</span><span class="p">,</span> <span class="s1">&#39;GMP_1&#39;</span><span class="p">,</span>
         <span class="s1">&#39;GMPl_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Mo_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Mo_1&#39;</span><span class="p">,</span> <span class="s1">&#39;Gran_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Gran_1&#39;</span><span class="p">,</span> <span class="s1">&#39;Gran_2&#39;</span><span class="p">]</span>
<span class="n">links</span><span class="o">.</span><span class="n">palette</span> <span class="o">=</span> <span class="n">links</span><span class="o">.</span><span class="n">palette</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">order</span><span class="p">]</span>
<span class="n">links</span><span class="o">.</span><span class="n">palette</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>palette</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>MEP_0</td>
      <td>#0FCFC0</td>
    </tr>
    <tr>
      <td>Mk_0</td>
      <td>#9CDED6</td>
    </tr>
    <tr>
      <td>Ery_0</td>
      <td>#7D87B9</td>
    </tr>
    <tr>
      <td>Ery_1</td>
      <td>#BEC1D4</td>
    </tr>
    <tr>
      <td>Ery_2</td>
      <td>#D6BCC0</td>
    </tr>
    <tr>
      <td>Ery_3</td>
      <td>#BB7784</td>
    </tr>
    <tr>
      <td>Ery_4</td>
      <td>#8E063B</td>
    </tr>
    <tr>
      <td>Ery_5</td>
      <td>#4A6FE3</td>
    </tr>
    <tr>
      <td>Ery_6</td>
      <td>#8595E1</td>
    </tr>
    <tr>
      <td>Ery_7</td>
      <td>#B5BBE3</td>
    </tr>
    <tr>
      <td>Ery_8</td>
      <td>#E6AFB9</td>
    </tr>
    <tr>
      <td>Ery_9</td>
      <td>#E07B91</td>
    </tr>
    <tr>
      <td>GMP_0</td>
      <td>#D33F6A</td>
    </tr>
    <tr>
      <td>GMP_1</td>
      <td>#11C638</td>
    </tr>
    <tr>
      <td>GMPl_0</td>
      <td>#8DD593</td>
    </tr>
    <tr>
      <td>Mo_0</td>
      <td>#D5EAE7</td>
    </tr>
    <tr>
      <td>Mo_1</td>
      <td>#F3E1EB</td>
    </tr>
    <tr>
      <td>Gran_0</td>
      <td>#C6DEC7</td>
    </tr>
    <tr>
      <td>Gran_1</td>
      <td>#EAD3C6</td>
    </tr>
    <tr>
      <td>Gran_2</td>
      <td>#F0B98D</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
</div>
<div class="section" id="5.-Network-preprocessing">
<h4>5. Network preprocessing<a class="headerlink" href="#5.-Network-preprocessing" title="Permalink to this headline">¶</a></h4>
<div class="section" id="5.1.-Filter-network-edges">
<h5>5.1. Filter network edges<a class="headerlink" href="#5.1.-Filter-network-edges" title="Permalink to this headline">¶</a></h5>
<p>Celloracle leverages bagging ridge or Bayesian ridge regression for network inference. These methods provide a network edge strength as a distribution rather than point value. We can use the distribution to know the certainness of the connection.</p>
<p>We filter the network edges as follows.</p>
<ol class="arabic simple">
<li>Remove uncertain network edge based on the p-value.</li>
<li>Remove weak network edge. In this tutorial, we pick up top 2000 edges in terms of network strength.</li>
</ol>
<p>The raw network data is stored as an attribute, “links_dict,” while filtered network data is stored in “filtered_links.” So the filtering function keeps raw network information rather than overwriting data. You can come back to the filtering process to filter the data with different parameters if you want.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">links</span><span class="o">.</span><span class="n">filter_links</span><span class="p">(</span><span class="n">p</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;coef_abs&quot;</span><span class="p">,</span> <span class="n">thread_number</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="5.2.-Degree-distribution">
<h5>5.2. Degree distribution<a class="headerlink" href="#5.2.-Degree-distribution" title="Permalink to this headline">¶</a></h5>
<p>For the first step, we examine network degree distribution. Network degree, which is the number of edges for each node, is one of the important metrics to investigate the network structure (<a class="reference external" href="https://en.wikipedia.org/wiki/Degree_distribution">https://en.wikipedia.org/wiki/Degree_distribution</a>).</p>
<p>Please keep in mind that the degree distribution may change depending on the filtering threshold.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">links</span><span class="o">.</span><span class="n">plot_degree_distributions</span><span class="p">(</span><span class="n">plot_model</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">/degree_distribution/&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_1.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_3.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_3.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_3.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_5.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_5.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_5.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_7.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_7.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_7.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_9.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_9.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_9.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_5
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_11.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_11.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_11.png" style="width: 562px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_6
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_13.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_13.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_13.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_7
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_15.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_15.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_15.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_8
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_17.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_17.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_17.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_9
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_19.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_19.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_19.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GMP_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_21.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_21.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_21.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GMP_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_23.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_23.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_23.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GMPl_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_25.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_25.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_25.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Gran_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_27.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_27.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_27.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Gran_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_29.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_29.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_29.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Gran_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_31.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_31.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_31.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
MEP_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_33.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_33.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_33.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mk_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_35.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_35.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_35.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mo_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_37.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_37.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_37.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mo_1
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_39.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_39.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_52_39.png" style="width: 561px; height: 318px;" /></a>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="5.3.-Calculate-netowrk-score">
<h5>5.3. Calculate netowrk score<a class="headerlink" href="#5.3.-Calculate-netowrk-score" title="Permalink to this headline">¶</a></h5>
<p>Next, we calculate several network score using some R libraries. Please make sure that R libraries are installed in your PC before running the command below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># calculate network scores. It takes several minuts.</span>
<span class="n">links</span><span class="o">.</span><span class="n">get_score</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
processing... batch 1/5
Ery_0: finished.
Ery_1: finished.
Ery_2: finished.
Ery_3: finished.
processing... batch 2/5
Ery_4: finished.
Ery_5: finished.
Ery_6: finished.
Ery_7: finished.
processing... batch 3/5
Ery_8: finished.
Ery_9: finished.
GMP_0: finished.
GMP_1: finished.
processing... batch 4/5
GMPl_0: finished.
Gran_0: finished.
Gran_1: finished.
Gran_2: finished.
processing... batch 5/5
MEP_0: finished.
Mk_0: finished.
Mo_0: finished.
Mo_1: finished.
the scores are saved in ./louvain_annot/
</pre></div></div>
</div>
<p>The score is stored as a attribute “merged_score”, and score will also be saved in a folder in your computer.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">links</span><span class="o">.</span><span class="n">merged_score</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>degree_all</th>
      <th>degree_in</th>
      <th>degree_out</th>
      <th>clustering_coefficient</th>
      <th>clustering_coefficient_weighted</th>
      <th>degree_centrality_all</th>
      <th>degree_centrality_in</th>
      <th>degree_centrality_out</th>
      <th>betweenness_centrality</th>
      <th>closeness_centrality</th>
      <th>...</th>
      <th>assortative_coefficient</th>
      <th>average_path_length</th>
      <th>community_edge_betweenness</th>
      <th>community_random_walk</th>
      <th>community_eigenvector</th>
      <th>module</th>
      <th>connectivity</th>
      <th>participation</th>
      <th>role</th>
      <th>cluster</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Stat3</td>
      <td>91</td>
      <td>0</td>
      <td>91</td>
      <td>0.019780</td>
      <td>0.020122</td>
      <td>0.167279</td>
      <td>0.000000</td>
      <td>0.167279</td>
      <td>0</td>
      <td>0.000013</td>
      <td>...</td>
      <td>-0.172207</td>
      <td>2.475244</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>4.052023</td>
      <td>0.632291</td>
      <td>Connector Hub</td>
      <td>Ery_0</td>
    </tr>
    <tr>
      <td>Mycn</td>
      <td>28</td>
      <td>0</td>
      <td>28</td>
      <td>0.002646</td>
      <td>0.001828</td>
      <td>0.051471</td>
      <td>0.000000</td>
      <td>0.051471</td>
      <td>0</td>
      <td>0.000004</td>
      <td>...</td>
      <td>-0.172207</td>
      <td>2.475244</td>
      <td>2</td>
      <td>1</td>
      <td>5</td>
      <td>2</td>
      <td>2.347146</td>
      <td>0.625000</td>
      <td>Connector</td>
      <td>Ery_0</td>
    </tr>
    <tr>
      <td>Zbtb1</td>
      <td>27</td>
      <td>0</td>
      <td>27</td>
      <td>0.005698</td>
      <td>0.008576</td>
      <td>0.049632</td>
      <td>0.000000</td>
      <td>0.049632</td>
      <td>0</td>
      <td>0.000011</td>
      <td>...</td>
      <td>-0.172207</td>
      <td>2.475244</td>
      <td>3</td>
      <td>1</td>
      <td>5</td>
      <td>1</td>
      <td>2.275431</td>
      <td>0.666667</td>
      <td>Connector</td>
      <td>Ery_0</td>
    </tr>
    <tr>
      <td>E2f4</td>
      <td>186</td>
      <td>3</td>
      <td>183</td>
      <td>0.009474</td>
      <td>0.011623</td>
      <td>0.341912</td>
      <td>0.005515</td>
      <td>0.336397</td>
      <td>2788</td>
      <td>0.000010</td>
      <td>...</td>
      <td>-0.172207</td>
      <td>2.475244</td>
      <td>4</td>
      <td>1</td>
      <td>3</td>
      <td>0</td>
      <td>8.496354</td>
      <td>0.637760</td>
      <td>Connector Hub</td>
      <td>Ery_0</td>
    </tr>
    <tr>
      <td>Ybx1</td>
      <td>71</td>
      <td>9</td>
      <td>62</td>
      <td>0.027364</td>
      <td>0.027485</td>
      <td>0.130515</td>
      <td>0.016544</td>
      <td>0.113971</td>
      <td>1153</td>
      <td>0.000004</td>
      <td>...</td>
      <td>-0.172207</td>
      <td>2.475244</td>
      <td>5</td>
      <td>11</td>
      <td>4</td>
      <td>3</td>
      <td>5.433857</td>
      <td>0.564967</td>
      <td>Connector Hub</td>
      <td>Ery_0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 22 columns</p>
</div></div>
</div>
</div>
<div class="section" id="5.4.-Save">
<h5>5.4. Save<a class="headerlink" href="#5.4.-Save" title="Permalink to this headline">¶</a></h5>
<p>Save processed GRN. We use this file in the next notebook; “in silico simulation with GRNs”.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Save Links object.</span>
<span class="n">links</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="s2">&quot;links.celloracle.links&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># You can load files with the following command.</span>
<span class="n">links</span> <span class="o">=</span> <span class="n">co</span><span class="o">.</span><span class="n">load_hdf5</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="s2">&quot;links.celloracle.links&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<p>If you are not interested in the network score analysis below and just want to do simulation, you can skip them and go to the next notebook.</p>
</div>
</div>
<div class="section" id="6.-Network-analysis;-Network-score-for-each-gene">
<h4>6. Network analysis; Network score for each gene<a class="headerlink" href="#6.-Network-analysis;-Network-score-for-each-gene" title="Permalink to this headline">¶</a></h4>
<p>Links class has many functions to visualize network score. See the documentation for the details of the functions.</p>
<div class="section" id="6.1.-Network-score-in-each-cluster">
<h5>6.1. Network score in each cluster<a class="headerlink" href="#6.1.-Network-score-in-each-cluster" title="Permalink to this headline">¶</a></h5>
<p>We have calculated several network scores for centrality. We can use the centrality score to identify key regulatory genes because centrality is one of the important indicators of network structure (<a class="reference external" href="https://en.wikipedia.org/wiki/Centrality">https://en.wikipedia.org/wiki/Centrality</a>).</p>
<p>Let’s visualize genes with high network centrality.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check cluster name</span>
<span class="n">links</span><span class="o">.</span><span class="n">cluster</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>[&#39;Ery_0&#39;,
 &#39;Ery_1&#39;,
 &#39;Ery_2&#39;,
 &#39;Ery_3&#39;,
 &#39;Ery_4&#39;,
 &#39;Ery_5&#39;,
 &#39;Ery_6&#39;,
 &#39;Ery_7&#39;,
 &#39;Ery_8&#39;,
 &#39;Ery_9&#39;,
 &#39;GMP_0&#39;,
 &#39;GMP_1&#39;,
 &#39;GMPl_0&#39;,
 &#39;Gran_0&#39;,
 &#39;Gran_1&#39;,
 &#39;Gran_2&#39;,
 &#39;MEP_0&#39;,
 &#39;Mk_0&#39;,
 &#39;Mo_0&#39;,
 &#39;Mo_1&#39;]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Visualize top n-th genes that have high scores.</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_scores_as_rank</span><span class="p">(</span><span class="n">cluster</span><span class="o">=</span><span class="s2">&quot;MEP_0&quot;</span><span class="p">,</span> <span class="n">n_gene</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">/ranked_score&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_0.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_0.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_0.png" style="width: 304px; height: 319px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_1.png" style="width: 341px; height: 319px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_2.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_2.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_2.png" style="width: 304px; height: 319px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_3.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_3.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_3.png" style="width: 305px; height: 319px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_4.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_4.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_4.png" style="width: 296px; height: 319px;" /></a>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_5.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_5.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_65_5.png" style="width: 342px; height: 318px;" /></a>
</div>
</div>
</div>
<div class="section" id="6.2.-Network-score-comparison-between-two-clusters">
<h5>6.2. Network score comparison between two clusters<a class="headerlink" href="#6.2.-Network-score-comparison-between-two-clusters" title="Permalink to this headline">¶</a></h5>
<p>By comparing network scores between two clusters, we can analyze the difference of GRN structure.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">ticklabel_format</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s1">&#39;sci&#39;</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span><span class="n">scilimits</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_score_comparison_2D</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;eigenvector_centrality&quot;</span><span class="p">,</span>
                               <span class="n">cluster1</span><span class="o">=</span><span class="s2">&quot;MEP_0&quot;</span><span class="p">,</span> <span class="n">cluster2</span><span class="o">=</span><span class="s2">&quot;GMPl_0&quot;</span><span class="p">,</span>
                               <span class="n">percentile</span><span class="o">=</span><span class="mi">98</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">/score_comparison&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_68_0.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_68_0.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_68_0.png" style="width: 407px; height: 304px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>

<span></span><span class="n">plt</span><span class="o">.</span><span class="n">ticklabel_format</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s1">&#39;sci&#39;</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span><span class="n">scilimits</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_score_comparison_2D</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;betweenness_centrality&quot;</span><span class="p">,</span>
                               <span class="n">cluster1</span><span class="o">=</span><span class="s2">&quot;MEP_0&quot;</span><span class="p">,</span> <span class="n">cluster2</span><span class="o">=</span><span class="s2">&quot;GMPl_0&quot;</span><span class="p">,</span>
                               <span class="n">percentile</span><span class="o">=</span><span class="mi">98</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">/score_comparison&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_69_0.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_69_0.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_69_0.png" style="width: 416px; height: 304px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">ticklabel_format</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s1">&#39;sci&#39;</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span><span class="n">scilimits</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_score_comparison_2D</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;degree_centrality_all&quot;</span><span class="p">,</span>
                               <span class="n">cluster1</span><span class="o">=</span><span class="s2">&quot;MEP_0&quot;</span><span class="p">,</span> <span class="n">cluster2</span><span class="o">=</span><span class="s2">&quot;GMPl_0&quot;</span><span class="p">,</span>
                               <span class="n">percentile</span><span class="o">=</span><span class="mi">98</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">/score_comparison&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_70_0.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_70_0.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_70_0.png" style="width: 392px; height: 304px;" /></a>
</div>
</div>
</div>
<div class="section" id="6.3.-Network-score-dynamics">
<h5>6.3. Network score dynamics<a class="headerlink" href="#6.3.-Network-score-dynamics" title="Permalink to this headline">¶</a></h5>
<p>In the following session, we focus on how network score for gene changes during the differentiation.</p>
<p>We make some graph to show an example of how to use functions using Gata2 gene.</p>
<p>Gata2 is known to play an essential role in the early progenitor population; MEP, GMP. The following results can recapitulate the Gata2 feature.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Visualize Gata2 network score dynamics</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_score_per_cluster</span><span class="p">(</span><span class="n">goi</span><span class="o">=</span><span class="s2">&quot;Gata2&quot;</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">/network_score_per_gene/&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Gata2
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_72_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_72_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_72_1.png" style="width: 408px; height: 301px;" /></a>
</div>
</div>
</div>
<div class="section" id="6.4.-Gene-cartography-analysis">
<h5>6.4. Gene cartography analysis<a class="headerlink" href="#6.4.-Gene-cartography-analysis" title="Permalink to this headline">¶</a></h5>
<p>Gene cartography is a method for gene network analysis. The method classifies gene into several groups using the network module structure and connections. It provides us an insight about the role and regulatory mechanism per each gene. Please read the paper for the details of gene cartography (<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a>)</p>
<p>The gene cartography will be calculated for the GRN in each cluster. Thus we can know how it change by comparing the the score between clusters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot cartography as a scatter plot</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_cartography_scatter_per_cluster</span><span class="p">(</span><span class="n">scatter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                           <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                           <span class="n">gois</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Gata1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gata2&quot;</span><span class="p">,</span> <span class="s2">&quot;Sfpi1&quot;</span><span class="p">],</span>
                                           <span class="n">auto_gene_annot</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                           <span class="n">args_dot</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;n_levels&quot;</span><span class="p">:</span> <span class="mi">105</span><span class="p">},</span>
                                           <span class="n">args_line</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;c&quot;</span><span class="p">:</span><span class="s2">&quot;gray&quot;</span><span class="p">},</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">/cartography&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_1.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_3.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_3.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_3.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_5.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_5.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_5.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_7.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_7.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_7.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_9.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_9.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_9.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_5
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_11.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_11.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_11.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_6
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_13.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_13.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_13.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_7
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_15.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_15.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_15.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_8
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_17.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_17.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_17.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Ery_9
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_19.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_19.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_19.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GMP_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_21.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_21.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_21.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GMP_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_23.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_23.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_23.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GMPl_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_25.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_25.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_25.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Gran_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_27.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_27.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_27.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Gran_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_29.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_29.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_29.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Gran_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_31.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_31.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_31.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
MEP_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_33.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_33.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_33.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mk_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_35.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_35.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_35.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mo_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_37.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_37.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_37.png" style="width: 356px; height: 264px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mo_1
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_39.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_39.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_74_39.png" style="width: 362px; height: 264px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot the summary of cartography analysis</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_cartography_term</span><span class="p">(</span><span class="n">goi</span><span class="o">=</span><span class="s2">&quot;Gata2&quot;</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">/cartography&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Gata2
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_75_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_75_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_75_1.png" style="width: 408px; height: 346px;" /></a>
</div>
</div>
</div>
</div>
<div class="section" id="7.-Network-analysis;-Network-score-distribution">
<h4>7. Network analysis; Network score distribution<a class="headerlink" href="#7.-Network-analysis;-Network-score-distribution" title="Permalink to this headline">¶</a></h4>
<p>Next, we visualize the distribution of network score to get insight into the global trend of GRNs.</p>
<div class="section" id="7.1.-Distribution-of-network-degree">
<h5>7.1. Distribution of network degree<a class="headerlink" href="#7.1.-Distribution-of-network-degree" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.040</span><span class="p">])</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_score_discributions</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;degree_centrality_all&quot;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;boxplot&quot;</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">&quot;</span><span class="p">)</span>


</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
degree_centrality_all
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_78_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_78_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_78_1.png" style="width: 390px; height: 263px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.40</span><span class="p">])</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_score_discributions</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;eigenvector_centrality&quot;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;boxplot&quot;</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">&quot;</span><span class="p">)</span>



</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
eigenvector_centrality
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_79_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_79_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_79_1.png" style="width: 383px; height: 263px;" /></a>
</div>
</div>
</div>
<div class="section" id="7.2.-Distribution-of-netowrk-entolopy">
<h5>7.2. Distribution of netowrk entolopy<a class="headerlink" href="#7.2.-Distribution-of-netowrk-entolopy" title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">links</span><span class="o">.</span><span class="n">plot_network_entropy_distributions</span><span class="p">(</span><span class="n">save</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{save_folder}</span><span class="s2">&quot;</span><span class="p">)</span>


</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:2614: RuntimeWarning: invalid value encountered in true_divide
  pk = 1.0*pk / np.sum(pk, axis=0)
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/celloracle/network_analysis/links_object.py:345: RuntimeWarning: divide by zero encountered in log
  ent_norm.append(en/np.log(k[i]))
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/celloracle/network_analysis/links_object.py:345: RuntimeWarning: invalid value encountered in double_scalars
  ent_norm.append(en/np.log(k[i]))
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_81_1.png"><img alt="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_81_1.png" src="_images/notebooks_04_Network_analysis_Network_analysis_with_with_Paul_etal_2015_data_81_1.png" style="width: 393px; height: 263px;" /></a>
</div>
</div>
<p>Using the network score, we could pick up cluster-specific key TFs. “Gata2”, “Gata1”, “Klf1”, “E2f1”, for example, are known to play an essential role in MEP, and these TFs showed high network score in our GRN.</p>
<p>In this step, however, we cannot know the specific function or relationship between cell fate.</p>
<p>In the next analysis, we investigate their function and relationship between cell fate by the simulation with GRNs.</p>
</div>
</div>
</div>
</div>
<span id="document-tutorials/simulation"></span><div class="section" id="simulation-with-grns">
<span id="simulation"></span><h3>5. Simulation with GRNs<a class="headerlink" href="#simulation-with-grns" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">celloracle</span></code> leverage GRNs to simulate signal propagation inside a cell.
We can estimate the effect of gene perturbation by the simulation with GRNs.</p>
<p><code class="docutils literal notranslate"><span class="pre">celloracle</span></code> leverage GRNs to simulate signal propagation inside a cell. We can estimate the effect of gene perturbation by the simulation with GRNs.</p>
<p>Besides, we will combine the simulation for the signal propagation and the simulation of cell state transition, which is performed by a python library for RNA-velocity analysis, <code class="docutils literal notranslate"><span class="pre">velocyto</span></code> .
A series of simulations visualizes a complex system in which TF controls enormous target genes to determines cell fate.</p>
<p>In short, celloracle provides insight into the regulatory mechanism of cell state from the viewpoint of TF and GRN.</p>
<p>Python notebook</p>
<div class="toctree-wrapper compound">
<span id="document-notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data"></span><div class="section" id="0.-Import-libraries">
<h4>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">¶</a></h4>
<div class="section" id="0.1.-Import-public-libraries">
<h5>0.1. Import public libraries<a class="headerlink" href="#0.1.-Import-public-libraries" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">matplotlib.colors</span> <span class="k">as</span> <span class="nn">colors</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">celloracle</span> <span class="k">as</span> <span class="nn">co</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#plt.rcParams[&quot;font.family&quot;] = &quot;arial&quot;</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span><span class="mi">6</span><span class="p">]</span>
<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;savefig.dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">600</span>

<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
</div>
<div class="section" id="0.1.-Make-a-folder-to-save-graph">
<h5>0.1. Make a folder to save graph<a class="headerlink" href="#0.1.-Make-a-folder-to-save-graph" title="Permalink to this headline">¶</a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># make folder to save plots</span>
<span class="n">save_folder</span> <span class="o">=</span> <span class="s2">&quot;figures&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">save_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="1.-Load-data">
<h4>1. Load data<a class="headerlink" href="#1.-Load-data" title="Permalink to this headline">¶</a></h4>
<div class="section" id="1.1.-Load-processed-oracle-object">
<h5>1.1. Load processed oracle object<a class="headerlink" href="#1.1.-Load-processed-oracle-object" title="Permalink to this headline">¶</a></h5>
<p>Load the oracle object. See the previous notebook for the detil of how to prepare oracle object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">oracle</span> <span class="o">=</span> <span class="n">co</span><span class="o">.</span><span class="n">load_hdf5</span><span class="p">(</span><span class="s2">&quot;../04_Network_analysis/Paul_15_data.celloracle.oracle&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.2.-Load-inferred-GRNs">
<h5>1.2. Load inferred GRNs<a class="headerlink" href="#1.2.-Load-inferred-GRNs" title="Permalink to this headline">¶</a></h5>
<p>In the previous notebook, we already calculated GRNs. We use this GRNs for simulation.</p>
<p>We import GRNs which was saved in Links object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">links</span> <span class="o">=</span> <span class="n">co</span><span class="o">.</span><span class="n">load_hdf5</span><span class="p">(</span><span class="s2">&quot;../04_Network_analysis/links.celloracle.links&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="3.-Make-predictive-models-for-simulation">
<h4>3. Make predictive models for simulation<a class="headerlink" href="#3.-Make-predictive-models-for-simulation" title="Permalink to this headline">¶</a></h4>
<p>We construct ridge regression models again based on the filtered GRN structure. This process takes less time than the GRN inference in the previous notebook because we use only significant TFs to predict target gene instead of all regulatory candidate TFs.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">links</span><span class="o">.</span><span class="n">filter_links</span><span class="p">()</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">get_cluster_specific_TFdict_from_Links</span><span class="p">(</span><span class="n">links_object</span><span class="o">=</span><span class="n">links</span><span class="p">)</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">fit_GRN_for_simulation</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">use_cluster_specific_TFdict</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
calculating GRN using cluster specicif TF dict...
calculating GRN in Ery_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "14dfcd7cc22b4b53b240773f26d8a2be", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1074 genes
calculating GRN in Ery_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "e6146b0a06da47ccabd6eb5e209161f9", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1092 genes
calculating GRN in Ery_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "84bdf6ae2f7545c9917c255b030c404c", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1064 genes
calculating GRN in Ery_3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "5eeabe9cf08945278865668694541e49", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1105 genes
calculating GRN in Ery_4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8040356d76564851afaf6946a1878525", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1102 genes
calculating GRN in Ery_5
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "a4f9f8d4998f4b08856601f61e3298aa", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1116 genes
calculating GRN in Ery_6
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "005d0f6c78a6440ab5a7f74eb74039b1", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1097 genes
calculating GRN in Ery_7
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "39009d989a58456c8d3152f1ea07befd", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1062 genes
calculating GRN in Ery_8
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8be0075f3cfe4b93883b1f805714a316", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1117 genes
calculating GRN in Ery_9
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "2cd200598fe741d8881c94919b698998", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1121 genes
calculating GRN in GMP_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "176a12ac5bca4924a5608fe4c3aa4d21", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1107 genes
calculating GRN in GMP_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "edcd27b56c4b486c92bb5ec3dc450b64", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1104 genes
calculating GRN in GMPl_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "1b1b948f1f5747aba3930aea2be7bb2e", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1089 genes
calculating GRN in Gran_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "e0b30edace2045528bd368893d1a7e29", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1067 genes
calculating GRN in Gran_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "e8e7ec9e98d84a1da7cfeffaccc789ff", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1076 genes
calculating GRN in Gran_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "65eca3323d144d4e85bfcf5be0367a44", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1105 genes
calculating GRN in MEP_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "986ddcc973a5410ea0edff528194b929", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1152 genes
calculating GRN in Mk_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "183f54ddaaa043179b7042dd19a7f955", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1114 genes
calculating GRN in Mo_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "30b1881884604ad697f3b13d5e98d0bd", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1085 genes
calculating GRN in Mo_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "06bad2e6d2c6432b8ce366c3eb9fb0cd", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1074 genes
</pre></div></div>
</div>
</div>
<div class="section" id="4.-in-silico-Perturbation-simulation">
<h4>4. in silico Perturbation-simulation<a class="headerlink" href="#4.-in-silico-Perturbation-simulation" title="Permalink to this headline">¶</a></h4>
<p>Next, we simulate an effect of perturbation of a TF to investigate the function and regulatory mechanism of the TF.</p>
<p>See the celloracle paper for the details and scientific remnants on the algorithm.</p>
<p>In this notebook, we’ll show an example of the simulation; we’ll simulate knock-out of Gata1 gene in the hematopoiesis.</p>
<p>Celloracle’s simulation supposes to recapitulate previous findings of this gene.</p>
<p>The previous studies have shown that Gata1 is one of the decisive TFs that regulate cell fate in the myeloid progenitors and Gata1 have a positive effect on the erythroid cell differentiation.</p>
<div class="section" id="4.1.-Check-gene-expression-pattern.">
<h5>4.1. Check gene expression pattern.<a class="headerlink" href="#4.1.-Check-gene-expression-pattern." title="Permalink to this headline">¶</a></h5>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># check gene expression</span>
<span class="n">goi</span> <span class="o">=</span> <span class="s2">&quot;Gata1&quot;</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">goi</span><span class="p">,</span> <span class="n">oracle</span><span class="o">.</span><span class="n">cluster_column_name</span><span class="p">],</span>
                 <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;imputed_count&quot;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_15_0.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_15_0.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_15_0.png" style="width: 935px; height: 265px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot gene expression in histogram</span>
<span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">obs_df</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="p">[</span><span class="n">goi</span><span class="p">],</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;imputed_count&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">hist</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_16_0.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_16_0.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_16_0.png" style="width: 383px; height: 263px;" /></a>
</div>
</div>
</div>
<div class="section" id="4.1.-calculate-future-gene-expression-after-perturbation.">
<h5>4.1. calculate future gene expression after perturbation.<a class="headerlink" href="#4.1.-calculate-future-gene-expression-after-perturbation." title="Permalink to this headline">¶</a></h5>
<p>Although any arbitrary gene expression value can be used for input of perturbation, we recommend avoiding extreme value which is far from natural gene expression range. If you set Gata1 gene expression to 100, for example, it may lead to biologically infeasible results.</p>
<p>Here we simulate Gata1 KO; we predict what happens to the cells if Gata1 gene expression changed into 0.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Enter perturbation conditions to simulate signal propagation after the perturbation.</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">simulate_shift</span><span class="p">(</span><span class="n">perturb_condition</span><span class="o">=</span><span class="p">{</span><span class="n">goi</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">},</span>
                      <span class="n">n_propagation</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="4.2.-calculate-transition-probability-between-cells">
<h5>4.2. calculate transition probability between cells<a class="headerlink" href="#4.2.-calculate-transition-probability-between-cells" title="Permalink to this headline">¶</a></h5>
<p>In the step avobe, we simulated simulated future gene expression after perturbation. This prediction is based on itelative calculations of signal propagations inside the GRN. Here we performed just small number of the calculation, thus the we simulated short future.</p>
<p>Next step, we will calculate the propability of cell state transition based on the simulated data. Using this the transition probability between cells, we can predict how a cell changes after perturbation.</p>
<p>This transition probability will be used in two ways.</p>
<ol class="arabic simple">
<li>Visualization of directed trjectory graph.</li>
<li>Markof simulation.</li>
</ol>
<p>In Step 4.2 and 4.3, we use some functions imported from velocytoloom class in velocyto.py. Please see the documentation of VelocytoLoom for the detail of parameters in the functions. <a class="reference external" href="http://velocyto.org/velocyto.py/fullapi/api_analysis.html">http://velocyto.org/velocyto.py/fullapi/api_analysis.html</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># get transition probability</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">estimate_transition_prob</span><span class="p">(</span><span class="n">n_neighbors</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">knn_random</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sampled_fraction</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

<span class="c1"># calculate embedding</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">calculate_embedding_shift</span><span class="p">(</span><span class="n">sigma_corr</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">)</span>

<span class="c1"># calculate global trend of cell transition</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">calculate_grid_arrows</span><span class="p">(</span><span class="n">smooth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">steps</span><span class="o">=</span><span class="p">(</span><span class="mi">40</span><span class="p">,</span> <span class="mi">40</span><span class="p">),</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/k/anaconda3/envs/test/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3326: FutureWarning: arrays to stack must be passed as a &#34;sequence&#34; type such as list or tuple. Support for non-sequence iterables such as generators is deprecated as of NumPy 1.16 and will raise an error in the future.
  exec(code_obj, self.user_global_ns, self.user_ns)
WARNING:root:Nans encountered in corrcoef and corrected to 1s. If not identical cells were present it is probably a small isolated cluster converging after imputation.
</pre></div></div>
</div>
</div>
<div class="section" id="4.3.-Visualization">
<h5>4.3. Visualization<a class="headerlink" href="#4.3.-Visualization" title="Permalink to this headline">¶</a></h5>
<div class="section" id="4.3.1.-Detailed-directed-trajectory-graph">
<h6>4.3.1. Detailed directed trajectory graph<a class="headerlink" href="#4.3.1.-Detailed-directed-trajectory-graph" title="Permalink to this headline">¶</a></h6>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="kc">None</span><span class="p">,(</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
<span class="n">quiver_scale</span> <span class="o">=</span> <span class="mi">40</span>


<span class="n">ix_choice</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">/</span><span class="mf">1.</span><span class="p">),</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">embedding</span> <span class="o">=</span> <span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">oracle</span><span class="o">.</span><span class="n">embedding_name</span><span class="p">]</span>

<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">embedding</span><span class="p">[</span><span class="n">ix_choice</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">embedding</span><span class="p">[</span><span class="n">ix_choice</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
            <span class="n">c</span><span class="o">=</span><span class="s2">&quot;0.8&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">38</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">rasterized</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">quiver_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">headaxislength</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">headlength</span><span class="o">=</span><span class="mi">11</span><span class="p">,</span> <span class="n">headwidth</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
                   <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.35</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.0045</span><span class="p">,</span><span class="n">edgecolors</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">,</span>
                   <span class="n">color</span><span class="o">=</span><span class="n">oracle</span><span class="o">.</span><span class="n">colorandum</span><span class="p">[</span><span class="n">ix_choice</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">quiver</span><span class="p">(</span><span class="n">embedding</span><span class="p">[</span><span class="n">ix_choice</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">embedding</span><span class="p">[</span><span class="n">ix_choice</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
           <span class="n">oracle</span><span class="o">.</span><span class="n">delta_embedding</span><span class="p">[</span><span class="n">ix_choice</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">oracle</span><span class="o">.</span><span class="n">delta_embedding</span><span class="p">[</span><span class="n">ix_choice</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
           <span class="n">scale</span><span class="o">=</span><span class="n">quiver_scale</span><span class="p">,</span> <span class="o">**</span><span class="n">quiver_kwargs</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
<span class="c1">#plt.savefig(f&quot;{save_folder}/full_arrows{goi}.png&quot;,  transparent=True)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>(-10815.27020913708, 10950.84121716522, -10711.36365432337, 10949.477199695968)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_22_1.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_22_1.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_22_1.png" style="width: 398px; height: 357px;" /></a>
</div>
</div>
</div>
<div class="section" id="4.3.2.-Grid-graph">
<h6>4.3.2. Grid graph<a class="headerlink" href="#4.3.2.-Grid-graph" title="Permalink to this headline">¶</a></h6>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot whole graph</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="kc">None</span><span class="p">,(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_grid_arrows</span><span class="p">(</span><span class="n">quiver_scale</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
                        <span class="n">scatter_kwargs_dict</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;alpha&quot;</span><span class="p">:</span><span class="mf">0.35</span><span class="p">,</span> <span class="s2">&quot;lw&quot;</span><span class="p">:</span><span class="mf">0.35</span><span class="p">,</span>
                                              <span class="s2">&quot;edgecolor&quot;</span><span class="p">:</span><span class="s2">&quot;0.4&quot;</span><span class="p">,</span> <span class="s2">&quot;s&quot;</span><span class="p">:</span><span class="mi">38</span><span class="p">,</span>
                                              <span class="s2">&quot;rasterized&quot;</span><span class="p">:</span><span class="kc">True</span><span class="p">},</span>
                        <span class="n">min_mass</span><span class="o">=</span><span class="mf">0.015</span><span class="p">,</span> <span class="n">angles</span><span class="o">=</span><span class="s1">&#39;xy&#39;</span><span class="p">,</span> <span class="n">scale_units</span><span class="o">=</span><span class="s1">&#39;xy&#39;</span><span class="p">,</span>
                        <span class="n">headaxislength</span><span class="o">=</span><span class="mf">2.75</span><span class="p">,</span>
                        <span class="n">headlength</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">headwidth</span><span class="o">=</span><span class="mf">4.8</span><span class="p">,</span> <span class="n">minlength</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span>
                        <span class="n">plot_random</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">scale_type</span><span class="o">=</span><span class="s2">&quot;relative&quot;</span><span class="p">)</span>
<span class="c1">#plt.savefig(f&quot;{save_folder}/vectorfield_{goi}.png&quot;, transparent=True)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_24_0.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_24_0.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_24_0.png" style="width: 619px; height: 574px;" /></a>
</div>
</div>
</div>
</div>
<div class="section" id="4.4.-Markov-simulation-to-analyze-the-effects-of-perturbation-on-cell-fate-transition">
<h5>4.4. Markov simulation to analyze the effects of perturbation on cell fate transition<a class="headerlink" href="#4.4.-Markov-simulation-to-analyze-the-effects-of-perturbation-on-cell-fate-transition" title="Permalink to this headline">¶</a></h5>
<p>We can also simulate cell state transition with Markof simulation with the transition probability.</p>
<div class="section" id="4.4.1.-Do-simulation">
<h6>4.4.1. Do simulation<a class="headerlink" href="#4.4.1.-Do-simulation" title="Permalink to this headline">¶</a></h6>
<p>We simulate with the parameters, “n_steps=200” and “n_duplication=5” in the following example.</p>
<p>It means that</p>
<ol class="arabic simple">
<li>we will do 200 times of iterative simulations to predict how cell changes over time</li>
<li>and we will do 5 round of simulations for each cells.</li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[83]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%%time</span>
<span class="c1"># n_steps is the number of steps in markov simulation.</span>
<span class="c1"># n_duplication is the number of technical duplicate for the simulation</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">run_markov_chain_simulation</span><span class="p">(</span><span class="n">n_steps</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">n_duplication</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 1.33 s, sys: 0 ns, total: 1.33 s
Wall time: 1.33 s
</pre></div></div>
</div>
</div>
<div class="section" id="4.4.2.-Check-the-results-of-simulation-focusing-on-a-specific-cells">
<h6>4.4.2. Check the results of simulation focusing on a specific cells<a class="headerlink" href="#4.4.2.-Check-the-results-of-simulation-focusing-on-a-specific-cells" title="Permalink to this headline">¶</a></h6>
<p>Check the results of simulation. Pick up some cells and visualize their transition trajectory.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[88]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#np.random.seed(12)</span>
<span class="c1"># randomly pick up 3 cells</span>
<span class="n">cells</span> <span class="o">=</span> <span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">ixs_mcmc</span><span class="p">,</span> <span class="mi">3</span><span class="p">)]</span>

<span class="c1"># visualize the simulated results of cell transition after perturbation</span>
<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">cells</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;cell </span><span class="si">{k}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">50</span><span class="p">]):</span> <span class="c1"># time points</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">oracle</span><span class="o">.</span><span class="n">plot_mc_result_as_trajectory</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">i</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;simulation step: 0~</span><span class="si">{i}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell 1961
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_1.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_1.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_1.png" style="width: 573px; height: 209px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell 43
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_3.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_3.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_3.png" style="width: 573px; height: 209px;" /></a>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell 1567
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_5.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_5.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_32_5.png" style="width: 573px; height: 209px;" /></a>
</div>
</div>
</div>
<div class="section" id="4.4.3.-Summarize-the-results-of-simulation-by-plotting-sankey-diagram">
<h6>4.4.3. Summarize the results of simulation by plotting sankey diagram<a class="headerlink" href="#4.4.3.-Summarize-the-results-of-simulation-by-plotting-sankey-diagram" title="Permalink to this headline">¶</a></h6>
<p>Sankey diagram is useful when you want to visualize cell transition between some groups.</p>
<p>The function below can make a Sankey diagram with any arbitrary cluster unit.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[89]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot sankey diagram</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">])</span>
<span class="n">cl</span> <span class="o">=</span> <span class="s2">&quot;louvain_annot&quot;</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_mc_resutls_as_sankey</span><span class="p">(</span><span class="n">cluster_use</span><span class="o">=</span><span class="n">cl</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_35_0.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_35_0.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_35_0.png" style="width: 357px; height: 357px;" /></a>
</div>
</div>
<p>The sankey diagram above looks messy because cluster order is random.</p>
<p>Change cluster order and make plot again</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[90]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>

<span></span><span class="n">cl</span> <span class="o">=</span> <span class="s2">&quot;louvain_annot&quot;</span>
<span class="n">order</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;MEP_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Mk_0&#39;</span><span class="p">,</span><span class="s1">&#39;Ery_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_1&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_2&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_3&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_4&#39;</span><span class="p">,</span>
         <span class="s1">&#39;Ery_5&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_6&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_7&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_8&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_9&#39;</span><span class="p">,</span>
         <span class="s1">&#39;GMP_0&#39;</span><span class="p">,</span> <span class="s1">&#39;GMP_1&#39;</span><span class="p">,</span> <span class="s1">&#39;GMP_2&#39;</span><span class="p">,</span> <span class="s1">&#39;GMPl_0&#39;</span><span class="p">,</span> <span class="s1">&#39;GMPl_1&#39;</span><span class="p">,</span>
         <span class="s1">&#39;Mo_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Mo_1&#39;</span><span class="p">,</span> <span class="s1">&#39;Mo_2&#39;</span><span class="p">,</span> <span class="s1">&#39;Gran_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Gran_1&#39;</span><span class="p">,</span> <span class="s1">&#39;Gran_2&#39;</span><span class="p">,</span> <span class="s1">&#39;Gran_3&#39;</span><span class="p">]</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_mc_resutls_as_sankey</span><span class="p">(</span><span class="n">cluster_use</span><span class="o">=</span><span class="n">cl</span><span class="p">,</span>  <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">)</span>
<span class="c1">#plt.savefig(f&quot;{save_folder}/mcmc_{cl}.png&quot;)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_37_0.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_37_0.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_37_0.png" style="width: 226px; height: 357px;" /></a>
</div>
</div>
<p>Make another saneky diagram with different cluster unit.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[92]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">order</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Megakaryocytes&#39;</span><span class="p">,</span> <span class="s1">&#39;MEP&#39;</span><span class="p">,</span> <span class="s1">&#39;Erythroids&#39;</span><span class="p">,</span> <span class="s1">&#39;GMP&#39;</span><span class="p">,</span> <span class="s1">&#39;late_GMP&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocytes&#39;</span><span class="p">,</span> <span class="s1">&#39;Granulocytes&#39;</span><span class="p">]</span>
<span class="n">cl</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.35</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mf">0.65</span><span class="p">)</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_mc_resutls_as_sankey</span><span class="p">(</span><span class="n">cluster_use</span><span class="o">=</span><span class="n">cl</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="c1">#plt.savefig(f&quot;{save_folder}/mcmc_{cl}{goi}.png&quot;, transparent=True)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_39_0.png"><img alt="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_39_0.png" src="_images/notebooks_05_simulation_Gata1_KO_simulation_with_with_Paul_etal_2015_data_39_0.png" style="width: 346px; height: 360px;" /></a>
</div>
</div>
<p>Based on the results, we may conclude several things as follows.</p>
<p>Gata1 KO induced cell state transition from Erythroids to MEP, and from MEP to GMP. (1) These results suggest that Gata1 may play a role in the progression of Erythroid differentiation and cell state determination between MEP and GMP lineage.</p>
<ol class="arabic simple" start="2">
<li>Gata1 KO also induced cell state transition from granulocytes to late GMP, suggesting Gata1’s involvement in Granulocytes differentiation.</li>
</ol>
<p>These results agree with previous reports about Gata1. We could recapitulate Gata1’s cell-type-specific function regarding the cell fate decisions in hematopoiesis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="toctree-wrapper compound">
<span id="document-modules/index"></span><div class="section" id="api">
<h2>API<a class="headerlink" href="#api" title="Permalink to this headline">¶</a></h2>
<div class="section" id="command-line-api">
<h3>Command Line API<a class="headerlink" href="#command-line-api" title="Permalink to this headline">¶</a></h3>
<p>CellOracle have one command line API. This command can be used to convert scRNA-seq data.
If you have a scRNA-seq data which was processed with seurat and saved as Rds file, you can use the following command to make anndata from seurat object.
The anndata produced by this command can be used for an input of celloracle. Please go to step3 and step4 of the tutorial to see an example.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seuratToAnndata</span> <span class="n">YOUR_SEURAT_OBJECT</span><span class="o">.</span><span class="n">Rds</span> <span class="n">OUTPUT_PATH</span>
</pre></div>
</div>
</div>
<div class="section" id="python-api">
<h3>Python API<a class="headerlink" href="#python-api" title="Permalink to this headline">¶</a></h3>
<div class="toctree-wrapper compound">
<span id="document-modules/celloracle"></span><div class="section" id="custom-class-in-celloracle">
<h4>Custom class in celloracle<a class="headerlink" href="#custom-class-in-celloracle" title="Permalink to this headline">¶</a></h4>
<p>We define some custom classes in celloracle.</p>
<span class="target" id="module-celloracle"></span><dl class="class">
<dt id="celloracle.Oracle">
<em class="property">class </em><code class="descclassname">celloracle.</code><code class="descname">Oracle</code><a class="headerlink" href="#celloracle.Oracle" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">celloracle.trajectory.modified_VelocytoLoom_class.modified_VelocytoLoom</span></code></p>
<p>Oracle is the main class in CellOracle.
Oracle object imports scRNA-seq data (anndata) and TF information to infer cluster-specific GRNs.
It can predict future gene expression patterns and cell state transition after perturbations of TFs.
Please see the paper of CellOracle for details.</p>
<p>The code of the Oracle class was made of three components below.</p>
<p>(1) Anndata:  Gene expression matrix and metadata in single-cell RNA-seq are stored in anndata object.
Processed values, such as normalized counts and simulated values, are stored as layers of anndata.
Metadata (i.e., Cluster info) are saved in anndata.obs. Refer to scanpy/anndata documentation for detail.</p>
<p>(2) Net: Net is a custom class in celloracle.
Net object process several data to infer GRN.
See the documentation of Net class for detail.</p>
<p>(3) VelycytoLoom: Calculation of transition probability and visualization of directed trajectory graph will be performed in the same way as velocytoloom.
VelocytoLoom is class for the Velocyto, which is a python library for RNA-velocity analysis.
In the celloracle, we use almost the same functions as velocytoloom for the visualization, but celloracle use simulated gene expression values instead of RNA-velocity data.</p>
<p>Some CellOracle’s methods were inspired by velocyto analysis and some codes were made by modifying VelocytoLoom class.</p>
<dl class="attribute">
<dt id="celloracle.Oracle.adata">
<code class="descname">adata</code><a class="headerlink" href="#celloracle.Oracle.adata" title="Permalink to this definition">¶</a></dt>
<dd><p><em>anndata</em> – Imported anndata object</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Oracle.cluster_column_name">
<code class="descname">cluster_column_name</code><a class="headerlink" href="#celloracle.Oracle.cluster_column_name" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – The column name in adata.obs about cluster info</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Oracle.embedding_name">
<code class="descname">embedding_name</code><a class="headerlink" href="#celloracle.Oracle.embedding_name" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – The key name in adata.obsm about dimensional reduction cordinates</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.addTFinfo_dictionary">
<code class="descname">addTFinfo_dictionary</code><span class="sig-paren">(</span><em>TFdict</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.addTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Add new TF info to pre-existing TFdict.
Values in the old TF dictionary will remain.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>TFdict</strong> (<em>dictionary</em>) – Python dictionary of TF info.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.copy">
<code class="descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Deepcopy itself.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.fit_GRN_for_simulation">
<code class="descname">fit_GRN_for_simulation</code><span class="sig-paren">(</span><em>GRN_unit='cluster'</em>, <em>alpha=1</em>, <em>use_cluster_specific_TFdict=False</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.fit_GRN_for_simulation" title="Permalink to this definition">¶</a></dt>
<dd><p>Do GRN inference.
Please see the paper of CellOracle for details.</p>
<p>GRN can be constructed at an arbitrary cell group.
If you want to infer cluster-specific GRN, please set [GRN_unit=”cluster”].
GRN will be inferred for each cluster. You can select Cluster information when you import data (not when you run this method.).</p>
<p>If you set [GRN_unit=”whole”], GRN will be made using all cells.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>GRN_unit</strong> (<em>str</em>) – select “cluster” or “whole”</li>
<li><strong>alpha</strong> (<em>float</em><em> or </em><em>int</em>) – the strength of regularization.
If you set a lower value, the sensitivity increase, and you can detect a weak network connection, but it might get more noize.
With a higher value of alpha may reduce the chance of overfitting.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.get_cluster_specific_TFdict_from_Links">
<code class="descname">get_cluster_specific_TFdict_from_Links</code><span class="sig-paren">(</span><em>links_object</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.get_cluster_specific_TFdict_from_Links" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract TF and its target gene information from Links object.
This function can be used to reconstruct GRNs based on pre-existing GRNs saved in Links object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>links_object</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – Please see the explanation of Links class.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.get_links">
<code class="descname">get_links</code><span class="sig-paren">(</span><em>cluster_name_for_GRN_unit=None</em>, <em>alpha=10</em>, <em>bagging_number=20</em>, <em>verbose_level=1</em>, <em>test_mode=False</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.get_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Make GRN for each cluster and returns results as a Links object.
Several preprocessing should be done before using this function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cluster_name_for_GRN_unit</strong> (<em>str</em>) – Cluster name for GRN calculation. The cluster information should be stored in Oracle.adata.obs.</li>
<li><strong>alpha</strong> (<em>float</em><em> or </em><em>int</em>) – the strength of regularization.
If you set a lower value, the sensitivity increase, and you can detect a weak network connection, but it might get more noize.
With a higher value of alpha may reduce the chance of overfitting.</li>
<li><strong>bagging_number</strong> (<em>int</em>) – The number for bagging calculation.</li>
<li><strong>verbose_level</strong> (<em>int</em>) – if [verbose_level&gt;1], most detailed progress information will be shown.
if [verbose_level &gt; 0], one progress bar will be shown.
if [verbose_level == 0], no progress bar will be shown.</li>
<li><strong>test_mode</strong> (<em>bool</em>) – If test_mode is True, GRN calculation will be done for only one cluster rather than all clusters.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.import_TF_data">
<code class="descname">import_TF_data</code><span class="sig-paren">(</span><em>TF_info_matrix=None</em>, <em>TF_info_matrix_path=None</em>, <em>TFdict=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.import_TF_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Load data about potential-regulatory TFs.
You can import either TF_info_matrix or TFdict.
See the tutorial of celloracle or motif_analysis module for an example to make such files.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>TF_info_matrix</strong> (<em>pandas.DataFrame</em>) – TF_info_matrix.</li>
<li><strong>TF_info_matrix_path</strong> (<em>str</em>) – File path for TF_info_matrix (pandas.DataFrame).</li>
<li><strong>TFdict</strong> (<em>dictionary</em>) – Python dictionary of TF info.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.import_anndata_as_normalized_count">
<code class="descname">import_anndata_as_normalized_count</code><span class="sig-paren">(</span><em>adata</em>, <em>cluster_column_name=None</em>, <em>embedding_name=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.import_anndata_as_normalized_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Load scRNA-seq data. scRNA-seq data should be prepared as an anndata.
Preprocessing (cell and gene filtering, calculate DR and cluster, etc.) should be done before loading data.
The method will import NORMALIZED and LOG TRANSFORMED but NOT SCALED and NOT CENTERED DATA.
See the tutorial for the details for an example of how to process scRNA-seq data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>adata</strong> (<em>anndata</em>) – anndata object that store scRNA-seq data.</li>
<li><strong>cluster_column_name</strong> (<em>str</em>) – the column name about cluster info in anndata.obs.
Clustering data suppose to be in anndata.obs.</li>
<li><strong>embedding_name</strong> (<em>str</em>) – the key name about a dimensional reduction in anndata.obsm.
Dimensional reduction (or 2D trajectory graph) should be in anndata.obsm.</li>
<li><strong>transform</strong> (<em>str</em>) – The method for log-transformation. Chose one from “natural_log” or “log2”.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.import_anndata_as_raw_count">
<code class="descname">import_anndata_as_raw_count</code><span class="sig-paren">(</span><em>adata</em>, <em>cluster_column_name=None</em>, <em>embedding_name=None</em>, <em>transform='natural_log'</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.import_anndata_as_raw_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Load scRNA-seq data. scRNA-seq data should be prepared as an anndata.
Preprocessing (cell and gene filtering, calculate DR and cluster, etc.) should be done before loading data.
The method imports RAW GENE COUNTS because unscaled and uncentered gene expression data are required for the GRN inference and simulation.
See tutorial notebook for the details about how to process scRNA-seq data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>adata</strong> (<em>anndata</em>) – anndata object that stores scRNA-seq data.</li>
<li><strong>cluster_column_name</strong> (<em>str</em>) – the column name about cluster info in anndata.obs.
Clustering data suppose to be in anndata.obs.</li>
<li><strong>embedding_name</strong> (<em>str</em>) – the key name about a dimensional reduction in anndata.obsm.
Dimensional reduction (or 2D trajectory graph) should be in anndata.obsm.</li>
<li><strong>transform</strong> (<em>str</em>) – The method for log-transformation. Chose one from “natural_log” or “log2”.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.plot_mc_result_as_kde">
<code class="descname">plot_mc_result_as_kde</code><span class="sig-paren">(</span><em>n_time</em>, <em>args={}</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.plot_mc_result_as_kde" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up one timepoint in the cell state-transition simulation and plot as a kde plot.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>n_time</strong> (<em>int</em>) – the number in Markov simulation</li>
<li><strong>args</strong> (<em>dictionary</em>) – An argument for seaborn.kdeplot.
See seaborn documentation for detail (<a class="reference external" href="https://seaborn.pydata.org/generated/seaborn.kdeplot.html#seaborn.kdeplot">https://seaborn.pydata.org/generated/seaborn.kdeplot.html#seaborn.kdeplot</a>).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.plot_mc_result_as_trajectory">
<code class="descname">plot_mc_result_as_trajectory</code><span class="sig-paren">(</span><em>cell_name</em>, <em>time_range</em>, <em>args={}</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.plot_mc_result_as_trajectory" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up several timepoints in the cell state-transition simulation and plot as a line plot.
This function can be used to visualize how cell-state changes after perturbation focusing on a specific cell.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cell_name</strong> (<em>str</em>) – cell name. chose from adata.obs.index</li>
<li><strong>time_range</strong> (<em>list of int</em>) – the number in markov simulation</li>
<li><strong>args</strong> (<em>dictionary</em>) – dictionary for the arguments for matplotlib.pyplit.plot.
See matplotlib documentation for detail (<a class="reference external" href="https://matplotlib.org/api/_as_gen/matplotlib.pyplot.plot.html#matplotlib.pyplot.plot">https://matplotlib.org/api/_as_gen/matplotlib.pyplot.plot.html#matplotlib.pyplot.plot</a>).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.plot_mc_resutls_as_sankey">
<code class="descname">plot_mc_resutls_as_sankey</code><span class="sig-paren">(</span><em>cluster_use</em>, <em>start=0</em>, <em>end=-1</em>, <em>order=None</em>, <em>font_size=10</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.plot_mc_resutls_as_sankey" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the simulated cell state-transition as a Sankey-diagram after groping by the cluster.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cluster_use</strong> (<em>str</em>) – cluster information name in anndata.obs.
You can use any arbitrary cluster information in anndata.obs.</li>
<li><strong>start</strong> (<em>int</em>) – The starting point of Sankey-diagram. Please select a  step in the Markov simulation.</li>
<li><strong>end</strong> (<em>int</em>) – The end point of Sankey-diagram. Please select a  step in the Markov simulation.
if you set [end=-1], the final step of Markov simulation will be used.</li>
<li><strong>order</strong> (<em>list of str</em>) – The order of cluster name in sankey-diagram.</li>
<li><strong>font_size</strong> (<em>int</em>) – Font size for cluster name label in Sankey diagram.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.prepare_markov_simulation">
<code class="descname">prepare_markov_simulation</code><span class="sig-paren">(</span><em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.prepare_markov_simulation" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up cells for Markov simulation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>verbose</strong> (<em>bool</em>) – If True, it plots selected cells.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.run_markov_chain_simulation">
<code class="descname">run_markov_chain_simulation</code><span class="sig-paren">(</span><em>n_steps=500</em>, <em>n_duplication=5</em>, <em>seed=123</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.run_markov_chain_simulation" title="Permalink to this definition">¶</a></dt>
<dd><p>Do Markov simlation to predict cell transition after perturbation.
The transition probability between cells has been calculated
based on simulated gene expression values in the signal propagation process.
The cell state transition will be simulated based on the probability.
You can simulate the process for multiple times to get a robust outcome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>n_steps</strong> (<em>int</em>) – steps for Markov simulation. This value is equivalent to the time after perturbation.</li>
<li><strong>n_duplication</strong> (<em>int</em>) – the number for multiple calculations.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.simulate_shift">
<code class="descname">simulate_shift</code><span class="sig-paren">(</span><em>perturb_condition=None</em>, <em>GRN_unit='whole'</em>, <em>n_propagation=3</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.simulate_shift" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate signal propagation with GRNs. Please see the paper of CellOracle for details.
This function simulates a gene expression pattern in the near future.
Simulated values will be stored in anndata.layers: [“simulated_count”]</p>
<p>Three data below are used for the simulation.
(1) GRN inference results (coef_matrix).
(2) perturb_condition: You can set arbitrary perturbation condition.
(3) gene expression matrix: simulation starts from imputed gene expression data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>perturb_condition</strong> (<em>dictionary</em>) – condition for perturbation.
if you want to simulate knockout for GeneX, please set [perturb_condition={“GeneX”: 0.0}]
Although you can set any non-negative values for the gene condition, avoid setting biologically unfeasible values for the perturb condition.
It is strongly recommended to check actual gene expression values in your data before selecting perturb condition.</li>
<li><strong>n_propagation</strong> (<em>int</em>) – Calculation will be performed iteratively to simulate signal propagation in GRN.
you can set the number of steps for this calculation.
With a higher number, the results may recapitulate signal propagation for many genes.
However, a higher number of propagation may cause more error/noise.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.summarize_mc_results_by_cluster">
<code class="descname">summarize_mc_results_by_cluster</code><span class="sig-paren">(</span><em>cluster_use</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.summarize_mc_results_by_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>This function summarizes the simulated cell state-transition by groping the results into each cluster.
It returns sumarized results as a pandas.DataFrame.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>cluster_use</strong> (<em>str</em>) – cluster information name in anndata.obs.
You can use any arbitrary cluster information in anndata.obs.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.to_hdf5">
<code class="descname">to_hdf5</code><span class="sig-paren">(</span><em>file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.oracle’</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Oracle.updateTFinfo_dictionary">
<code class="descname">updateTFinfo_dictionary</code><span class="sig-paren">(</span><em>TFdict</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.updateTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Update a TF dictionary.
If a key in the new TF dictionary already existed in the old TF dictionary, old values will be replaced with a new one.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>TFdict</strong> (<em>dictionary</em>) – Python dictionary of TF info.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="celloracle.Links">
<em class="property">class </em><code class="descclassname">celloracle.</code><code class="descname">Links</code><span class="sig-paren">(</span><em>name</em>, <em>links_dict={}</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This is a class for the processing and visualization of GRNs.
Links object stores cluster-specific GRNs and metadata.
Please use “get_links” function in Oracle object to generate Links object.</p>
<dl class="attribute">
<dt id="celloracle.Links.links_dict">
<code class="descname">links_dict</code><a class="headerlink" href="#celloracle.Links.links_dict" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – Dictionary that store unprocessed network data.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Links.filtered_links">
<code class="descname">filtered_links</code><a class="headerlink" href="#celloracle.Links.filtered_links" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – Dictionary that store filtered network data.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Links.merged_score">
<code class="descname">merged_score</code><a class="headerlink" href="#celloracle.Links.merged_score" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.dataframe</em> – Network scores.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Links.cluster">
<code class="descname">cluster</code><a class="headerlink" href="#celloracle.Links.cluster" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list of str</em> – List of cluster name.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Links.name">
<code class="descname">name</code><a class="headerlink" href="#celloracle.Links.name" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – Name of clustering unit.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Links.palette">
<code class="descname">palette</code><a class="headerlink" href="#celloracle.Links.palette" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.dataframe</em> – DataFrame that store color information.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.add_palette">
<code class="descname">add_palette</code><span class="sig-paren">(</span><em>cluster_name_list</em>, <em>color_list</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.add_palette" title="Permalink to this definition">¶</a></dt>
<dd><p>Add color information for each cluster.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cluster_name_list</strong> (<em>list of str</em>) – cluster names.</li>
<li><strong>color_list</strong> (<em>list of stf</em>) – list of color for each cluster.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.filter_links">
<code class="descname">filter_links</code><span class="sig-paren">(</span><em>p=0.001</em>, <em>weight='coef_abs'</em>, <em>thread_number=10000</em>, <em>genelist_source=None</em>, <em>genelist_target=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.filter_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter network edges.
In most case inferred GRN has non-significant random edges.
We have to remove such random edges before analyzing network structure.
You can do the filtering either of three ways below.</p>
<ol class="arabic simple">
<li>Do filtering based on the p-value of the network edge.
Please enter p-value for thresholding.</li>
<li>Do filtering based on network edge number.
If you set the number, network edges will be filtered based on the order of a network score. The top n-th network edges with network weight will remain, and the other edges will be removed.
The network data have several types of network weight, so you have to select which network weight do you want to use.</li>
<li>Do filtering based on an arbitrary gene list. You can set a gene list for source nodes or target nodes.</li>
</ol>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>p</strong> (<em>float</em>) – threshold for p-value of the network edge.</li>
<li><strong>weight</strong> (<em>str</em>) – Please select network weight name for the filtering</li>
<li><strong>genelist_source</strong> (<em>list of str</em>) – gene list to remain in regulatory gene nodes. Default is None.</li>
<li><strong>genelist_target</strong> (<em>list of str</em>) – gene list to remain in target gene nodes. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.get_network_entropy">
<code class="descname">get_network_entropy</code><span class="sig-paren">(</span><em>value='coef_abs'</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.get_network_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate network entropy scores.
Please select a type of network weight for the entropy calculation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>value</strong> (<em>str</em>) – Default is “coef_abs”.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.get_score">
<code class="descname">get_score</code><span class="sig-paren">(</span><em>test_mode=False</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.get_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Get several network sores using R libraries.
Make sure all dependant R libraries are installed in your environment before running this function.
You can check the installation for the R libraries by running test_installation() in network_analysis module.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.plot_cartography_scatter_per_cluster">
<code class="descname">plot_cartography_scatter_per_cluster</code><span class="sig-paren">(</span><em>gois=None</em>, <em>clusters=None</em>, <em>scatter=False</em>, <em>kde=True</em>, <em>auto_gene_annot=False</em>, <em>percentile=98</em>, <em>args_dot={}</em>, <em>args_line={}</em>, <em>args_annot={}</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_cartography_scatter_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a plot of gene network cartography.
Please read the original paper of gene network cartography for the principle of gene network cartography.
<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>gois</strong> (<em>list of srt</em>) – List of Gene name to highlight.</li>
<li><strong>clusters</strong> (<em>list of str</em>) – List of cluster name to analyze. If None, all clusters in Links object will be analyzed.</li>
<li><strong>scatter</strong> (<em>bool</em>) – Whether to make a scatter plot.</li>
<li><strong>auto_gene_annot</strong> (<em>bool</em>) – Whether to pick up genes to make an annotation.</li>
<li><strong>percentile</strong> (<em>float</em>) – Genes with a network score above the percentile will be shown with annotation. Default is 98.</li>
<li><strong>args_dot</strong> (<em>dictionary</em>) – Arguments for scatter plot.</li>
<li><strong>args_line</strong> (<em>dictionary</em>) – Arguments for lines in cartography plot.</li>
<li><strong>args_annot</strong> (<em>dictionary</em>) – Arguments for annotation in plots.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.plot_cartography_term">
<code class="descname">plot_cartography_term</code><span class="sig-paren">(</span><em>goi</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_cartography_term" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the summary of gene network cartography like a heatmap.
Please read the original paper of gene network cartography for the principle of gene network cartography.
<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>gois</strong> (<em>list of srt</em>) – List of Gene name to highlight.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.plot_degree_distributions">
<code class="descname">plot_degree_distributions</code><span class="sig-paren">(</span><em>plot_model=False</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_degree_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of network degree (the number of edge per gene).
The network degree will be visualized in both linear scale and log scale.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>plot_model</strong> (<em>bool</em>) – Whether to plot linear approximation line.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.plot_network_entropy_distributions">
<code class="descname">plot_network_entropy_distributions</code><span class="sig-paren">(</span><em>update_network_entropy=False</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_network_entropy_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of network entropy.
See the CellOracle paper for the detail.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<em>Links object</em>) – See network_analisis.Links class for detail.</li>
<li><strong>values</strong> (<em>list of str</em>) – The list of score to visualize. If it is None, all network score (listed above) will be used.</li>
<li><strong>update_network_entropy</strong> (<em>bool</em>) – Whether to recalculate network entropy.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.plot_score_comparison_2D">
<code class="descname">plot_score_comparison_2D</code><span class="sig-paren">(</span><em>value</em>, <em>cluster1</em>, <em>cluster2</em>, <em>percentile=99</em>, <em>annot_shifts=None</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_score_comparison_2D" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a scatter plot that shows the relationship of a specific network score in two groups.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>value</strong> (<em>srt</em>) – The network score type.</li>
<li><strong>cluster1</strong> (<em>str</em>) – Cluster name. Network scores in the cluster1 will be visualized in the x-axis.</li>
<li><strong>cluster2</strong> (<em>str</em>) – Cluster name. Network scores in the cluster2 will be visualized in the y-axis.</li>
<li><strong>percentile</strong> (<em>float</em>) – Genes with a network score above the percentile will be shown with annotation. Default is 99.</li>
<li><strong>annot_shifts</strong> (<em>(</em><em>float</em><em>, </em><em>float</em><em>)</em>) – Annotation visualization setting.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.plot_score_discributions">
<code class="descname">plot_score_discributions</code><span class="sig-paren">(</span><em>values=None</em>, <em>method='boxplot'</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_score_discributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of network scores.
An individual data point is a network edge (gene) of GRN in each cluster.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See Links class for detail.</li>
<li><strong>values</strong> (<em>list of str</em>) – The list of score to visualize. If it is None, all network score will be used.</li>
<li><strong>method</strong> (<em>str</em>) – Plotting method. Select either of “boxplot” or “barplot”.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.plot_score_per_cluster">
<code class="descname">plot_score_per_cluster</code><span class="sig-paren">(</span><em>goi</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_score_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot network score for a gene.
This function visualizes the network score of a specific gene between clusters to get an insight into the dynamics of the gene.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>goi</strong> (<em>srt</em>) – Gene name.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.plot_scores_as_rank">
<code class="descname">plot_scores_as_rank</code><span class="sig-paren">(</span><em>cluster</em>, <em>n_gene=50</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_scores_as_rank" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up top n-th genes wich high-network scores and make plots.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>cluster</strong> (<em>str</em>) – Cluster nome to analyze.</li>
<li><strong>n_gene</strong> (<em>int</em>) – Number of genes to plot. Default is 50.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Links.to_hdf5">
<code class="descname">to_hdf5</code><span class="sig-paren">(</span><em>file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.links’</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="celloracle.Net">
<em class="property">class </em><code class="descclassname">celloracle.</code><code class="descname">Net</code><span class="sig-paren">(</span><em>gene_expression_matrix</em>, <em>gem_standerdized=None</em>, <em>TFinfo_matrix=None</em>, <em>cellstate=None</em>, <em>TFinfo_dic=None</em>, <em>annotation=None</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Net is a custom class for inferring sample-specific GRN from scRNA-seq data.
This class is used inside Oracle class for GRN inference.
This class requires two information below.</p>
<ol class="arabic simple">
<li>single-cell RNA-seq data
The Net class needs processed scRNA-seq data.
Gene and cell filtering, quality check, normalization, log-transformation (but not scaling and centering) have to be done before starting GRN calculation with this class.
You can also use any arbitrary metadata (i.e., mRNA count, cell-cycle phase) for GRN input.</li>
<li>potential regulatory connection
This method uses the list of potential regulatory TFs as input.
This information can be calculated from Atac-seq data using the motif-analysis module.
If sample-specific ATAC-seq data is not available,
you can use general TF-binding info made from public ATAC-seq dataset of various tissue/cell type.</li>
</ol>
<dl class="attribute">
<dt id="celloracle.Net.linkList">
<code class="descname">linkList</code><a class="headerlink" href="#celloracle.Net.linkList" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> – the result of GRN inference.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.all_genes">
<code class="descname">all_genes</code><a class="headerlink" href="#celloracle.Net.all_genes" title="Permalink to this definition">¶</a></dt>
<dd><p><em>numpy.array</em> – an array of all genes that exist in input gene expression matrix</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.embedding_name">
<code class="descname">embedding_name</code><a class="headerlink" href="#celloracle.Net.embedding_name" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – the key name name in adata.obsm about dimensional reduction coordinates</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.annotation">
<code class="descname">annotation</code><a class="headerlink" href="#celloracle.Net.annotation" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – annotation. you can add an arbitrary annotation.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.coefs_dict">
<code class="descname">coefs_dict</code><a class="headerlink" href="#celloracle.Net.coefs_dict" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – coefs of linear regression.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.stats_dict">
<code class="descname">stats_dict</code><a class="headerlink" href="#celloracle.Net.stats_dict" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – statistic values about coefs.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.fitted_genes">
<code class="descname">fitted_genes</code><a class="headerlink" href="#celloracle.Net.fitted_genes" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list of str</em> – list of genes with which regression model was successfully calculated.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.failed_genes">
<code class="descname">failed_genes</code><a class="headerlink" href="#celloracle.Net.failed_genes" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list of str</em> – list of genes that were not able to get coefs</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.cellstate">
<code class="descname">cellstate</code><a class="headerlink" href="#celloracle.Net.cellstate" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> – metadata for GRN input</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.TFinfo">
<code class="descname">TFinfo</code><a class="headerlink" href="#celloracle.Net.TFinfo" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> – information about potential regulatory TFs.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.gem">
<code class="descname">gem</code><a class="headerlink" href="#celloracle.Net.gem" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> – merged matrix made with gene_expression_matrix and cellstate matrix.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.gem_standerdized">
<code class="descname">gem_standerdized</code><a class="headerlink" href="#celloracle.Net.gem_standerdized" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> – almost the same as gem, but the gene_expression_matrix was standardized.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.library_last_update_date">
<code class="descname">library_last_update_date</code><a class="headerlink" href="#celloracle.Net.library_last_update_date" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – last update date of this code (this info is for code development. it will be deleted in the future)</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.Net.object_initiation_date">
<code class="descname">object_initiation_date</code><a class="headerlink" href="#celloracle.Net.object_initiation_date" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – the date when this object is made.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.addAnnotation">
<code class="descname">addAnnotation</code><span class="sig-paren">(</span><em>annotation_dictionary</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.addAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a new annotation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>annotation_dictionary</strong> (<em>dictionary</em>) – e.g. {“sample_name”: “NIH 3T3 cell”}</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.addTFinfo_dictionary">
<code class="descname">addTFinfo_dictionary</code><span class="sig-paren">(</span><em>TFdict</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.addTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a new TF info to pre-exiting TFdict.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>TFdict</strong> (<em>dictionary</em>) – python dictionary of TF info.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.addTFinfo_matrix">
<code class="descname">addTFinfo_matrix</code><span class="sig-paren">(</span><em>TFinfo_matrix</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.addTFinfo_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Load TF info dataframe.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>TFinfo</strong> (<em>pandas.DataFrame</em>) – information about potential regulatory TFs.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.copy">
<code class="descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Deepcopy itself</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.fit_All_genes">
<code class="descname">fit_All_genes</code><span class="sig-paren">(</span><em>bagging_number=200</em>, <em>scaling=True</em>, <em>model_method='bagging_ridge'</em>, <em>command_line_mode=False</em>, <em>log=None</em>, <em>alpha=1</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.fit_All_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Make ML models for all genes.
The calculation will be performed in parallel using scikit-learn bagging function.
You can select a modeling method (bagging_ridge or bayesian_ridge).  The calculation usually takes a long time.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</li>
<li><strong>scaling</strong> (<em>bool</em>) – Whether to scale regulatory gene expression values.</li>
<li><strong>model_method</strong> (<em>str</em>) – ML model name. “bagging_ridge” or “bayesian_ridge”</li>
<li><strong>command_line_mode</strong> (<em>bool</em>) – Please select False if the calculation is performed on jupyter notebook.</li>
<li><strong>log</strong> (<em>logging object</em>) – log object to output log</li>
<li><strong>alpha</strong> (<em>int</em>) – Strength of regularization.</li>
<li><strong>verbose</strong> (<em>bool</em>) – Whether to show a progress bar.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.fit_All_genes_parallel">
<code class="descname">fit_All_genes_parallel</code><span class="sig-paren">(</span><em>bagging_number=200</em>, <em>scaling=True</em>, <em>log=None</em>, <em>verbose=10</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.fit_All_genes_parallel" title="Permalink to this definition">¶</a></dt>
<dd><p>!! This function has some bug now and currently unavailable.</p>
<p>Make ML models for all genes.
The calculation will be performed in parallel using joblib parallel module.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</li>
<li><strong>scaling</strong> (<em>bool</em>) – Whether to scale regulatory gene expression values.</li>
<li><strong>log</strong> (<em>logging object</em>) – log object to output log</li>
<li><strong>verbose</strong> (<em>int</em>) – verbose for joblib parallel</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.fit_genes">
<code class="descname">fit_genes</code><span class="sig-paren">(</span><em>target_genes</em>, <em>bagging_number=200</em>, <em>scaling=True</em>, <em>model_method='bagging_ridge'</em>, <em>save_coefs=False</em>, <em>command_line_mode=False</em>, <em>log=None</em>, <em>alpha=1</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.fit_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Make ML models for genes of interest.
The calculation will be performed in parallel using scikit-learn bagging function.
You can select a modeling method (bagging_ridge or bayesian_ridge).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>target_genes</strong> (<em>list of str</em>) – gene list</li>
<li><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</li>
<li><strong>scaling</strong> (<em>bool</em>) – Whether to scale regulatory gene expression values.</li>
<li><strong>model_method</strong> (<em>str</em>) – ML model name. “bagging_ridge” or “bayesian_ridge”</li>
<li><strong>save_coefs</strong> (<em>bool</em>) – Whether to store details of coef values in bagging model.</li>
<li><strong>command_line_mode</strong> (<em>bool</em>) – Please select False if the calculation is performed on jupyter notebook.</li>
<li><strong>log</strong> (<em>logging object</em>) – log object to output log</li>
<li><strong>alpha</strong> (<em>int</em>) – Strength of regularization.</li>
<li><strong>verbose</strong> (<em>bool</em>) – Whether to show a progress bar.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.plotCoefs">
<code class="descname">plotCoefs</code><span class="sig-paren">(</span><em>target_gene</em>, <em>sort=True</em>, <em>threshold_p=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.plotCoefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of Coef values (network edge weights).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>target_gene</strong> (<em>str</em>) – gene name</li>
<li><strong>sort</strong> (<em>bool</em>) – Whether to sort genes by its strength</li>
<li><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</li>
<li><strong>threshold_p</strong> (<em>float</em>) – the threshold for p-values. TFs will be filtered based on the p-value.
if None, no filtering is applied.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.to_hdf5">
<code class="descname">to_hdf5</code><span class="sig-paren">(</span><em>file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.net’</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.updateLinkList">
<code class="descname">updateLinkList</code><span class="sig-paren">(</span><em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.updateLinkList" title="Permalink to this definition">¶</a></dt>
<dd><p>Update linkList.
LinkList is a data frame that store information about inferred GRN.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>verbose</strong> (<em>bool</em>) – Whether to show a progress bar</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.Net.updateTFinfo_dictionary">
<code class="descname">updateTFinfo_dictionary</code><span class="sig-paren">(</span><em>TFdict</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.updateTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Update TF info matrix</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>TFdict</strong> (<em>dictionary</em>) – A python dictionary in which a key is Target gene, value are potential regulatory genes for the target gene.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="celloracle.load_hdf5">
<code class="descclassname">celloracle.</code><code class="descname">load_hdf5</code><span class="sig-paren">(</span><em>file_path</em>, <em>object_class_name=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.load_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Load an object of celloracle’s custom class that was saved as hdf5.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>file_path</strong> (<em>str</em>) – file_path.</li>
<li><strong>object_class_name</strong> (<em>str</em>) – Types of object.
If it is None, object class will be identified from the extension of file_name.
Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="modules-for-atac-seq-analysis">
<h4>Modules for ATAC-seq analysis<a class="headerlink" href="#modules-for-atac-seq-analysis" title="Permalink to this headline">¶</a></h4>
<blockquote>
<div><div class="toctree-wrapper compound">
<span id="document-modules/celloracle.motif_analysis"></span><div class="section" id="module-celloracle.motif_analysis">
<span id="celloracle-motif-analysis-module"></span><h5>celloracle.motif_analysis module<a class="headerlink" href="#module-celloracle.motif_analysis" title="Permalink to this headline">¶</a></h5>
<p>The <a class="reference internal" href="#module-celloracle.motif_analysis" title="celloracle.motif_analysis"><code class="xref py py-mod docutils literal notranslate"><span class="pre">motif_analysis</span></code></a> module implements transcription factor motif scan.</p>
<p>Genomic activity information (peak of ATAC-seq or Chip-seq) is extracted first.
Then the peak DNA sequence will be subjected to TF motif scan.
Finally we will get list of TFs that potentially binds to a specific gene.</p>
<dl class="function">
<dt id="celloracle.motif_analysis.is_genome_installed">
<code class="descclassname">celloracle.motif_analysis.</code><code class="descname">is_genome_installed</code><span class="sig-paren">(</span><em>ref_genome</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.is_genome_installed" title="Permalink to this definition">¶</a></dt>
<dd><p>Celloracle motif_analysis module uses gimmemotifs and genomepy internally.
Reference genome files should be installed in the PC to use gimmemotifs and genomepy.
This function checks the installation status of the reference genome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>ref_genome</strong> (<em>str</em>) – names of reference genome. i.e., “mm10”, “hg19”</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.motif_analysis.peak2fasta">
<code class="descclassname">celloracle.motif_analysis.</code><code class="descname">peak2fasta</code><span class="sig-paren">(</span><em>peak_ids</em>, <em>ref_genome</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.peak2fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert peak_id into fasta object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>peak_id</strong> (<em>str</em><em> or </em><em>list of str</em>) – Peak_id.  e.g. “chr5_0930303_9499409”
or it can be a list of peak_id.  e.g. [“chr5_0930303_9499409”, “chr11_123445555_123445577”]</li>
<li><strong>ref_genome</strong> (<em>str</em>) – Reference genome name.   e.g. “mm9”, “mm10”, “hg19” etc</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">DNA sequence in fasta format</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">gimmemotifs fasta object</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.motif_analysis.read_bed">
<code class="descclassname">celloracle.motif_analysis.</code><code class="descname">read_bed</code><span class="sig-paren">(</span><em>bed_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.read_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Load bed file and return as dataframe.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>bed_path</strong> (<em>str</em>) – File path.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">bed file in dataframe.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">pandas.dataframe</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.motif_analysis.load_TFinfo_from_parquets">
<code class="descclassname">celloracle.motif_analysis.</code><code class="descname">load_TFinfo_from_parquets</code><span class="sig-paren">(</span><em>folder_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.load_TFinfo_from_parquets" title="Permalink to this definition">¶</a></dt>
<dd><p>Load TFinfo object which was saved with the function; “save_as_parquet”.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>folder_path</strong> (<em>str</em>) – folder path</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Loaded TFinfo object.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="index.html#celloracle.motif_analysis.TFinfo" title="celloracle.motif_analysis.TFinfo">TFinfo</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.motif_analysis.make_TFinfo_from_scanned_file">
<code class="descclassname">celloracle.motif_analysis.</code><code class="descname">make_TFinfo_from_scanned_file</code><span class="sig-paren">(</span><em>path_to_raw_bed</em>, <em>path_to_scanned_result_bed</em>, <em>ref_genome</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.make_TFinfo_from_scanned_file" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is currently an available.</p>
</dd></dl>

<dl class="class">
<dt id="celloracle.motif_analysis.TFinfo">
<em class="property">class </em><code class="descclassname">celloracle.motif_analysis.</code><code class="descname">TFinfo</code><span class="sig-paren">(</span><em>peak_data_frame</em>, <em>ref_genome</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This is a custom class for motif analysis in celloracle.
TFinfo object performs motif scan using the TF motif database in gimmemotifs and several functions of genomepy.
Analysis results can be exported as a python dictionary or dataframe.
These files; python dictionary of dataframe of TF binding information, are needed in GRN inference.</p>
<dl class="attribute">
<dt id="celloracle.motif_analysis.TFinfo.peak_df">
<code class="descname">peak_df</code><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.peak_df" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.dataframe</em> – dataframe about DNA peak and target gene data.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.motif_analysis.TFinfo.all_target_gene">
<code class="descname">all_target_gene</code><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.all_target_gene" title="Permalink to this definition">¶</a></dt>
<dd><p><em>array of str</em> – target genes.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.motif_analysis.TFinfo.ref_genome">
<code class="descname">ref_genome</code><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.ref_genome" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – reference genome name that was used in DNA peak generation.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.motif_analysis.TFinfo.scanned_df">
<code class="descname">scanned_df</code><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.scanned_df" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – Results of motif scan. Key is a peak name. Value is a dataframe of motif scan.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.motif_analysis.TFinfo.dic_targetgene2TFs">
<code class="descname">dic_targetgene2TFs</code><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.dic_targetgene2TFs" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – Final product of Motif scan. Key is a target gene. Value is a list of regulatory candidate genes.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.motif_analysis.TFinfo.dic_peak2Targetgene">
<code class="descname">dic_peak2Targetgene</code><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.dic_peak2Targetgene" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – Dictionary. Key is a peak name. Value is a list of the target gene.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.motif_analysis.TFinfo.dic_TF2targetgenes">
<code class="descname">dic_TF2targetgenes</code><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.dic_TF2targetgenes" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – Final product of Motif scan. Key is a TF. Value is a list of potential target genes of the TF.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.copy">
<code class="descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Deepcoty itself.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.filter_motifs_by_score">
<code class="descname">filter_motifs_by_score</code><span class="sig-paren">(</span><em>threshold</em>, <em>method='cumlative_score'</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.filter_motifs_by_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove motifs with low binding scores.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>method</strong> (<em>str</em>) – thresholding method. Select either of [“indivisual_score”, “cumlative_score”]</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.filter_peaks">
<code class="descname">filter_peaks</code><span class="sig-paren">(</span><em>peaks_to_be_remained</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.filter_peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter peaks.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>peaks_to_be_remained</strong> (<em>array of str</em>) – list of peaks. Peaks that are NOT in the list will be removed.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.make_TFinfo_dataframe_and_dictionary">
<code class="descname">make_TFinfo_dataframe_and_dictionary</code><span class="sig-paren">(</span><em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.make_TFinfo_dataframe_and_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the final step of motif_analysis.
Convert scanned results into a data frame and dictionaries.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>verbose</strong> (<em>bool</em>) – Whether to show a progress bar.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.reset_dictionary_and_df">
<code class="descname">reset_dictionary_and_df</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.reset_dictionary_and_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset TF dictionary and TF dataframe.
The following attributes will be erased; TF_onehot, dic_targetgene2TFs, dic_peak2Targetgene, dic_TF2targetgenes.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.reset_filtering">
<code class="descname">reset_filtering</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.reset_filtering" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset filtering information.
You can use this function to stat over the filtering step with new conditions.
The following attributes will be erased; TF_onehot, dic_targetgene2TFs, dic_peak2Targetgene, dic_TF2targetgenes.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.save_as_parquet">
<code class="descname">save_as_parquet</code><span class="sig-paren">(</span><em>folder_path=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.save_as_parquet" title="Permalink to this definition">¶</a></dt>
<dd><p>Save itself. Some attributes are saved as parquet file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>folder_path</strong> (<em>str</em>) – folder path</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.scan">
<code class="descname">scan</code><span class="sig-paren">(</span><em>background_length=200</em>, <em>fpr=0.02</em>, <em>n_cpus=-1</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.scan" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan DNA sequences searching for TF binding motifs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>background_length</strong> (<em>int</em>) – background length. This is used for the calculation of the binding score.</li>
<li><strong>fpr</strong> (<em>float</em>) – False positive rate for motif identification.</li>
<li><strong>n_cpus</strong> (<em>int</em>) – number of CPUs for parallel calculation.</li>
<li><strong>verbose</strong> (<em>bool</em>) – Whether to show a progress bar.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.to_dataframe">
<code class="descname">to_dataframe</code><span class="sig-paren">(</span><em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Return results as a dataframe.
Rows are peak_id, and columns are TFs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>verbose</strong> (<em>bool</em>) – Whether to show a progress bar.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">TFinfo matrix.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">pandas.dataframe</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.to_dictionary">
<code class="descname">to_dictionary</code><span class="sig-paren">(</span><em>dictionary_type='targetgene2TFs'</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.to_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Return TF information as a python dictionary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dictionary_type</strong> (<em>str</em>) – Type of dictionary. Select from [“targetgene2TFs”, “TF2targetgenes”].
If you chose “targetgene2TFs”, it returns a dictionary in which a key is a target gene, and a value is a list of regulatory candidate genes (TFs) of the target.
If you chose “TF2targetgenes”, it returns a dictionary in which a key is a TF and a value is a list of potential target genes of the TF.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">dictionary.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dictionary</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.motif_analysis.TFinfo.to_hdf5">
<code class="descname">to_hdf5</code><span class="sig-paren">(</span><em>file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.TFinfo.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.tfinfo’</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="celloracle.motif_analysis.get_tss_info">
<code class="descclassname">celloracle.motif_analysis.</code><code class="descname">get_tss_info</code><span class="sig-paren">(</span><em>peak_str_list</em>, <em>ref_genome</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.get_tss_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Get annotation about Transcription Starting Site (TSS).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>peak_str_list</strong> (<em>list of str</em>) – list of peak_id. e.g., [“chr5_0930303_9499409”, “chr11_123445555_123445577”]</li>
<li><strong>ref_genome</strong> (<em>str</em>) – reference genome name.</li>
<li><strong>verbose</strong> (<em>bool</em>) – verbosity.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.motif_analysis.integrate_tss_peak_with_cicero">
<code class="descclassname">celloracle.motif_analysis.</code><code class="descname">integrate_tss_peak_with_cicero</code><span class="sig-paren">(</span><em>tss_peak</em>, <em>cicero_connections</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.motif_analysis.integrate_tss_peak_with_cicero" title="Permalink to this definition">¶</a></dt>
<dd><p>Process output of cicero data and returns DNA peak information for motif analysis in celloracle.
Please see the tutorial of celloracle documentation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>tss_peak</strong> (<em>pandas.dataframe</em>) – dataframe about TSS information. Please use the function, “get_tss_info” to get this dataframe.</li>
<li><strong>cicero_connections</strong> (<em>dataframe</em>) – dataframe that stores the results of cicero analysis.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">DNA peak about promoter/enhancer and its annotation about target gene.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pandas.dataframe</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>
</div></blockquote>
</div>
<div class="section" id="modules-for-network-analysis">
<h4>Modules for Network analysis<a class="headerlink" href="#modules-for-network-analysis" title="Permalink to this headline">¶</a></h4>
<blockquote>
<div><div class="toctree-wrapper compound">
<span id="document-modules/celloracle.network_analysis"></span><div class="section" id="module-celloracle.network_analysis">
<span id="celloracle-network-analysis-module"></span><h5>celloracle.network_analysis module<a class="headerlink" href="#module-celloracle.network_analysis" title="Permalink to this headline">¶</a></h5>
<p>The <a class="reference internal" href="#module-celloracle.network_analysis" title="celloracle.network_analysis"><code class="xref py py-mod docutils literal notranslate"><span class="pre">network_analysis</span></code></a> module implements Network analysis.</p>
<dl class="function">
<dt id="celloracle.network_analysis.get_links">
<code class="descclassname">celloracle.network_analysis.</code><code class="descname">get_links</code><span class="sig-paren">(</span><em>oracle_object</em>, <em>cluster_name_for_GRN_unit=None</em>, <em>alpha=10</em>, <em>bagging_number=20</em>, <em>verbose_level=1</em>, <em>test_mode=False</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.get_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Make GRN for each cluster and returns results as a Links object.
Several preprocessing should be done before using this function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>oracle_object</strong> (<a class="reference internal" href="index.html#celloracle.Oracle" title="celloracle.Oracle"><em>Oracle</em></a>) – See Oracle module for detail.</li>
<li><strong>cluster_name_for_GRN_unit</strong> (<em>str</em>) – Cluster name for GRN calculation. The cluster information should be stored in Oracle.adata.obs.</li>
<li><strong>alpha</strong> (<em>float</em><em> or </em><em>int</em>) – the strength of regularization.
If you set a lower value, the sensitivity increase, and you can detect a weak network connection, but it might get more noize.
With a higher value of alpha may reduce the chance of overfitting.</li>
<li><strong>bagging_number</strong> (<em>int</em>) – The number for bagging calculation.</li>
<li><strong>verbose_level</strong> (<em>int</em>) – if [verbose_level&gt;1], most detailed progress information will be shown.
if [verbose_level &gt; 0], one progress bar will be shown.
if [verbose_level == 0], no progress bar will be shown.</li>
<li><strong>test_mode</strong> (<em>bool</em>) – If test_mode is True, GRN calculation will be done for only one cluster rather than all clusters.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.network_analysis.test_R_libraries_installation">
<code class="descclassname">celloracle.network_analysis.</code><code class="descname">test_R_libraries_installation</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.test_R_libraries_installation" title="Permalink to this definition">¶</a></dt>
<dd><p>CellOracle.network_analysis use some R libraries for network analysis.
This is a test function to check instalation of necessary R libraries.</p>
</dd></dl>

<dl class="function">
<dt id="celloracle.network_analysis.load_links">
<code class="descclassname">celloracle.network_analysis.</code><code class="descname">load_links</code><span class="sig-paren">(</span><em>file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.load_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Load links object saved as a hdf5 file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>file_path</strong> (<em>str</em>) – file path.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">loaded links object.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links">Links</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="celloracle.network_analysis.Links">
<em class="property">class </em><code class="descclassname">celloracle.network_analysis.</code><code class="descname">Links</code><span class="sig-paren">(</span><em>name</em>, <em>links_dict={}</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This is a class for the processing and visualization of GRNs.
Links object stores cluster-specific GRNs and metadata.
Please use “get_links” function in Oracle object to generate Links object.</p>
<dl class="attribute">
<dt id="celloracle.network_analysis.Links.links_dict">
<code class="descname">links_dict</code><a class="headerlink" href="#celloracle.network_analysis.Links.links_dict" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – Dictionary that store unprocessed network data.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network_analysis.Links.filtered_links">
<code class="descname">filtered_links</code><a class="headerlink" href="#celloracle.network_analysis.Links.filtered_links" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dictionary</em> – Dictionary that store filtered network data.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network_analysis.Links.merged_score">
<code class="descname">merged_score</code><a class="headerlink" href="#celloracle.network_analysis.Links.merged_score" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.dataframe</em> – Network scores.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network_analysis.Links.cluster">
<code class="descname">cluster</code><a class="headerlink" href="#celloracle.network_analysis.Links.cluster" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list of str</em> – List of cluster name.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network_analysis.Links.name">
<code class="descname">name</code><a class="headerlink" href="#celloracle.network_analysis.Links.name" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – Name of clustering unit.</p>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network_analysis.Links.palette">
<code class="descname">palette</code><a class="headerlink" href="#celloracle.network_analysis.Links.palette" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.dataframe</em> – DataFrame that store color information.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.add_palette">
<code class="descname">add_palette</code><span class="sig-paren">(</span><em>cluster_name_list</em>, <em>color_list</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.add_palette" title="Permalink to this definition">¶</a></dt>
<dd><p>Add color information for each cluster.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cluster_name_list</strong> (<em>list of str</em>) – cluster names.</li>
<li><strong>color_list</strong> (<em>list of stf</em>) – list of color for each cluster.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.filter_links">
<code class="descname">filter_links</code><span class="sig-paren">(</span><em>p=0.001</em>, <em>weight='coef_abs'</em>, <em>thread_number=10000</em>, <em>genelist_source=None</em>, <em>genelist_target=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.filter_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter network edges.
In most case inferred GRN has non-significant random edges.
We have to remove such random edges before analyzing network structure.
You can do the filtering either of three ways below.</p>
<ol class="arabic simple">
<li>Do filtering based on the p-value of the network edge.
Please enter p-value for thresholding.</li>
<li>Do filtering based on network edge number.
If you set the number, network edges will be filtered based on the order of a network score. The top n-th network edges with network weight will remain, and the other edges will be removed.
The network data have several types of network weight, so you have to select which network weight do you want to use.</li>
<li>Do filtering based on an arbitrary gene list. You can set a gene list for source nodes or target nodes.</li>
</ol>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>p</strong> (<em>float</em>) – threshold for p-value of the network edge.</li>
<li><strong>weight</strong> (<em>str</em>) – Please select network weight name for the filtering</li>
<li><strong>genelist_source</strong> (<em>list of str</em>) – gene list to remain in regulatory gene nodes. Default is None.</li>
<li><strong>genelist_target</strong> (<em>list of str</em>) – gene list to remain in target gene nodes. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.get_network_entropy">
<code class="descname">get_network_entropy</code><span class="sig-paren">(</span><em>value='coef_abs'</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.get_network_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate network entropy scores.
Please select a type of network weight for the entropy calculation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>value</strong> (<em>str</em>) – Default is “coef_abs”.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.get_score">
<code class="descname">get_score</code><span class="sig-paren">(</span><em>test_mode=False</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.get_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Get several network sores using R libraries.
Make sure all dependant R libraries are installed in your environment before running this function.
You can check the installation for the R libraries by running test_installation() in network_analysis module.</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.plot_cartography_scatter_per_cluster">
<code class="descname">plot_cartography_scatter_per_cluster</code><span class="sig-paren">(</span><em>gois=None</em>, <em>clusters=None</em>, <em>scatter=False</em>, <em>kde=True</em>, <em>auto_gene_annot=False</em>, <em>percentile=98</em>, <em>args_dot={}</em>, <em>args_line={}</em>, <em>args_annot={}</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_cartography_scatter_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a plot of gene network cartography.
Please read the original paper of gene network cartography for the principle of gene network cartography.
<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>gois</strong> (<em>list of srt</em>) – List of Gene name to highlight.</li>
<li><strong>clusters</strong> (<em>list of str</em>) – List of cluster name to analyze. If None, all clusters in Links object will be analyzed.</li>
<li><strong>scatter</strong> (<em>bool</em>) – Whether to make a scatter plot.</li>
<li><strong>auto_gene_annot</strong> (<em>bool</em>) – Whether to pick up genes to make an annotation.</li>
<li><strong>percentile</strong> (<em>float</em>) – Genes with a network score above the percentile will be shown with annotation. Default is 98.</li>
<li><strong>args_dot</strong> (<em>dictionary</em>) – Arguments for scatter plot.</li>
<li><strong>args_line</strong> (<em>dictionary</em>) – Arguments for lines in cartography plot.</li>
<li><strong>args_annot</strong> (<em>dictionary</em>) – Arguments for annotation in plots.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.plot_cartography_term">
<code class="descname">plot_cartography_term</code><span class="sig-paren">(</span><em>goi</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_cartography_term" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the summary of gene network cartography like a heatmap.
Please read the original paper of gene network cartography for the principle of gene network cartography.
<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>gois</strong> (<em>list of srt</em>) – List of Gene name to highlight.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.plot_degree_distributions">
<code class="descname">plot_degree_distributions</code><span class="sig-paren">(</span><em>plot_model=False</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_degree_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of network degree (the number of edge per gene).
The network degree will be visualized in both linear scale and log scale.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>plot_model</strong> (<em>bool</em>) – Whether to plot linear approximation line.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.plot_network_entropy_distributions">
<code class="descname">plot_network_entropy_distributions</code><span class="sig-paren">(</span><em>update_network_entropy=False</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_network_entropy_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of network entropy.
See the CellOracle paper for the detail.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<em>Links object</em>) – See network_analisis.Links class for detail.</li>
<li><strong>values</strong> (<em>list of str</em>) – The list of score to visualize. If it is None, all network score (listed above) will be used.</li>
<li><strong>update_network_entropy</strong> (<em>bool</em>) – Whether to recalculate network entropy.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.plot_score_comparison_2D">
<code class="descname">plot_score_comparison_2D</code><span class="sig-paren">(</span><em>value</em>, <em>cluster1</em>, <em>cluster2</em>, <em>percentile=99</em>, <em>annot_shifts=None</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_score_comparison_2D" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a scatter plot that shows the relationship of a specific network score in two groups.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>value</strong> (<em>srt</em>) – The network score type.</li>
<li><strong>cluster1</strong> (<em>str</em>) – Cluster name. Network scores in the cluster1 will be visualized in the x-axis.</li>
<li><strong>cluster2</strong> (<em>str</em>) – Cluster name. Network scores in the cluster2 will be visualized in the y-axis.</li>
<li><strong>percentile</strong> (<em>float</em>) – Genes with a network score above the percentile will be shown with annotation. Default is 99.</li>
<li><strong>annot_shifts</strong> (<em>(</em><em>float</em><em>, </em><em>float</em><em>)</em>) – Annotation visualization setting.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.plot_score_discributions">
<code class="descname">plot_score_discributions</code><span class="sig-paren">(</span><em>values=None</em>, <em>method='boxplot'</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_score_discributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of network scores.
An individual data point is a network edge (gene) of GRN in each cluster.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See Links class for detail.</li>
<li><strong>values</strong> (<em>list of str</em>) – The list of score to visualize. If it is None, all network score will be used.</li>
<li><strong>method</strong> (<em>str</em>) – Plotting method. Select either of “boxplot” or “barplot”.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.plot_score_per_cluster">
<code class="descname">plot_score_per_cluster</code><span class="sig-paren">(</span><em>goi</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_score_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot network score for a gene.
This function visualizes the network score of a specific gene between clusters to get an insight into the dynamics of the gene.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>goi</strong> (<em>srt</em>) – Gene name.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.plot_scores_as_rank">
<code class="descname">plot_scores_as_rank</code><span class="sig-paren">(</span><em>cluster</em>, <em>n_gene=50</em>, <em>save=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_scores_as_rank" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up top n-th genes wich high-network scores and make plots.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analisis.Links class for detail.</li>
<li><strong>cluster</strong> (<em>str</em>) – Cluster nome to analyze.</li>
<li><strong>n_gene</strong> (<em>int</em>) – Number of genes to plot. Default is 50.</li>
<li><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="celloracle.network_analysis.Links.to_hdf5">
<code class="descname">to_hdf5</code><span class="sig-paren">(</span><em>file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.links’</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="celloracle.network_analysis.transfer_scores_from_links_to_adata">
<code class="descclassname">celloracle.network_analysis.</code><code class="descname">transfer_scores_from_links_to_adata</code><span class="sig-paren">(</span><em>adata</em>, <em>links</em>, <em>method='median'</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.transfer_scores_from_links_to_adata" title="Permalink to this definition">¶</a></dt>
<dd><p>Transfer the summary of network scores (median or mean) per group from Links object into adata.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>adata</strong> (<em>anndata</em>) – anndata</li>
<li><strong>links</strong> (<a class="reference internal" href="index.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – likns object</li>
<li><strong>method</strong> (<em>str</em>) – The method to summarize data.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.network_analysis.linkList_to_networkgraph">
<code class="descclassname">celloracle.network_analysis.</code><code class="descname">linkList_to_networkgraph</code><span class="sig-paren">(</span><em>filteredlinkList</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.linkList_to_networkgraph" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert linkList into Graph object in NetworkX.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filteredlinkList</strong> (<em>pandas.DataFrame</em>) – GRN saved as linkList.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Network X graph objenct.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">Graph object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.network_analysis.draw_network">
<code class="descclassname">celloracle.network_analysis.</code><code class="descname">draw_network</code><span class="sig-paren">(</span><em>linkList</em>, <em>return_graph=False</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.draw_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot network graph.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>linkList</strong> (<em>pandas.DataFrame</em>) – GRN saved as linkList.</li>
<li><strong>return_graph</strong> (<em>bool</em>) – Whether to return graph object.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Network X graph objenct.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">Graph object</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>
</div></blockquote>
</div>
<div class="section" id="other-modules">
<h4>Other modules<a class="headerlink" href="#other-modules" title="Permalink to this headline">¶</a></h4>
<blockquote>
<div><div class="toctree-wrapper compound">
<span id="document-modules/celloracle.go_analysis"></span><div class="section" id="celloracle-go-analysis-module">
<h5>celloracle.go_analysis module<a class="headerlink" href="#celloracle-go-analysis-module" title="Permalink to this headline">¶</a></h5>
<blockquote>
<div><span class="target" id="module-celloracle.go_analysis"></span><p>The <a class="reference internal" href="#module-celloracle.go_analysis" title="celloracle.go_analysis"><code class="xref py py-mod docutils literal notranslate"><span class="pre">go_analysis</span></code></a> module implements Gene Ontology analysis.
This module use goatools internally.</p>
<dl class="function">
<dt id="celloracle.go_analysis.geneSymbol2ID">
<code class="descclassname">celloracle.go_analysis.</code><code class="descname">geneSymbol2ID</code><span class="sig-paren">(</span><em>symbols</em>, <em>species='mouse'</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.go_analysis.geneSymbol2ID" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert gene symbol into Entrez gene id.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>symbols</strong> (<em>array of str</em>) – gene symbol</li>
<li><strong>species</strong> (<em>str</em>) – Select species. Either “mouse” or “human”</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Entrez gene id</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list of str</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.go_analysis.geneID2Symbol">
<code class="descclassname">celloracle.go_analysis.</code><code class="descname">geneID2Symbol</code><span class="sig-paren">(</span><em>IDs</em>, <em>species='mouse'</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.go_analysis.geneID2Symbol" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert Entrez gene id into gene symbol.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>IDs</strong> (<em>array of str</em>) – Entrez gene id.</li>
<li><strong>species</strong> (<em>str</em>) – Select species. Either “mouse” or “human”.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Gene symbol</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list of str</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.go_analysis.get_GO">
<code class="descclassname">celloracle.go_analysis.</code><code class="descname">get_GO</code><span class="sig-paren">(</span><em>gene_query</em>, <em>species='mouse'</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.go_analysis.get_GO" title="Permalink to this definition">¶</a></dt>
<dd><p>Get Gene Ontologies (GOs).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gene_query</strong> (<em>array of str</em>) – gene list.</li>
<li><strong>species</strong> (<em>str</em>) – Select species. Either “mouse” or “human”</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">GO analysis results as dataframe.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pandas.dataframe</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div></blockquote>
</div>
<span id="document-modules/celloracle.utility"></span><div class="section" id="celloracle-utility-module">
<h5>celloracle.utility module<a class="headerlink" href="#celloracle-utility-module" title="Permalink to this headline">¶</a></h5>
<blockquote>
<div><span class="target" id="module-celloracle.utility"></span><p>The <a class="reference internal" href="#module-celloracle.utility" title="celloracle.utility"><code class="xref py py-mod docutils literal notranslate"><span class="pre">utility</span></code></a> module has several functions that support celloracle.</p>
<dl class="class">
<dt id="celloracle.utility.makelog">
<em class="property">class </em><code class="descclassname">celloracle.utility.</code><code class="descname">makelog</code><span class="sig-paren">(</span><em>file_name=None</em>, <em>directory=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.makelog" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This is a class for making log.</p>
<dl class="method">
<dt id="celloracle.utility.makelog.info">
<code class="descname">info</code><span class="sig-paren">(</span><em>comment</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.makelog.info" title="Permalink to this definition">¶</a></dt>
<dd><p>Add comment into the log file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>comment</strong> (<em>str</em>) – comment.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="celloracle.utility.save_as_pickled_object">
<code class="descclassname">celloracle.utility.</code><code class="descname">save_as_pickled_object</code><span class="sig-paren">(</span><em>obj</em>, <em>filepath</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.save_as_pickled_object" title="Permalink to this definition">¶</a></dt>
<dd><p>Save any object using pickle.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>obj</strong> (<em>any python object</em>) – python object.</li>
<li><strong>filepath</strong> (<em>str</em>) – file path.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.utility.load_pickled_object">
<code class="descclassname">celloracle.utility.</code><code class="descname">load_pickled_object</code><span class="sig-paren">(</span><em>filepath</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.load_pickled_object" title="Permalink to this definition">¶</a></dt>
<dd><p>Load pickled object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filepath</strong> (<em>str</em>) – file path.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">loaded object.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">python object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.utility.intersect">
<code class="descclassname">celloracle.utility.</code><code class="descname">intersect</code><span class="sig-paren">(</span><em>list1</em>, <em>list2</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.intersect" title="Permalink to this definition">¶</a></dt>
<dd><p>Intersect two list and get components that exists in both list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>list1</strong> (<em>list</em>) – input list.</li>
<li><strong>list2</strong> (<em>list</em>) – input list.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">intersected list.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.utility.exec_process">
<code class="descclassname">celloracle.utility.</code><code class="descname">exec_process</code><span class="sig-paren">(</span><em>commands</em>, <em>message=True</em>, <em>wait_finished=True</em>, <em>return_process=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.exec_process" title="Permalink to this definition">¶</a></dt>
<dd><p>Excute a command. This is a wrapper of “subprocess.Popen”</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>commands</strong> (<em>str</em>) – command.</li>
<li><strong>message</strong> (<em>bool</em>) – Whether to return a message or not.</li>
<li><strong>wait_finished</strong> (<em>bool</em>) – Whether to wait for the process finished. If False, the function finish immediately.</li>
<li><strong>return_process</strong> (<em>bool</em>) – Whether to return “process”.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.utility.standard">
<code class="descclassname">celloracle.utility.</code><code class="descname">standard</code><span class="sig-paren">(</span><em>df</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.standard" title="Permalink to this definition">¶</a></dt>
<dd><p>Standerdize value.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>df</strong> (<em>padas.dataframe</em>) – dataframe.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">data after standerdization.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">pandas.dataframe</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.utility.load_hdf5">
<code class="descclassname">celloracle.utility.</code><code class="descname">load_hdf5</code><span class="sig-paren">(</span><em>file_path</em>, <em>object_class_name=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.load_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Load an object of celloracle’s custom class that was saved as hdf5.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>file_path</strong> (<em>str</em>) – file_path.</li>
<li><strong>object_class_name</strong> (<em>str</em>) – Types of object.
If it is None, object class will be identified from the extension of file_name.
Default is None.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="celloracle.utility.inverse_dictionary">
<code class="descclassname">celloracle.utility.</code><code class="descname">inverse_dictionary</code><span class="sig-paren">(</span><em>dictionary</em>, <em>verbose=True</em>, <em>return_value_as_numpy=False</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.utility.inverse_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Make inversed dictionary.
See examples below for detail.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>dictionary</strong> (<em>dict</em>) – python dictionary</li>
<li><strong>verbose</strong> (<em>bool</em>) – Whether to show progress bar.</li>
<li><strong>return_value_as_numpy</strong> (<em>bool</em>) – Whether to convert values into numpy array.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Python dictionary.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dict</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dic</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;a&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;b&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">inverse_dictionary</span><span class="p">(</span><span class="n">dic</span><span class="p">)</span>
<span class="go">{1: [&#39;a&#39;], 2: [&#39;a&#39;, &#39;b&#39;], 3: [&#39;a&#39;, &#39;b&#39;], 4: [&#39;b&#39;]}</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dic</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;a&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;b&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">inverse_dictionary</span><span class="p">(</span><span class="n">dic</span><span class="p">,</span> <span class="n">return_value_as_numpy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">{1: array([&#39;a&#39;], dtype=&#39;&lt;U1&#39;),</span>
<span class="go"> 2: array([&#39;a&#39;, &#39;b&#39;], dtype=&#39;&lt;U1&#39;),</span>
<span class="go"> 3: array([&#39;a&#39;, &#39;b&#39;], dtype=&#39;&lt;U1&#39;),</span>
<span class="go"> 4: array([&#39;b&#39;], dtype=&#39;&lt;U1&#39;)}</span>
</pre></div>
</div>
</dd></dl>

</div></blockquote>
</div>
<span id="document-modules/celloracle.data"></span><div class="section" id="celloracle-data-module">
<h5>celloracle.data module<a class="headerlink" href="#celloracle-data-module" title="Permalink to this headline">¶</a></h5>
<blockquote>
<div><span class="target" id="module-celloracle.data"></span><p>The <a class="reference internal" href="#module-celloracle.data" title="celloracle.data"><code class="xref py py-mod docutils literal notranslate"><span class="pre">data</span></code></a> module implements data download and loading.</p>
<dl class="function">
<dt id="celloracle.data.load_TFinfo_df_mm9_mouse_atac_atlas">
<code class="descclassname">celloracle.data.</code><code class="descname">load_TFinfo_df_mm9_mouse_atac_atlas</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.data.load_TFinfo_df_mm9_mouse_atac_atlas" title="Permalink to this definition">¶</a></dt>
<dd><p>Load Transcription factor binding information made from mouse scATAC-seq atlas dataset.
mm9 genome was used for the reference genome.</p>
<p>Args:</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">TF binding info.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">pandas.dataframe</td>
</tr>
</tbody>
</table>
</dd></dl>

</div></blockquote>
</div>
<span id="document-modules/celloracle.data_conversion"></span><div class="section" id="celloracle-data-conversion-module">
<h5>celloracle.data_conversion module<a class="headerlink" href="#celloracle-data-conversion-module" title="Permalink to this headline">¶</a></h5>
<blockquote>
<div><span class="target" id="module-celloracle.data_conversion"></span><p>The <a class="reference internal" href="#module-celloracle.data_conversion" title="celloracle.data_conversion"><code class="xref py py-mod docutils literal notranslate"><span class="pre">data_conversion</span></code></a> module implements data conversion between different platform.</p>
<dl class="function">
<dt id="celloracle.data_conversion.seurat_object_to_anndata">
<code class="descclassname">celloracle.data_conversion.</code><code class="descname">seurat_object_to_anndata</code><span class="sig-paren">(</span><em>file_path_seurat_object</em>, <em>delete_tmp_file=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.data_conversion.seurat_object_to_anndata" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert seurat object into anndata.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>file_path_seurat_object</strong> (<em>str</em>) – File path of seurat object. Seurat object should be saved as Rds format.</li>
<li><strong>delete_tmp_file</strong> (<em>bool</em>) – Whether to delete temporary file.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">anndata object.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">anndata</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div></blockquote>
</div>
</div>
</div></blockquote>
</div>
</div>
</div>
</div>
</div>
<div class="toctree-wrapper compound">
<span id="document-changelog"></span><div class="section" id="changelog">
<span id="id1"></span><h2>Changelog<a class="headerlink" href="#changelog" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="toctree-wrapper compound">
<span id="document-license/index"></span><div class="section" id="license">
<span id="id1"></span><h2>License<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h2>
<p>The software is provided under xxxxxxx License.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">License</span> <span class="n">xxxxxxxx</span>

<span class="n">Copyright</span> <span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="mi">2019</span> <span class="n">Kenji</span> <span class="n">Kamimoto</span><span class="p">,</span> <span class="n">Christy</span> <span class="n">Hoffmann</span><span class="p">,</span> <span class="n">Samantha</span> <span class="n">Morris</span>


<span class="n">This</span> <span class="ow">is</span> <span class="n">a</span> <span class="n">placeholder</span><span class="o">.</span> <span class="n">License</span> <span class="n">description</span> <span class="n">will</span> <span class="n">be</span> <span class="n">shown</span> <span class="n">here</span><span class="o">.</span>
</pre></div>
</div>
</div>
</div>
<div class="toctree-wrapper compound">
<span id="document-citation/index"></span><div class="section" id="authors-and-citations">
<span id="citing"></span><h2>Authors and citations<a class="headerlink" href="#authors-and-citations" title="Permalink to this headline">¶</a></h2>
<div class="section" id="cite-celloracle">
<h3>Cite celloracle<a class="headerlink" href="#cite-celloracle" title="Permalink to this headline">¶</a></h3>
<p>If you use velocyto please cite our bioarxiv preprint <code class="docutils literal notranslate"><span class="pre">TITLE</span></code>.</p>
</div>
<div class="section" id="celloracle-software-development">
<h3>celloracle software development<a class="headerlink" href="#celloracle-software-development" title="Permalink to this headline">¶</a></h3>
<p>celloracle is developed and maintained by the effort of celloracle team in Samantha Morris Lab.
Please post troubles or questions on  <a class="reference external" href="https://cole-trapnell-lab.github.io/cicero-release/">the Github repository</a></p>
</div>
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Samantha Morris Lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>