

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>celloracle.chip_analysis.integrate_chip_data_with_grn_results &mdash; celloracle 0.1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'../../../',
              VERSION:'0.1.0',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Welcome to celloracleâ€™s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>celloracle.chip_analysis.integrate_chip_data_with_grn_results</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for celloracle.chip_analysis.integrate_chip_data_with_grn_results</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">This file contains custom functions for the analysis of ATAC-seq data.</span>
<span class="sd">Genomic activity information (peak of ATAC-seq) will be extracted first.</span>
<span class="sd">Then the peak DNA sequence will be subjected to TF motif scan.</span>
<span class="sd">Finally we will get list of TFs that potentially binds to a specific gene.</span>

<span class="sd">Codes were written by Kenji Kamimoto.</span>


<span class="sd">&#39;&#39;&#39;</span>

<span class="c1">###########################</span>
<span class="c1">### 0. Import libralies ###</span>
<span class="c1">###########################</span>


<span class="c1"># 0.1. libraries for fundamental data science and data processing</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>

<span class="c1">##</span>
<span class="kn">from</span> <span class="nn">..utility</span> <span class="k">import</span> <span class="n">exec_process</span>
<span class="kn">from</span> <span class="nn">..</span> <span class="k">import</span> <span class="n">network</span> <span class="k">as</span> <span class="n">tn</span>

<span class="c1">#########################</span>
<span class="c1">### prepare chip data ###</span>
<span class="c1">#########################</span>


<span class="nd">@np</span><span class="o">.</span><span class="n">vectorize</span>
<span class="k">def</span> <span class="nf">intersect_chip_bed_with_atac_bed</span><span class="p">(</span><span class="n">base_bed_file_path</span><span class="p">,</span> <span class="n">chip_bed_file_path</span><span class="p">):</span>

    <span class="n">file_name</span> <span class="o">=</span> <span class="n">chip_bed_file_path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">GEO_name</span> <span class="o">=</span> <span class="n">file_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">output_file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">chip_bed_file_path</span><span class="p">)</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;/</span><span class="si">{GEO_name}</span><span class="s2">_intersected.bed&quot;</span>

    <span class="n">command</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;bedtools intersect -a </span><span class="si">{base_bed_file_path}</span><span class="s2"> -b </span><span class="si">{chip_bed_file_path}</span><span class="s2"> -wa &gt; </span><span class="si">{output_file_path}</span><span class="s2">&quot;</span>
    <span class="n">exec_process</span><span class="p">(</span><span class="n">commands</span><span class="o">=</span><span class="n">command</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">wait_finished</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;processed: </span><span class="si">{file_name}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">output_file_path</span>
<span class="n">intersect_chip_bed_with_atac_bed</span><span class="o">.</span><span class="n">excluded</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>


<div class="viewcode-block" id="load_and_merge_chip_bed_files"><a class="viewcode-back" href="../../../modules/celloracle.chip_analysis.html#celloracle.chip_analysis.integrate_chip_data_with_grn_results.load_and_merge_chip_bed_files">[docs]</a><span class="k">def</span> <span class="nf">load_and_merge_chip_bed_files</span><span class="p">(</span><span class="n">base_bed_file_path</span><span class="p">,</span>
                                  <span class="n">chip_bed_file_paths</span><span class="p">,</span>
                                  <span class="n">chip_type</span><span class="p">,</span>
                                  <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>


    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;1/3 loading base atac bed file...&quot;</span><span class="p">)</span>
    <span class="n">merged</span> <span class="o">=</span> <span class="n">_read_peaks</span><span class="p">(</span><span class="n">base_bed_file_path</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;2/3 processing chip files...&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chip_bed_file_path</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">chip_bed_file_paths</span><span class="p">):</span>


        <span class="n">file_name</span> <span class="o">=</span> <span class="n">chip_bed_file_path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">GEO_name</span> <span class="o">=</span> <span class="n">file_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sample_name</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;</span><span class="si">{GEO_name}</span><span class="s2">_</span><span class="si">{chip_type}</span><span class="s2">&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot; processing </span><span class="si">{file_name}</span><span class="s2">...&quot;</span><span class="p">)</span>

        <span class="n">chip_bed_df</span> <span class="o">=</span> <span class="n">_read_peaks</span><span class="p">(</span><span class="n">chip_bed_file_path</span><span class="p">)</span>
        <span class="n">chip_bed_df</span><span class="p">[</span><span class="n">sample_name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">chip_bed_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot; empty data&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">merged</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">merged</span><span class="p">,</span> <span class="n">chip_bed_df</span><span class="p">[[</span><span class="n">sample_name</span><span class="p">]]],</span>
                           <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;3/3 merging files...&quot;</span><span class="p">)</span>
    <span class="n">merged</span> <span class="o">=</span> <span class="n">merged</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">merged</span> <span class="o">=</span> <span class="n">merged</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;peak_chr&quot;</span><span class="p">,</span> <span class="s2">&quot;peak_start&quot;</span><span class="p">,</span> <span class="s2">&quot;peak_end&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">save_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">save_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">chip_bed_file_paths</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;/processed_chip_data.parquet&quot;</span>

    <span class="n">merged</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="n">save_path</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;processed file was saved in </span><span class="se">\n</span><span class="si">{save_path}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">merged</span></div>

<span class="k">def</span> <span class="nf">_read_peaks</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;peak_chr&quot;</span><span class="p">,</span> <span class="s2">&quot;peak_start&quot;</span><span class="p">,</span> <span class="s2">&quot;peak_end&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">tmp</span>

    <span class="n">tmp</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;peak_chr&quot;</span><span class="p">,</span> <span class="s2">&quot;peak_start&quot;</span><span class="p">,</span> <span class="s2">&quot;peak_end&quot;</span><span class="p">]</span>
    <span class="n">ind</span> <span class="o">=</span> <span class="p">[</span><span class="n">tmp</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;peak_chr&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span>
           <span class="nb">str</span><span class="p">(</span><span class="n">tmp</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;peak_start&quot;</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span>
           <span class="nb">str</span><span class="p">(</span><span class="n">tmp</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;peak_end&quot;</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tmp</span><span class="p">))]</span>
    <span class="n">tmp</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">ind</span>

    <span class="c1"># remove duplicated values</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="o">~</span><span class="n">tmp</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)]</span>

    <span class="k">return</span> <span class="n">tmp</span>

<span class="c1">#####################################################</span>
<span class="c1">### integrate Chip data and GRN inference results ###</span>
<span class="c1">###  fucusing on a TF                             ###</span>
<span class="c1">#####################################################</span>

<div class="viewcode-block" id="integrate_Chip_and_TN"><a class="viewcode-back" href="../../../modules/celloracle.chip_analysis.html#celloracle.chip_analysis.integrate_chip_data_with_grn_results.integrate_Chip_and_TN">[docs]</a><span class="k">def</span> <span class="nf">integrate_Chip_and_TN</span><span class="p">(</span><span class="n">tn_object</span><span class="p">,</span> <span class="n">Chip_data</span><span class="p">,</span> <span class="n">TF_of_interest</span><span class="p">,</span>
                          <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;gene_mode&quot;</span><span class="p">):</span>

    <span class="c1"># version190605</span>


    <span class="c1"># extract grn data</span>
    <span class="n">grnDF_mean</span> <span class="o">=</span> <span class="n">tn</span><span class="o">.</span><span class="n">getDF_peakxTF</span><span class="p">(</span><span class="n">tn_object</span><span class="p">,</span> <span class="s2">&quot;coef_abs&quot;</span><span class="p">)</span>
    <span class="n">grnDF_ps</span> <span class="o">=</span> <span class="n">tn</span><span class="o">.</span><span class="n">getDF_peakxTF</span><span class="p">(</span><span class="n">tn_object</span><span class="p">,</span> <span class="s2">&quot;-logp&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">TF_of_interest</span> <span class="ow">in</span> <span class="n">grnDF_mean</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">grnDF_mean</span><span class="p">[</span><span class="n">TF_of_interest</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">grnDF_ps</span><span class="p">[</span><span class="n">TF_of_interest</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">grnDF_mean</span> <span class="o">=</span> <span class="n">grnDF_mean</span><span class="p">[[</span><span class="s2">&quot;gene_short_name&quot;</span><span class="p">,</span> <span class="n">TF_of_interest</span><span class="p">]]</span>
    <span class="n">grnDF_ps</span> <span class="o">=</span> <span class="n">grnDF_ps</span><span class="p">[[</span><span class="n">TF_of_interest</span><span class="p">]]</span>
    <span class="n">grnDF</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">grnDF_mean</span><span class="p">,</span> <span class="n">grnDF_ps</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">grnDF</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_short_name&quot;</span><span class="p">,</span> <span class="s2">&quot;coef_abs&quot;</span><span class="p">,</span> <span class="s2">&quot;-logp&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;gene_mode&quot;</span><span class="p">:</span>
        <span class="n">grnDF</span> <span class="o">=</span> <span class="n">grnDF</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;gene_short_name&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
        <span class="n">promoter_enhancer_peak</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;peak_mode&quot;</span><span class="p">:</span>
        <span class="n">grnDF</span> <span class="o">=</span> <span class="n">grnDF</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;gene_short_name&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">promoter_enhancer_peak</span> <span class="o">=</span> <span class="n">grnDF</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>

    <span class="n">Chip_data</span> <span class="o">=</span> <span class="n">Chip_data</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">grnDF</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">integrated</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">Chip_data</span><span class="p">,</span> <span class="n">grnDF</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>



    <span class="c1">## change data structure</span>

    <span class="n">chip_sample_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">Chip_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="n">chip_sample_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chip_sample_names</span> <span class="k">if</span> <span class="ow">not</span> <span class="s2">&quot;gene&quot;</span> <span class="ow">in</span> <span class="n">i</span><span class="p">]</span>

    <span class="n">li</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chip_sample_names</span><span class="p">:</span>

        <span class="n">tmp</span> <span class="o">=</span> <span class="n">integrated</span><span class="p">[[</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;coef_abs&quot;</span><span class="p">,</span> <span class="s2">&quot;-logp&quot;</span><span class="p">]]</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">i</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">tmp</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;gene_short_name&quot;</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">tmp</span><span class="p">[</span><span class="s2">&quot;chip_score_order&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>

        <span class="n">tmp</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="s2">&quot;gene_short_name&quot;</span><span class="p">]</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;gene_short_name&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="n">tmp</span><span class="p">[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">i</span>

        <span class="n">tmp</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">i</span> <span class="p">:</span> <span class="s2">&quot;chip_bind&quot;</span><span class="p">})</span>
        <span class="n">li</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">)</span>


    <span class="n">integrated</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">li</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>


    <span class="k">return</span> <span class="n">integrated</span><span class="p">,</span> <span class="n">promoter_enhancer_peak</span></div>


<div class="viewcode-block" id="loadAndProcessData"><a class="viewcode-back" href="../../../modules/celloracle.chip_analysis.html#celloracle.chip_analysis.integrate_chip_data_with_grn_results.loadAndProcessData">[docs]</a><span class="k">def</span> <span class="nf">loadAndProcessData</span><span class="p">(</span><span class="n">save_dir</span><span class="p">,</span> <span class="n">TF_of_interest</span><span class="p">,</span> <span class="n">chip_data</span><span class="p">,</span> <span class="n">tn_paths</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">save_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">tn_path</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">tn_paths</span><span class="p">):</span>
        <span class="c1"># define save path</span>
        <span class="n">save_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_dir</span><span class="p">,</span>
                                 <span class="n">tn_path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_result.parquet&quot;</span><span class="p">)</span>
        <span class="c1"># load TransNetObject</span>
        <span class="n">tn1</span> <span class="o">=</span> <span class="n">tn</span><span class="o">.</span><span class="n">load_compressed_TN_object</span><span class="p">(</span><span class="n">tn_path</span><span class="p">)</span>
        <span class="c1"># integrate Chip seq data with TN results</span>
        <span class="n">integrated</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">integrate_Chip_and_TN</span><span class="p">(</span><span class="n">tn1</span><span class="p">,</span> <span class="n">chip_data</span><span class="p">,</span> <span class="n">TF_of_interest</span><span class="p">)</span>
        <span class="c1"># save integrated object</span>
        <span class="n">integrated</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_path</span><span class="p">))</span></div>

<span class="c1">#####################################################</span>
<span class="c1">### integrate Chip data and GRN inference results ###</span>
<span class="c1">###  fucusing on a Histone Marker                 ###</span>
<span class="c1">#####################################################</span>

<div class="viewcode-block" id="integrate_Chip_and_TN_for_HM_analysus"><a class="viewcode-back" href="../../../modules/celloracle.chip_analysis.html#celloracle.chip_analysis.integrate_chip_data_with_grn_results.integrate_Chip_and_TN_for_HM_analysus">[docs]</a><span class="k">def</span> <span class="nf">integrate_Chip_and_TN_for_HM_analysus</span><span class="p">(</span><span class="n">tn_object</span><span class="p">,</span> <span class="n">Chip_data</span><span class="p">):</span>

    <span class="c1"># version190607</span>

    <span class="c1"># extract grn data</span>
    <span class="n">grnDF</span> <span class="o">=</span> <span class="n">tn</span><span class="o">.</span><span class="n">getDF_peakxTF</span><span class="p">(</span><span class="n">tn_object</span><span class="p">,</span> <span class="s2">&quot;coef_abs&quot;</span><span class="p">)</span>

    <span class="c1"># get sum of absolute coef value.</span>
    <span class="c1"># this values sppose to represent activity score for this gene locus</span>
    <span class="n">grnDF</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">grnDF</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;gene_short_name&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
                         <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;coef_abs_sum&quot;</span><span class="p">])</span>
    <span class="c1">#return grnDF</span>
    <span class="n">grnDF</span> <span class="o">=</span> <span class="n">grnDF</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;peak_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
    <span class="n">grnDF</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">promoter_enhancer_peak</span> <span class="o">=</span> <span class="n">grnDF</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>

    <span class="c1"># integragte gene activity score with Chip data</span>
    <span class="n">integrated</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">Chip_data</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">grnDF</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
                            <span class="n">grnDF</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1">#return integrated</span>
    <span class="c1"># change data format</span>
    <span class="n">integrated</span> <span class="o">=</span> <span class="n">integrated</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">integrated</span> <span class="o">=</span> <span class="n">integrated</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="s2">&quot;coef_abs_sum&quot;</span><span class="p">])</span>
    <span class="n">integrated</span> <span class="o">=</span> <span class="n">integrated</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;index&quot;</span><span class="p">)</span>
    <span class="n">integrated</span> <span class="o">=</span> <span class="n">integrated</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;variable&quot;</span><span class="p">:</span> <span class="s2">&quot;sample_id&quot;</span><span class="p">,</span>
                                            <span class="s2">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;chip_bind&quot;</span><span class="p">})</span>
    <span class="k">return</span> <span class="n">integrated</span><span class="p">,</span> <span class="n">promoter_enhancer_peak</span></div>



<div class="viewcode-block" id="loadAndProcessData_for_HM_analysis"><a class="viewcode-back" href="../../../modules/celloracle.chip_analysis.html#celloracle.chip_analysis.integrate_chip_data_with_grn_results.loadAndProcessData_for_HM_analysis">[docs]</a><span class="k">def</span> <span class="nf">loadAndProcessData_for_HM_analysis</span><span class="p">(</span>
        <span class="n">save_dir</span><span class="p">,</span> <span class="n">chip_data</span><span class="p">,</span> <span class="n">tn_paths</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">save_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">tn_path</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">tn_paths</span><span class="p">):</span>
        <span class="c1"># define save path</span>
        <span class="n">save_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_dir</span><span class="p">,</span> <span class="n">tn_path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_result.parquet&quot;</span><span class="p">)</span>
        <span class="c1"># load TransNetObject</span>
        <span class="n">tn1</span> <span class="o">=</span> <span class="n">tn</span><span class="o">.</span><span class="n">load_net_from_patquets</span><span class="p">(</span><span class="n">tn_path</span><span class="p">)</span>
        <span class="c1"># integrate Chip seq data with TN results</span>
        <span class="n">integrated</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">integrate_Chip_and_TN_for_HM_analysus</span><span class="p">(</span><span class="n">tn1</span><span class="p">,</span> <span class="n">chip_data</span><span class="p">)</span>
        <span class="c1"># save integrated object</span>
        <span class="n">integrated</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_path</span><span class="p">))</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Samantha Morris Lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>